{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bc_epitopes_HLA_overlap-05-06-20.txt\r\n",
      "'Extract minimal epitope predictions.ipynb'\r\n",
      "Frequency_table_27mer.csv\r\n",
      "Frequency_table_27mer_all_ethnic_groups.csv\r\n",
      "HLA_freq_formated.txt\r\n",
      "HLA_freq_formatted.csv\r\n",
      "'Select vaccine peptides.ipynb'\r\n",
      "'Summarize Vaccine Peptides.ipynb'\r\n",
      "Table_S5_final_vaccine-peptides-merged.csv\r\n",
      "Window_15mer_all_proteins.txt\r\n",
      "Window_21mer_all_proteins.txt\r\n",
      "Window_27mer_all_proteins.txt\r\n",
      "final-vaccine-peptides-15mer.csv\r\n",
      "final-vaccine-peptides-15mer.fa\r\n",
      "final-vaccine-peptides-15mer.html\r\n",
      "final-vaccine-peptides-15mer.log\r\n",
      "final-vaccine-peptides-15mer.tex\r\n",
      "final-vaccine-peptides-21mer.csv\r\n",
      "final-vaccine-peptides-21mer.fa\r\n",
      "final-vaccine-peptides-21mer.html\r\n",
      "final-vaccine-peptides-21mer.tex\r\n",
      "final-vaccine-peptides-27mer.csv\r\n",
      "final-vaccine-peptides-27mer.fa\r\n",
      "final-vaccine-peptides-27mer.html\r\n",
      "final-vaccine-peptides-27mer.tex\r\n",
      "final-vaccine-peptides-h2b-15mer.csv\r\n",
      "final-vaccine-peptides-h2b-15mer.fa\r\n",
      "final-vaccine-peptides-h2b-15mer.html\r\n",
      "final-vaccine-peptides-h2b-15mer.tex\r\n",
      "final-vaccine-peptides-h2b-21mer.csv\r\n",
      "final-vaccine-peptides-h2b-21mer.fa\r\n",
      "final-vaccine-peptides-h2b-21mer.html\r\n",
      "final-vaccine-peptides-h2b-21mer.tex\r\n",
      "final-vaccine-peptides-h2b-27mer.csv\r\n",
      "final-vaccine-peptides-h2b-27mer.fa\r\n",
      "final-vaccine-peptides-h2b-27mer.html\r\n",
      "final-vaccine-peptides-h2b-27mer.tex\r\n",
      "final-vaccine-peptides-h2d-15mer.csv\r\n",
      "final-vaccine-peptides-h2d-15mer.fa\r\n",
      "final-vaccine-peptides-h2d-15mer.html\r\n",
      "final-vaccine-peptides-h2d-15mer.tex\r\n",
      "final-vaccine-peptides-h2d-21mer.csv\r\n",
      "final-vaccine-peptides-h2d-21mer.fa\r\n",
      "final-vaccine-peptides-h2d-21mer.html\r\n",
      "final-vaccine-peptides-h2d-21mer.tex\r\n",
      "final-vaccine-peptides-h2d-27mer.csv\r\n",
      "final-vaccine-peptides-h2d-27mer.fa\r\n",
      "final-vaccine-peptides-h2d-27mer.html\r\n",
      "final-vaccine-peptides-h2d-27mer.tex\r\n",
      "final-vaccine-peptides-murine-15mer.csv\r\n",
      "final-vaccine-peptides-murine-15mer.fa\r\n",
      "final-vaccine-peptides-murine-15mer.html\r\n",
      "final-vaccine-peptides-murine-15mer.tex\r\n",
      "final-vaccine-peptides-murine-21mer.csv\r\n",
      "final-vaccine-peptides-murine-21mer.fa\r\n",
      "final-vaccine-peptides-murine-21mer.html\r",
      "\r\n",
      "final-vaccine-peptides-murine-21mer.tex\r\n",
      "final-vaccine-peptides-murine-27mer.csv\r\n",
      "final-vaccine-peptides-murine-27mer.fa\r\n",
      "final-vaccine-peptides-murine-27mer.html\r\n",
      "final-vaccine-peptides-murine-27mer.tex\r\n",
      "final_vaccine-peptides-merged.csv\r\n",
      "human_27mers.pptx\r\n",
      "manufacturability-violin-plot.png\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-h2b.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-h2d.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-human-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-human-h2b.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-human-h2d.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-human.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides-murine.csv\r\n",
      "minimal-epitopes-in-15mer-vaccine-peptides.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-h2b.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-h2d.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-human-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-human-h2b.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-human-h2d.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-human.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides-murine.csv\r\n",
      "minimal-epitopes-in-21mer-vaccine-peptides.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-h2b.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-h2d.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-human-h2b-h2d.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-human-h2b.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-human-h2d.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-human.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides-murine.csv\r\n",
      "minimal-epitopes-in-27mer-vaccine-peptides.csv\r\n",
      "murine_27mers.pptx\r\n",
      "old-order.txt\r\n",
      "order-new-sorted.txt\r\n",
      "order-new.txt\r\n",
      "order-old-sorted.txt\r\n",
      "selected-bcell-15mer.csv\r\n",
      "selected-bcell-21mer.csv\r\n",
      "selected-bcell-27mer.csv\r\n",
      "selected-bcell-cd4-15mer.csv\r\n",
      "selected-bcell-cd4-21mer.csv\r\n",
      "selected-bcell-cd4-27mer.csv\r\n",
      "selected-bcell-cd4-cd8-15mer.csv\r\n",
      "selected-bcell-cd4-cd8-21mer.csv\r\n",
      "selected-bcell-cd4-cd8-27mer.csv\r\n",
      "selected-bcell-cd4-h2b-15mer.csv\r\n",
      "selected-bcell-cd4-h2b-21mer.csv\r\n",
      "selected-bcell-cd4-h2b-27mer.csv\r\n",
      "selected-bcell-cd4-h2d-27mer.csv\r\n",
      "selected-bcell-cd8-15mer.csv\r\n",
      "selected-bcell-cd8-21mer.csv\r\n",
      "selected-bcell-cd8-27mer.csv\r\n",
      "selected-bcell-cd8-h2b-15mer.csv\r\n",
      "selected-bcell-cd8-h2b-21mer.csv\r\n",
      "selected-bcell-cd8-h2b-27mer.csv\r\n",
      "selected-bcell-cd8-h2b-h2d-15mer.csv\r\n",
      "selected-bcell-cd8-h2b-h2d-21mer.csv\r\n",
      "selected-bcell-cd8-h2b-h2d-27mer.csv\r\n",
      "selected-bcell-cd8-h2d-15mer.csv\r\n",
      "selected-bcell-cd8-h2d-21mer.csv\r\n",
      "selected-bcell-cd8-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-15mer.csv\r\n",
      "selected-tcell-cd4-21mer.csv\r\n",
      "selected-tcell-cd4-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-h2b-15mer.csv\r\n",
      "selected-tcell-cd4-h2b-21mer.csv\r\n",
      "selected-tcell-cd4-h2b-27mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-h2d-27mer.csv\r\n",
      "selected-tcell-cd8-15mer.csv\r\n",
      "selected-tcell-cd8-21mer.csv\r\n",
      "selected-tcell-cd8-27mer.csv\r\n",
      "selected-tcell-cd8-h2b-15mer.csv\r\n",
      "selected-tcell-cd8-h2b-21mer.csv\r\n",
      "selected-tcell-cd8-h2b-27mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd8-h2d-15mer.csv\r\n",
      "selected-tcell-cd8-h2d-21mer.csv\r\n",
      "selected-tcell-cd8-h2d-27mer.csv\r\n",
      "summary-15mer.csv\r\n",
      "summary-15mer.tex\r\n",
      "summary-21mer.csv\r\n",
      "summary-21mer.tex\r\n",
      "summary-27mer.csv\r\n",
      "summary-27mer.tex\r\n",
      "tables.aux\r\n",
      "tables.log\r\n",
      "tables.pdf\r\n",
      "tables.tex\r\n",
      "vaccine-peptides-15mer-annotated-05-18-20.csv\r\n",
      "vaccine-peptides-15mer-annotated.csv\r\n",
      "vaccine-peptides-21mer-annotated-05-18-20.csv\r\n",
      "vaccine-peptides-21mer-annotated.csv\r\n",
      "vaccine-peptides-27mer-annotated-05-18-20.csv\r\n",
      "vaccine-peptides-27mer-annotated.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence,Protein,Start,End,HLA-I_peptides,HLA-I_haplotypes,HLA-I_pop_freq,HLA-II_peptides,HLA-II_haplotypes,HLA-II_pop_freq,Mouse MHC-I b,Mouse MHC-I d,Mouse MHC-II b,Mouse MHC-II d,Total_frequency\r\n",
      "MESLVPGFNEKTHVQ,orf1ab,1,15,\"ESLVPGFNEK,SLVPGFNEK\",\"HLA-A*11:01,HLA-A*03:01\",0.299196077180553,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "ESLVPGFNEKTHVQL,orf1ab,2,16,\"ESLVPGFNEK,SLVPGFNEK\",\"HLA-A*11:01,HLA-A*03:01\",0.299196077180553,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "SLVPGFNEKTHVQLS,orf1ab,3,17,SLVPGFNEK,\"HLA-A*03:01,HLA-A*11:01\",0.299196077180553,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "LVPGFNEKTHVQLSL,orf1ab,4,18,\"\",\"\",0,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "VPGFNEKTHVQLSLP,orf1ab,5,19,\"\",\"\",0,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "PGFNEKTHVQLSLPV,orf1ab,6,20,\"\",\"\",0,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "GFNEKTHVQLSLPVL,orf1ab,7,21,HVQLSLPVL,HLA-C*03:04,0.143681590569932,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "FNEKTHVQLSLPVLQ,orf1ab,8,22,HVQLSLPVL,HLA-C*03:04,0.143681590569932,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n",
      "NEKTHVQLSLPVLQV,orf1ab,9,23,\"HVQLSLPVL,QLSLPVLQV,VQLSLPVLQV\",\"HLA-C*03:04,HLA-A*02:01\",0.501188874694642,\"\",\"\",0,\"\",\"\",\"\",\"\",0\r\n"
     ]
    }
   ],
   "source": [
    "!head Window_15mer_all_proteins.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_15mer = pd.read_csv(\"Window_15mer_all_proteins.txt\")\n",
    "df_21mer = pd.read_csv(\"Window_21mer_all_proteins.txt\")\n",
    "df_27mer = pd.read_csv(\"Window_27mer_all_proteins.txt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>HLA-I_peptides</th>\n",
       "      <th>HLA-I_haplotypes</th>\n",
       "      <th>HLA-I_pop_freq</th>\n",
       "      <th>HLA-II_peptides</th>\n",
       "      <th>HLA-II_haplotypes</th>\n",
       "      <th>HLA-II_pop_freq</th>\n",
       "      <th>Mouse MHC-I b</th>\n",
       "      <th>Mouse MHC-I d</th>\n",
       "      <th>Mouse MHC-II b</th>\n",
       "      <th>Mouse MHC-II d</th>\n",
       "      <th>Total_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESLVPGFNEKTHVQ</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>ESLVPGFNEK,SLVPGFNEK</td>\n",
       "      <td>HLA-A*11:01,HLA-A*03:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ESLVPGFNEKTHVQL</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>ESLVPGFNEK,SLVPGFNEK</td>\n",
       "      <td>HLA-A*11:01,HLA-A*03:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SLVPGFNEKTHVQLS</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>SLVPGFNEK</td>\n",
       "      <td>HLA-A*03:01,HLA-A*11:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LVPGFNEKTHVQLSL</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VPGFNEKTHVQLSLP</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sequence Protein  Start  End        HLA-I_peptides  \\\n",
       "0  MESLVPGFNEKTHVQ  orf1ab      1   15  ESLVPGFNEK,SLVPGFNEK   \n",
       "1  ESLVPGFNEKTHVQL  orf1ab      2   16  ESLVPGFNEK,SLVPGFNEK   \n",
       "2  SLVPGFNEKTHVQLS  orf1ab      3   17             SLVPGFNEK   \n",
       "3  LVPGFNEKTHVQLSL  orf1ab      4   18                   NaN   \n",
       "4  VPGFNEKTHVQLSLP  orf1ab      5   19                   NaN   \n",
       "\n",
       "          HLA-I_haplotypes  HLA-I_pop_freq HLA-II_peptides HLA-II_haplotypes  \\\n",
       "0  HLA-A*11:01,HLA-A*03:01        0.299196             NaN               NaN   \n",
       "1  HLA-A*11:01,HLA-A*03:01        0.299196             NaN               NaN   \n",
       "2  HLA-A*03:01,HLA-A*11:01        0.299196             NaN               NaN   \n",
       "3                      NaN        0.000000             NaN               NaN   \n",
       "4                      NaN        0.000000             NaN               NaN   \n",
       "\n",
       "   HLA-II_pop_freq Mouse MHC-I b Mouse MHC-I d Mouse MHC-II b Mouse MHC-II d  \\\n",
       "0              0.0           NaN           NaN            NaN            NaN   \n",
       "1              0.0           NaN           NaN            NaN            NaN   \n",
       "2              0.0           NaN           NaN            NaN            NaN   \n",
       "3              0.0           NaN           NaN            NaN            NaN   \n",
       "4              0.0           NaN           NaN            NaN            NaN   \n",
       "\n",
       "   Total_frequency  \n",
       "0              0.0  \n",
       "1              0.0  \n",
       "2              0.0  \n",
       "3              0.0  \n",
       "4              0.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_15mer.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "extremely_unstable_dipeptides_unc = {\n",
    "  \"DG\",\n",
    "  \"NG\",\n",
    "}\n",
    "problematic_dipeptides_unc = {\n",
    "   \"DS\", \n",
    "   \"DN\", \n",
    "   \"DD\", \n",
    "   \"NN\", \n",
    "   \"ND\", \n",
    "   \"NS\",\n",
    "}\n",
    "\n",
    "problematic_dipeptides_jpt = {\"NP\"}\n",
    "\n",
    "problematic_dipeptides = problematic_dipeptides_unc.union(problematic_dipeptides_jpt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9701"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_15mer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9701"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_21mer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9701"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_27mer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    9701.000000\n",
       "mean       26.963818\n",
       "std         0.798730\n",
       "min         1.000000\n",
       "25%        27.000000\n",
       "50%        27.000000\n",
       "75%        27.000000\n",
       "max        27.000000\n",
       "Name: Sequence, dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_27mer.Sequence.str.len().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9687 9681 9675\n"
     ]
    }
   ],
   "source": [
    "# in case the tiling of sequences included all shorter sequences at the start/end of the whole proteome\n",
    "# drop those sequences \n",
    "df_15mer = df_15mer[df_15mer.Sequence.str.len() == 15]\n",
    "df_21mer = df_21mer[df_21mer.Sequence.str.len() == 21]\n",
    "df_27mer = df_27mer[df_27mer.Sequence.str.len() == 27]\n",
    "print(len(df_15mer), len(df_21mer), len(df_27mer))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "hydropathy_dict = {\n",
    "    \"A\": 1.8,\n",
    "    \"C\": 2.5,\n",
    "    \"D\": -3.5,\n",
    "    \"E\": -3.5,\n",
    "    \"F\": 2.8,\n",
    "    \"G\": -0.4,\n",
    "    \"H\": -3.2,\n",
    "    \"I\": 4.5,\n",
    "    \"K\": -3.9,\n",
    "    \"L\": 3.8,\n",
    "    \"M\": 1.9,\n",
    "    \"N\": -3.5,\n",
    "    \"P\": -1.6,\n",
    "    \"Q\": -3.5,\n",
    "    \"R\": -4.5,\n",
    "    \"S\": -0.8,\n",
    "    \"T\": -0.7,\n",
    "    \"V\": 4.2,\n",
    "    \"W\": -0.9,\n",
    "    \"Y\": -1.3\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gravy_score(amino_acids):\n",
    "    \"\"\"\n",
    "    Mean amino acid hydropathy averaged across residues of a peptide\n",
    "    or protein sequence.\n",
    "    \"\"\"\n",
    "    total = 0 \n",
    "    for aa in amino_acids:\n",
    "        total += hydropathy_dict[aa]\n",
    "    return total / len(amino_acids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def max_kmer_gravy_score(amino_acids, k):\n",
    "    \"\"\"\n",
    "    Returns max GRAVY score of any kmer in the amino acid sequence,\n",
    "    used to determine if there are any extremely hydrophobic regions within a\n",
    "    longer amino acid sequence.\n",
    "    \"\"\"\n",
    "    return max(\n",
    "        gravy_score(amino_acids[i:i + k])\n",
    "        for i in range(len(amino_acids) - k + 1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def max_short_subsequence_gravy_score(amino_acids):\n",
    "    gravy_scores = [\n",
    "        max_kmer_gravy_score(amino_acids, k)\n",
    "        for k in [5, 6, 7, 8]\n",
    "    ]\n",
    "    return max(gravy_scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def difficult_c_terminal(amino_acids):\n",
    "    return amino_acids[-1] in {\"P\", \"C\", \"H\"}\n",
    "\n",
    "def difficult_n_terminal(amino_acids):\n",
    "    \"\"\"\n",
    "    Is the N-terminus one of {Gln, Glu, Cys}?\n",
    "    ---\n",
    "    Priority I: avoid N-terminal Gln, Glu, Cys\n",
    "    Priority II: Asparagine at the N-terminus of a peptide is also hard\n",
    "    to synthesize, though not as bad as {Gln, Glu, Cys}\n",
    "    \"\"\"\n",
    "    return amino_acids[0] in {\"Q\", \"E\", \"C\", \"N\"}\n",
    "\n",
    "\n",
    "def cysteine_count(amino_acids):\n",
    "    \"\"\"\n",
    "    How many cysteines are in the amino acid sequence?\n",
    "    Problem with cysteine residues: They can form disulfide bonds across\n",
    "    distant parts of the peptide\n",
    "    \"\"\"\n",
    "    return amino_acids.count(\"C\")\n",
    "\n",
    "def methionine_count(amino_acids):\n",
    "    \"\"\"\n",
    "    How many methionies?\n",
    "    \"\"\"\n",
    "    return amino_acids.count(\"M\")\n",
    "\n",
    "def thiol_count(amino_acids):\n",
    "    return cysteine_count(amino_acids) + methionine_count(amino_acids)\n",
    "\n",
    "def num_extremely_unstable_dipeptides(amino_acids):\n",
    "    total = 0\n",
    "    for dipeptide in extremely_unstable_dipeptides_unc:\n",
    "        total += amino_acids.count(dipeptide)\n",
    "    return total\n",
    "    extremely_unstable_dipeptides_unc\n",
    "\n",
    "def num_moderately_unstable_dipeptides(amino_acids):\n",
    "    total = 0\n",
    "    for dipeptide in problematic_dipeptides_unc.union(problematic_dipeptides_jpt):\n",
    "        total += amino_acids.count(dipeptide)\n",
    "    return total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_manufacturability(df):\n",
    "    df = df.copy()\n",
    "    df[\"num_extremely_unstable_dipeptides\"] = df.Sequence.map(num_extremely_unstable_dipeptides)\n",
    "    df[\"num_moderately_unstable_dipeptides\"] = df.Sequence.map(num_moderately_unstable_dipeptides)\n",
    "    df[\"num_cysteines\"] = df.Sequence.map(cysteine_count)\n",
    "    df[\"num_methionines\"] = df.Sequence.map(methionine_count)\n",
    "    df[\"num_thiols\"] = df.Sequence.map(thiol_count)\n",
    "    df[\"difficult_c_terminal\"] = df.Sequence.map(difficult_c_terminal)\n",
    "    df[\"difficult_n_terminal\"] = df.Sequence.map(difficult_n_terminal)\n",
    "    df[\"gravy\"] = df.Sequence.map(gravy_score)\n",
    "    df[\"max_kmer_gravy\"] = df.Sequence.map(max_short_subsequence_gravy_score)\n",
    "    df[\"extreme_local_hydrophobicity\"] = df.max_kmer_gravy > 2.5\n",
    "    df[\"difficult_local_hydrophobicity\"] = df.max_kmer_gravy > 1.5\n",
    "    df[\"too_many_cysteines\"] = df.num_cysteines > 1\n",
    "    df[\"manufacturability_difficulty_score\"] = (\n",
    "        1.0  * (df.gravy > 2.0) +        # hydrophobicity of whole peptide\n",
    "        1.0 * df.difficult_n_terminal +  # lowest priority: N-terminal residues\n",
    "        2.0 * df.difficult_c_terminal + \n",
    "        2.0 * df.num_thiols + \n",
    "        2.0 * df.difficult_local_hydrophobicity + \n",
    "        3.0 * df.num_moderately_unstable_dipeptides + \n",
    "        5.0 * df.too_many_cysteines +          # disulfide bonds\n",
    "        10.0 * df.extreme_local_hydrophobicity + \n",
    "        10.0 * df.num_extremely_unstable_dipeptides \n",
    "    )\n",
    "    return df\n",
    "\n",
    "df_27mer_annot = annotate_manufacturability(df_27mer);\n",
    "df_27mer_annot.to_csv(\"vaccine-peptides-27mer-annotated.csv\", index=False)\n",
    "df_21mer_annot = annotate_manufacturability(df_21mer);\n",
    "df_21mer_annot.to_csv(\"vaccine-peptides-21mer-annotated.csv\", index=False)\n",
    "df_15mer_annot = annotate_manufacturability(df_15mer)\n",
    "df_15mer_annot.to_csv(\"vaccine-peptides-15mer-annotated.csv\", index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_concat = pd.concat([\n",
    "    df_15mer_annot,\n",
    "    df_21mer_annot,\n",
    "    df_27mer_annot,\n",
    "])\n",
    "df_concat[\"Length\"] = df_concat.Sequence.str.len()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAs4AAAHgCAYAAAC1ouv3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOzdeXhcZ3n///czo13e9z12VmcPIUBJ+6UlLQVKQvkGSn/9lkIpVxfaQhNoVhIgCYEsUEIIlABJIM5GFrIZCqQJgSQkJLbjJfESr7Jk2ZYlWdY2I83MuX9/zIw8sSVrRp4zZ+bo87quuUaz34klzUf33Od5nJkhIiIiIiJHFgm6ABERERGRSqDgLCIiIiKSBwVnEREREZE8KDiLiIiIiORBwVlEREREJA8KziIiIiIieagKuoB8zZgxwxYvXhx0GSIiIiISYitXrmw3s5nD3VYxwXnx4sWsWLEi6DJEREREJMScc00j3aZRDRERERGRPCg4i4iIiIjkQcFZRERERCQPCs4iIiIiInlQcBYRERERyYOCs4iIiIhIHhScRURERETyoOAsIiIiIpIHBWcRERERkTwoOIuIiIiI5EHBWUREREQkDwrOIiIiIiJ5UHAWEREREcmDgrOIiIiISB4UnEVERERE8qDgLCIiIiKSBwVnEREREZE8KDiLiIiIFNHOnTtZvnx50GWID6qCLkBEREQkTK655ho2b97M+eefH3QpUmTqOIuIiIgU0ebNm4MuQXyi4CwiIiIikgcFZxERERGRPCg4i4iIiPjAzIIuQYpMwVlERETEBwrO4aPgLCIiIuIDBefwUXAWERER8YGCc/goOIuIiIj4QME5fBScRURERHyg4Bw+Cs4iIiIiPlBwDh8FZxEREREfKDiHj4KziIiIiA88zwu6BCkyBWcRERERkTwoOIuIiIj4QB3n8FFwFhEREfGBZpzDR8FZRERExAepVCroEqTIFJxFRERERPKg4CwiIiLiA804h4+Cs4iIiIgPFJzDR8FZRERExAcKzuGj4CwiIiLiA62qET4KziIiIiI+UHAOHwVnERERER8oOIePgrOIiIiIDzTjHD4KziIiIiI+UMc5fBScRURERHyg4Bw+Cs4iIiIiPlBwDh8FZxEREREfaMY5fBScRURERETyoOAsIiIi4gONaoSPgrOIiIiIDzSqET4KziIiIZVMJkkkEkGXITJuqeMcPgrOIiIh9YUvfIGLLroo6DJExi0F5/CpCroAERHxx4svvhh0CSLjmoJz+KjjLCIiIuIDBefwUXAWERER8YEODgwfBWcRERERH6jjHD4KziIiIiIieVBwFhEREfGBRjXCR8FZRERExAca1QgfBWcRERERHyg4h4+Cs4iIiIgPFJzDR8FZRERExAcKzuGj4CwiIiIikgcFZxEREREfqOMcPgrOIiIiIj5QcA4fBWcRERERkTwoOIuIiIj4QB3n8FFwFhERERHJg4KziIiIiEgeFJxFRERERPKg4CwiIiLiA804h4+Cs4iIiIgPFJzDR8FZRERExAcKzuFTkuDsnIs65151zi3PXF7inPu9c26Lc+4nzrmaUtQhIiIiIjJWpeo4/wewIefyjcA3zex4YD/wqRLVISIiIlISnucFXYIUme/B2Tm3APgA8MPMZQecBzycucuPgQ/5XYeIiIhIKWlUI3xK0XG+BbgUyP7ZNR3oMrNk5nILML8EdYiIiIiUjIJz+PganJ1z5wNtZrZyjI//J+fcCufcin379hW5OhERERH/aFQjfPzuOP8h8EHn3A7gAdIjGt8CpjjnqjL3WQDsGu7BZvZ9MzvHzM6ZOXOmz6WKiIiIFE8qlQq6BCkyX4OzmV1hZgvMbDHw/wHPmNnfAr8GPpK52yeAx/2sQ0RERKTU1HEOn6DWcb4M+Jxzbgvpmec7AqpDRERExBfqOIdP1eh3KQ4zexZ4NvP1NuDtpXptERERkVJTcA4f7RwoIiIi4oNkMjn6naSiKDiLiIiI+EDBOXwUnEVERER8oOAcPgrOIiIiIj5QcA4fBWcRERGRIsldgi6RSARYifhBwVlERESkSHLDslbVCB8FZxEREZEiyQ3LGtUIHwVnERERkSLJ7ThrVCN8FJxFREREiiS3y6yOc/goOIuIiIgUSW5Y1oxz+Cg4i4iIiBRJbljWqEb4KDiLiIiIFIlGNcJNwVlEJOTMLOgSRMYNjWqEm4KziEjIKTiLlI6Cc7gpOIuIhJyCs0jpaFQj3BScRURCTsFZpHRyu8zqOIePgrOISMh5nhd0CSLjhnYODDcFZxGRkFPHWaR0NKoRbgrO4rstW7bQ1dUVdBkiIiK+y4Zlc6bgHEIKzuK7f/iHf+Caa64JugyRcUsdZ5HSGQrLVeo4h5GCs5TEypUrgy5BZNzSjLNI6QyF5ah2DgwjBWcRERGRIsntOA8mBoMtRopOwVlEJOQ0qiFSOkNd5mp1nMNIwVlEJIRyw7JGNURKJ7fjnEgqOIeNgrOISAjlhmV1nEVKZ3AwM55RlfO1hIaCs4hICKnjLBKMbFi2GtOoRggpOIuIhFDu7mUKziKlM9RlrobEoIJz2Cg4i6/0EbFIMHLDsoKzSOkMDg6m01WVgnMYKTiLr3K7XiJSOrk/e/o5FCmdgYEBXNRBNP2zp01QwkXBWXylTpdIMHLfrPXGLVI68XgcV5UOzpAO0hIeCs7iKwVnkWAoOIsEY2BgIB2aq9KX4/F4oPVIcSk4i68UnEWCkTueoeAsUjqxWAyrsqGOs4JzuCg4i68UnEWCkbt+rJbEEimdeDyORSwdnlFwDhsFZ/GVgrNIMHK7zArOIqXT39+PxQzX5IYuS3goOIuvFJxFgqGOs0gw+vr7IAWuPx2cY7FYwBVJMSk4i68UnEWCkRuWFZxFSqe/vx8c6RPqOIeNgrP4ShugiAQjNyzndp9FxF/9sTcHZ3Wcw0XBWXyl4CwSjNywrOAsUjrx/rg6ziGm4Cy+0qiGSDA04yxSeslkMv3zpuAcWgrO4it1nEWCoRlnkdIbCsmZ4Owijr6+vkBrkuJScBZfKTiLBEMzziKlNxSSM91mV+PUcQ4ZBWfxlYKzSDDUcRYpvaGQnE1XVRrVCBsFZ/GVgrNIMBScRUrvTaMagFWbgnPIKDiLiIRQbljO3UVQRPxz6KiGF/Xo7e0NriApOgVn8ZU6ziLByA3LCs4ipXFocKYaevsUnMNEwVl8peXoRIKR7ThXR3RwoEipHBqcrcrUcQ4ZBWfxlYKzSDCSySQOqIk6dZxFSmQoOGfTVTX09Ws5ujBRcBZfaVRDJBiJRIKqCFRFNKohUip9fX0HxzQAqiHWpy23w0TBWXyVSqWCLkFkXEomk1RFHdGIVtUQKZXe3l4i1ZE3zTgnk0kGBgYCrUuKR8FZfKXgLBKMRCJB1EGVMwVnkRLp6emBmpwranKul1BQcBZfacZZJBiJRIJqjWqIlFRPTw9WnTOiWJ0+0wGC4VFQcHbO1TvnTvKrGAkfvWGLBCORSBCNmDrOIiXU09ODV32wYWQ1NnS9hEPewdk5dwGwGvhF5vJZzrkn/CpMwkHBWSQYiUSCagdVEU/BWaREug50DYVlYGhUo7u7O5iCpOgK6Th/GXg70AVgZquBJT7UJCGiGWeRYKRX1Uh3nLWOs0hpHDhwAGpzrqjNuV5CoZDgnDCzQ//ltdaYHJE6XSLBGBwcpCbiURMxBgd1RL+I31KpFH29fcMG566urkBqkuIrJDi/7pz7f0DUOXeCc+7bwO98qktCQsFZJBgDAwNUO4/qiDEQjwddjkjo9fT0pPcuyF1VIwou6hScQ6SQ4PwZ4FRgALgPOABc5EdREh65wVmboYiUzuBAnOoo1ETRGrIiJbB///70F7kdZweuTsE5TKryuZNzLgpca2b/CXzB35IkTHKDczKZpLq6OsBqRMaPeDzOtIhREzEFZ5ES6OzsBMDqDZezfaBX6w3dJpUvr46zmaWAP/K5Fgmh3OCssQ2R0onFYtRGjbqoEdeohojvOjo60l/Uvfl6qzXa29tLX5D4Iq+Oc8armeXnHgL6slea2U+LXpWEhoKzSDDi8Rg1jelRjXhcHWcRvw11lQ8NznVGe4eCc1gUEpzrgA7gvJzrDFBwlhEdOqohIqURjw9QO8mojRqDiQSpVIpoNBp0WSKh1dHRgatyhyereujp7iGZTFJVVUjsknKU97+gmX3Sz0IknHLXcVZwFikNMyM+MEhdZlQD0gcINjQ0BFyZSHh1dHTg6h05481pdemfyf379zNz5sxAapPiKWTnwAXOuUedc22Z0yPOuQV+FieVLzcsKziLlMbAwABm6W5zbSY49/f3B1yVSLjt27ePVO3hm35ZffpnUHPO4VDIcnR3AU8A8zKnJzPXiYzI87xhvxYR/8RiMQDqolAbffN1IuKPtva2oZD8JvXpMwXncCgkOM80s7vMLJk5/QjQZw5yRLlhWes4i5RGNiTndpwVnEX8Y2Z0dnQedmAgMHTd0KobUtEKCc4dzrmPOeeimdPHSB8sKDKi3LCsjrNIaWSXn6s9ZMZZRPzR39/PQHxgqLv8JnWAS49ySOUrJDj/A/BRYA+wG/gIoAMG5Yhyg7M6ziKlkRuc1XEW8d/QGMZwwdlBpD6ijnNIFLKqRhPwQR9rkRByzg37tYj452BwPjjjrE1QRPyTDc5WN3yDyKvz1HEOiUJW1fixc25KzuWpzrk7/SlLRETGariOs4KziH+GusnDdZxJB+p97QrOYVDIqMYZZtaVvWBm+4G3FL8kCZNIJDLs1yLin6HgHDFqIppxFvHbSNttZ1mdaVQjJApJMhHn3NTsBefcNEYZ9XDO1TnnXnbOrXHOve6cuyZz/RLn3O+dc1uccz9xztWMrXwpd7k7lWnXMpHSyIbk6mh6y21Qx1nET0O7BlaPcIe6g7sHSmUrJDh/A3jROXedc+4rwO+Am0Z5zABwnpmdCZwFvM859wfAjcA3zex4YD/wqcJLl0qgjrNI6Q3XcVZwFvHP0K6BI8mMcHR2dpamIPFN3knGzO4GLgT2kl5Z40IzWzbKY8zMejMXqzMnA84DHs5c/2PgQwXWLRUit8us4CxSGrkzztURcE7BWcRPHR0dw+4amJU9aFDBufIVcnDgccBWM7sNeA34s9yDBY/wuKhzbjXQBjwFbAW6zCz7eUULML/gyqUi5IZljWqIlEY2JNdE06G5Nuo04yzio33t+0ZcUQPQJighUkgL8BEg5Zw7HrgdWAjcN9qDzCxlZmcBC4C3A0vzfUHn3D8551Y451ZoGZfKlBuWq6ryXv1QRI5CPB6nOgqRzCfHtVVax1nET52dI+wamFWXcz+paIUEZy/TJb4QuM3MLgHm5vvgzIocvwbeCUxxzmVT1AJg1wiP+b6ZnWNm58ycqd29K5FGNURKLxaLUVd1cN6yLmoKziI+GRgYINYfyys4q+Nc+QpJMgnn3N8AHweWZ64b6fhRAJxzM7PjHM65euA9wAbSAfojmbt9Ani8kKKlcmgDFJHS6+vro6Hq4MfG9VGP/v7+ACsSCa+hXQOPFJwjEKmLHLyvVKxCgvMnSXeLrzez7c65JcARDw4k3ZH+tXNuLfAK8JSZLQcuAz7nnNsCTAfuKLx0EREZTl9fH/VRb+hyfTRFX1/vER4hImM1tGtg/RFmnDO3KzhXvkK23F4PfDbn8nbSy8oB4Jx7xMw+fMhj1jLMJilmto30vLOEnLrMIqXX19dLXeTgEf71UeNAd3eAFYmE11AYHmHXwCyv1qNtX5v/BYmvijl0emwRn0tCwvMOdr3MjvzXuIgUR9f+/UysOfjzNrHGONB9IMCKRMJraPGCUYKz1Rta6KDyFTM4KxXJYVKpg12v3BAtIv7p6trPpJqDP2+Tajy6urr1x6uID/bu3YurPsKugVkNcKDrgJaGrHBa5kB8lRucc78WEX8kk0m6e/qYVG0s21TPsk31TKo2kqkUvb2acxYpttbWVmgERptMnJA+2717t98liY+KGZw1zCqHSSQSw34tIv7o6urCzJhc69HUE6WpJ8rk2nT3WUthiRRf865mvMbRP1G1xvQnPq2trX6XJD4qZOfAC5xzR7r/ZUWoR0JmcHBw2K9FxB979uwBYEbdwTfy7Nd79+4NpCaRsPI8j7279w6F4iNqTJ/t2jXs1hVSIQrpOP81sNk5d5Nz7rDd/8zsV8UrS8IiNyxrrkvEf9ngPHOY4Jy9TUSKY/fu3elPUyfmcedaiNRG2LFjh99liY/yDs5m9jHSS8ttBX7knHsxsyV2Pt8uMk7l7lamnctE/JcNx9PrDwbnKbVGNKKPiEWKbevWrQDYlDw6zg5Sk1Js2brF56rETwXNOJtZN/Aw8ADpzU3+L7DKOfcZH2qTEFBwFimt5uZmptQ56g7udk/EwawGaGlpCa4wkRDasiUTgifld3+bbGzbtk2rTFWwQmac/9I59yjwLOlFV95uZu8HzgQ+7095Uulyt/nVlr8i/tuxYzvz6g8/EHde/SBNO7YHUJFIeG3ZsgU3yeW/ndwUGIgP6NOfClZIx/lC4JtmdrqZ3WxmbQBm1g98ypfqpOL19fVhNY1DX4uIf8yMph1NzGtMHnbbvMYUu1pbSSYPv01ECmdmvL7+dVJT8l9q1aamRzrWr1/vV1nis0KC8x4z+23uFc65GwHM7OmiViWh0d3dg1eTXrxSa8iK+Ku9vZ3+WIx5wyyNNb/RI5XyNK4hUiS7d+9mf+d+mFHAgyaDq3GsW7fOt7rEX4UE5/cMc937i1WIhFN3Tw9W0wDO0dPTE3Q5IqG2bds2ABZOOLwDlr0uezCTiBydbPi16QXsyOnAm+qxZu0an6oSv40anJ1zn3bOrQOWOufW5py2A2v9L1EqWW9vL1ZVi6uq1aiGiM+yoXjRMMF5XmOKqDsYrkXk6Kxbtw5X42ByYY+zGemRKjWTKlM+Hef7gAuAxzPn2dNbM0vUiQzLzOjv78OqaqGqRr8kRHy2ZcsWptdDY/XhHbDqCMydYAdXARCRo/LKilfwpnsF75tsMw0zY/Xq1f4UJr7KJzhHgW7g34CenBPOuWn+lSaVLhaL4aVSEK3Fi9YqOIv4bOuWzSxsHHmHzkWNg2zZvKmEFYmE0+7du9nduhubXcCYRtZ0cFWOV155pfiFie/yWUBlJZD9zjj07yoDji1qRRIa2YMBraoGL1JNT48ODhTxy+DgIDubm/nAwpGP8F80IcXvtnTS09PDxInau0pkrFauXAkwtuAcAW+Gx8uvvFzkqqQURg3OZrakFIVI+GRnmi1ag0Vr6OlVx1nEL01NTaRS3rDzzVm5BwieddZZpSpNJHReeeUVXIPLb6vtYdhso3VNK3v27GHOnDnFLU58VcgGKO8a7uRncVLZhpafi9ZgVTXqOIv4aOjAwIlH6DhnbtOcs8jYpVIpXn7lZVIzUwXPN2fZnHSnWuMalSffvW4ALsn5ug54O+kxjvOKWpGERrbjXLXvDaymgVivdg4U8cv27dupisDs+pG38p1SY0yscezYsaN0hYmEzKZNm+jr7YNTj+JJJoJrcLz88stccMEFRatN/Jd3cDazN/3LOucWArcUvSIJjWxwdgPdePWTGYjHMTOcG+Of6CIyoh07djCv0Yge4XNE52BeQ4Lt27UknchYvfxyejZ5TPPNWQ5Ss9Od62QySVVVIX1MCVIhG6AcqgU4uViFSPjEYrHMVw4i1ZhZznUiUkw7tm1lXkNi1PvNb0yxY/t2zI7iTV9kHHvp9y/BNKD26J7HZhux/hgbNmwoSl1SGoXMOH/bOXdr5nQb8Bywyr/SpNL192dGM5zDotVvvk5EiiYej7OnbR/zG0eeb86a35iip7eP/fv3l6AykXDp7e1l44aNeLNHHonK2+z0meacK0shHecVpGeaVwIvApdpAxQ5kqHusotg0fTHUPF4PMCKRMJp165dmBlz8wjO2fs0Nzf7XZZI6KxevRrP87BZRfjEpgaYBitWrjj655KSKWSo5mEgbmYpAOdc1DnXYGZqIcqw3hSSI9WHXyciRZENwXMbRu+Czcncp7m5mTPPPNPXukTCZuXKlbgqB9OL83zeLI/169fT399PQ0NDcZ5UfFVIx/lpoD7ncj3wv8UtR8IkFoulj0aCoY6zRjVEiq+lpQWA2fWjd5xn1HlURQ4+RkTyN7TNdrQ4z2ezDC/lsWbNmuI8ofiukOBcZ2ZDC/FmvtafRzKidEjOfItFanKuE5FiamlpYWqdoy6PzxAjDmY1mEY1RAq0f/9+djbtLM6YRtYMcBHH6tWri/ec4qtCgnOfc+7s7AXn3FsBLZEgI+rv78/pOOvgQBG/tLQ0M6tu9BU1smbXJWjdpeAsUohsV9hmFjE4R8GmGate1VoLlaKQGeeLgIecc62k98qZA/y1L1VJKPT19WHZbZUUnEV8s6ulhVPyGNPImlXvsWHXbq2rLlKA1atXp+ebpxb3eb2ZHps3bdacc4XIu+NsZq8AS4FPA/8CnGxmK/0qTCpfd3cPuPS3mFWlF7zs6ekJsiSR0InH43R07j/ijoGHmt3gER8YoLOz08fKRMLl1dWvpuebj2YHjGHYDMPzPF577bXiPrH4YtR/fufceZnzC4ELgBMzpwsy14kMq+vAgaFRDSJV4CJ0d3cHW5RIyOzatQuAOQ35d5yzBxFmHysiR9bT08OO7TuwGT5sHDQDcLBu3briP7cUXT6jGu8CniEdmg9lwE+LWpGERk9PN5bpOOMcrqZOwVmkyIZW1MhjKbqs7H1bWlo444wzfKlLJEzWr1+PmWHTfQjOVeCmONauW1v855aiyyc4Z7eXusPMnvezGAmPWCxGPBaD2olD11lVAx0dHQFWJRI+2a7x7AI6zjPrPKJOHWeRfL322mvpo7uKtH7zoVLTU6xfv55kMklVVSGHn0mp5TOp88nM+a1+FiLh0t7env7CHfwWS1bX09a2L6CKRMKppaWFSbWOhgLea6MRmNmg3QNF8rV23VqYQmFLKhRiOgzEB9i6datPLyDFks+3wAbn3GZgvnMu93MEB5iZ6XM+OUw2OJs7uEq8VTewr31vUCWJhFLzzp3Mqc9/KbqsOfWDNO9s8qEikXBJpdLdYG9+/uNQhcqOgLz++uucdNJJvr2OHL1Rg7OZ/Y1zbg7wS+CD/pckYbBnz570F5Gc4Fw7gQP7NhGPx6mrqwuoMpFwaW5u4vQCxjSy5jR4bNy1C8/ziESKvEyASIg0NTUxEB+AaT6+SANE6iNs2LDBxxeRYshnVY2nzWwP8Eszazr0VIIapQK1tramV9TICc5e7SQgJ1SLyFHp6+ujc/+Bguabs+Y0pIgPDB4cqxKRYa1fvx7AnwMDsxykpqZY97pW1ih3+bQZ5jrnziW9/NxbnHNn5578LlAq065du3C1E950ndVNHLpNRI5edkWNOQWs4ZyVfUz2OURkeBs2bMDVOJgw+n2Phk0zWltatd9BmctnxvmLwNXAAuC/DrnNgPOKXZRUvp3NzSRrJoIdfEPPdpx1QJJIcbS2tgKFLUWXlX1Ma2srZ5+tHojISDZu2og3xQOfN9m0aemO9ubNm/UzWcZG7Tib2cNm9n7gJjN79yEnhWY5jJnRvLMZr37Km2+orsNV1yk4ixRJ9tObWQVst501vc4jGtEnQCJHkkwm2b59OzbFxzGNrMnpsy1btvj/WjJmo3acnXNLzWwj8LPhRjPMbJUvlUnF6uzsJBbrx5s1mUj/m7f0TdVNpqlpZ0CViYRLa2srk2sd9WNYIiviYGa9grPIkezcuZNkIpleis5vdRBpiLB58+YSvJiMVT6/bj8P/CPwjWFu06iGHCbbUba6yYfdlqqbTNNOBWeRYti7dy/Ta5NjfvyM2gRte3WwrshIst3fknScgdSkFJve2FSS15KxyWc5un/MnL/b/3IkDLLB2as/PDhb3WQO7HuDvr4+GhsbS12aSKi07d3NrNrCxzSyptV6vL5Xa6uLjKSpqSk92zxx1LsWhU02Wra1kEqliEajoz9ASi6fUY0Lj3S7mf20eOVIGDQ3N0MkitUcfgiyl+lC79y5k5NPPrnUpYmEyr597Zw0Y+ybMkyr8+jc06VtfkVG0NLSQmRChFRk7H+gFmQCJBNJ2tramDt3bmleUwqSz2/KCzLns4BzgWcyl98N/A5QcJY32bVrF9RNSq/jfAivLr2yRmtrq4KzyFHo7++nPxZnau3Yg/PUWg8zo7Ozk1mzZhWxOpFwaGpqIjWhRKEZsInpkZDm5mYF5zKVz6oanzSzTwLVwClm9mEz+zBwauY6kTdp2dVKcphuM4DVpj/v2r17dylLEgmdrq4uACbVjH32MvvYAwcOFKUmkTAxM1p2tWATSjPfDAyNhGj1qfJVyD6rC80sN+3sBRYVuR6pcGbG7t2tQwH5MNFqXE2DgrPIUcqG3YnVY39Tzz5WwVnkcAcOHGBwYBBKeThOLbioY6+OPShbhQy1Pe2c+yVwf+byXwP/W/ySpJL19vYSj8XwZo68xZJX06htt0WO0sHgPPZRjexjs91rETkoG16toYQdZweuwdHW1la615SC5B2czezfnXP/F3hX5qrvm9mj/pQllaqjowMAqxn5T/RUdQNt+/aVqiSRUOru7gag8Sg6zhMyj9UWvyKHGwqv9aV93VRdSh3nMlbQYdSZoDxsWHbOvWhm7yxKVVKx2tvbAbDqhhHvY9UNtLe3lKokkVDq7e0Fji4411fZm55LRA7al23wjPx25gtrMPa06VPZclXIjPNo6or4XFKhOjvTOwUeMTjXNNDX20MikShVWSKhMxScq8YenGuiUB1VcBYZTnt7ezol1Zb4hethf+d+PG/sY1jin2IG5xIOAUm5ys5dWtXIv2myt2U/ahaRwvX09FAThaqj/C0+odppVENkGO3t7UTqI+kNUEqpDryUp4N2y1Qxg7PIwTfgqpoR75MNznqzFhm7np4eJtQc/Tt6Y7WnP2JFhtHe3o53FOukj5XV29DrS/kpZnAu9d9kUoa6u7tx1bXgRv7WUsdZ5OgdOHCACVVH/6beGE3Rrc6WyGHa9rVhdQF8mJ45GFHBuTzlHZydc59xzk09wl3+rlw1bTcAACAASURBVAj1SIXr7++H6MjdZgAi6dtjsVgJKhIJp+4DB5hQlTzq55lQ7XHggJajE8llZrS1tZV2KbqsTHDWknTlqZCO82zgFefcg8659zn35v2Uzey14pYmlSgWi2GRIy/WYtGqofuKyNh0de0fWk7uaEyqMa3jLHKInp4e4rF4aTc/yaoHImi/gzKVd3A2s6uAE4A7gL8HNjvnvuqcO86n2qQCxWIxvFGCM5HqofuKyNh0dnYypQjzl5NrPA4c6CaVShWhKpFwyIbWQDrODiKNEa3lXKYKmnE2MwP2ZE5JYCrwsHPuJh9qkwo0ODiIueiR7xSJDt1XRAo3MDBAX3+MKTVH/6Y+ucbwTF1nkVxD3d4Sr+GclapPsat1VzAvLkdUyIzzfzjnVgI3AS8Ap5vZp4G3Ah/2qT6pMIODiSMeGAhgkUjmvgrOImOR3aFzchE6ztmudfY5RQR27tyZ/mJiMK9vE4ydzTtJ9yulnBSyc+A04EIza8q90sw859z5xS1LKlUikcAio3ScMx1pbYAiMjbZo+2nFiE4T1VwFjlMU1MTkYYIqeqARpgmQd+2Prq6upg69UjrMkipFTKqceyhodk5twzAzDYUtSqpWMlUilFXJswcV6pdkUTGJrsVcDGD89D2wiLCjh07SE0Ibu7fJqY7zU1NTaPcU0qtkOB8au4F51yU9JiGyBDzvKFgPLL0t52Cs8jYZLvDU2uLM+Ps0JqxIllmRtPOJmxSgGMSk9Jn27dvD64GGdaowdk5d4Vzrgc4wznXnTn1AG3A475XKBXFM4/RO86Z+yo4i4zJvn37qI5CY9XRv7FXRWBSnVPHWSSjtbU1vRTd5ACLqAdX49iyZUuARchwRg3OZvY1M5sI3GxmkzKniWY23cyuKEGNElJu1M60iAyno6ODqXV5fLiTp6k1KTo7O4vzZCIV7o033gDApgbYcXbgTfHY9Mam4GqQYY16cKBzbqmZbQQecs6dfejtZrbKl8pERGRYHR0dTC7CroFZk6tTdLSr4ywC6eDsIm5oXCIoNsXYtm0byWSSqqpC1nIQP+XzL/F54B+BbwxzmwHnFbUiqWjpLvIof6VrdR2Ro9LR3sbsIhwYmDWl1qNFM84iAGzatCk9pjHKAlG+mwrJN5Ls2LGD448/PuBiJGvU4Gxm/5g5f7f/5Uilc+QRnDO3a1RDZGw6O/dz0tQiBucaj/27D+B5HpFIQftiiYSK53ms37Ce1Ozgd9K0aen3yg0bNig4l5F8RjUuPNLtZvbT4pUjlc5FIkB+b+h6gxYpXCKRoLevn0mzi/fRzaTM7oHd3d1MmTKlaM8rUmmam5vp7+tP71wRtEaI1EZ4/fXXueCCC4KuRjLyGdU40r+WAQrOMiQScWCjBWfNaoiMVXZr7Ek1xes4Z5+rq6tLwVnGtfXr1wNg08vgfcpBamqK115/LehKJEc+oxqfLEUhEg7p7UE1giHil/379wPpLnGxTM48V2dnJ4sXLy7a84pUmtdffx1X4wLbavtQNt1oXt9MT08PEyeWSVHjXD6jGh8zs3ucc58b7nYz+6/ilyWVKp2bRwvOCtYiY3Vw85PidZwnZ55LS9LJePfq6lfxpnll8zZlMwwz47XXXuOd73xn0OUI+e0c2Jg5nzjCaUTOuYXOuV8759Y75153zv1H5vppzrmnnHObM+faiD0kLK9NTbTltshYZXf4m1LE4Kxtt0XSn+Y072zGZpbBmEbWdHARx+rVq4OuRDLyGdW4PXN+zRiePwl83sxWOecmAiudc08Bfw88bWY3OOcuBy4HLhvD80uZSSQT4KqPfKdMRzqRSJSgIpFwyYbbKUUc1aiPQm2V07bbMq6tWbMGoLyCczQ9rrFy1cqgK5GMvJc1cM4d65x70jm3zznX5px73Dl37JEeY2a7sxukmFkPsAGYD/wl8OPM3X4MfGhs5Uu5iccHsOjowdlFqxkYGChNUSIh0tbWxuRaR1URF6VxDqbVGXv37i3ek4pUmFWrVuGqHJTZZ+DeDI8tm7fQ29sbdClCAcEZuA94EJgLzAMeAu7P98HOucXAW4DfA7PNbHfmpj3A7ALqkDIWj8chksdiLdGq9H1FpCDbtm5hfsNg0Z93fv0g27duKfrzilQCM+P5F57Hm+kVloxKwOYYnufx+9//PuhShMK+PRrMbJmZJTOne4C6fB7onJsAPAJcZGbdubdZehmGYT8Xcc79k3NuhXNuhWbvyp/necRi/VhklI4zQLRafz2LFMjzPHbs2MHCCcXfnGHhhBS7WnfrkyAZlzZv3kz7vnZsfhmNaWRNB1fneP7554OuRMgjOGcO5JsG/I9z7nLn3GLn3DHOuUuBn+fx+GrSofnenM1S9jrn5mZunwu0DfdYM/u+mZ1jZufMnDkz3/8mCUh3dzdeKoXV1I9631RVvY7gFylQa2sr8YFBFvkUnD0ztm/fXvTnFil3zz//PDiweWUYnB2k5qT43Yu/07FBZSCfjvNKYAXwUeCfgV8DzwKfBv76SA906T2V7wA2HLJs3RPAJzJffwJ4vKCqpSxll8my6oZR7+tV1dO2TwciiRQie2T9sZOTRX/u7HPq6H0Zb8yMZ599FmYAtUFXMzybb8T6Y6xatSroUsa9UYOzmS0xs2Mz54eejnhwIPCHwN8B5znnVmdOfwHcALzHObcZ+LPMZalw2QOLrKZxlHuC1Tayr61NS9KJFOC5555jRj0saCz+z82MOmPRRI/nn3uu6M8tUs5ee+01duzYgbewjN+PZqfHNR5/XH3GoOWz5fYQ59xpwCnkzDab2d0j3d/MnmfkZcT/tJDXlvK3detWALz60Q9J9uqnMrAnzp49e5g3b57fpYlUvFgsxsoVK/iTOfHR9xgao7NnDPDEa69p620ZVx599FFcjcOOKcMxjawopBaneOGFF9i7dy+zZ2tNhaAUshzdl4BvZ07vBm4CPuhTXVKBtm7diqubBFU1o97Xa5g29BgRGd2LL77IYCLBW2f6N+P41pkJPDOeU9dZxonOzk5+/etfkzomVWArsfTsWMMwdZ0DVsiqGh8h3SXeY2afBM4EJvtSlVSk19dvIJFHtxkyXWnn2Lhxo89ViVQ+M+MnD9zPrAZj6ZTizzdnLZ6YYtFEjwd/8oDGqGRceOyxx0ilUthxZdxtzmoEm2s8/sTj9Pf3B13NuFVIcI6ZmQcknXOTSK+EsdCfsqTStLS0sHfPblKT5uf3gGg13oTZvPSS1qUUGc2qVavYsHETH1jUT9THNWadgwuO6adpZ7OWvpLQa29v5/4H7k8vQTcx6Gry4y316Onu4f77895GQ4qskF/BK5xzU4AfkF5pYxXwoi9VScXJLsyemrIg78ckJy9g8+Y3tCydyCjuvfceptTC/5lb/I1PDvX2WQlmNRj33rOM9DL7IuH0wx/+kMHBQbwzKujTlengLfS47/77tNNnQPIKzpll5b5mZl1m9j3gPcAnMiMbIvz2t89B/WSsblLej8mG7BdeeMGvskQq3ssvv8yKFSt538J+aqL+v140Aucv6mfDxk3pJbpEQmjz5s38z//8D6njUzAh6GoKY6cbyVSSH/zgB0GXMi7lFZwzu/v9POfyDjNb61tVUlF27drFq6+uYnD68QU9zmuYDg1TefyJJ3yqTKSyxeNxvvH1m5nbaPz5wtLt6PfH8wZZPMnjW7d8k56enpK9rkgppFIpbvnWLVADdnIFfqrSCKkTUvzqV79i3bp1QVcz7hQyqrHKOfc23yqRirV8+XJwjuTMEwt7oHMMzDyJNzZt4o033vCnOJEKdtddd7F7z17+YWlvSbrNWdEIfGppL11dXdx+++2le2GREnjggQdYt3YdqTNSMPoiUGXJlhpuguOaa6+hr68v6HLGlUKC8zuAF51zW51za51z65xz6jqPc7FYjCefXE5y8sK8Nj45VHLGCbhIFY888ogP1YlUrk2bNvGTn/yEP5k3wMlT/VtJYyRLJqV438I4TzzxhHYTlNDYtGkTP/zhD7EFVt7rNo+mGpJvS9LW1sYtt9wSdDXjSiHB+b3AccB5wAXA+ZlzGcceffRRursPkJh3xtieoKqWwZkn8stf/pKWlpbiFidSoXp7e/nyl77IlBqPvzkhFlgdFx4XY3ajce01X6arqyuwOkSKIR6Pc8011+DVenhv9Ubenq1SzADvZI9f/vKXPP3000FXM24UEpxthJOMU/39/dxz732kJi/AmzhnzM+TmHcWRoQf/ehHxStOpEKZGTfeeAO7d+/m30/rprE6uF+zdVH47Gk9HNjfyXXXXksqlQqsFpGjYWZ861vfoqWlheQ5yYod0TiUnWwwHW7++s3s2rUr6HLGhUKC88+A5Znzp4FtwP/4UZRUhgceeIDenm4GF7z1qJ7HahoYnHUyTz31FNu2bStSdSKV6ZFHHuE3v/ktf318PydOCT6oHjMxxcdP6uOVFSu45557gi5HZEwee+wxfvazn+Et9SBMu1VHIPWOFLFkjMuvuFwbo5RA3sHZzE43szMy5ycAb0frOI9bu3fv5t577yM5bQnehJlH/XyD886Eqhq+ecstWjtWxq01a9bw3e98h7NnJPiLRaVbRWM0fzJvkD+cM8Cdd97BSy+9FHQ5IgVZvXo137r1W9hcw04L4ftLIyTfkaSpqYnrv3q9dv302Zj3oDKzVaQPGJRx6Nvf/jZJzxhcVKRvgeo64vPfyprVq3nmmWeK85wiFaS1tZUvXHkFM+uS/POpfbgymr90Dj55cj+LJnh8+UtfZPv27UGXJJKXvXv3ctXVV2GNhveOEMw1j2Q2eGd4PPfb51i2bFnQ1YRa3sHZOfe5nNN/OufuA1p9rE3K1EsvvcTzzz/PwLyzsNrirRyfnLUUa5zBt2+7TcvryLjS19fH5ZdeQmqgj8+fGexc80jqovC5M7up9uJcftmlOlhQyl4sFuOKK6+gJ9ZD8twkVAddkb/sBMNb5HHHHXfw3HPPBV1OaBXScZ6Yc6olPev8l34UJeWrv7+fm7/+DWiYQmLOacV9chchvvhcOjs6+f73v1/c5xYpU6lUimuu+TLNzc189rRu5jSU78es0+uMi8/opr1tL1df9QUSiUTQJYkMy/M8rr/+erZs2ULy7cl0cgk7B3aOwTS49rpr2bJlS9AVhVIhwXm9mV2TOV1vZvei5ejGnTvvvJN9bXuJLf4jiBR/RwZvwiwSs0/m0cce47XXXiv684uUm9tuu42XXvo9Hz+pj1OnlX695kIdPznFP57Sy5q167j55pt1TIKUpTvvvJPf/va3eGd4MDfoakooCqlzUwxGBrn0skvp6OgIuqLQKSQ4X5HndRJSGzdu5KGHHiIxa+lRLT83msGF50BNIzfeeJM6WhJqDz/8MI888gjvXxTnTxcMBl1O3s6dk+DDx8b4xS9+wd133x10OSJv8tRTT3H33XfjLfGwE8bhH3b1kDw3Scf+Dq688koGBsrnQOMwGDU4O+fe75z7NjDfOXdrzulHQPm3R6QokskkX7vhBqhpYHChzzuvR2uIH3MuTU07uPfee/19LZGAvPDCC9z27W/z1pmDgW5yMlYfWhLnj+YMcMcdd/DUU08FXY4IABs2bOBrN3wNZoKdbYEcDOhWO+gCuiDybCR9udSmpncW3LBhgz4ZKrJ8Os6twAogDqzMOT1BejdBGQceeOABtm/bRmzRO6Gq1vfXS01dRHL6sfz4xz9mx44dvr+eSCm98cYbXPPlL7N4UopPn9ZHpAKP9HcOPnVKP0unJrnha19l7dq1QZck49z+/fu58gtXkqpJkXpn6ijWDTs6rsvhEpnTPofrCugHfAF4p3r86le/4pFHHgmmhhAa9dvKzNaY2Y+B04F7zOzHmcuPA+r/jwPNzc3cedddJKcuJjVtccled+CYd+JFqrnhxhu1LqWExr59+7js0ktojAxw8Rnd1BX/UIGSqY7ARWf0MqM2yZVXXK6dyyQwyWSSq794NZ37O0m+M5lewkCwkw2bZ9x2222sXr066HJCoZC/x34F1Odcrgf+t7jlSLnxPI8bbriRFBEGF59b2hevrie+8B2sf/11Hn300dK+togPYrEYl192KX3dXXz+jG6m1lb+x6cTqo3Pn9mNN9DHZZdeQk9PT9AlyTj03//936xds5bUW1MwNehqyogD7+0e1mhc9cWraGtrC7qiildIcK4zs97shczXDcUvScrJk08+ybp1a4kveDtWU/p/7uSM40lNns/3br+dvXv3lvz1RYrF8zyuu+46tm7dyr+f1s2iicFvp10scxo8Ljq9m127Wvji1VeTTOrwFymdp59+moceegjveA87pvL/GC266vTBgj19PVx19VX6+TxKhQTnPufc2dkLzrm3ApV3RIvkra2tje9897ukJs8jOfPEYIpwjoElf8RgIsVNOsBBKtjtt9/O888/z9+e0M9ZM8L3xrV0apJPLe1j5apVfPOb39TPqpREa2srN918E0wHO1PfcyOaBMlzkmzcsJE77rgj6GoqWiHB+SLgIefcc86554GfAP/uT1kSNM/z+OpXv8rAYJKBxX9EkPv/Wu1E4gvO4ZWXX2b58uWB1SEyVj//+c+5//77+dMFcf58YXgPDXnXvEEuWBzjySef5KGHHgq6HAm5ZDLJl6/5MvFknNQ7gjsYsGIsAO9Yj3vvu5cVK1YEXU3FyvvbzMxeAZYCnwb+BTjZzFb6VZgE6+GHH2bVqlXEF70Dq5sUdDkkZ59CavI8br3127S0tARdjkjeVq9ezddvvpnTpiX5+ImxIP8GLYm/Oi7O22YO8t3vfIcXX3wx6HIkxO644w42bthI8uwkNAZdTWWwMw030XHtddeyf//+oMupSIX+fXYScApwNvA3zrmPF78kCdq2bdv43u23p5eEm3lS0OWkOcfAkj9m0DOuvfY6zWhJRWhpaeELV17BrPoknzm9l+g46IhFHPzzaX0smpjimi9/ia1btwZdkoTQqlWruO+++/CWeLAw6GoqSBUk/yDJge4DfO1rX9NI1Rjk/WvcOfcl4NuZ07uBm4AP+lSXBKSnp4crv/AFUq6a+JJgRzQOZbWNxI45l40bN3DbbbcFXY7IEXV3d3PZJZdgg/187oxuGqvHzxtUXRQ+d0Y3tRbnsksvob29PeiSJET6+vq4/qvXwwSws8bPz1XRTIbU6Sleeuklfv7znwddTcUppP/xEeBPgT1m9kngTGCyL1VJIJLJJF/80pdo3b2b/uPPg+ryWzQlNf04EnNO46c//SlPPvlk0OWIDGtwcJAvXHklu3fv4uIzDjCnYfytQz6tzvjcGd0c2N/BFZdfRiymY8mlOL7zne+wb98+km9LQlXQ1VQmO95gFnzr1m+xZ8+eoMupKIUE55iZeUDSOTcJaEMfkITK9773PVauWMHAMX+IN3FO0OWMaHDR20lNXsB//dd/sWbNmqDLEXkTM+Omm25izdq1/NMpvZw0JTzLzhVqyaQU/3ZqD5s3b+a6a68llRq//y+kOF566SWWL1+Od6IH04OupoI5SJ2TYiA5wA033KBNxgpQSHBe4ZybAvyA9JbbqwAd+RES99xzDw8++CCJ2aeQnFUmc80jcRHix7+bVM1ELrv8cjZt2hR0RSJDfvSjH/GrX/2KDx8b49w5iaDLCdzZMxP87Qn9PP/CC3z3u98NuhypYL29vdxw4w24yQ47VSMaR60RUmekWLVqlT7BLcCowdk594eZLy82sy4z+x7wHuATmZENqXD33Xcf3//+90lOP47BY/4g6HLyU1VL/0nvpT8V4aKLLuaNN94IuiIRli9fzl133cX/mTvAh5bEgy6nbLx30QB/vjDOQw89xIMPPhh0OVKhbr/9djo7O0mek4QK3qq+nNgSg9nw3f/+Lvv27Qu6nIqQT8f51sz5UHfZzHaY2Vp/SpJSeuCBB/je975HcvqxDBz3x+Aq57B/q51I/0l/QV/K8R8XXcyWLVuCLknGsZdeeomvf/3rnD49yadO7i+n42rLwsdOjPG2mYPcdtttPPPMM0GXIxVm3bp1PP7443jHezAt6GpCxEHq7BTxgTi3fOuWoKupCPmkpIRz7vvAAufcrYee/C5Q/GFm3HnnnXz3u98lOW0JA8f9SUWF5iyrm0j/Se+nP2F85jOfZd26dUGXJOPQxo0b+eLVV7NoQpLPnt5DVeX9KPku4uDTp/Vx4pQU13/lOlavXh10SVIhBgcHufGmG3GNDjtNIxpFNwFSp6R47rfP8dxzzwVdTdnL59f7+cAzpLfXXjnMSSpMMpnk5ptv5kc/+hGJGScwcNy7KzI0Z1ndJPqWfoA+L8pFF12sH3wpqT179nDZpZcwITLA58/spl5H+Y+oJgqfO7OHmXVJrrzicnbu3Bl0SVIBHnzwQXY27ST5Fq2i4Rc70XBTHN/4r28Qj2vM7EhGTUtm1m5mDwAfNLMfH3oqQY1SRPF4nKuuvprly5czOO8sBo99F0QqNzRnWd1E+k4+n8HaKVx11VU88cQTQZck40Bvby+XXXoJA33dXHLWAabWqhs2mgnVxiVndhNJ9nPpJf9JV1dX0CVJGTtw4ADL7lmGzTWYG3Q1IRaB5FuSdHZ08vDDDwddTVkrJDFd5Jy789CTb5VJ0XV3d3PRxRfzuxd+x8Ax55JYeE5ZbXBy1Krr6V/6fpKTF/D1r3+du+66S7siiW+SySRf/tKX2NnUxGdP72Z+o5ZzytfMeo/PndHNvra9XHnlFQwMDARdkpSpZcuWEYvF8E7Xz5fvZoDNM5bds4wDBw4EXU3ZKiQ4Lwd+ljk9DUwCev0oSopv7969/Ou//hsbNmwkfvx5JOecEnRJ/ohWEz/hPSRmnMBdd93FN77xDa0dK7649dZbefmVV/jk0j5Om6Yt4At1/OQU/3JKD6+99jo33XSj/siVw+zZs4dHfvoI3jGetlsrEe80j1h/jHvuuSfoUspW3sHZzB7JOd0LfBQ4x7/SpFh27NjBv3z60zS37qb/xPeSmr4k6JL8FYkweOy7GJx7Jk888QRf/NKX1NGSonr22Wd57LHH+Itj4vzJ/MGgy6lY75id4CPHxnjqqf/V1r9ymGXLluGZpzWbS2kyeIs9HvnpI+zfvz/oasrS0Qy3ngDMKlYh4o+tW7fyr//6b3R299O/9AN4k+cFXVJpOEdi0dsYWPQHPPfb33LZ5ZcrPEtRtLW1cfNNN3LsZI+PHqdtpI/WB5fEOWVakm/dcgvNzc1BlyNlIpFI8PQzT5NakIKGoKsZX+xEI5lI8pvf/CboUspS3sHZOdfjnOvOngNPApf5V5ocrV27dnHx5z5HX8Kj7+Tz8RrH3/6kybmnMXDsu1i1ciXXXHMNyaQ+Upex8zyP67/yFQbj/fzrqVp2rhgiDv75lF6iNsh11+pnVNJWrlxJf18/tkDd5pKbBG6S45lfa7314RQyqjHRzCblnJ9oZo/4WZyMXXt7OxdddDEHemP0nfg+rG5S0CUFJjnzRAaO+QOef/55brrpJjxPB5nI2Dz66KO8uno1Hz+hlzkN+j4qlul1xqdO6mHjpje47777gi5HysCzzz6Lq3EwO+hKxiEHqfkp1qxeQ2dnZ9DVlJ2C+iXOufnOuXOdc+/KnvwqTMZuYGCAz3/+P2nr6KD/xD/HGqYGXVLgknNOY3D+W/jFL37BnXdqMRgp3ODgIPcsu5ulU5O8a175zzUv21RPU0+Upp4oX1kxgWWb6oMu6YjePjvB2TMS/OSB++nv7w+6HAmQmfGb3/6G1JyUttYOiC00zIzf/e53QZdSdgoZ1bgReAG4Crgkc/pPn+qSo/CDH/yA7du3ETvuPLwJGkPPSsw/m8SME1i2bJl2GJSC/fKXv6Sjcz9/uThWEas4NvVEiaUixFIRNnZV09RT/gnkg0ti9PT28fjjjwddigRocHCQvt6+9NpdEoyJ6bOOjo5g6yhDhXScPwScZGZ/YWYXZE4f9KswGZtVq1bx4IMPkph9CqkpC4Mup7w4x+Ax78RqJ3DdV65XV0vylkwmufeeZRw7ydPScz46fnKKU6cl+ckD9+tg3nEsFsscdKtdAoMTARd1ep8cRiHBeRtQ7VchcvTi8TjXf/WrUD+ZwYVvC7qc8lRVQ2zJu9izeze333570NVIhXj11Vdp3b2H84/pr4hucyU7/5gYnfu79BHxODa05bOCc6BctTv4R4wMKSQ49wOrnXO3O+duzZ78KkwK9+ijj7KvrY3Y4j+CqP7GGYk3aS6JWUt5/PHHaW1tDbocqQCrV68m4uCM6YmgSwm9k6cmqa1yrFmzJuhSJCBDwbn8p4vCLZLzbyFDCgnOTwDXAb8DVuacpAzEYjHuvfc+UpPn402aG3Q5ZS8x/ywMx9133x10KVIB1qxZzZJJKerUAfNdVQROmDTImtWvBl2KBGT27NlEohHoCrqScWwAvH6PhQs18nmoQpaj+/FwJz+Lk/w99thjdHcfYHD+2UGXUhGsppHBmUv5xS9+wa5du4IuR8rYwMAAG9av56Qp6jaXytIpSbZt30F3d3fQpUgA6uvrWXrSUiLtWig9MO3ps7POOivYOspQIatqnOCce9g5t945ty178rM4yU8qleKhhx4mNWke3kQtepmvxLwz8Cw94iIyko6ODhLJFAsaU0GXMm4smJDCzNi9e3fQpUhA3vKWt8B+QMfiBsLtc1TXVLN06dKgSyk7hfw5dxfw36S/jd8N3A3c40dRUpgXX3yR9vZ9JGafEnQpFcVqGklOPYaf/eznmuOSEWUPjqmLagezUqnN/L/Wyhrj11lnnQUe0BZ0JeOQQXRvlNNOPY3qah0vdahCgnO9mT0NODNrMrMvAx/wpywpxKOPPoqrbSQ1dVHQpVScxOyT6evr5ZlntLWoDC8b3moVnEsm+/9af9COX295y1uYPmM60U1R0I9eae0G6zbe9773cfKnVAAAIABJREFUBV1JWSokOA845yLAZufcvzvn/i8wwae6JE/d3d2sWLGCgekngNM8WKG8iXOhfrKCs4woG5yr9ONVMtWZ/9fqOI9fNTU1fOxvP5aetd0XdDXjiEF0Q5TZc2bznve8J+hqylIhbwX/ATQAnwXeCnwM+LgfRUn+VqxYgZlps5Oxco7EpAW8unq13qRlWLNnp48baIspOZfK3sz/6+z/exmfzj//fKZMnUJ0g9alK5m9QCd84uOfoKpKywgNp5B3AgOWkV6W7hzgROAHfhQl+Xv55ZdxVbV4E2YGXUrFSk1ZQGJwkLVr1wZdipShOXPm0FBfz84K2LI6LHb2RIlGIyxevDjoUiRAtbW1/O3/+9v0nPOeoKsZBzyIvh5lxqwZvPe97w26mrJVSHC+l/QBgh8Gzs+cLvCjKMnf1q1bSTZM15jGUUhl/ujYunVrwJVIOYpEIhx3/HE09ar7UipNPVUcs2gRNTU1QZciAfvQhz7EgoULqFpZBYNBVxNubqODTvjMv31GBwUeQSFpa5+ZPWFm2zMHBzaZWZNvlUleGhoawLRM1tFwXvr/X319fcCVSLk65ZRT2dZdTUdc+237rWfQselADSefcmrQpUgZqK2t5YtXfxE34HCv6ufPN50QWR/hz97zZ7z73e8OupqyVkhw/pJz7ofOub9xzl2YPflWmeRlwoQJRL3y3JihpulFIv0dRPo7qFu/nJqmF4MuaXipdBtjwgQd6yrDu/DCC8FFeGJHXdClhN7Pd9YykDI++tGPBl2KlImlS5fy95/4eyI7I7hmheeiS0HVK1VMnz6diy+6OOhqyl4hwfmTwFnA+0iPaFxAelxDAjRt2jRcvAcSsaBLOUykrwOXSuBSCaI9e4j0dQRd0rCivelDtqdNmxZwJVKu5s6dy1984AM821pHuw4S9E33oONXLQ2cd96fsmTJkqDLkTLysY99jJOWnkR0VRR6gq5mFIn0J5gf+chH0p9klmdvK83ArXJYt3HVF65i4sSJQVdU9gp5B3ibmZ1jZp8ws09mTv/gW2WSl7/6q7/C4VHTvCLoUipTKkHdrpWctHQpZ555ZtDVSBn7u7/7OyKRKA9trcO0rqwvHt32/7d353E2l/0fx1/X2WZfMIOx78MwjChb0cKtUqmIRJaUVJZ0J5Hs2lCUklI/hRZ1tyh3+2IthFHWbGMZ+zD72c/1+2Ombt13ajAz37N8no9HD7OfN833nPe5zrWE4/bCwIEDjY4i/IzFYmHypMlEh0djWWMBf94EyQ3dunVjxIgRdOvWza+Ls9qlMGWY6N+/P61atTI6TkA4n+K8ViklR9P5mVq1atGzRw+sJ3/FVHDK6DgBx3okHe0s4MGRIzGZZCRRnFuVKlW4vc8drDkWxucHw4yOE3RWZNr46nA43W++mdq1axsdR/ihatWq8dSTT2F2mDGvNYO/Lu+xwvLly3n++edZvnw5+Ok6O3VIYfrFxDXXXMPgwYONjhMwzqcptAXSlVK7lFI/K6V+UUrJ/l1+YMCAAcTExhCx9zuUM9/oOAHDnLUf29Ff6NKlC02bykIk8fcGDx5Mx45X8NbuSH464aePhgFoa5aF13dGcWnr1gwfPtzoOMKPpaam8ti4x+AUqJ+Uf54qaAW73c7777+P3W73z+KcBeYNZpqlNuPRRx9FKZk7XlLnU5yvBRoC/+A/85tlOzo/EBMTw1NPPkmE8hC5cznKkWN0JL9nOfkr4Xu/JSWlCQ8++KDRcUSAMJlMjB//OI0bJ/PSthj25sjezhfrcL6JOVtjqVW7NpOnTJFDF8TfuuaaaxgyZEjRYsFf/LQ8+7NcsKy1ULVKVZ6Y/gRhYfIK2vkocXE+ews62Y7O/6SmpvL8nNnE2BRRO5ajCk8bHclvWY5tI2zfSi655BKee/ZZWQwhzkt4eDhPPvU0FRMSmbklln25Up4v1OF8E0+nxxERHcczM2bKzjaixPr27Uv37t0x7TIV7T8sSqYALKssxIbHMmvmLOLj441OFHBkUmcQSU5O5sW5c4mPjiBq53LM2YeMjuRftA/bgXWEHfiBDh068PRTT8nezeKCVKxYkWefm01Uhco8sSmOrVkySnq+fs02M3VTPKbICjz73Gw5XlucF6UUo0aN4h//+AemrSbUbinPf8sOlpUWosxRzJk9hxo1ahidKCBJcQ4yderU4eV5L1G3Vk3Cd32JNXMzsgUA4LYTsfMzrMd+4ZZbbmHq1Kny8pS4KDVq1OCleS+TVKMWM7fEsO64P05k9E9bTll4anMc8QlVeWney9SrV8/oSCIAmUwmHn30Ua644gpM6SbUfinP5+QsKs02j41ZM2fJNXcRpDgHoaSkJF6e9xJdunTGdngj4bu/Ak/onlVqyj9B9LaPCbNnMW7cOEaNGiXzKEWpSEhIYO6LL9E4pSlzf4nmm8NyRPTfWXPUxrNbYqhTr37RE4+kJKMjiQBmsViYOHEirVu3xrTRBJlGJ/JDHrCstmCxW5jxzAyaNGlidKKAJsU5SIWHhzN+/HhGjhyJNTeTqO3LQnLRoPnUbiJ2LCcxPop5817i2muvNTqSCDIxMTE8++xztG3Xlv/bGcVH+2Sf53P54mAY87ZFkdqiBXOef4EKFSoYHUkEAZvNxvTp02ncuDGWdRaQnVn/wwfmH8yobMXUqVNJS0szOlHAk+IcxJRS9OjRgzmzZxNt8RG1/RNMuceMjlU+tMZ6eCPhe1fQonkqry1YQKNGjYxOJYJUeHg406c/QZcuXXh/XwRLfo3AJ+X5d1rD+3vDWfRrJJdffjkzZswkKirK6FgiiERERPDM089QLakalrUWyDU6kR/QxVv2HYOHH36Y9u3bG50oKEhxDgEtWrTg1VfmU61KAhG7/o3l5G6jI5Utn4ewvd9jy9zMddddx7OzZhEXF2d0KhHkLBYLjz32GD179uTzQ+G8si0Sj8/oVMbzaXhjVwQf7Y/g+uuvZ8qUKbK+QJSJ+Ph4Zs2cRWxELJbVFrAbnchYaqvCdMDEXXfdxQ033GB0nKAhxTlEVK9enVfmzyeteQvC9q3AeiRIz67xuonY9QWWrL0MGTKERx99FKtVFm2J8mEymRg+fDiDBw9m9bEwXvglGm8Il2efhle2R/L14XD69OnDmDFjZH2BKFPVqlVj1sxZ2Ly2opHnEL3+1AGFaaeJG2+8kQEDBhgdJ6hIcQ4hMTExzJo1k6uvvhrbofVYTuw0OlLp8nkJ3/MN5rxjTJgwgX79+slpSKLcKaUYMGAAw4cPZ+NJK2/+GhGyc57f2xvO6qNhDBo0iPvuu0+uR1EuGjVqxOPjH0ef1qjtIfg7Vwjm9KJTAUeNGiXXXSkr0+KslHpdKXVCKbX1rI9VVEp9pZTaXfynrA4pR1arlfHjx3NZmzaE7V+NOWuf0ZFKh/YRtncF5uzDjB49ms6dOxudSIS42267jT59+vDN4XCWHwi9qQnfHrbxSUYEN954IwMHDjQ6jggxHTt25Nprr8W00wRZRqcpR7roKG2bycb4x8bLKzxloKxHnBdSdFT32R4FvtFaNwS+KX5flCOLxcK0qVNp2qwZ4fu+x5R71OhIF812cD2W0/sYOnSozOUSfuPee+/lqquu4p09kfwYQvs8p5+ysHBXFG3btpERL2GYkSNHkpiYiGWDBTxGpykfareCE/DgyAepVq2a0XGCUpkWZ631SuC/z37uDrxR/PYbwM1lmUH8ufDwcJ55+mmqJ1Ujct8K8DiNjnTBzNmHsB7bSo8ePbjjjjuMjiPE70wmE+PGjaN5airzt0dz0h78s+NynIoXt8XQoEEDJk2aLCNewjBRUVFFUzbyQ2TKhh3MW82079Ce66+/3ug0QcuIe/EqWuvfhjiPAec8Z1UpNUQp9ZNS6qeTJ0+WT7oQEhMTw+OPj0e5CwnL+MHoOBfG7SBi/ypq16nD0KFDjU4jxP8ICwvj8QkTMJmtvL07+I94f29vBC6fiQkTJxEZGWl0HBHi0tLS6NixI+YMM3iNTlO21H6F9mqGDxsur/KUIUOHP7TWGjjnshmt9Sta69Za69aJiYnlmCx0pKSk0L9/fyxZezCf3m90nPMWlrEGk9fJxAkTZIsr4beqVKlC3353sv6EjZ1ngncENiPXzIojYfTo0ZNatWoZHUcIAG668Sa0U6Myg7hMajBnmGnVqhXVq1c3Ok1QM6I4H1dKJQEU/3nCgAziLP3796duvXqEH94AvsDZu8eUdwLL6f0MGDCABg0aGB1HiL90++23UzkxgUW/RgXl4Shaw+LdUcTGxsj2V8KvtG7dmqpJVTHtD+KpUsdAF2i6d+9udJKgZ8Rv0TLgt3vVAcDHBmQQZ7FYLAy55x6w52I5FTiHo9gyNxIbG0evXr2MjiLE3woPD2fQXYM5kGciI89sdJxSd9xuYucZM/3u7E9MTIzRcYT4nclk4qYbbyoapis0Ok3ZUAcUsfGxXH755UZHCXplvR3d28APQLJS6rBSajDwFNBFKbUb6Fz8vjBY+/btadQombCj6eDz/4lgprxjmHMy6devr8yjFAGjXbt2AGw9HXzTNX77O3Xo0MHgJEL8r+Tk5KI3CozNUVZMBSYaNWgki3HLQVnvqtFHa52ktbZqrWtorV/TWmdpra/RWjfUWnfWWv/3rhvCAEWHNvQHRx7m3CNGx/lblhM7iYqK5uabZVMWETgqVqxI/Xp12XraZnSUUrc1y0qVyokyv1L4pQoVio+MCNwNpP6SyWX6z99RlKkgnvAjzlebNm0ICw/HfDrD6Ch/Tfuw5Rzi8ss7EB4ebnQaIc5L60sv49dsC+7AWU7wt7SGHdlhtL70MlnNL/xSfHw8AMoZnL+f2qGlOJcTKc7idzabjfbt2mHLOYQ/nxFsyj2GdjtlLpcISPHx8Xh8lPkCQbtHERERQc+ePYmIiMDuKbvCoBQ4PPLALfzX7/Pug3HE2Qvao2VtQTmR4iz+4NJLL0W7ClHOPKOjnJM5v2gjlksvvdTgJEKcP6ez6JHbWsb3voUeRbdu3RgxYgTdunWjsAyLM4DNonA4HGV6G0JcqMzMzKI3oozNUSbMYIo0cfjwYaOThASZRS7+IDo6uugNnx+fT+pzYzKZZFGgCEhOpxObWWEq41eMIy2a5cuXA7B8+XIqW8p2iDvM/J8nBUL4m127dgGgK/jvq6kXwxvnZfuO7UbHCAky4iz+4Lc5w8qPi7PyebHJYSciQB0/fpxIa9k/eEdYNHa7nffffx+73U5EGRfnKIuP48ePl+ltCHGhdu7cibIqCNLZDLqi5vChwxQWBul+e35EirP4g99HjPz8SbnX68XtdhsdQ4jzYrfbWbtmDS0rBd/IbFolJxs3/kR2drbRUYT4A6016VvS0fEagnNtILqCRmvN1q1bjY4S9KQ4iz9Yv349ymLDF1XJ6Cjn5I1Jwu1y8fPPPxsdRYjzsmbNGhxOJ+2ruoyOUuraV3Xh9fpYsWKF0VGE+IOtW7eyf99+fDWDaCub/1YZVLjiww8/NDpJ0JPiLH6ntWb16jW4Y6uDyX9PNvPGVUOZzKxZs8boKEKcly+++IKKEZAc779ToS5UrWgv1aM1X3z+GdqPd+URoee9995D2RS6dhD/XprBW9fL2rVrOXLE/89iCGRSnMXv1q5dy5kzp/HG1zI6yl8zW/HEVuOrr78hL89/d/8Q4mzr169n3bp1XJVkL/OFgUZQCq6uZmfrtu18//33RscRAihaU7By5Uq8db1Bvx2Crq/RSvPBBx8YHSWoSXEWABQUFDBj5kyIrIinUj2j4/wtV/VLyMnJZt68eUZHEeJvFRYWMuOZp0mK0nSrHbxbtnWu4aRurI/Zz84iNzfX6DhC8M477+DTPnT9IB5t/k0E+Kr7WPbJMllrUIakOAsA5s2bx+nTp7HXvdyvp2n8xhediKtqMz799FM2btxodBwh/tJrr73G8RMnubtJHjb/v7wumNkEdzfJJyc3lxdffNHoOCLEHThwgA8/+hBfHV9w7t/8J3SKxuFwsGDBAqOjBC0pzoKvv/6aZcuW4a7SFF90ZaPjlJi7eiuIiGPa9OmyDZbwWxs2bOD999+jcw0HyfFeo+OUudoxXrrVtvPZZ5/JlA1hqBfmvoA2a3SzEBht/k0s+Br4+OSTT9izZ4/RaYKSFOcQt2rVKqZNm4YvNglXzdZGxzk/Zgv2+ldxOjuPESNHcurUKaMTCfEH+/bt4/Hx46kZ7aN3A7vRccrNLXUdNIz3Mn3aVLZvl0MZRPn74YcfWL9uPd4mXgg3Ok350ikabDDn+TmyULcMSHEOYevXr2fCxIl4IxOwN+oCpsBbOeGLSqCw0T84dvwkDz44SuZ1Cb+RlZXFI6MfJkzb+WeLXCIC7/K6YDYzjGqeR5zFzdgxj3D06FGjI4kQ4nQ6mfP8HFSMQjcIweJoA29TL1vSt/Dtt98anSboSHEOUStXrmTsuHF4wuIoTO4KZpvRkS6YL6YKhY26cOhwJiNGjpRpG8JwdrudR8c8Qs6ZLB5qnkul8NB78I61aR5ukYOrMI8xox+WHXBEuVmyZAlHMo/gaekJ2Zaj62lURcWc5+fItVfKQvRXKnRprVm0aBHjx4/HZYunMPlasAT+8dW+2GoUNurCgUNHGHz3PXJ6kjCM0+lk3Nix7N69mwea5lE3NvjnNZ9LtSgfI1NzOXT4EKMffliOAxZl7sCBAyxatAhfLR9UMTqNgRR4LvGQnZ3N/PnzjU4TVKQ4hxCn08nUqdN49dVX8VSqT2GT68EaYXSsUuOLq05Byo3kunwMHzGCL774wuhIIsQ4nU4eGzeOTZs2MSSlgEsS5Vj4lIoehjfLZ9fOHYwe/TB2e+jM9RblS2vNjBkz8Fl86Bah9yrP/6hQtFBw2bJlMphUiqQ4h4jMzEweGDaMr7/+CleN1jjrXxmQc5r/jo6Ip6DJTbgjE5k+fTpz587F5Qq+442F/3G73UyY8DjrN2zg7iYFXJ4kv3e/aV3Zzf3N8tm2dSuPjhmDwxG8e1kL43z++ef8/PPPeJuF3oLAc9FNNSpS8cyMZ/B4gu/EUiNIcQ5yWmu++OILBg4axO69GTgadsZdPa3omK9gZQ3Hnnwd7iopLF26lCH3DiUjI8PoVCKIud1uJk6cyA8//MigxgV0qi6l+b+1qeJmaNN80tPTGTv2USnPolTl5eXx4ksvQgLoujLa/DsreNI8ZOzP4MMPPzQ6TVCQ4hzE8vPzmTp1KtOnT8dui6eg2S14K9YxOlb5MJlw1WmPo1EX9h/MZPDdd7Ns2TLZmkeUOqfTyfjHHmP16tX0Ty7kmhpSms+lfVU396QUsGnjJh55ZLTMeRalZsGCBeTm5uJt6YUgHhe6INWAqrDgtQWybWspkOIcpDZu3MjAQYP4+ptvcNVohb3x9eiwaKNjlTtvhdoUNLsFR0QiM2fOZNy4cXLHIUqN3W7n0UfH8OOPP3JX4wL+UdNpdCS/17Gai/ua5fPzli3886GHyM/PNzqSCHC7du3io48+wlffB/FGp/FDCrxpXhwOB/PmzTM6TcCT4hxkcnNzefLJJxk1ahQncu3Ym9yAu3pLUKH7v1rbInEkX4uzVhvW/rCOvv3uZNmyZfh8PqOjiQBWUFDA6IcfZnPxQsCrZaS5xNpXdTM8NZ+dO7fz4IMjycnJMTqSCFBaa2bPmQ1hRfN5xTnEgDfZy1dffcUvv/xidJqAFrptKshorfnmm2+4o28/Pvv8c1xJLShodiu+mFDej+csSuFJSqUg9RbyLbHMnDmT4cNHcPDgQaOTiQCUnZ3NqAdHsnXrL9zfNJ8rqklpPl+XVnbzYPM89u/dzYjhw+SVIHFBVq5cybat2/A29ULgHkdQLnRjjYpQvPjiizJt8SJIcQ4CmZmZjBkzhsmTJ5PjtWJvejPuWpcG5a4ZF0uHx+FofD3Oeh3ZunMXAwYOZOHChTid8hK7KJnjx48z7IH72btnNyOb59O2qmw5d6FaJnh4uEUeRw8f5P77hnL48GGjI4kA4vF4mPfyPFSsQteRIvi3LOBN8bJ9+3ZWrFhhdJqAJcU5gBUWFvLyyy/T7847WbdhI85abShMuRFfVCWjo/k3pfAkNiI/tQfO2Fq8/vrr3NG3L9999508Cxd/6cCBAzxw/1BOHstkTFourWSf5ovWtKKHcZfkUHDmBA/cN5Rff/3V6EgiQCxbtqzohMDU0D0h8HzpuhoVp3hp3ku43XL/dSHkVy0A+Xw+PvvsM27v04e33noLR3xdCpr3xJOUGtJzmc+bNRJnw6uxN+nGiQIvEydOZNiw4fLALf7U9u3beeD++3DmZjH+khwaV5A9UUtLvVgvj7fKweTKZcTw4WzevNnoSMLP2e12Xl/4OiQCSUanCSAKPKkejh09xqeffmp0moAkLSvAbNmyhSFD7uXJJ5/kjMeKvelNuOp3QtuijI4WsHyxSRQ27Y6z7uVs3bmbe+65h6efflrmXIrfrV69mgdHjiDMm8/jrXKoHRO6x2iXlWpRPia0yibeXMjD/3yIr7/+2uhIwo8tW7aM3OzcosNOZPu581MVSIA3F70pB4RdACnOAWLPnj2MHv0Iw4cPZ/eBwzjqd6KwyY34oisbHS04KBOeyo3Jb94TV5WmLP/3Z/S+/Xbmz59PXl6e0emEgT744APGP/YY1cIdTGqVTdVI2Y2lrFQK10xslUP9GBdTpkxhyZIlMn1K/A+n08niJYuhCpBgdJoApMDbxEvWqSw+//xzo9MEHCnOfi4zM5MpU6Zw1113sX7TZlw1LyU/9Ta8CQ2D+/Q/o1jCcNVuS2HznthjarJkyRJu69WLJUuWyElnIcbn8/HSSy8xe/ZsWlRyMe6SHOLCpMSVtSirZkzLXNpVcTF//nxmzZolRwWLP/jkk0/Iyc7B20Re+blgVYCK8Mabb8j1dZ5k2wU/derUKd58802WffIJGoWrWgvcSc3BEmZ0tJCgw2NxNrgKd1JzvId/Yv78+Sxd+h533TWI66+/HqvVanREUYacTifTp0/n+++/p0sNB3cm2zHJ89RyYzXBfc0KSIjwsmzZMk6cOM6kSZOJjIw0OpowmMfjYclbS4rmNicanSaAqaIdNk6uPsnXX3/Ntddea3SigCEjzn4mKyuLF154gd69b+ejj5fhrNSIgua9cNe8VEqzAXxRlXAkd8Xe5AayPFZmzZpFnzvuYPny5fIsPUjl5OTw0KgH+f777+nTsJD+UpoNYVLQu4GDQY0LWL9unez1LABYu3YtWaey8DaS0eaLVhVUjOKjjz4yOklAkeLsJ7Kzs3nppZfo1bs3773/PoXxdShs3hNX3Q5om4yyGM0XWxV7kxtwJHfleIGPp59+mjv69uXzzz+XAh1Ejhw5wv1Dh7JzxzaGpebTrbZTZkQZ7JoaLka1yOPA/r3cN/ReMjIyjI4kDPTRRx+hIlXRAjdxcRR46xXt67xnzx6j0wQMKc4Gy8nJYf78+dx2Wy/eefddCmNrFRXmeh3R4bFGxxNnUwpvfE0KU27C0agLR3PdPPHEE/S7806+/PJLvF4ZAQlk27dvZ+i9Qzhz8giPtsylbRXZ49RftEzwMP6SHBw5J7n/vqGkp6cbHUkYIDMzk59++glvXa+0l1Ki62iUWbFs2TKjowQM+dUzSH5+Pq+//nrxwrO3KIiuTmFqD1z1r0SHxxkdr3R4XURERNCzZ08iIiLAGyTb3iiFt0JtCpt2x9GwM5nZDqZNm8aAAQP5/vvv8flk14VAk56ezoMjR2J15zKhVTbJ8fIkyN/UjfUyqXUOsRTwz38+xLp164yOJMrZ559/DqroEA9RSmzgreHl8y8+lwNRSkiKczmz2+0sXryY23r1YuHCheSHV6Ew9VacDa5CR8QbHa9UKY+Lbt26MWLECLp164byBElx/o1SeCvWobDpzTgaXM3BUzlMmDCBwXffzQ8//CDbaAWI9PR0Hhk9mopWBxNaZVMtSp74+KvECB+Pt8qheoSLcePGSnkOMes3rIeKQITRSYKLrq5x2B3s2LHD6CgBQXbVKCdOp5OPP/6YNxctJjcnG298TVzNOuOLCt5NKLXFxvLlywFYvnw52hKk93ZK4a1Uj4KKdbCc2svew5sZM2YMTZqkMGTIPbRq1crohOIczi7N41rmEC/bzfm9GJvm0Za5PLU5lnHjxvLEE0/Spk0bo2OJMlZYWMiunbvwJcsT21JXvDvJpk2baN68ubFZAoCMOJcxrTXffPMNd/Ttx9y5czmjI7Gn3IgjuWtQl2YAzDbsdjvvv/8+drsdzDajE5UtZcKT2JCC1J44617Ozn0HGTVqFP98+GH2799vdDrxX3755RcpzQEq2lpUnqtHuBg39lHWr19vdCRRxrZs2YLP50Mn+v91quM12lr8X6JGx/t5ZhuoCopNmzYZnSQgSHEuQ1u3bmXo0PuYPHkyJwo82Btfh6PJ9fhiqhgdTZQl039OIXTWuoyfNm1h4MCBzJw5k9OnTxudTlC0i82Ex8cTZ5bSHKh+K89JEW4mT5rIyZMnjY4kytDWrVuLGksAjDfpNA3xQDz4rvQVve/nvAletm7dKovcS0CKcxk4cuQIEydO5P7772fnvgM463WksGl3fHHVjY4mypPJgiepOfnNb8NVuQnLPvmU2/v0YfHixTidTqPThSytNU8//RQ52WcYkZonpTmARVs1w5vl4XIUMm3qVHnQD2K5ubmYbCYwG50kSEUUHS4jj01/T4pzKXK5XCxcuJB+d97J9ytX46rekvzUnngSG4GSf+qQZQ3HVac9ham3kh+eyCuvvEK/fnfKwiaDfPTRR6xZs5bbGxRSO0aKVqBLivLRv1E+m9PTefvtt42OI8qIw+GQVVllqfgJiRTnvydtrpRs2rSJAQMG8vrrr+OIrUlB8564a7QCsxzNLIroiHicjf6BvfH1HM91MHqAPKNjAAAgAElEQVT0aCZOnCinoZWjo0eP8uLcubSo5KZrTXmACBYdk1y0reJiwYIFsp4gSDmdThltLkvFT0ocDoexOQKAFOeLdObMGaZNm8aDDz5IZlYujuRrcTa4Gm2LMjqa8FO+uGoUNLsFV41WfL9iJX379uNf//qXvMxcDj799FM8HjeDmhTIiYBBRCkYkFyIGR8ff/yx0XFEGXC5XGgl06rKTPH9ocsVZNvGlgEpzhdIa82///1v7ujbly+//hpXtTQKUm/FG1/D6GgiEJjMuKu3pCD1VvJtFZgzZw73Dh3K3r17jU4WtDweD/9e/impFd0khMsDcLCJsWlaJ7r48ovP5eXmIFSjRg3IB2Q3urKRByaTiapV5SzzvyPF+QJkZmYyatQonnrqKfJUNIXNbsFdszWYZAKWOD86PA5H8rU46l/F7v0HGXz33bz66qvywF8G1q9fT9bpM1xZXf5tg9WV1Z3kFxSyYsUKo6OIUtaoUSO0R0Oe0UmCkzqjqF2nNmFhYUZH8XtSnM+Dx+Ph7bffpv+AAWz+eSvOOh2wN+mGjqhgdDQRyJTCm1Cf/GY9cFaox6JFixg4cBDp6elGJwsqq1evJsqmaJkQGsfK1o7xEmH2EWH20TjeHRILIZtU8FApouj/tQgujRo1AooKnihlGszZZhonNzY6SUCQIdIS2rVrF08/8wx7du/GU6EWrjodZB6zKF3WcFz1O+FJqE9mxlpGjBjBjTfeyNChQ4mJiTE6XcDLysoiMdyLJUSGC+5MtnMgr2g11fjW+QanKR8mBVXC3WRlyYLbYFOzZk3CwsOwn7JDHaPTBJl88Dl8vz85EX8tRB5CLlxhYSFz585lyJAh7D2QiaPB1TgbdpHSLMqML64GBc1uwV01lU8+/ZQ7+vbj22+/RWuZl3sxzpzOIsYa/KOuoS7GpsmWg4aCjtls5uqrrsZ8yAyyfq1Uqb0Kk9lEp06djI4SEKQ4/4UffviBfnf2Z+nSpbgSk8lP7YG3Uj1kOb4oc2YrrtptsDftTrbHwqRJkxgzZgzHjh0zOlnAyj5zhjirrCwKdrFWH9nZ2UbHEGXg1ltvRXs0KkMeg0uNB8wHzFzZ6UoSEgLgWEY/IMX5T+Tl5TF58mTGjBnDqQI39pQbcNW9HCwyaV6UL19UAoUpN+Ks1YZ1GzbS7847+fDDD2X0+QLYwsIo8MgDbrAr9CisNpvRMUQZSE5OJqVpCuZ9ZpC7wFKhDiq0S9OjRw+jowQMKc7/5eeff2bAwEF88+13uKpfQkHTm/HFyPYswkDKhCcplYLUW7GHJ/Lcc88xduxYGVU7T81Sm7M7NwyfPOAGtV9zw0ht3sLoGKKM9OzRE52nUYfkSfBF84L5VzP1G9SnWbNmRqcJGFKci3k8Hl5//XWGDx/OqTwH9pQbcNe4BExyVJHwDzosBkdyV5y12vLDj+sYMGAgGzduNDpWwGjRogX5Ls3RArnbC1ZZDsXJQmjevLnRUUQZueqqq0hunIw53QxyyN1FUVsVOk8z7IFhKJmCWmLyCAIUFBQwYsRIFi5ciKtS/aJR5ujKRscS4n8phSepGYUpN3HG6WPUQw/xxhtvGJ0qIKSmpgKw8aS8jB+sNp+yAv/5fy2Cj9ls5rFxj2HymlCbpexdsCww7TZx44030qpVK6PTBBQpzsC8efPYum0rznqdcNW/EizywCr8my+qEgUp3fFUrMdrr70mI88lUL16ddq2bcNHGZEcK5S7vmBz2qFYujea1NRmsq1WkKtTpw53DboL02ETHDY6TQDygmWjhYSEBO6//36j0wSckH/0SE9PZ9myZbirNMWT2NDoOEKUnNmKs94VEBHL08/MwOGQ1y3/ilKK0aMfwRoWzqs7omWucxDRGl7fGYVXWXj00bGYTCH/0Bb0+vTpQ8NGDbFsskCu0WkCiAa1SaFzNI+OeZSoKNla93yF9L2Ly+Xi6aefgfBYXDXkpQoRgEwW7LUv59jRIyxcuNDoNH4vMTGR4SNGsuuMmS8PyS45wWLVURvpp6zcM+ReatasaXQcUQ4sFgtTJk8hNiIWyyoLFBidKABoUD8rTBkmBg4cyGWXXWZ0ooAU0sU5NzeXI0cyccdUBbPV6DhCXBBfTGWwhrNt2zajowSE6667jnZt2/LW7kg+PxiG7OwX2L7PtPHajiiap6bSs2dPo+OIclS9enVmPzebcMKLyrPT6ET+Te1SmH41ccsttzBo0CCj4wSskC7OCQkJ3HTTTVhP7UYVnjE6jhAXxHr0F3A7GDx4sNFRAoJSikmTJ9O+w+Us/jWShTsj8ci5KAHHp+Gt3REs2BHFJa1a89TTT8sUjRBUv359ZjwzA4vTgmW1BdxGJ/JPap/C9IuJa665hpEjR8ouGhch5O9lBg8eTEREJGEHfwSfPHqKwKIceYQd/ZmOHTuRlpZmdJyAERERwbRp0+jbty/fZIYxIz2GArc8kAQKhwdmb4nm3wfCueWWW3j6mWeIjo42OpYwSPPmzZk2dRoqR2FeKdvU/YEGtVth2mjisjaXMW7cOHmCeZFC/l8vPj6eIffcjTknk6htH2LKOWJ0JCH+ns+L5ejPRG37AJvFzP3332d0ooBjMpm49957GTt2LLtywpj0Uxx7cmTfdn93IM/MlI1xpGfZGDlyJKNGjcJisRgdSxisXbt2TJ82HVuBDcu3FsgxOpEf8IFKV5jSTVxxxRVMmzoNq1WmpV6skC/OALfeeitPPPEElWNsROz8N2G7v0U5ZaWB8E+m4id5YQfX0/ayS1m48P+oVq2a0bEC1nXXXcdzs2fjDq/E5A2xvLEzgkKP0anEf3N44e3dETy+PpZcUzzPzJghxwSLP+jQoQMvzn2ROFsclu8tcNzoRAbygGmtCdMeE71792bKlCmEh4cbnSooKB0gK2Nat26tf/rppzK9DafTyVtvvcXixUvw+DTOamm4qzYFk4xmXAjbgR+wnPwVAF9kJXxRlXDVbmdwqsClHHnYDq7DciaDqklJjHrwQdq1k3/P0lJQUMCCBQv44IN/ER8GdzbM59LKbgJ5KuC0n4qmL4xvnW9wkouTfsrCwl0xnLLDDTfcwNChQ4mNjTU6lvBTx48fZ/Qjo8k4kIHvEh+6bvn3HNP3ReOSvisNmAJqB8saC2TDqFGjuPnmm8s/Q4BTSm3UWrf+089Jcf5fR44cYe7cuaxevRpli8BZpRnuyk3kYJQLEL79UwAcKTcYnCRwKfsZrEd+xpq1B6vVyoD+/enduzdhYbKdWlnYsWMHM2c8w+49e0lLcDMwuZCEiMBc/xDoxfmMU7FoVyTrT9ioU6smDz8yRo7TFiVSUFDAhAkT2LBhA74GPnQLXa6vsRtWnE+B5UcLNm1j6pSptG3btnxvP0hIcb5A6enpLFq0iA0bNqAsYTgrN8FdtRlY5eWOkpLifOFMBaewZqZjOXMAq83Kzd2707t3bypXluPgy5rH4+GDDz5gwauv4vO4uKF2Id1qOwgLsCnQgVqc3T744mAYH2dE4VFmBgwYSJ8+fWR+pjgvHo+Hl19+maVLl0ICeNt6IaJ8brvci7MGtUdh+tlEUtUknpj+BPXr1y+f2w5Cf1WcZQ7CX0hLSyMtLY2dO3eyePFiVq5cSdjxbbgSk3EnpaJtcuKOKH2m3KPYjmzBnHOYiMgobut/Jz179iQ+Pt7oaCHDYrHQq1cvOnXqxEsvvcQH333HiqMR3F4/n7ZVAnv6hj/TGjaetPL2niiOFyrat2/HsGHDqFGjhtHRRACyWCwMGzaMlJQUnnzqSVzfuPC08UCi0clKmQfUTwrTIRPtO7TnsXGPERMTY3SqoCUjzuchIyODJUuW8NVXX6Ex4aySgrtaC7DIS+bnIiPOJWcqOIXt0AbMOZnExsXT5/be3HzzzXIkqh9IT0/nhefnsHvPXhrFe+nXqIB6sV6jY/2tQBpxPpRvYvGvUWw7baFOrZoMGzFSTjYTpSYjI4Ox48aSmZmJL9WHbqShDJ8Al9uIcx5Yfig6dvyee+7hjjvukO3mSoFM1ShlR44c4f/+7//48ssvwWLDWbUF7qopsojwT0hx/nvKkYvt8EYsWXuJio5hQP87ueWWW2QOs5/xer189tlnvDr/ZbJzcrkiyUnvBnbiwvz3PjQQinOeS/GvfeF8mxlOVFQUdw2+m+7du8sWc6LUFRQU8MSTT7Bq5Sp8NXzoS3WZve5eLsX5CFg2WIgOj2bypMm0bv2nPU9cACnOZWTv3r28/PJ81q37ERUWjaNaSzyJDUHJs73fSHH+C247tsx0rCd3YLVY6N2rF3369JGX2PxcQUEBb775Ju8tXYrV5KNH3QK61HBi9sPL3p+Ls08XHZe9dF8UhR4TN998M4MGDSIuLs7oaCKIaa15++23mT9/PsSCp50HyuAut0yLswa1TWHaYaJho4Y8Mf0JqlSpUvq3E8KkOJexzZs389K8eezauRMdVQlH7fb4YuSXGKQ4/yntw3J8B+GZG8Hrplu3bgwaNIjExGCbeBfcDh06xJzZs1m/YQM1Y3wMaFRA4wr+tQG0vxbnvTlm3tgVzb5cEy2ap/LgqIdkIZMoVxs2bGDipIkUOAvwXOqBUt4Kv8yKswtM602oo4rrrruOhx56SF6dLANSnMuB1prvv/+e5194gaxTp3AnNsJV87KQ34FDivMfmfJOEH5gDaogi1atWzNyxAjq1KljdCxxgbTWrFq1ihfmzOb4yVO0r+qkT0M7Ffxk+oa/Fec8l+LdPRGsOBJGxYoVuP+BYXTu3Bklqy2FAY4ePcq4x8axd89efCk+dErpzXsuk+KcUzSf2VRoYuTIkXTv3l2unTIiu2qUA6UUV111FW3atOGNN97g3aVLsWUfxFGjNZ7EZGQZfohzO7AdWo/15K9UrFSJkY9M5sorr5Q7vQCnlKJjx45cdtllLF68mLffWsLmU+HcWreAf9T0z+kbRvBp+C7TxtJ90Ti8il69b2PgwIGy8FUYKikpiZfnvcyMGTP48ssv0TkaXxsf+OO2k0eL9meOi4lj+vPTSU1NNTpRyDJsxFkpdS0wh6Jf0QVa66f+6uv9fcT5v+3fv59Zzz7Lz1u2oKMq4Uxqjrdi3ZCb/xzyI85uB9bj2wg7sR3lddOrVy8GDhxIZGSk0clEGTh8+DBz5sxm3br11IrxMTA5n0bxxu2+4Q8jzvtzzSzcFc3eHBNpaS146KF/yqsswq9orXn33XeZN28euqLG294LF/licWmOOKs9ClO6iQYNGvDM08+QkJBw0T9T/DW/m6qhlDIDvwJdgMPABqCP1nr7ub4n0IozFF2M33zzDa+9/n9kHj4EEXE4qzbHk9AATP74lLb0hWpxVq4CrEd/wXZyF9rrpn2HDgy55x7q1atndDRRxn6bvjFn9nOcPJVFp2pObm9gJ8ZW/ve1RhbnArfi/b3hfJMZTlxcHMOGj5BpGcKvrVixgilTp+CxefB08MBFnOpeKsVZg9qiMO020a59OyZNnERERDmd4BLi/HGqxmXAHq31PgCl1DtAd+CcxTkQKaXo3LkzV111FatWreLNRYvYs3sV4Uc24aiSiqdyMpjlJKxgohw5WI/8jC1rDwrNNddcQ9++faUwh5Dfpm+0bt2aN954g6VL32XjqTB61y+gUzUXpiDvjVrD2mM23toTRZ5LcfMttzB48GDZLUb4vU6dOvFC4gs88ugj5H2XV7TjhlEHtXrAtM6EOqLo2bMnDzzwAGZzaAy4+TujRpx7Atdqre8ufv9OoI3Weth/fd0QYAhArVq1Wh04cKDcs5YmrTUbNmxg0eLFbElPR1nDcVWog7dCHbyx1SAINy23HfgBAFftdgYnKUNuB5YzB7CcycCccxiLxcIN3bpx++23U61aKS/VFgFn3759PDtrFj//8gv1Yn3c2SifhuU0fWPRrqLRqTuT7eVye/tzzSz6NYpfs800Tm7EPx8eTXJycrncthCl5ejRo4x+ZDSHDh0qOmnwAg6uVOlFz5B12gV0LBeYV5tRpxUjRoygR48e5/8zxEXxx6kaJSrOZwvEqRp/ZevWrbz33nusXfsDTqcDZQ3DFVezqETH1QCzrNv0Z8pZgPlMRlFZzjsGWpNYuTL/6NKFnj17UqlSJaMjCj+itearr75i3ksvknX6DB2qFh2eUjHcP3bfuFg5LsV7xbtlxMbGcM+Qe+nWrZuMkImAlZeXx+jRo9m+Yzu+S33o2uV0rTrBssqCKdfEpEmT6NSpU/ncrvgDfyzO7YBJWuuuxe+PBdBaP3mu7wm24vwbp9PJhg0bWLlyJatWraagIB9ltuKOrY6nYh288bXAYjM6pqBoGobldAbWMwdQ+ScAqFmrNld26kjHjh1p1KiRzN8Uf6mwsJAlS5bwzjtvY/J5ualOIdfVcmAL0H7p8cGXh8L4MCMKl0/Ro0dPBgwYINMyRFAoLCxk7NixbN68Gd8lPnT9Mu5L9qLSbCm08MQTT9CmTZuyvT1xTv5YnC0ULQ68BsikaHHgHVrrbef6nmAtzmfzeDxs2bKFFStWsGLFSs6cOQ0mE96YJLyx1fDGJuGLSgi5nTkM43FhzjuGOfco1txMKDwNQMNGyVx1ZSeuuOIKateubXBIEYiOHDnCSy+9yMqVq0iMhNvr53NZZXfA7FqpNaSfsvLWniiOFijatLmM4cNHUKtWLaOjCVGqnE4nj094nB9/+BFfcx86uYw6UwFYVlqweWzMeGYGaWlpZXM7okT8rjgDKKWuB2ZTtB3d61rr6X/19aFQnM/m8/nYvn07q1atYvWatRw6WDS/W1msuKOq4ItNkiJd2s4qypb8Y6iCU6A1FquVZk2bccUVl9OxY0c52lSUmo0bN/LC83PYtz+DhvFe7mhQUG7zny9URq6Zt/ZEsv20hZrVqzNsxAjatQviNQwi5Hk8HqZOncp3331XNuW5ECzfW4hUkcyaOYuUlJTS/fnivPllcT5foVac/9vp06fZsmULmzdvZuOmzWcVaRvu6Cr4Yn4r0pWkSJdUcVE25R3FmndWUbZYaNq0KZdccgktW7akSZMmcqSpKDNer5fPPvuMBa++wukz2bSp4qJ3fTuVI0v5qN6LlOVQvLc3gjVHw4iJiWbQXYPp3r07FousxxDBz+v1MmXKlKLy3MqHrldK3ckBlhUWwj3hPD/neVlM6yekOAeh06dPk56eTnp6Ohs3buLQoYPA2UW6alGRjkwIyt06LojHiTnv+P8UZbPFQrOmTWnZsiUtW7YkJSVFirIod4WFhbzzzju88/ZbuN0u/lHDQfe6DqKtxt5HF3rg04xwPjsUASYLt93Wi379+hEdHW1oLiHKm9vtZuzYsaxfvx5fWx+65kVem66i6RmWAgvPznqWFi1alE5QcdGkOIeArKys/xTpTZs4fOgQAMpsxRNduXiedFV8UYkhc/gKbkfR1Iu8Y1h+K8qA2WIhpUkKrVpdQlpaGk2bNpWiLPzGqVOnWLBgAZ999m8irXBbvQKurl7++z/7NKw+auOdvVHkOqFz587cc889JCUllW8QIfyIw+HgoX8+xLZt2/C098CFXg4eMK8yYz5j5qmnnpKFgH5GinMIOntqx6bN6Rw8kAGAMlvwRFXG+9uIdHQimILkpVa3HXPuMcx5R7HmH4OCosV8VquNps2a0jItjbS0NBlRFgFh7969PD9nDpvT06kTW3R8d4O48pn/nJFr5o1fo9idbaZpSgojRo6kSZMm5XLbQvi7/Px8ho8Yzr6MfXiu9kDcef4ADepHhTnTzKRJk7jqqqvKJKe4cFKcBdnZ2WzZsoX09HQ2b05n//59aK1RJjPeqEQ8sUl442sVLzYMkKX9XjfmnCOYcw5hzT8OhWcAsIWFkZqa+ntRbty4MTabbOknAo/Wmm+//Za5LzxP1ukzdKpWtP9zbBkd3332MdmxsbHcd/8DdO3aFZNM9xLiD06dOsVdg+8ix5OD5xoPnMchwOpXhWmLiaFDh3LHHXeUXUhxwaQ4i/+Rl5f3hyK9Z8/uoiIdFoUrtgbeCrXwxlb3u4NYlKsA85mDWLIPYsk9ivZ5CAsPp0Xz5rRs2ZK0tDSSk5NlwZIIKoWFhb8f3x1u8tGzXiHX1HCW2vQNrWFV8bSMfDkmW4gS2bJlCyNHjsSb5MXXzgcluR5PgnmFmSsuv4Jp06bJ3v9+Soqz+FvZ2dn8+OOPrFmzhnXr1+Ow21FmC+6YJLwVauONr4W2RZZ/MK0xFWZhPnMAa/ah3+cpV6lSlSuuuJwOHTrQvHlzrNbzeLovRIDKyMhg9nPPsWnzZlpUcjO0aQExFzn6XOiBV7dHseGEjWZNUxj10D9p2LBhKSUWIri9++67vPjiiyXbps4Olm8sJFVMYsGrC2SBrR+T4izOi9vtJj09nbVr17Jy1SpOnig6JU9HJ+KOr4UnoSE6rAwveK0x5R3DkrUXW84htLMApRRNUlK4vEMHOnToQJ06deSZughJWms+/vhjXnh+DnFWL8Ob5VL/Auc+H8o3MeeXWE7Yzdx777307t1bpmUIcR601kyYMIEVK1fg7eL9y/nOpjUmwrLCeGX+K9SrV6/8QorzJsVZXDCtNfv27WPt2rWsXr2GnTt3oAFPxXq4k5oX7Rtdajfmw3x6P2HHtqLyTxIWFk6bNpfRoUMH2rZtS4UKFUrvtoQIcDt27GDC+MfIyjpFv0YFXFPddV7LE1YftfH6zmiiY+OYNHmKnFQmxAXKycmh9+29KYwuxHuF98+nbBwH88qiJ6h9+/Yt94zi/EhxFqXm+PHjvPfee3y8bBlOhwNvXHXcVVPxxlW/8EWFXjeWk7sIO74NHHlUr16DPn1up2vXrrL7hRB/IScnh2nTprJu3XouT3IyJKXwb+c9aw1v7orgq8PhpLVozsRJk6lUqRSfAAsRgpYuXcrcuXPxXu793y3qNFi+tpAYlsiSxUvkcS0ASHEWpS4vL49ly5axdOl7nDlzGh1VEVflFLT1/OZBm/KPE3ZyJ9rtpGmzZtzRpw8dOnSQl4uFKCGfz8fChQtZuHAh19Vy0LeR/S+//sN94fxrXwS9evVi6NChspBWiFLgdru5s/+dHM07iqeLB856CFP7FKaNJqZMmcKVV15pWEZRcn9VnOUeU1yQmJgY+vbtS69evfj666956623ObB/9Xn/HKUUV1zRkdtv702zZs3KIKkQwc1kMnHXXXeRm5vLBx98QLUoL1dVd/3p1/543Mq/9kXQtWtXHnjgAVknIEQpsVqtDHtgGOPGjUMdUujaxYOSGsw7zDRNbUqnTp2MDSlKhRRncVGsVivXXXcd1157Lfv27cPtdp/X91esWJHKlSuXUTohQsewYcM4fOgQC3/aQJUIHykVPX/4/N4cM/O3x5DaLIXRo0dLaRailHXo0IHKVSpz/ODx/xTnk6ALNbf1vE2uuSAhxVmUCqUU9evXNzqGECHLYrEwafJk7ht6L09sOvSnX1O1SiLTn3hSDgQSogwopej6j64sWrwIHEA4qAOK8Ihw2rdvb3Q8UUqkOAshRJCIjo7m2edm89lnn+Hx/HHE2Ww206VLF+Lj4w1KJ0Tw69y5M4sWLULtVugkjfmImSuvvlIWBAYRKc5CCBFEEhMT6d+/v9ExhAhJdevWpX6D+uzduRd2gkbTtWtXo2OJUiTFWQghhBCilMx4ZgZ79+4FICIigtTUVIMTidIkxVkIIYQQopQkJCSQkJBgdAxRRmSzXCGEEEIIIUpAirMQQgghhBAlIMVZCCGEEEKIEpDiLIQQQgghRAlIcRZCCCGEEKIEpDgLIYQQQghRAlKchRBCCCGEKAEpzkIIIYQQQpSAFGchhBBCCCFKQIqzEEIIIYQQJSDFWQghhBBCiBKQ4iyEEEIIIUQJSHEWQgghhBCiBKQ4CyGEEEIIUQJSnIUQQgghhCgBKc5CCCGEEEKUgBRnIYQQQgghSkBprY3OUCJKqZPAAaNziAuSAJwyOoQQIUquPyGMIdde4KqttU78s08ETHEWgUsp9ZPWurXROYQIRXL9CWEMufaCk0zVEEIIIYQQogSkOAshhBBCCFECUpxFeXjF6ABChDC5/oQwhlx7QUjmOAshhBBCCFECMuIshBBCCCFECUhxFqVKKfW6UuqEUmrrWR+bpJTKVEqlF/93vZEZhQhGSqmaSqnvlFLblVLblFIjiz9+W/H7PqWUrPAXopT9xbX37lmPexlKqXSjs4qLJ1M1RKlSSnUE8oE3tdbNij82CcjXWs80MpsQwUwplQQkaa03KaVigI3AzYAGfMB84GGt9U8GxhQi6Jzr2tNabz/ra2YBOVrrKUblFKXDYnQAEVy01iuVUnWMziFEqNFaHwWOFr+dp5TaAVTXWn8FoJQyMp4QQetc1x6wHUAVXXy9gKsNCylKjUzVEOVlmFLq5+KpHBWMDiNEMCt+8toSWGdsEiFCyzmuvSuA41rr3UZkEqVLirMoD/OA+kAaRc/KZxkbR4jgpZSKBv4FPKi1zjU6jxCh4i+uvT7A28akEqVNpmqIMqe1Pv7b20qpV4FPDYwjRNBSSlkpeuBeorX+wOg8QoSKc117SikLcCvQyqhsonTJiLMoc8ULJ35zC7D1XF8rhLgwxfMoXwN2aK2fNTqPEKHib669zsBOrfXh8k8myoLsqiFKlVLqbeBKIAE4Dkwsfj+NotX9GcC9xYsphBClRCl1ObAK+IWiXTQAxgFhwAtAIpANpGutuxoSUoggdK5rT2v9b6XUQuBHrfXLRuUTpUuKsxBCCCGEECUgUzWEEEIIIYQoASnOQgghhBBClIAUZyGEEEIIIUpAirMQQgghhBAlIMVZCCGEEEKIEpDiLIQQfkoplV/GP3+gUqraWe9nKKUSyvI2hRAikElxFkKI0DUQqPZ3XySEEKKIHLkthBABRCmVCLwM1Cr+0INa6zVKqUnFH6tX/OdsrfXzxX9SvrwAAAGwSURBVN/zONAPOAkcAjZSdBhRa2CJUsoOtCv+ecOVUjcCVuA2rfXO8vh7CSFEIJARZyGECCxzgOe01pcCPYAFZ32uMdAVuAyYqJSyKqV++7oWwHUUlWW01u8DPwF9tdZpWmt78c84pbW+BJgHPFwefyEhhAgUMuIshBCBpTOQopT67f1YpVR08dvLtdZOwKmUOgFUAToAH2utHYBDKfXJ3/z8D4r/3AjcWrrRhRAisElxFkKIwGIC2hYX4d8VF2nnWR/ycmH38b/9jAv9fiGECFoyVUMIIQLLl8Dw395RSqX9zdevAW5USoUXj0zfcNbn8oCY0o8ohBDBSUYThBDCf0UqpQ6f9f6zwAjgRaXUzxTdh68Ehp7rB2itNyillgE/A8eBX4Cc4k8vBF7+r8WBQgghzkFprY3OIIQQogwppaK11vlKqUiKivYQrfUmo3MJIUSgkRFnIYQIfq8opVKAcOANKc1CCHFhZMRZCCGEEEKIEpDFgUIIIYQQQpSAFGchhBBCCCFKQIqzEEIIIYQQJSDFWQghhBBCiBKQ4iyEEEIIIUQJSHEWQgghhBCiBP4f3uDNs2l2gKAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn \n",
    "import matplotlib \n",
    "from matplotlib import pyplot\n",
    "fig = pyplot.figure(figsize=(12,8))\n",
    "\n",
    "seaborn.violinplot(data=df_concat, x=\"Length\", y=\"manufacturability_difficulty_score\")\n",
    "\n",
    "fig.savefig(\"manufacturability-violin-plot.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcell_epitopes = pd.read_csv(\"../bcell/accessible-linear-bcell-epitopes-grouped-merged-filtered.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_sources</th>\n",
       "      <th>sources</th>\n",
       "      <th>accessible_subsequence_start</th>\n",
       "      <th>accessible_subsequence_end</th>\n",
       "      <th>accessible_subsequence</th>\n",
       "      <th>accessible_subsequence_length</th>\n",
       "      <th>accessible_subsequence_is_4mer_or_longer</th>\n",
       "      <th>IgG</th>\n",
       "      <th>IgA</th>\n",
       "      <th>IgG_and_IgA</th>\n",
       "      <th>accessible_subsequence_contains_polymorphism</th>\n",
       "      <th>accessible_subsequence_contains_glycosite</th>\n",
       "      <th>near_any_feature</th>\n",
       "      <th>in_any_feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>Charite 2020 S459-473; Wang 2020 S456-460; Dah...</td>\n",
       "      <td>456</td>\n",
       "      <td>473</td>\n",
       "      <td>FRKSNLKPFERDISTEIY</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5</td>\n",
       "      <td>Charite 2020 S569-583; Charite 2020 S571-585; ...</td>\n",
       "      <td>580</td>\n",
       "      <td>583</td>\n",
       "      <td>QTLE</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7</td>\n",
       "      <td>Poh 2020 S809-826; Wang 2020 S806-820; Charite...</td>\n",
       "      <td>809</td>\n",
       "      <td>812</td>\n",
       "      <td>PSKP</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    num_sources                                            sources  \\\n",
       "5             3  Charite 2020 S459-473; Wang 2020 S456-460; Dah...   \n",
       "7             5  Charite 2020 S569-583; Charite 2020 S571-585; ...   \n",
       "11            7  Poh 2020 S809-826; Wang 2020 S806-820; Charite...   \n",
       "\n",
       "    accessible_subsequence_start  accessible_subsequence_end  \\\n",
       "5                            456                         473   \n",
       "7                            580                         583   \n",
       "11                           809                         812   \n",
       "\n",
       "   accessible_subsequence  accessible_subsequence_length  \\\n",
       "5      FRKSNLKPFERDISTEIY                             18   \n",
       "7                    QTLE                              4   \n",
       "11                   PSKP                              4   \n",
       "\n",
       "    accessible_subsequence_is_4mer_or_longer   IgG   IgA  IgG_and_IgA  \\\n",
       "5                                       True  True  True         True   \n",
       "7                                       True  True  True         True   \n",
       "11                                      True  True  True         True   \n",
       "\n",
       "    accessible_subsequence_contains_polymorphism  \\\n",
       "5                                          False   \n",
       "7                                          False   \n",
       "11                                         False   \n",
       "\n",
       "    accessible_subsequence_contains_glycosite  near_any_feature  \\\n",
       "5                                       False              True   \n",
       "7                                       False              True   \n",
       "11                                      False              True   \n",
       "\n",
       "    in_any_feature  \n",
       "5             True  \n",
       "7            False  \n",
       "11           False  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcell_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count      258\n",
      "unique      50\n",
      "top       QTLE\n",
      "freq        54\n",
      "Name: overlapping_bcell_epitope, dtype: object\n",
      "\n",
      "count    258.000000\n",
      "mean     591.034884\n",
      "std      144.914448\n",
      "min      456.000000\n",
      "25%      457.000000\n",
      "50%      580.000000\n",
      "75%      809.000000\n",
      "max      812.000000\n",
      "Name: overlapping_bcell_epitope_start, dtype: float64\n",
      "\n",
      "count    258.000000\n",
      "mean     596.383721\n",
      "std      142.341632\n",
      "min      456.000000\n",
      "25%      473.000000\n",
      "50%      583.000000\n",
      "75%      811.000000\n",
      "max      812.000000\n",
      "Name: overlapping_bcell_epitope_end, dtype: float64\n",
      "\n",
      "count    29043.000000\n",
      "mean         0.056399\n",
      "std          0.751685\n",
      "min          0.000000\n",
      "25%          0.000000\n",
      "50%          0.000000\n",
      "75%          0.000000\n",
      "max         18.000000\n",
      "Name: overlapping_bcell_epitope_length, dtype: float64\n",
      "\n",
      "count    29043.000000\n",
      "mean         0.006508\n",
      "std          0.074768\n",
      "min          0.000000\n",
      "25%          0.000000\n",
      "50%          0.000000\n",
      "75%          0.000000\n",
      "max          1.000000\n",
      "Name: fraction_of_full_bcell_epitope, dtype: float64\n",
      "\n",
      "count                    258\n",
      "unique                     3\n",
      "top       FRKSNLKPFERDISTEIY\n",
      "freq                     114\n",
      "Name: full_bcell_epitope, dtype: object\n",
      "\n",
      "count    258.000000\n",
      "mean       0.007752\n",
      "std        0.087874\n",
      "min        0.000000\n",
      "25%        0.000000\n",
      "50%        0.000000\n",
      "75%        0.000000\n",
      "max        1.000000\n",
      "Name: bcell_epitope_distance_from_edge, dtype: float64\n",
      "\n",
      "count     29043\n",
      "unique        2\n",
      "top        True\n",
      "freq      28785\n",
      "Name: bcell_epitope_more_than_1aa_from_edge, dtype: object\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>HLA-I_peptides</th>\n",
       "      <th>HLA-I_haplotypes</th>\n",
       "      <th>HLA-I_pop_freq</th>\n",
       "      <th>HLA-II_peptides</th>\n",
       "      <th>HLA-II_haplotypes</th>\n",
       "      <th>HLA-II_pop_freq</th>\n",
       "      <th>...</th>\n",
       "      <th>manufacturability_difficulty_score</th>\n",
       "      <th>Length</th>\n",
       "      <th>overlapping_bcell_epitope</th>\n",
       "      <th>overlapping_bcell_epitope_start</th>\n",
       "      <th>overlapping_bcell_epitope_end</th>\n",
       "      <th>overlapping_bcell_epitope_length</th>\n",
       "      <th>fraction_of_full_bcell_epitope</th>\n",
       "      <th>full_bcell_epitope</th>\n",
       "      <th>bcell_epitope_distance_from_edge</th>\n",
       "      <th>bcell_epitope_more_than_1aa_from_edge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESLVPGFNEKTHVQ</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>ESLVPGFNEK,SLVPGFNEK</td>\n",
       "      <td>HLA-A*11:01,HLA-A*03:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ESLVPGFNEKTHVQL</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>ESLVPGFNEK,SLVPGFNEK</td>\n",
       "      <td>HLA-A*11:01,HLA-A*03:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SLVPGFNEKTHVQLS</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>SLVPGFNEK</td>\n",
       "      <td>HLA-A*03:01,HLA-A*11:01</td>\n",
       "      <td>0.299196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LVPGFNEKTHVQLSL</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VPGFNEKTHVQLSLP</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9670</th>\n",
       "      <td>AFPFTIYSLLLCRMNSRNYIAQVDVVN</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>952</td>\n",
       "      <td>978</td>\n",
       "      <td>AFPFTIYSL,CRMNSRNYI,FPFTIYSL,FPFTIYSLL</td>\n",
       "      <td>HLA-A*24:02,HLA-C*06:02,HLA-B*07:02,HLA-B*08:0...</td>\n",
       "      <td>0.648355</td>\n",
       "      <td>IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,SLLLCRMNSRNYIA...</td>\n",
       "      <td>HLA-DRB1*13:01</td>\n",
       "      <td>0.106080</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9671</th>\n",
       "      <td>FPFTIYSLLLCRMNSRNYIAQVDVVNF</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>953</td>\n",
       "      <td>979</td>\n",
       "      <td>CRMNSRNYI,FPFTIYSL,FPFTIYSLL,NYIAQVDVVNF,YIAQV...</td>\n",
       "      <td>HLA-C*06:02,HLA-B*07:02,HLA-B*08:01,HLA-B*35:0...</td>\n",
       "      <td>0.648355</td>\n",
       "      <td>IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.315083</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9672</th>\n",
       "      <td>PFTIYSLLLCRMNSRNYIAQVDVVNFN</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>954</td>\n",
       "      <td>980</td>\n",
       "      <td>CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF</td>\n",
       "      <td>HLA-C*06:02,HLA-A*24:02,HLA-B*35:01</td>\n",
       "      <td>0.379828</td>\n",
       "      <td>IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.315083</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9673</th>\n",
       "      <td>FTIYSLLLCRMNSRNYIAQVDVVNFNL</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>955</td>\n",
       "      <td>981</td>\n",
       "      <td>AQVDVVNFNL,CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF</td>\n",
       "      <td>HLA-A*02:01,HLA-C*06:02,HLA-A*24:02,HLA-B*35:01</td>\n",
       "      <td>0.638745</td>\n",
       "      <td>IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.315083</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9674</th>\n",
       "      <td>TIYSLLLCRMNSRNYIAQVDVVNFNLT</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>956</td>\n",
       "      <td>982</td>\n",
       "      <td>AQVDVVNFNL,CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF</td>\n",
       "      <td>HLA-A*02:01,HLA-C*06:02,HLA-A*24:02,HLA-B*35:01</td>\n",
       "      <td>0.638745</td>\n",
       "      <td>IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.315083</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29043 rows  37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Sequence Protein  Start  End  \\\n",
       "0                 MESLVPGFNEKTHVQ  orf1ab      1   15   \n",
       "1                 ESLVPGFNEKTHVQL  orf1ab      2   16   \n",
       "2                 SLVPGFNEKTHVQLS  orf1ab      3   17   \n",
       "3                 LVPGFNEKTHVQLSL  orf1ab      4   18   \n",
       "4                 VPGFNEKTHVQLSLP  orf1ab      5   19   \n",
       "...                           ...     ...    ...  ...   \n",
       "9670  AFPFTIYSLLLCRMNSRNYIAQVDVVN   ORF10    952  978   \n",
       "9671  FPFTIYSLLLCRMNSRNYIAQVDVVNF   ORF10    953  979   \n",
       "9672  PFTIYSLLLCRMNSRNYIAQVDVVNFN   ORF10    954  980   \n",
       "9673  FTIYSLLLCRMNSRNYIAQVDVVNFNL   ORF10    955  981   \n",
       "9674  TIYSLLLCRMNSRNYIAQVDVVNFNLT   ORF10    956  982   \n",
       "\n",
       "                                         HLA-I_peptides  \\\n",
       "0                                  ESLVPGFNEK,SLVPGFNEK   \n",
       "1                                  ESLVPGFNEK,SLVPGFNEK   \n",
       "2                                             SLVPGFNEK   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "9670             AFPFTIYSL,CRMNSRNYI,FPFTIYSL,FPFTIYSLL   \n",
       "9671  CRMNSRNYI,FPFTIYSL,FPFTIYSLL,NYIAQVDVVNF,YIAQV...   \n",
       "9672                   CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF   \n",
       "9673        AQVDVVNFNL,CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF   \n",
       "9674        AQVDVVNFNL,CRMNSRNYI,NYIAQVDVVNF,YIAQVDVVNF   \n",
       "\n",
       "                                       HLA-I_haplotypes  HLA-I_pop_freq  \\\n",
       "0                               HLA-A*11:01,HLA-A*03:01        0.299196   \n",
       "1                               HLA-A*11:01,HLA-A*03:01        0.299196   \n",
       "2                               HLA-A*03:01,HLA-A*11:01        0.299196   \n",
       "3                                                   NaN        0.000000   \n",
       "4                                                   NaN        0.000000   \n",
       "...                                                 ...             ...   \n",
       "9670  HLA-A*24:02,HLA-C*06:02,HLA-B*07:02,HLA-B*08:0...        0.648355   \n",
       "9671  HLA-C*06:02,HLA-B*07:02,HLA-B*08:01,HLA-B*35:0...        0.648355   \n",
       "9672                HLA-C*06:02,HLA-A*24:02,HLA-B*35:01        0.379828   \n",
       "9673    HLA-A*02:01,HLA-C*06:02,HLA-A*24:02,HLA-B*35:01        0.638745   \n",
       "9674    HLA-A*02:01,HLA-C*06:02,HLA-A*24:02,HLA-B*35:01        0.638745   \n",
       "\n",
       "                                        HLA-II_peptides  \\\n",
       "0                                                   NaN   \n",
       "1                                                   NaN   \n",
       "2                                                   NaN   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "9670  IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,SLLLCRMNSRNYIA...   \n",
       "9671  IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...   \n",
       "9672  IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...   \n",
       "9673  IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...   \n",
       "9674  IYSLLLCRMNSRNYI,LLLCRMNSRNYIAQV,MNSRNYIAQVDVVN...   \n",
       "\n",
       "                                 HLA-II_haplotypes  HLA-II_pop_freq  ...  \\\n",
       "0                                              NaN         0.000000  ...   \n",
       "1                                              NaN         0.000000  ...   \n",
       "2                                              NaN         0.000000  ...   \n",
       "3                                              NaN         0.000000  ...   \n",
       "4                                              NaN         0.000000  ...   \n",
       "...                                            ...              ...  ...   \n",
       "9670                                HLA-DRB1*13:01         0.106080  ...   \n",
       "9671  HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01         0.315083  ...   \n",
       "9672  HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01         0.315083  ...   \n",
       "9673  HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01         0.315083  ...   \n",
       "9674  HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*04:01         0.315083  ...   \n",
       "\n",
       "     manufacturability_difficulty_score Length overlapping_bcell_epitope  \\\n",
       "0                                   4.0     15                      None   \n",
       "1                                   3.0     15                      None   \n",
       "2                                   2.0     15                      None   \n",
       "3                                   2.0     15                      None   \n",
       "4                                   2.0     15                      None   \n",
       "...                                 ...    ...                       ...   \n",
       "9670                               19.0     27                      None   \n",
       "9671                               19.0     27                      None   \n",
       "9672                               19.0     27                      None   \n",
       "9673                               19.0     27                      None   \n",
       "9674                               19.0     27                      None   \n",
       "\n",
       "     overlapping_bcell_epitope_start  overlapping_bcell_epitope_end  \\\n",
       "0                                NaN                            NaN   \n",
       "1                                NaN                            NaN   \n",
       "2                                NaN                            NaN   \n",
       "3                                NaN                            NaN   \n",
       "4                                NaN                            NaN   \n",
       "...                              ...                            ...   \n",
       "9670                             NaN                            NaN   \n",
       "9671                             NaN                            NaN   \n",
       "9672                             NaN                            NaN   \n",
       "9673                             NaN                            NaN   \n",
       "9674                             NaN                            NaN   \n",
       "\n",
       "      overlapping_bcell_epitope_length  fraction_of_full_bcell_epitope  \\\n",
       "0                                    0                             0.0   \n",
       "1                                    0                             0.0   \n",
       "2                                    0                             0.0   \n",
       "3                                    0                             0.0   \n",
       "4                                    0                             0.0   \n",
       "...                                ...                             ...   \n",
       "9670                                 0                             0.0   \n",
       "9671                                 0                             0.0   \n",
       "9672                                 0                             0.0   \n",
       "9673                                 0                             0.0   \n",
       "9674                                 0                             0.0   \n",
       "\n",
       "      full_bcell_epitope  bcell_epitope_distance_from_edge  \\\n",
       "0                   None                               NaN   \n",
       "1                   None                               NaN   \n",
       "2                   None                               NaN   \n",
       "3                   None                               NaN   \n",
       "4                   None                               NaN   \n",
       "...                  ...                               ...   \n",
       "9670                None                               NaN   \n",
       "9671                None                               NaN   \n",
       "9672                None                               NaN   \n",
       "9673                None                               NaN   \n",
       "9674                None                               NaN   \n",
       "\n",
       "      bcell_epitope_more_than_1aa_from_edge  \n",
       "0                                      True  \n",
       "1                                      True  \n",
       "2                                      True  \n",
       "3                                      True  \n",
       "4                                      True  \n",
       "...                                     ...  \n",
       "9670                                   True  \n",
       "9671                                   True  \n",
       "9672                                   True  \n",
       "9673                                   True  \n",
       "9674                                   True  \n",
       "\n",
       "[29043 rows x 37 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from collections import OrderedDict\n",
    "\n",
    "def annotate_overlapping_bcell_epitopes(df, bcell_epitopes, verbose=False):\n",
    "    cols = OrderedDict([\n",
    "        (\"overlapping_bcell_epitope\", []),\n",
    "        (\"overlapping_bcell_epitope_start\", []),\n",
    "        (\"overlapping_bcell_epitope_end\", []),\n",
    "        (\"overlapping_bcell_epitope_length\", []),\n",
    "        (\"fraction_of_full_bcell_epitope\", []),\n",
    "        (\"full_bcell_epitope\", []),\n",
    "        (\"bcell_epitope_distance_from_edge\", []),\n",
    "        (\"bcell_epitope_more_than_1aa_from_edge\", []),\n",
    "    ])\n",
    "\n",
    "    for (prot, start, end) in zip(df_concat.Protein, df_concat.Start, df_concat.End):\n",
    "        found_bcell_epitope = False\n",
    "        if prot == \"S\":\n",
    "            for (epitope, epitope_start, epitope_end) in zip(\n",
    "                    bcell_epitopes.accessible_subsequence,\n",
    "                    bcell_epitopes.accessible_subsequence_start,\n",
    "                    bcell_epitopes.accessible_subsequence_end):\n",
    "                if end >= epitope_start and start <= epitope_end:\n",
    "                    found_bcell_epitope = True\n",
    "                    overlap_start = max(epitope_start, start)\n",
    "                    overlap_end = min(epitope_end, end)\n",
    "                    overlap_length = overlap_end - overlap_start + 1\n",
    "                    overlapping_bcell_epitope = epitope[\n",
    "                        overlap_start - epitope_start:\n",
    "                        overlap_end - epitope_start + 1]\n",
    "                    overlap_percent_of_full_bcell_epitope = overlap_length / len(epitope)\n",
    "\n",
    "                    cols[\"overlapping_bcell_epitope\"].append(overlapping_bcell_epitope)\n",
    "                    cols[\"overlapping_bcell_epitope_start\"].append(overlap_start)\n",
    "                    cols[\"overlapping_bcell_epitope_end\"].append(overlap_end)\n",
    "                    cols[\"overlapping_bcell_epitope_length\"].append(overlap_length)\n",
    "                    cols[\"fraction_of_full_bcell_epitope\"].append(overlap_percent_of_full_bcell_epitope)\n",
    "                    cols[\"full_bcell_epitope\"].append(epitope)\n",
    "                    distance_from_start = overlap_start - epitope_start\n",
    "                    distance_from_end = epitope_end - overlap_end\n",
    "                    distance_from_edge = min(distance_from_start, distance_from_end)\n",
    "                    cols[\"bcell_epitope_distance_from_edge\"].append(distance_from_edge)\n",
    "                    cols[\"bcell_epitope_more_than_1aa_from_edge\"].append(distance_from_edge > 1)\n",
    "                    break\n",
    "        if not found_bcell_epitope:\n",
    "            cols[\"overlapping_bcell_epitope\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_start\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_end\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_length\"].append(0)\n",
    "            cols[\"fraction_of_full_bcell_epitope\"].append(0.0)\n",
    "            cols[\"full_bcell_epitope\"].append(None)\n",
    "            cols[\"bcell_epitope_distance_from_edge\"].append(None)\n",
    "            cols[\"bcell_epitope_more_than_1aa_from_edge\"].append(True)\n",
    "        \n",
    "    df_with_bcell_epitopes = df.copy()\n",
    "    for (col_name, col_values) in cols.items():\n",
    "        df_with_bcell_epitopes[col_name] = col_values\n",
    "        if verbose:\n",
    "            print(\"%s\\n\" % (df_with_bcell_epitopes[col_name].describe(),))\n",
    "    return df_with_bcell_epitopes\n",
    "\n",
    "df_with_bcell_epitopes = annotate_overlapping_bcell_epitopes(df_concat, bcell_epitopes, verbose=True);\n",
    "df_with_bcell_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
      "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
      "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
      "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
      "       'num_extremely_unstable_dipeptides',\n",
      "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
      "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
      "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
      "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
      "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df_concat.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Residue Proper Numbering</th>\n",
       "      <th>SiteSpecific Accessiblity</th>\n",
       "      <th>Nude Accessibility</th>\n",
       "      <th>Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>1142</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1116</th>\n",
       "      <td>1143</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>1144</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1118</th>\n",
       "      <td>1145</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119</th>\n",
       "      <td>1146</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1120 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Residue Proper Numbering  SiteSpecific Accessiblity  Nude Accessibility  \\\n",
       "0                           27                       0.94                0.99   \n",
       "1                           28                       0.99                1.00   \n",
       "2                           29                       0.15                0.26   \n",
       "3                           30                       0.36                0.76   \n",
       "4                           31                       0.00                0.00   \n",
       "...                        ...                        ...                 ...   \n",
       "1115                      1142                       0.78                0.89   \n",
       "1116                      1143                       0.66                0.70   \n",
       "1117                      1144                       0.78                0.79   \n",
       "1118                      1145                       0.85                0.86   \n",
       "1119                      1146                       1.00                1.00   \n",
       "\n",
       "      Difference  \n",
       "0           0.05  \n",
       "1           0.01  \n",
       "2           0.11  \n",
       "3           0.40  \n",
       "4           0.00  \n",
       "...          ...  \n",
       "1115        0.11  \n",
       "1116        0.04  \n",
       "1117        0.01  \n",
       "1118        0.00  \n",
       "1119        0.00  \n",
       "\n",
       "[1120 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_accessibility = pd.read_csv(\"../solvent-accessibility/Woods-Glycans-MD-Site-Specific-Accessibility.csv\");\n",
    "df_accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def num_contiguous_accessible_residues(protein, start, end, aa_num_to_accessibility, cutoff=0.2):\n",
    "    if protein != \"S\":\n",
    "        return 0\n",
    "    kmers = []\n",
    "    curr_kmer = 0\n",
    "    for aa_num in range(start, end + 1):\n",
    "        accessibility = aa_num_to_accessibility.get(aa_num, 0)\n",
    "        if accessibility >= cutoff:\n",
    "            curr_kmer += 1\n",
    "        elif curr_kmer > 0:\n",
    "            kmers.append(curr_kmer)\n",
    "            curr_kmer = 0\n",
    "    if curr_kmer > 0:\n",
    "        kmers.append(curr_kmer)\n",
    "    if len(kmers) == 0:\n",
    "        return 0\n",
    "    else:\n",
    "        return max(kmers)\n",
    "\n",
    "def annotate_accesibility(df, cutoff=0.2):\n",
    "    df = df.copy()\n",
    "    df_accessibility = pd.read_csv(\"../solvent-accessibility/Woods-Glycans-MD-Site-Specific-Accessibility.csv\");\n",
    "    aa_num_to_accessibility = dict(\n",
    "        zip(df_accessibility[\"Residue Proper Numbering\"],\n",
    "            df_accessibility[\"SiteSpecific Accessiblity\"]))\n",
    "    accessibility_table_start = min(aa_num_to_accessibility.keys())\n",
    "    accessibility_table_end = max(aa_num_to_accessibility.keys())\n",
    "    transmembrane_start = 1208\n",
    "    \n",
    "    df[\"in_or_after_spike_transmembrane\"] = [\n",
    "        (p == \"S\" and s >= transmembrane_start)\n",
    "        for (p, s) in zip(df.Protein, df.Start)\n",
    "    ]\n",
    "    df[\"in_spike\"] = [p == \"S\" for p in df.Protein]\n",
    "    df[\"not_in_spike\"] = ~df.in_spike\n",
    "    df[\"in_spike_accessibility_table\"] = [\n",
    "        (p == \"S\" and s >= accessibility_table_start and e <= accessibility_table_end)\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    \n",
    "    df[\"longest_contiguous_accessible_region_in_spike\"] = [\n",
    "        num_contiguous_accessible_residues(p, s, e, aa_num_to_accessibility, cutoff=cutoff)\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    \n",
    "    df[\"max_accessibility\"] = [\n",
    "        max([aa_num_to_accessibility.get(aa_num, 0) for aa_num in range(s, e + 1)])\n",
    "        if p == \"S\"\n",
    "        else 0\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    \n",
    "    df[\"known_accessibility\"] = (\n",
    "        df.not_in_spike | \n",
    "        df.in_spike_accessibility_table | \n",
    "        df.in_or_after_spike_transmembrane\n",
    "    )\n",
    "    df[\"unknown_accessibility\"] = ~df.known_accessibility\n",
    "    df[\"num_unknown_accessibility\"] = [\n",
    "        0 if (p != \"S\" or s >= transmembrane_start) else max(\n",
    "            max(0, accessibility_table_start - s),\n",
    "            max(0, e - accessibility_table_end))\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    \n",
    "    return df\n",
    "\n",
    "def annotate_selection_criteria(df, rounding_decimals=2):\n",
    "    df = df.copy()\n",
    "    df[\"manufacturability_feasibility_score\"] = 2.0 ** (\n",
    "        -df.manufacturability_difficulty_score / 10.0)\n",
    "    df[\"has mouse MHC-I b\"] = ~df[\"Mouse MHC-I b\"].isnull()\n",
    "    df[\"has mouse MHC-II b\"] = ~df[\"Mouse MHC-II b\"].isnull()\n",
    "    df[\"has mouse MHC-I d\"] = ~df[\"Mouse MHC-I d\"].isnull()\n",
    "    df[\"has mouse MHC-II d\"] = ~df[\"Mouse MHC-II d\"].isnull()\n",
    "    df[\"mouse H2-b count\"] = (\n",
    "        df[\"has mouse MHC-I b\"].astype(int) + \n",
    "        df[\"has mouse MHC-II b\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse H2-d count\"] = (\n",
    "        df[\"has mouse MHC-I d\"].astype(int) + \n",
    "        df[\"has mouse MHC-II d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse class I count\"] = (\n",
    "        df[\"has mouse MHC-I b\"].astype(int) + \n",
    "        df[\"has mouse MHC-I d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse class II count\"] = (\n",
    "        df[\"has mouse MHC-II b\"].astype(int) + \n",
    "        df[\"has mouse MHC-II d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse count\"] = df[\"mouse H2-b count\"] + df[\"mouse H2-d count\"]\n",
    "    df[\"HLA-I_pop_freq_rounded\"] = df[\"HLA-I_pop_freq\"].round(rounding_decimals)\n",
    "    df[\"HLA-II_pop_freq_rounded\"] = df[\"HLA-II_pop_freq\"].round(rounding_decimals)\n",
    "    df[\"Total_frequency_rounded\"] = df[\"Total_frequency\"].round(rounding_decimals)\n",
    "    df[\"manufacturability_feasibility_score_rounded\"] = \\\n",
    "        df[\"manufacturability_feasibility_score\"].round(rounding_decimals)\n",
    "\n",
    "    return annotate_accesibility(df)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_concat_annot = annotate_selection_criteria(df_concat);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot.known_accessibility[df_concat_annot.num_unknown_accessibility > 0].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0840534171249018"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot[df_concat_annot.Protein == \"S\"].unknown_accessibility.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
       "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
       "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
       "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
       "       'num_extremely_unstable_dipeptides',\n",
       "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
       "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
       "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
       "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
       "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length',\n",
       "       'manufacturability_feasibility_score', 'has mouse MHC-I b',\n",
       "       'has mouse MHC-II b', 'has mouse MHC-I d', 'has mouse MHC-II d',\n",
       "       'mouse H2-b count', 'mouse H2-d count', 'mouse class I count',\n",
       "       'mouse class II count', 'mouse count', 'HLA-I_pop_freq_rounded',\n",
       "       'HLA-II_pop_freq_rounded', 'Total_frequency_rounded',\n",
       "       'manufacturability_feasibility_score_rounded',\n",
       "       'in_or_after_spike_transmembrane', 'in_spike', 'not_in_spike',\n",
       "       'in_spike_accessibility_table',\n",
       "       'longest_contiguous_accessible_region_in_spike', 'max_accessibility',\n",
       "       'known_accessibility', 'unknown_accessibility',\n",
       "       'num_unknown_accessibility', 'overlapping_bcell_epitope',\n",
       "       'overlapping_bcell_epitope_start', 'overlapping_bcell_epitope_end',\n",
       "       'overlapping_bcell_epitope_length', 'fraction_of_full_bcell_epitope',\n",
       "       'full_bcell_epitope', 'bcell_epitope_distance_from_edge',\n",
       "       'bcell_epitope_more_than_1aa_from_edge'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell = annotate_overlapping_bcell_epitopes(df_concat_annot, bcell_epitopes);\n",
    "df_concat_annot_bcell.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_alleles(alleles):\n",
    "    if type(alleles) in (list, set, tuple):\n",
    "        alleles = set(alleles)\n",
    "    elif type(alleles) is str:\n",
    "        alleles = set(alleles.split(\",\"))\n",
    "    else:\n",
    "        alleles = set()\n",
    "    return alleles\n",
    "\n",
    "    \n",
    "def population_coverage(alleles, allele_frequencies):\n",
    "    alleles = parse_alleles(alleles)\n",
    "\n",
    "    if len(alleles) > 0:\n",
    "\n",
    "        individual_allele_frequencies = np.array([allele_frequencies[a] for a in alleles])\n",
    "        prob_not_allele_haploid = 1.0 - (individual_allele_frequencies / 100.0)\n",
    "        prob_not_allele_diploid = prob_not_allele_haploid ** 2\n",
    "        prob_missing_all = np.prod(prob_not_allele_diploid)\n",
    "        prob_atleast_one = 1.0 - prob_missing_all\n",
    "\n",
    "        return prob_atleast_one\n",
    "    else:\n",
    "        return 0.0 \n",
    "\n",
    "def update_marginal_coverage(\n",
    "        df, \n",
    "        allele_frequencies, \n",
    "        chosen_class1_alleles=[], \n",
    "        chosen_class2_alleles=[], \n",
    "        decimals=2,\n",
    "        prefix=\"marginal_\"):\n",
    "    \"\"\"\n",
    "    Create or update columns named 'marginal_class1_coverage' & 'marginal_class2_coverage'\n",
    "    by excluding a blacklist of class1 & class2 alleles which have already been \n",
    "    included in a partially constructed list of vaccine peptides.\n",
    "    \n",
    "    Expects allele_frequencies to be a dictionary with values between [0, 100]\n",
    "    and will update DataFrame with values in the same range. \n",
    "    \"\"\"\n",
    "    # based on which class1 and class2 alleles we've already chosen\n",
    "    # compute updated \n",
    "    df = df.copy()\n",
    "    chosen_class1_alleles = set(chosen_class1_alleles)\n",
    "    chosne_class2_alleles = set(chosen_class2_alleles)\n",
    "    class1_coverages = []\n",
    "    class2_coverages = []\n",
    "    for class1_alleles, class2_alleles in zip(df['HLA-I_haplotypes'], df['HLA-II_haplotypes']):\n",
    "        class1_alleles = parse_alleles(class1_alleles)\n",
    "        class2_alleles = parse_alleles(class2_alleles)\n",
    "        extra_class1_alleles = class1_alleles.difference(chosen_class1_alleles)\n",
    "        extra_class2_alleles = class2_alleles.difference(chosen_class2_alleles)\n",
    "        class1_coverages.append(population_coverage(extra_class1_alleles, allele_frequencies))\n",
    "        class2_coverages.append(population_coverage(extra_class2_alleles, allele_frequencies))\n",
    "    \n",
    "    class1_key = \"%sclass1_coverage\" % prefix\n",
    "    df[class1_key] = class1_coverages\n",
    "    df[class1_key + \"_rounded\"] = df[class1_key].round(decimals)\n",
    "    class2_key = \"%sclass2_coverage\" % prefix\n",
    "    df[class2_key] = class2_coverages\n",
    "    df[class2_key + \"_rounded\"] = df[class2_key].round(decimals)\n",
    "    combined_key = \"%scombined_coverage\" % prefix\n",
    "    df[combined_key] = np.array(class1_coverages) * np.array(class2_coverages)\n",
    "    df[combined_key + \"_rounded\"] = df[combined_key].round(decimals)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class\n",
      "I     164.655275\n",
      "II    139.423328\n",
      "Name: Frequency, dtype: float64\n",
      "Class\n",
      "I     17\n",
      "II    15\n",
      "Name: Frequency, dtype: int64\n",
      "Alleles in frequency table, not in vaccine peptides:\n",
      "==> HLA-DQA1*03:01/DQB1*03:02 (9.62)\n",
      "==> HLA-DQA1*01:01/DQB1*05:01 (10.85)\n",
      "==> HLA-DQA1*02:01/DQB1*02:02 (11.08)\n",
      "==> HLA-B*15:01 (4.02)\n",
      "Alleles in vaccine peptides, not in frequency table:\n"
     ]
    }
   ],
   "source": [
    "df_freq = pd.read_csv(\"HLA_freq_formated.txt\");\n",
    "df_freq[\"Class\"] = df_freq.Haplotype.map(lambda x: \"II\" if \"D\" in x else \"I\");\n",
    "print(df_freq.groupby(\"Class\")[\"Frequency\"].sum())\n",
    "print(df_freq.groupby(\"Class\")[\"Frequency\"].count())\n",
    "\n",
    "allele_to_frequency = {a: f for (a, f) in zip(df_freq[\"Haplotype\"], df_freq[\"Frequency\"])}\n",
    "unique_alleles_in_vaccine_peptides = set([])\n",
    "for class1, class2 in zip(df_concat[\"HLA-I_haplotypes\"], df_concat[\"HLA-II_haplotypes\"]):\n",
    "    if type(class1) is str:\n",
    "        unique_alleles_in_vaccine_peptides.update(class1.split(\",\"))\n",
    "    if type(class2) is str:\n",
    "        unique_alleles_in_vaccine_peptides.update(class2.split(\",\"))\n",
    "len(unique_alleles_in_vaccine_peptides), len(allele_to_frequency)\n",
    "print(\"Alleles in frequency table, not in vaccine peptides:\")\n",
    "for a, f in allele_to_frequency.items():\n",
    "    if a not in unique_alleles_in_vaccine_peptides:\n",
    "        print(\"==> %s (%0.2f)\" % (a, f))\n",
    "print(\"Alleles in vaccine peptides, not in frequency table:\")\n",
    "for a in unique_alleles_in_vaccine_peptides:\n",
    "    if a not in allele_to_frequency:\n",
    "        print(\"==> %s\" % a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Should be near zero 7.771561172376096e-16 6.661338147750939e-16\n",
      "Should be >0.4 0.41749339987056 6.661338147750939e-16\n"
     ]
    }
   ],
   "source": [
    "\n",
    "        \n",
    "df_concat_with_marginal_freqs = update_marginal_coverage(\n",
    "    df=df_concat,\n",
    "    allele_frequencies=allele_to_frequency)\n",
    "# sanity check that the population frequency calculation matches the one \n",
    "# that Christof annotated vaccine peptides with in the absence of any \n",
    "# \"chosen\" / excluded alleles.\n",
    "print(\n",
    "    \"Should be near zero\",\n",
    "    (df_concat_with_marginal_freqs[\"HLA-I_pop_freq\"] - df_concat_with_marginal_freqs[\"marginal_class1_coverage\"]).abs().max(),\n",
    "    (df_concat_with_marginal_freqs[\"HLA-II_pop_freq\"] - df_concat_with_marginal_freqs[\"marginal_class2_coverage\"]).abs().max()\n",
    ")\n",
    "df_concat_with_marginal_freqs_a0201 = update_marginal_coverage(\n",
    "    df=df_concat,\n",
    "    chosen_class1_alleles={\"HLA-A*02:01\"},\n",
    "    allele_frequencies=allele_to_frequency)\n",
    "# sanity check that the population frequency calculation matches the one \n",
    "# that Christof annotated vaccine peptides with in the absence of any \n",
    "# \"chosen\" / excluded alleles.\n",
    "print(\n",
    "    \"Should be >0.4\",\n",
    "    (df_concat_with_marginal_freqs[\"HLA-I_pop_freq\"] - df_concat_with_marginal_freqs_a0201[\"marginal_class1_coverage\"]).abs().max(),\n",
    "    (df_concat_with_marginal_freqs[\"HLA-II_pop_freq\"] - df_concat_with_marginal_freqs_a0201[\"marginal_class2_coverage\"]).abs().max()\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
       "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
       "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
       "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
       "       'num_extremely_unstable_dipeptides',\n",
       "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
       "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
       "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
       "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
       "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length',\n",
       "       'overlapping_bcell_epitope', 'overlapping_bcell_epitope_start',\n",
       "       'overlapping_bcell_epitope_end', 'overlapping_bcell_epitope_length',\n",
       "       'fraction_of_full_bcell_epitope', 'full_bcell_epitope',\n",
       "       'bcell_epitope_distance_from_edge',\n",
       "       'bcell_epitope_more_than_1aa_from_edge'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_with_bcell_epitopes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29043"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "def choose_sort_columns(class1, class2, bcell): \n",
    "    class1_sort_columns = [\n",
    "        \"marginal_class1_coverage_rounded\", \n",
    "        \"HLA-I_pop_freq_rounded\",\n",
    "        \"mouse class I count\",\n",
    "    ]\n",
    "    class2_sort_columns = [\n",
    "        \"marginal_class2_coverage_rounded\", \n",
    "        \"HLA-II_pop_freq_rounded\",\n",
    "        \"mouse class II count\",\n",
    "    ]\n",
    "    \n",
    "    if class1 and class2:\n",
    "        primary_sort_columns = [\n",
    "            \"marginal_combined_coverage_rounded\", \n",
    "            \"Total_frequency_rounded\",\n",
    "            \"mouse count\",\n",
    "        ]\n",
    "        \n",
    "        secondary_sort_columns = class1_sort_columns + class2_sort_columns\n",
    "    elif class1:\n",
    "        primary_sort_columns = class1_sort_columns\n",
    "        secondary_sort_columns = class2_sort_columns\n",
    "    elif class2:\n",
    "        primary_sort_columns = class2_sort_columns\n",
    "        secondary_sort_columns = class1_sort_columns\n",
    "    else:\n",
    "        primary_sort_columns = []\n",
    "        secondary_sort_columns = []\n",
    "        \n",
    "    if bcell:\n",
    "        primary_sort_columns = [\"fraction_of_full_bcell_epitope\", \"bcell_epitope_more_than_1aa_from_edge\"] + primary_sort_columns \n",
    "    \n",
    "    tie_breakers = [\n",
    "        \"marginal_combined_coverage\",\n",
    "        \"marginal_class1_coverage\",\n",
    "        \"marginal_class2_coverage\",\n",
    "        \"Total_frequency\",\n",
    "        \"HLA-I_pop_freq\",\n",
    "        \"HLA-II_pop_freq\",\n",
    "        \"manufacturability_feasibility_score\",\n",
    "    ]\n",
    "    if bcell:\n",
    "        tie_breakers = [\"bcell_epitope_distance_from_edge\"] + tie_breakers \n",
    "        \n",
    "    tie_breakers = tie_breakers + [\"Start\"]\n",
    "    \n",
    "    return (\n",
    "        primary_sort_columns + \n",
    "        [\"manufacturability_feasibility_score_rounded\"] + \n",
    "        secondary_sort_columns + \n",
    "        tie_breakers\n",
    "    )\n",
    "    \n",
    "    \n",
    "def apply_filters(df, class1, class2, bcell, H2b, H2d, length, exclude_accessible_spike_regions, verbose):\n",
    "    masks = []\n",
    "    \n",
    "    if H2b:\n",
    "        if class1:\n",
    "            H2b_class1 = df[\"has mouse MHC-I b\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class I H2b\" % (H2b_class1.sum(), len(df)))\n",
    "            masks.append(H2b_class1)\n",
    "        if class2:\n",
    "            H2b_class2 = df[\"has mouse MHC-II b\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class II H2b\" % (H2b_class2.sum(), len(df)))\n",
    "            masks.append(H2b_class2)\n",
    "    if H2d:\n",
    "        if class1:\n",
    "            H2d_class1 = df[\"has mouse MHC-I d\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class I H2d\" % (H2d_class1.sum(), len(df)))\n",
    "            masks.append(H2d_class1)\n",
    "        if class2:\n",
    "            H2d_class2 = df[\"has mouse MHC-II d\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class II H2d\" % (H2d_class2.sum(), len(df)))\n",
    "            masks.append(H2d_class2)\n",
    "            \n",
    "    if length:\n",
    "        length_mask = df.Sequence.str.len() == length\n",
    "        if verbose:\n",
    "            print(\"==> %d/%d peptides are %dmer\" % (length_mask.sum(), len(df), length))\n",
    "        masks.append(length_mask)\n",
    "    \n",
    "    if bcell:\n",
    "        bcell_mask = df.overlapping_bcell_epitope_length >= 4\n",
    "        if verbose:\n",
    "            print(\"==> %d/%d peptides overlap B-cell epitopes\" % (bcell_mask.sum(), len(bcell_mask)))\n",
    "        masks.append(bcell_mask)\n",
    "        \n",
    "    if exclude_accessible_spike_regions:\n",
    "        accessible_spike_mask = (\n",
    "            ((df.longest_contiguous_accessible_region_in_spike + df.num_unknown_accessibility) > 3) |\n",
    "            (df.max_accessibility > 0.6)\n",
    "        )\n",
    "        not_accessible_mask = ~accessible_spike_mask\n",
    "        if verbose:\n",
    "            print(\"%d/%d peptides do not contain accessible regions of the spike protein\" % (\n",
    "                not_accessible_mask.sum(),\n",
    "                len(not_accessible_mask)))\n",
    "        masks.append(not_accessible_mask)\n",
    "        \n",
    "    combined_mask = np.ones(len(df), dtype=bool)\n",
    "    for m in masks:\n",
    "        combined_mask = combined_mask & m.values\n",
    "    df_subset = df[combined_mask]\n",
    "    if verbose:\n",
    "        print(\"==> Keeping %d/%d peptides\" % (len(df_subset), len(df)))\n",
    "    return df_subset\n",
    "\n",
    "def select_vaccine_peptides(\n",
    "        df, \n",
    "        length=None,\n",
    "        class1=True,\n",
    "        class2=True,\n",
    "        H2b=True,\n",
    "        H2d=True,\n",
    "        bcell=False,\n",
    "        bcell_epitopes=None,\n",
    "        target_population_coverage=0.9,\n",
    "        min_per_protein={\"S\": 1, \"M\": 1, \"N\": 1},\n",
    "        max_kmer_overlap=8,\n",
    "        verbose=False):\n",
    "    \n",
    "    df = annotate_selection_criteria(df)\n",
    "    if bcell:\n",
    "        assert bcell_epitopes is not None\n",
    "        df = annotate_overlapping_bcell_epitopes(df, bcell_epitopes=bcell_epitopes)\n",
    "        exclude_accessible_spike_regions = False\n",
    "    else:\n",
    "        exclude_accessible_spike_regions = True\n",
    "        \n",
    "    sort_columns = choose_sort_columns(class1, class2, bcell)\n",
    "    df = apply_filters(\n",
    "        df=df,\n",
    "        class1=class1, \n",
    "        class2=class2, \n",
    "        bcell=bcell, \n",
    "        H2b=H2b, \n",
    "        H2d=H2d, \n",
    "        length=length,\n",
    "        exclude_accessible_spike_regions=exclude_accessible_spike_regions,\n",
    "        verbose=verbose)\n",
    "    \n",
    "    if verbose:\n",
    "        print(\"==> protein counts of candidate peptides: %s\" % (dict(df.Protein.value_counts()),))\n",
    "    \n",
    "    k = max_kmer_overlap + 1\n",
    "\n",
    "    \n",
    "    class1_population_coverage = 0.0\n",
    "    class2_population_coverage = 0.0\n",
    "    \n",
    "    selected_rows = []\n",
    "    seen_kmers = set()\n",
    "    protein_counts = Counter()\n",
    " \n",
    "    chosen_class1_alleles = set()\n",
    "    chosen_class2_alleles = set()\n",
    "    need_more_class1 = class1\n",
    "    need_more_class2 = class2\n",
    "    missing_proteins = set(min_per_protein.keys())\n",
    "    all_bcell_epitopes = set(df.full_bcell_epitope.unique())\n",
    "    seen_bcell_epitopes = set()\n",
    "    \n",
    "    if bcell:\n",
    "        missing_bcell_epitopes = all_bcell_epitopes\n",
    "    else:\n",
    "        missing_bcell_epitopes = set()\n",
    "    \n",
    "    while need_more_class1 or need_more_class2 or len(missing_proteins) > 0 or len(missing_bcell_epitopes) > 0:\n",
    "        df_with_marginal_coverage = update_marginal_coverage(\n",
    "            df=df,\n",
    "            allele_frequencies=allele_to_frequency, \n",
    "            chosen_class1_alleles=chosen_class1_alleles, \n",
    "            chosen_class2_alleles=chosen_class2_alleles)\n",
    "        \n",
    "        \n",
    "        df_sorted = df_with_marginal_coverage.sort_values(sort_columns, ascending=False)\n",
    "        \n",
    "        \n",
    "        missing_bcell_epitopes_mask = df_sorted.full_bcell_epitope.isin(missing_bcell_epitopes).values\n",
    "       \n",
    "        if need_more_class1 and need_more_class2:\n",
    "            marginal_coverage_mask = (\n",
    "                (df_sorted.marginal_class1_coverage > 0).values &\n",
    "                (df_sorted.marginal_class2_coverage > 0).values\n",
    "            )\n",
    "        elif need_more_class1:\n",
    "            marginal_coverage_mask = (df_sorted.marginal_class1_coverage > 0).values\n",
    "        elif need_more_class2:\n",
    "            marginal_coverage_mask = (df_sorted.marginal_class2_coverage > 0).values\n",
    "        else:\n",
    "            marginal_coverage_mask = np.zeros(len(df_sorted), dtype=bool)\n",
    "                    \n",
    "        \n",
    "        \n",
    "        if bcell:\n",
    "            # first try adding peptides which contribute both B-cell epitopes and \n",
    "            # T-cell coverage, then try adding just B-cell epitopes,\n",
    "            # and lastly try just adding additional coverage\n",
    "            candidate_subset_order_masks = [    \n",
    "                missing_bcell_epitopes_mask & marginal_coverage_mask,\n",
    "                missing_bcell_epitopes_mask & ~marginal_coverage_mask,\n",
    "                ~missing_bcell_epitopes_mask & marginal_coverage_mask,\n",
    "             \n",
    "            ]\n",
    "        else:\n",
    "            missing_proteins_mask = df_sorted.Protein.isin(missing_proteins)     \n",
    "            # first try adding peptides which contribute both missing proteins and \n",
    "            # T-cell allele population coverage, then try adding just additional \n",
    "            # T-cell allele coverage, and lastly try adding just missing proteins\n",
    "            candidate_subset_order_masks = [\n",
    "                missing_proteins_mask & marginal_coverage_mask,\n",
    "                ~missing_proteins_mask & marginal_coverage_mask,\n",
    "                missing_proteins_mask & ~marginal_coverage_mask,\n",
    "            ]\n",
    "            \n",
    "        # if possible avoid extremely unmanufacturable peptides\n",
    "        manufacturable_mask = (df_sorted.manufacturability_feasibility_score > 0.25).values\n",
    "        candidate_subset_order_masks = (\n",
    "            [manufacturable_mask & mask for mask in candidate_subset_order_masks] + \n",
    "            [~manufacturable_mask & mask for mask in candidate_subset_order_masks]\n",
    "        )   \n",
    "        candidate_subset_order = [df_sorted[mask] for mask in candidate_subset_order_masks]\n",
    "            \n",
    "            \n",
    "        added_peptide = False\n",
    "        # first try adding peptides from proteins we don't have enough of before \n",
    "        # looking at peptides from other proteins\n",
    "        \n",
    "        for df_candidates in candidate_subset_order:\n",
    "            \n",
    "            if added_peptide:\n",
    "                    break\n",
    "            \n",
    "            for i, row in df_candidates.iterrows():\n",
    "                if added_peptide:\n",
    "                    break\n",
    "                \n",
    "                peptide = row.Sequence\n",
    "                    \n",
    "                peptide_kmers = {\n",
    "                    peptide[peptide_start:peptide_start + k]\n",
    "                    for peptide_start in range(len(peptide) - k + 1)\n",
    "                }\n",
    "                already_seen_sequence = any([kmer in seen_kmers for kmer in peptide_kmers])\n",
    "                \n",
    "                if already_seen_sequence:\n",
    "                    continue\n",
    "    \n",
    "                # accept peptide \n",
    "                seen_kmers.update(peptide_kmers)\n",
    "                \n",
    "                protein_counts[row.Protein] += 1\n",
    "                missing_proteins = {\n",
    "                    p \n",
    "                    for p, min_count in min_per_protein.items() \n",
    "                    if protein_counts.get(p, 0) < min_count \n",
    "                }\n",
    "        \n",
    "                selected_rows.append(row)\n",
    "                \n",
    "                class1_alleles = parse_alleles(row[\"HLA-I_haplotypes\"])\n",
    "                chosen_class1_alleles.update(class1_alleles)\n",
    "\n",
    "                class2_alleles = parse_alleles(row[\"HLA-II_haplotypes\"])\n",
    "                chosen_class2_alleles.update(class2_alleles)\n",
    "                \n",
    "                seen_bcell_epitopes = {row.full_bcell_epitope for row in selected_rows}\n",
    "                missing_bcell_epitopes = missing_bcell_epitopes.difference(seen_bcell_epitopes)\n",
    "                \n",
    "                added_peptide = True\n",
    "                \n",
    "\n",
    "        class1_population_coverage = population_coverage(\n",
    "            chosen_class1_alleles,\n",
    "            allele_frequencies=allele_to_frequency)\n",
    "        class2_population_coverage = population_coverage(\n",
    "            chosen_class2_alleles,\n",
    "            allele_frequencies=allele_to_frequency)\n",
    "                \n",
    "        if added_peptide:\n",
    "            if verbose:\n",
    "                print(\"Selected %d peptides, covering %0.2f%% of Class I (%d alleles) and %0.2f%% of Class II (%d alleles)\" % (\n",
    "                    len(selected_rows),\n",
    "                    class1_population_coverage * 100,\n",
    "                    len(chosen_class1_alleles),\n",
    "                    class2_population_coverage * 100,\n",
    "                    len(chosen_class2_alleles),\n",
    "                ))\n",
    "                if bcell:\n",
    "                    print(\"-- Selected B-cell epitopes: %s (missing %s)\" % (\n",
    "                            all_bcell_epitopes,\n",
    "                            missing_bcell_epitopes if missing_bcell_epitopes else \"{}\",\n",
    "                    ))\n",
    "            need_more_class1 = class1 and (class1_population_coverage < target_population_coverage)\n",
    "            need_more_class2 = class2 and (class2_population_coverage < target_population_coverage)\n",
    "        else:\n",
    "            if verbose:\n",
    "                print(\n",
    "                    \"Stopped population coverage class1=%0.2f, class2=%0.2f (target=%0.2f)\" % (\n",
    "                        class1_population_coverage,\n",
    "                        class2_population_coverage,\n",
    "                        target_population_coverage))\n",
    "                if bcell:\n",
    "                    print(\"-- Stopped with B-cell epitopes: %s (missing %s)\" % (\n",
    "                            all_bcell_epitopes,\n",
    "                            missing_bcell_epitopes if missing_bcell_epitopes else \"{}\"\n",
    "                    ))\n",
    "            need_more_class1 = need_more_class2 = False\n",
    "            # give up on filling out protein list\n",
    "            missing_proteins = set()\n",
    "    \n",
    "    if len(selected_rows) == 0:\n",
    "        print(\"Failed to select any peptides\")\n",
    "        return None \n",
    "\n",
    "    df_result = pd.DataFrame.from_records(selected_rows).sort_values([\"Length\", \"Protein\", \"Start\"])\n",
    "    print(\"\\t%d peptides\\n\\tClass I coverage = %0.2f%% (%d alleles)\\n\\tClass II coverage %0.2f%% (%d alleles)\\n\\tB-cell epitopes: %s\\n\\tProtein counts: %s\" % (\n",
    "                len(df_result),\n",
    "                class1_population_coverage * 100,\n",
    "                len(chosen_class1_alleles),\n",
    "                class2_population_coverage * 100,\n",
    "                len(chosen_class2_alleles),\n",
    "                seen_bcell_epitopes,\n",
    "                dict(protein_counts.items())\n",
    "            ))\n",
    "    seen_bcell_epitopes = {e for e in seen_bcell_epitopes if e is not None}\n",
    "    return df_result, class1_population_coverage, class2_population_coverage, seen_bcell_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def peptide_set_latex_symbol(bcell, cd4, cd8, h2b, h2d):\n",
    "    if bcell:\n",
    "        if cd4 and cd8:\n",
    "             c = \"\\\\setlength{\\\\fboxsep}{0.5pt} \\\\boxed{\\\\circledast}\"\n",
    "        elif cd4:\n",
    "            c = \"\\\\boxcircle\"\n",
    "        elif cd8:\n",
    "            c = \"\\\\boxast\"\n",
    "        else:\n",
    "            c = \"\\\\boxempty\"\n",
    "    else:\n",
    "        if cd4 and cd8:\n",
    "            c = \"\\\\circledast\"\n",
    "        elif cd4:\n",
    "            c = \"\\\\circ\"\n",
    "        elif cd8:\n",
    "            c = \"\\\\ast\"\n",
    "        else:\n",
    "            raise ValueError(\"Expected either CD4 or CD8\")\n",
    "\n",
    "    if h2b and h2d:\n",
    "        c += \"^{bd}\"\n",
    "    elif h2b:\n",
    "        c += \"^b\"\n",
    "    elif h2d:\n",
    "        c += \"^d\"\n",
    "    return \"$ %s $\" % c\n",
    "\n",
    "def peptide_set_name(bcell, cd4, cd8, h2b, h2d):\n",
    "    if bcell:\n",
    "        peptide_set_name = \"B-Cell\"\n",
    "    else:\n",
    "        peptide_set_name = \"\"\n",
    "    if cd4:\n",
    "        if peptide_set_name:\n",
    "            peptide_set_name += \"/\"\n",
    "        peptide_set_name += \"CD4+\"\n",
    "    if cd8:\n",
    "        if peptide_set_name:\n",
    "            peptide_set_name += \"/\"\n",
    "        peptide_set_name += \"CD8+\"\n",
    "    if h2b and h2d:\n",
    "        peptide_set_name += \" (H2$^b$ and H2$^d$ ligands)\"\n",
    "    elif h2b:\n",
    "        peptide_set_name += \" (H2$^b$ ligands)\"\n",
    "    elif h2d:\n",
    "        peptide_set_name += \" (H2$^d$ ligands)\"\n",
    "    return peptide_set_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1) Length=15, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t7 peptides\n",
      "\tClass I coverage = 95.31% (14 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1, 'orf1ab': 3, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-15mer.csv\n",
      "\n",
      "2) Length=15, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=True, \n",
      "\t6 peptides\n",
      "\tClass I coverage = 91.34% (10 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 2, 'M': 1, 'N': 1, 'orf1ab': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-15mer.csv\n",
      "\n",
      "3) Length=15, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "\t8 peptides\n",
      "\tClass I coverage = 91.66% (10 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 2, 'S': 1, 'orf1ab': 3, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-15mer.csv\n",
      "\n",
      "4) Length=15, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 92.28% (11 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'orf1ab': 2, 'S': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-15mer.csv\n",
      "\n",
      "5) Length=15, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-15mer.csv\n",
      "\n",
      "6) Length=15, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 86.68% (11 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-15mer.csv\n",
      "\n",
      "7) Length=15, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-15mer.csv\n",
      "\n",
      "8) Length=15, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 86.68% (11 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-15mer.csv\n",
      "\n",
      "9) Length=15, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 95.77% (15 alleles)\n",
      "\tClass II coverage 38.45% (3 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd8-15mer.csv\n",
      "\n",
      "10) Length=15, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 94.65% (13 alleles)\n",
      "\tClass II coverage 22.63% (1 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-15mer.csv\n",
      "\n",
      "11) Length=15, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 46.49% (3 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-15mer.csv\n",
      "\n",
      "12) Length=15, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 46.49% (3 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-15mer.csv\n",
      "\n",
      "13) Length=15, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 45.76% (4 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 4}\n",
      "\tFilename: selected-bcell-cd4-cd8-15mer.csv\n",
      "\n",
      "14) Length=15, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "15) Length=15, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "Failed to select any peptides\n",
      "\n",
      "16) Length=15, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "17) Length=15, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 5}\n",
      "\tFilename: selected-bcell-cd4-15mer.csv\n",
      "\n",
      "18) Length=15, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "19) Length=15, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 0.00% (0 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'PSKP'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-15mer.csv\n",
      "\n",
      "20) Length=15, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "21) Length=15, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t6 peptides\n",
      "\tClass I coverage = 84.21% (9 alleles)\n",
      "\tClass II coverage 29.90% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 6}\n",
      "\tFilename: selected-bcell-cd8-15mer.csv\n",
      "\n",
      "22) Length=15, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 20.45% (2 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2d-15mer.csv\n",
      "\n",
      "23) Length=15, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 72.54% (6 alleles)\n",
      "\tClass II coverage 20.45% (2 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd8-h2b-15mer.csv\n",
      "\n",
      "24) Length=15, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 20.45% (2 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2b-h2d-15mer.csv\n",
      "\n",
      "25) Length=15, B-Cell=True, Class1=False, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 43.97% (3 alleles)\n",
      "\tClass II coverage 11.88% (1 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-15mer.csv\n",
      "\n",
      "26) Length=21, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 95.13% (14 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'orf1ab': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-21mer.csv\n",
      "\n",
      "27) Length=21, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.90% (13 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'orf1ab': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-21mer.csv\n",
      "\n",
      "28) Length=21, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "\t6 peptides\n",
      "\tClass I coverage = 95.13% (14 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'ORF3a': 1, 'orf1ab': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-21mer.csv\n",
      "\n",
      "29) Length=21, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.83% (12 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'orf1ab': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-21mer.csv\n",
      "\n",
      "30) Length=21, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.80% (13 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-21mer.csv\n",
      "\n",
      "31) Length=21, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.80% (13 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-21mer.csv\n",
      "\n",
      "32) Length=21, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=False, \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t5 peptides\n",
      "\tClass I coverage = 83.46% (9 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'orf1ab': 1, 'ORF3a': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-21mer.csv\n",
      "\n",
      "33) Length=21, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 93.80% (13 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-21mer.csv\n",
      "\n",
      "34) Length=21, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 95.13% (14 alleles)\n",
      "\tClass II coverage 62.15% (6 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-21mer.csv\n",
      "\n",
      "35) Length=21, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 68.92% (6 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-21mer.csv\n",
      "\n",
      "36) Length=21, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 68.92% (6 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-21mer.csv\n",
      "\n",
      "37) Length=21, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 68.92% (6 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-21mer.csv\n",
      "\n",
      "38) Length=21, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 84.21% (9 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 4}\n",
      "\tFilename: selected-bcell-cd4-cd8-21mer.csv\n",
      "\n",
      "39) Length=21, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "40) Length=21, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "Failed to select any peptides\n",
      "\n",
      "41) Length=21, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "42) Length=21, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 84.21% (9 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 4}\n",
      "\tFilename: selected-bcell-cd4-21mer.csv\n",
      "\n",
      "43) Length=21, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "44) Length=21, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 37.18% (2 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'PSKP'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-21mer.csv\n",
      "\n",
      "45) Length=21, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "46) Length=21, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 88.91% (9 alleles)\n",
      "\tClass II coverage 22.63% (1 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-cd8-21mer.csv\n",
      "\n",
      "47) Length=21, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 38.45% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2d-21mer.csv\n",
      "\n",
      "48) Length=21, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 87.21% (8 alleles)\n",
      "\tClass II coverage 22.63% (1 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'PSKP'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd8-h2b-21mer.csv\n",
      "\n",
      "49) Length=21, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 77.22% (7 alleles)\n",
      "\tClass II coverage 38.45% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2b-h2d-21mer.csv\n",
      "\n",
      "50) Length=21, B-Cell=True, Class1=False, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 78.05% (7 alleles)\n",
      "\tClass II coverage 40.72% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-21mer.csv\n",
      "\n",
      "51) Length=27, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 95.77% (15 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'orf1ab': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-27mer.csv\n",
      "\n",
      "52) Length=27, B-Cell=False, Class1=True, Class2=True, H2b=False, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 92.28% (11 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-27mer.csv\n",
      "\n",
      "53) Length=27, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 95.13% (14 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'orf1ab': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-27mer.csv\n",
      "\n",
      "54) Length=27, B-Cell=False, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 92.28% (11 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-27mer.csv\n",
      "\n",
      "55) Length=27, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 85.43% (9 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-27mer.csv\n",
      "\n",
      "56) Length=27, B-Cell=False, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.05% (12 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1, 'ORF3a': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-27mer.csv\n",
      "\n",
      "57) Length=27, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=False, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 81.46% (8 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'orf1ab': 1, 'ORF3a': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-27mer.csv\n",
      "\n",
      "58) Length=27, B-Cell=False, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "\t5 peptides\n",
      "\tClass I coverage = 93.80% (13 alleles)\n",
      "\tClass II coverage 89.72% (12 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'orf1ab': 1, 'ORF3a': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-27mer.csv\n",
      "\n",
      "59) Length=27, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 95.77% (15 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-27mer.csv\n",
      "\n",
      "60) Length=27, B-Cell=False, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 95.13% (14 alleles)\n",
      "\tClass II coverage 65.84% (6 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-27mer.csv\n",
      "\n",
      "61) Length=27, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 95.77% (15 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-27mer.csv\n",
      "\n",
      "62) Length=27, B-Cell=False, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 72.62% (7 alleles)\n",
      "\tB-cell epitopes: {None}\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-27mer.csv\n",
      "\n",
      "63) Length=27, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 88.91% (9 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-cd4-cd8-27mer.csv\n",
      "\n",
      "64) Length=27, B-Cell=True, Class1=True, Class2=True, H2b=False, H2d=True, \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to select any peptides\n",
      "\n",
      "65) Length=27, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=False, \n",
      "Failed to select any peptides\n",
      "\n",
      "66) Length=27, B-Cell=True, Class1=True, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "67) Length=27, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 88.91% (9 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-cd4-27mer.csv\n",
      "\n",
      "68) Length=27, B-Cell=True, Class1=False, Class2=True, H2b=False, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 66.19% (4 alleles)\n",
      "\tClass II coverage 39.86% (3 alleles)\n",
      "\tB-cell epitopes: {'PSKP'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd4-h2d-27mer.csv\n",
      "\n",
      "69) Length=27, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 64.77% (5 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'PSKP'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-27mer.csv\n",
      "\n",
      "70) Length=27, B-Cell=True, Class1=False, Class2=True, H2b=True, H2d=True, \n",
      "Failed to select any peptides\n",
      "\n",
      "71) Length=27, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 90.80% (10 alleles)\n",
      "\tClass II coverage 57.66% (5 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-cd8-27mer.csv\n",
      "\n",
      "72) Length=27, B-Cell=True, Class1=True, Class2=False, H2b=False, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 81.78% (8 alleles)\n",
      "\tClass II coverage 38.45% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2d-27mer.csv\n",
      "\n",
      "73) Length=27, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=False, \n",
      "\t2 peptides\n",
      "\tClass I coverage = 89.39% (9 alleles)\n",
      "\tClass II coverage 46.48% (4 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'PSKP'}\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd8-h2b-27mer.csv\n",
      "\n",
      "74) Length=27, B-Cell=True, Class1=True, Class2=False, H2b=True, H2d=True, \n",
      "\t1 peptides\n",
      "\tClass I coverage = 81.78% (8 alleles)\n",
      "\tClass II coverage 38.45% (3 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY'}\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd8-h2b-h2d-27mer.csv\n",
      "\n",
      "75) Length=27, B-Cell=True, Class1=False, Class2=False, H2b=False, H2d=False, \n",
      "\t3 peptides\n",
      "\tClass I coverage = 81.78% (8 alleles)\n",
      "\tClass II coverage 52.84% (5 alleles)\n",
      "\tB-cell epitopes: {'FRKSNLKPFERDISTEIY', 'QTLE', 'PSKP'}\n",
      "\tProtein counts: {'S': 3}\n",
      "\tFilename: selected-bcell-27mer.csv\n"
     ]
    }
   ],
   "source": [
    "from collections import OrderedDict \n",
    "\n",
    "\n",
    "i = 0\n",
    "for length in df_concat.Length.unique():\n",
    "    summary = OrderedDict([\n",
    "        (\"Symbol\", []),\n",
    "        (\"Set\", []),\n",
    "        (\"\\\\# Peptides\", []),\n",
    "        (\"HLA-I Coverage\", []),\n",
    "        (\"HLA-II Coverage\", []),\n",
    "        (\"Total Coverage\", []),\n",
    "        (\"\\\\# B-cell Epitope Regions\", []) \n",
    "    ])\n",
    "    for bcell in [False, True]:\n",
    "        for class2 in [True, False]:\n",
    "            for class1 in [True, False]:\n",
    "                if not (class1 or class2 or bcell):\n",
    "                    continue\n",
    "                for h2b in [False, True]:\n",
    "                    if h2b and not (class1 or class2):\n",
    "                        continue\n",
    "                    for h2d in [False, True]:\n",
    "                            if h2d and not (class1 or class2):\n",
    "                                continue \n",
    "                            print(\"\\n%d) Length=%d, B-Cell=%s, Class1=%s, Class2=%s, H2b=%s, H2d=%s, \" % (\n",
    "                                i + 1,\n",
    "                                length,\n",
    "                                bcell,\n",
    "                                class1, \n",
    "                                class2, \n",
    "                                h2b, \n",
    "                                h2d, \n",
    "                                ))\n",
    "                            i += 1\n",
    "                            result = \\\n",
    "                                select_vaccine_peptides(\n",
    "                                    df_with_bcell_epitopes,\n",
    "                                    class1=class1,\n",
    "                                    class2=class2,\n",
    "                                    bcell=bcell,\n",
    "                                    bcell_epitopes=bcell_epitopes,\n",
    "                                    H2b=h2b,\n",
    "                                    H2d=h2d,\n",
    "                                    length=length,\n",
    "                                    target_population_coverage=0.85 if bcell else 0.9,\n",
    "                                    max_kmer_overlap=length - 5 if bcell else 8,\n",
    "                                    verbose=False)\n",
    "                            if result is None:\n",
    "                                continue\n",
    "                            df_vaccine_peptides, class1_population_coverage, class2_population_coverage, seen_bcell_epitopes = result\n",
    "                        \n",
    "                            if bcell:\n",
    "                                prefix = \"selected-bcell\"\n",
    "                            else:\n",
    "                                prefix = \"selected-tcell\"\n",
    "                            if class2:\n",
    "                                prefix += \"-cd4\"\n",
    "                            if class1:\n",
    "                                prefix += \"-cd8\"\n",
    "                            if h2b:\n",
    "                                prefix += \"-h2b\"\n",
    "                            if h2d: \n",
    "                                prefix += \"-h2d\"\n",
    "                            prefix += \"-%dmer\" % length\n",
    "                            filename = prefix + \".csv\"\n",
    "                            print(\"\\tFilename: %s\" % filename)\n",
    "                            df_vaccine_peptides.to_csv(filename, index=False)\n",
    "                            \n",
    "                            summary[\"Symbol\"].append(peptide_set_latex_symbol(\n",
    "                                bcell=bcell,\n",
    "                                cd4=class2,\n",
    "                                cd8=class1,\n",
    "                                h2b=h2b,\n",
    "                                h2d=h2d))\n",
    "                            \n",
    "                            \n",
    "                            summary[\"Set\"].append(peptide_set_name(bcell=bcell, cd4=class2, cd8=class1, h2b=h2b, h2d=h2d))\n",
    "                            summary[\"\\\\# Peptides\"].append(len(df_vaccine_peptides))\n",
    "                            summary[\"HLA-I Coverage\"].append(class1_population_coverage)\n",
    "                            summary[\"HLA-II Coverage\"].append(class2_population_coverage)\n",
    "                            summary[\"Total Coverage\"].append(class1_population_coverage * class2_population_coverage)\n",
    "                            summary[\"\\\\# B-cell Epitope Regions\"].append(len(seen_bcell_epitopes))\n",
    "    df_summary = pd.DataFrame(summary)\n",
    "    summary_prefix = \"summary-%dmer\" % length\n",
    "    summary_csv_filename = summary_prefix + \".csv\"\n",
    "    df_summary.to_csv(summary_csv_filename)\n",
    "    summary_latex_filename = summary_prefix + \".tex\"\n",
    "    with pd.option_context(\"max_colwidth\", 1000):\n",
    "        df_summary.to_latex(\n",
    "            summary_latex_filename,\n",
    "            formatters = {\n",
    "                \"HLA-I Coverage\": lambda x: \"%0.1f\\\\%%\" % (100 * x),\n",
    "                \"HLA-II Coverage\": lambda x: \"%0.1f\\\\%%\" % (100 * x),\n",
    "                \"Total Coverage\": lambda x: \"%0.1f\\\\%%\" % (100 * x), \n",
    "            },\n",
    "            column_format=\"ll\" + \"c\" * (len(df_summary.columns) - 2),\n",
    "            escape=False,\n",
    "            index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_concat_annot_S = df_concat_annot[df_concat_annot.Protein == \"S\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.36213668499607227"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df_concat_annot_S.longest_contiguous_accessible_region_in_spike > 3).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
