{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bc_epitopes_HLA_overlap-05-06-20.txt\r\n",
      "HLA_freq_formated.txt\r\n",
      "Select vaccine peptides.ipynb\r\n",
      "Window_15mer-05-06-20.txt\r\n",
      "Window_15mer-05-07-20.csv\r\n",
      "Window_15mer-05-08-20.csv\r\n",
      "Window_21mer-05-06-20.txt\r\n",
      "Window_21mer-05-07-20.csv\r\n",
      "Window_21mer-05-08-20.csv\r\n",
      "Window_27mer-05-06-20.txt\r\n",
      "Window_27mer-05-07-20.csv\r\n",
      "Window_27mer-05-08-20.csv\r\n",
      "selected-bcell-cd4-15mer.csv\r\n",
      "selected-bcell-cd4-21mer.csv\r\n",
      "selected-bcell-cd4-27mer.csv\r\n",
      "selected-bcell-cd4-h2b-15mer.csv\r\n",
      "selected-bcell-cd4-h2b-21mer.csv\r\n",
      "selected-bcell-cd4-h2b-27mer.csv\r\n",
      "selected-bcell-cd4-h2d-15mer.csv\r\n",
      "selected-bcell-cd4-h2d-21mer.csv\r\n",
      "selected-bcell-cd4-h2d-27mer.csv\r\n",
      "selected-bcell-vaccine-15mer.csv\r\n",
      "selected-bcell-vaccine-21mer.csv\r\n",
      "selected-bcell-vaccine-27mer.csv\r\n",
      "selected-tcell-cd4-15mer.csv\r\n",
      "selected-tcell-cd4-21mer.csv\r\n",
      "selected-tcell-cd4-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-cd8-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-h2b-15mer.csv\r\n",
      "selected-tcell-cd4-h2b-21mer.csv\r\n",
      "selected-tcell-cd4-h2b-27mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd4-h2d-15mer.csv\r\n",
      "selected-tcell-cd4-h2d-21mer.csv\r\n",
      "selected-tcell-cd4-h2d-27mer.csv\r\n",
      "selected-tcell-cd8-15mer.csv\r\n",
      "selected-tcell-cd8-21mer.csv\r\n",
      "selected-tcell-cd8-27mer.csv\r\n",
      "selected-tcell-cd8-h2b-15mer.csv\r\n",
      "selected-tcell-cd8-h2b-21mer.csv\r\n",
      "selected-tcell-cd8-h2b-27mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-15mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-21mer.csv\r\n",
      "selected-tcell-cd8-h2b-h2d-27mer.csv\r\n",
      "selected-tcell-cd8-h2d-15mer.csv\r\n",
      "selected-tcell-cd8-h2d-21mer.csv\r\n",
      "selected-tcell-cd8-h2d-27mer.csv\r\n",
      "vaccine-peptides-15mer-annotated-05-07-20.csv\r\n",
      "vaccine-peptides-21mer-annotated-05-07-20.csv\r\n",
      "vaccine-peptides-27mer-annotated-05-07-20.csv\r\n",
      "vaccine-peptides-with-bcell-epitopes-and-manufacturability.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_15mer = pd.read_csv(\"Window_15mer-05-08-20.csv\")\n",
    "df_21mer = pd.read_csv(\"Window_21mer-05-08-20.csv\")\n",
    "df_27mer = pd.read_csv(\"Window_27mer-05-08-20.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "extremely_unstable_dipeptides_unc = {\n",
    "  \"DG\",\n",
    "  \"NG\",\n",
    "}\n",
    "problematic_dipeptides_unc = {\n",
    "   \"DS\", \n",
    "   \"DN\", \n",
    "   \"DD\", \n",
    "   \"NN\", \n",
    "   \"ND\", \n",
    "   \"NS\",\n",
    "}\n",
    "\n",
    "problematic_dipeptides_jpt = {\"NP\"}\n",
    "\n",
    "problematic_dipeptides = problematic_dipeptides_unc.union(problematic_dipeptides_jpt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "hydropathy_dict = {\n",
    "    \"A\": 1.8,\n",
    "    \"C\": 2.5,\n",
    "    \"D\": -3.5,\n",
    "    \"E\": -3.5,\n",
    "    \"F\": 2.8,\n",
    "    \"G\": -0.4,\n",
    "    \"H\": -3.2,\n",
    "    \"I\": 4.5,\n",
    "    \"K\": -3.9,\n",
    "    \"L\": 3.8,\n",
    "    \"M\": 1.9,\n",
    "    \"N\": -3.5,\n",
    "    \"P\": -1.6,\n",
    "    \"Q\": -3.5,\n",
    "    \"R\": -4.5,\n",
    "    \"S\": -0.8,\n",
    "    \"T\": -0.7,\n",
    "    \"V\": 4.2,\n",
    "    \"W\": -0.9,\n",
    "    \"Y\": -1.3\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gravy_score(amino_acids):\n",
    "    \"\"\"\n",
    "    Mean amino acid hydropathy averaged across residues of a peptide\n",
    "    or protein sequence.\n",
    "    \"\"\"\n",
    "    total = 0 \n",
    "    for aa in amino_acids:\n",
    "        total += hydropathy_dict[aa]\n",
    "    return total / len(amino_acids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def max_kmer_gravy_score(amino_acids, k):\n",
    "    \"\"\"\n",
    "    Returns max GRAVY score of any kmer in the amino acid sequence,\n",
    "    used to determine if there are any extremely hydrophobic regions within a\n",
    "    longer amino acid sequence.\n",
    "    \"\"\"\n",
    "    return max(\n",
    "        gravy_score(amino_acids[i:i + k])\n",
    "        for i in range(len(amino_acids) - k + 1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def max_short_subsequence_gravy_score(amino_acids):\n",
    "    gravy_scores = [\n",
    "        max_kmer_gravy_score(amino_acids, k)\n",
    "        for k in [5, 6, 7, 8]\n",
    "    ]\n",
    "    return max(gravy_scores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def difficult_c_terminal(amino_acids):\n",
    "    return amino_acids[-1] in {\"P\", \"C\", \"H\"}\n",
    "\n",
    "def difficult_n_terminal(amino_acids):\n",
    "    \"\"\"\n",
    "    Is the N-terminus one of {Gln, Glu, Cys}?\n",
    "    ---\n",
    "    Priority I: avoid N-terminal Gln, Glu, Cys\n",
    "    Priority II: Asparagine at the N-terminus of a peptide is also hard\n",
    "    to synthesize, though not as bad as {Gln, Glu, Cys}\n",
    "    \"\"\"\n",
    "    return amino_acids[0] in {\"Q\", \"E\", \"C\", \"N\"}\n",
    "\n",
    "\n",
    "def cysteine_count(amino_acids):\n",
    "    \"\"\"\n",
    "    How many cysteines are in the amino acid sequence?\n",
    "    Problem with cysteine residues: They can form disulfide bonds across\n",
    "    distant parts of the peptide\n",
    "    \"\"\"\n",
    "    return amino_acids.count(\"C\")\n",
    "\n",
    "def methionine_count(amino_acids):\n",
    "    \"\"\"\n",
    "    How many methionies?\n",
    "    \"\"\"\n",
    "    return amino_acids.count(\"M\")\n",
    "\n",
    "def thiol_count(amino_acids):\n",
    "    return cysteine_count(amino_acids) + methionine_count(amino_acids)\n",
    "\n",
    "def num_extremely_unstable_dipeptides(amino_acids):\n",
    "    total = 0\n",
    "    for dipeptide in extremely_unstable_dipeptides_unc:\n",
    "        total += amino_acids.count(dipeptide)\n",
    "    return total\n",
    "    extremely_unstable_dipeptides_unc\n",
    "\n",
    "def num_moderately_unstable_dipeptides(amino_acids):\n",
    "    total = 0\n",
    "    for dipeptide in problematic_dipeptides_unc.union(problematic_dipeptides_jpt):\n",
    "        total += amino_acids.count(dipeptide)\n",
    "    return total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_manufacturability(df):\n",
    "    df = df.copy()\n",
    "    df[\"num_extremely_unstable_dipeptides\"] = df.Sequence.map(num_extremely_unstable_dipeptides)\n",
    "    df[\"num_moderately_unstable_dipeptides\"] = df.Sequence.map(num_moderately_unstable_dipeptides)\n",
    "    df[\"num_cysteines\"] = df.Sequence.map(cysteine_count)\n",
    "    df[\"num_methionines\"] = df.Sequence.map(methionine_count)\n",
    "    df[\"num_thiols\"] = df.Sequence.map(thiol_count)\n",
    "    df[\"difficult_c_terminal\"] = df.Sequence.map(difficult_c_terminal)\n",
    "    df[\"difficult_n_terminal\"] = df.Sequence.map(difficult_n_terminal)\n",
    "    df[\"gravy\"] = df.Sequence.map(gravy_score)\n",
    "    df[\"max_kmer_gravy\"] = df.Sequence.map(max_short_subsequence_gravy_score)\n",
    "    df[\"extreme_local_hydrophobicity\"] = df.max_kmer_gravy > 2.5\n",
    "    df[\"difficult_local_hydrophobicity\"] = df.max_kmer_gravy > 1.5\n",
    "    df[\"too_many_cysteines\"] = df.num_cysteines > 1\n",
    "    df[\"manufacturability_difficulty_score\"] = (\n",
    "        1.0  * (df.gravy > 2.0) +        # hydrophobicity of whole peptide\n",
    "        1.0 * df.difficult_n_terminal +  # lowest priority: N-terminal residues\n",
    "        2.0 * df.difficult_c_terminal + \n",
    "        2.0 * df.num_thiols + \n",
    "        2.0 * df.difficult_local_hydrophobicity + \n",
    "        3.0 * df.num_moderately_unstable_dipeptides + \n",
    "        5.0 * df.too_many_cysteines +          # disulfide bonds\n",
    "        10.0 * df.extreme_local_hydrophobicity + \n",
    "        10.0 * df.num_extremely_unstable_dipeptides \n",
    "    )\n",
    "    return df\n",
    "\n",
    "df_27mer_annot = annotate_manufacturability(df_27mer);\n",
    "df_27mer_annot.to_csv(\"vaccine-peptides-27mer-annotated-05-07-20.csv\", index=False)\n",
    "df_21mer_annot = annotate_manufacturability(df_21mer);\n",
    "df_21mer_annot.to_csv(\"vaccine-peptides-21mer-annotated-05-07-20.csv\", index=False)\n",
    "df_15mer_annot = annotate_manufacturability(df_15mer)\n",
    "df_15mer_annot.to_csv(\"vaccine-peptides-15mer-annotated-05-07-20.csv\", index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_concat = pd.concat([\n",
    "    df_15mer_annot,\n",
    "    df_21mer_annot,\n",
    "    df_27mer_annot,\n",
    "])\n",
    "df_concat[\"Length\"] = df_concat.Sequence.str.len()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1109aec50>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAs4AAAHgCAYAAAC1ouv3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOzdd3jc93Xn+/eZwaASYCdFilU9jmQrtixblu1ELombLFv2JvbaG9nJk+z1pjg38W7izd67m7UTx95UO7lWo9VlFapLtoqpYomiCilKpEiKBAvABpAESKID0879Y2YgiAKJGXJ+88P88Hk9zzyYPocFmA/OnN/3a+6OiIiIiIicWCzsAkREREREqoGCs4iIiIhIERScRURERESKoOAsIiIiIlIEBWcRERERkSIoOIuIiIiIFKEm7AKKNWfOHF+2bFnYZYiIiIhIhK1bt67L3eeOd1vVBOdly5axdu3asMsQERERkQgzs/bj3aZRDRERERGRIig4i4iIiIgUQcFZRERERKQIgc84m1kb0AdkgLS7X2Rms4A7gWVAG/Db7n4k6FpERERERE5WpTrOl7n7he5+Uf7yXwGr3P1sYFX+soiIiIjIpBXWqMYVwE358zcBnwupDhERERGRolQiODvwuJmtM7M/zF8339078uc7gfkVqENERERE5KRVYh3nD7r7PjObBzxhZm+MvdHd3cx8vAfmg/YfAixZsiT4SkVEREREjiPwjrO778t/PQjcB1wMHDCzBQD5rweP89hr3f0id79o7txxN3AREREREamIQIOzmTWZWXPhPPCbwOvAg8BV+btdBTwQZB0iIiIiIqcq6FGN+cB9ZlZ4rdvd/VEzexm4y8x+H2gHfjvgOkRERERETkmgwdnddwLvGuf6buCjQb62iIiIiEg5aedAEREREZEiKDiLiIiIiBRBwVlEREREpAgKziIiIiIiRVBwFhEREREpgoKziIiISBm1t7fz0EMPhV2GBKASW26LiIiITBl/8zd/w/bt27n88svDLkXKTB1nERERkTLavn172CVIQBScRURERALg7mGXIGWm4CwiIiIiUgQFZxEREZEAZLPZsEuQMlNwFhEREQmARjWiR8FZRERERKQICs4iIiIiAVDHOXoUnEVEREQCoBnn6FFwFhEREREpgoKziIiISADUcY4eBWcRERGRAGjGOXoUnEVEREQCoI5z9Cg4i4iIiARAHefoUXAWERERCYA6ztGj4CwiIiISAHWco0fBWURERCQAmUwm7BKkzBScRURERAKgUY3oUXAWERERCYCCc/QoOIuIiIgEQME5ehScRURERAKg4Bw9Cs4iIiIiAVBwjh4FZxEREZEAaFWN6FFwFhEREQmAgnP0KDiLiIiIBEDBOXoUnEVEREQCoBnn6FFwFhEREQmAOs7Ro+AsIiIiEgAF5+hRcBYREREJQDqdDrsEKTMFZxEREZEAqOMcPQrOIiIiIgFQcI4eBWcRERGRAGhUI3oUnEVEImrLli1s2LAh7DJEpiwF5+ipCbsAEREJxje+8Q2y2Sy//OUvwy5FZErSqEb0qOMsIhJR2nxBJFzqOEePgrOIiIhIABSco0fBWURERCQACs7Ro+AsIiIiEgAF5+hRcBYREREJgIJz9Cg4i4iIiJTJ2INyFZyjR8FZREREpEzGhmUF5+hRcBYREREpEwXnaFNwFhERESmTsWE5lUqFWIkEQcFZREREpEzGhmV1nKNHwVlERESkTDSqEW0KziIiIiJlouAcbQrOIiIiImWiUY1oU3AWERERKZNMJjN6XsE5ehScRURERMpkbMdZq2pEj4KziIiISJloxjnaFJxFREREymTsqMbY8xINCs4iIiIiZTK2y6zgHD0KziIiIiJlooMDo03BWURERKRMCsHZY042mw25Gik3BWcRERGRMhkNyzGNakSRgrOIiIhImYyG5TikMxrViBoFZxGRiHP3sEsQmTJGO84G2YxGNaJGwVlEJOI0ZylSOaO/qMb0S2sUKTiLiEScjuwXqZy3dJxdv7RGjYKziEjEqeMsUjmjXWbT914UKTiLiETQ2I+IdWS/SOW8peOs4Bw5FQnOZhY3s/Vm9nD+8nIze9HMtpvZnWZWW4k6RESmCm37KxIyC7sACUKlOs7fBLaMufx94J/d/SzgCPD7FapDRGRK0O5lIuFzdHBg1AQenM1sEfBp4Pr8ZQM+AqzM3+Um4HNB1yEiMpWo4ywiUn6V6Dj/C/DfgMKgz2zgqLsXWiB7gdMrUIeIyJQxtsusjrNISNRwjpxAg7OZfQY46O7rTvLxf2hma81s7aFDh8pcnYhIdI0Ny+o4i4REc86RE3TH+VLgs2bWBtxBbkTjX4EZZlaTv88iYN94D3b3a939Ine/aO7cuQGXKiISHRrVEAlHbiIVcDAl58gJNDi7+7fdfZG7LwO+BDzp7l8BngK+mL/bVcADQdYhIjLVaFRDJBxvCc6m4Bw1Ya3j/JfAn5vZdnIzzytCqkNEJJIUnEXCoeAcbTUT36U83P1p4On8+Z3AxZV6bRGRqUbL0YmEY2xwjsfj4RYjZaedA0VEIkgdZ5FwxGL5aOUQM8WsqNG/qIhIBGlVDZFwvGVUI6ZRjahRcBYRiSB1nEXCMdpxzqrjHEX6FxURiSDNOIuEY+yohmaco0fBWUQkgjSqIRKOt3ScY4pZUaN/URGRCFLHWSQcWo4u2hScRUQiSDPOIuEYO6qh4Bw9Cs4iIhGkLbdFwlEIy+amUY0I0r+oiEgEKTiLhGNsWFZwjh79i4qIRJAODhQJx9jxDI1qRI+Cs4hIBOngQJFwjO0yKzhHj4KzBO6HP/whzz77bNhliEwp2Wx23PMiUjka1Yge/YtK4FauXMlf//Vfh12GyJSiGWeRcKjLHG0KziIiETQ2LKvjLFI5mnGONgVnEZEIcvfR8wrOIpWj4BxtCs4iIhE0NiyPDdEiInLyFJwlUOp0iYRDHWcRkfJTcJZA6Q1bJBxaVUNEpPwUnCVQOppfJBwazxAJn74Po0fBWQKljRdEwqc3b5HK0fEF0abgLIFSx1lERKYSheVoU3CWQKnjLCIiU4kOzI02BWcJlIKziIhMJTowN9oUnCVQCs4iIjKVaNfOaFNwlkBpxllERKaSQlj2mOs9MIIUnCVQqVQq7BJEREQqZrTLHFPzKIoUnCVQGtUQCZ+ZhV2CyJQxGpbjkEqreRQ1Cs4SKAVnERGZSkY/aa3Re2AUKThLoPQxlYiITCWjYVkd50hScJZAKTiLiMhUMhqcayCdUsc5ahScJVAKziLh005mIpUzdlQjmUyGW4yUnYKzBErBWUREppLRsJzQylJRpOAsgVJwFgnH2C6zOs4ilVMIzp5wdZwjSMFZAqXgLBKOsTuWKTiLVI46ztGm4CyB0hu2SDjGBmdt+ytSOSMjI7l0VQOpZErvgxGj4CyB0hu2SDjGvlnrkx+RyhkZGSFWE4N47vtQazlHi4KzBEq/aYuEoxCWa2L6PhSppJGREagB4rnLw8PDodYj5aXgLIFSx1kkHIXgnIibOs4iFTQyMpILzQrOkaTgLIFSp0skHJlMhphB3DSqIVJJw8PDeNxzXWcUnKNGwVkqRiFapHLS6TTxfHDWjKVI5QwNDeHDDnvevCzRoeAsgdJasiLhSKfT1MSNeEzBWaSSBocG8Yxjgwao4xw1Cs4SKIVlkXCkUilqDGpirrVkRSpocHAQjNypcFkiQ8FZKkYhWqRy0uk0NTGo0aiGSEUNDA68JThrVCNaFJwlUBrVEAlHKpUiHnNqTNv+ilTS4MBbO84DAwOh1iPlpeAsgVJwFglHMpkkEYNELKtRDZEKcfdch1mjGpFVUnA2swYzOzeoYiR6tO2vSDhSqRQJc2osq46zSIUkk0mymayCc4QVHZzN7HLgVeDR/OULzezBoAqTaFDHWSQcqVSKmliWRAxSKQVnkUoYHcuIAQZWYxrViJhSOs7/C7gYOArg7q8CywOoSSJEHWeRcCSTSRLmJGJOSh1nkYoYDcn5brPVKjhHTSnBOeXuPcdcpxainNDYHcu0e5lI5SSTIyRiTk0svwWwiASuv78/dyYfnEno4MCoKSU4bzKz/wjEzexsM/sR8HxAdUlEjA3L6jiLVE5yJBecEzGtqiFSKcd2nLPx7JthWiKhlOD8J8CvAiPA7UAP8GdBFCXRoY6zSDiSySQ1MUjEUHAWqZDRkJxPV55w+vr7witIyq6mmDuZWRz43+7+LeCvgy1JomRsWNYmDCKVk1uOzrVzoEgFHTuq4bVOX5+Cc5QU1XF29wzwwYBrkQgaG5YVnEUqJ53fcjuhnQNFKmY0JI+ZcVZwjpaiOs556/PLz90NjE66u/u9Za9KIkOjGiLhSKVSJOqhJoY6ziIV0t/f/5Y1nEnkdhJ0d8zsRA+VKlFKcK4HuoGPjLnOAQVnOS4FZ5FwpNJpWnviuXWc0xm9cYtUQH9/P7FEjIzl3+9qc5/4JJNJ6urqwi1OyqLo4OzuXw+yEIkmjWqIhCOdTtNvMWKWWzU0k8lQU1NKr0REStXf3w+1Y65I5L709fUpOEdEKTsHLjKz+8zsYP50j5ktCrI4qX7qOIuEI53JYPjoJ8b6xVUkeP39/XhizBYXiTevl2goZTm6G4AHgYX500P560SOS8FZpPLcnUwmixkUpjM05ywSvL6+PrI1b+5Z4LU+er1EQynBea673+Du6fzpRmBuQHVJRGg5OpHKG/t9Z+NcJyLB6O3rHe0yA6PntXtgdJQSnLvN7KtmFs+fvkruYEGR49LOgSKVV/i+G3sooL7/RILXP9A/2mUGNKoRQaUE598DfhvoBDqALwI6YFBOaOybtbuf4J4iUi5jv+9snOtEJBgD/QPjdpwVnKOjlFU12oHPBliLRNDYsKw3bpHKKHyv2Zj1ZDWqIRKsdDrNyPDIW1fVyJ/XjHN0lLKqxk1mNmPM5Zlm9pNgyhIRkZM19hdWrdwsUhmjc8xjO84xsJip4xwhpYxqvNPdjxYuuPsR4NfKX5KIiJyK8cai9ImPSLBGw/HYjrOB1Sk4R0kpwTlmZjMLF8xsFqXtPChTnGacRSpjvO81ff+JBKswjvGWdZwBEppxjpJSgu8/AmvM7G5yn/59EfjbQKqSSNJ2vyKVUQjJ+o4TqZzRcJx46/XZmqxmnCOklIMDbzaztcBH8ldd6e6bgylLREROljrOIpU3Go5r33q917qCc4SUcnDgmcAOd/834HXgY2MPFjzOY+rN7CUze83MNpnZ3+SvX25mL5rZdjO708xqT/Q8Ur3GdpnVcRYRkajq7e3NnRknOPf09lS+IAlEKTPO9wAZMzsLuAZYDNw+wWNGgI+4+7uAC4FPmNn7ge8D/+zuZwFHgN8vuXKpCrFYbNzzIlJZ6jiLBOt4HWdqtRxdlJSSZLLungauBP7N3f8rsOBED/CcwkR8In9ycuMeK/PX3wR8rqSqpWqo4yxSeeMtR6fgLBKsvr4+LG4QP+aGBAwODGot9YgoJTinzOzLwO8CD+evS5zg/gDkt+d+FTgIPAHsAI7mQzjAXuD0EuqQKqLgLFJ54y09p+AsEqyenh6sbpz3ubrc95+6ztFQSnD+OnAJ8LfuvsvMlgO3TPQgd8+4+4XAIuBi4LxiX9DM/tDM1prZ2kOHDpVQqkwWCs4ilTfeqhpax1kkWEePHsXrxvkFte7N26X6FR2c3X2zu/+pu/80f3mXu3+/cLuZ3TPB448CT5EL3zPMrLCixyJg33Eec627X+TuF82dO7fYUkVEprTRkGz5bbdRcBYJ2uEjh8nWjvNpTz5MKzhHQzmP1jrj2CvMbG5h5Q0zawA+DmwhF6C/mL/bVcADZaxDJil1nEUqozBLqY6zSOUcPnxYHecpoJw7/403QLcAuMnM4uRC+l3u/rCZbQbuMLPvAuuBFWWsQyYpzViKVEY6nR49b+NcJyLl5e4cOXIElo5zY33uy+HDhytakwQj0C2z3X0D8GvjXL+T3LyzRJzCskjlje04W76noeAsEpz+/n5SyRQ0jHNjHWDQ1dVV6bIkAOUc1dDn8PI2Y4OzQrRIZaRSKSA331yYkFJwFglOd3d37sx4wdkg1hhTcI6IUnYOvNzMTnT/vyxDPRIxY8OyZixFKmNkZASAGD7a0ShcJyLlVwjF3jB+gyhbl0Wrg0VDKR3n3wFazewHZva2JeXc/fHylSVRMXbBdwVnkcpIJpNArtscs7deJyLl19nZmTvTOP7t3uh0HuisXEESmFKWo/squXnlHcCNZrYmv85yc2DVSdUbG5YVnEUqo9BdHjuqMTw8HGJFItHW2dmZG1gdb1SDXHA+cOCA3gcjoKQZZ3fvJbdV9h3kVsz4PPCKmf1JALVJBIz9IaHtRkUqY3BwEMjt/Fv4IT80NBRaPSJR19HRQawpdvxU1QTpVFora0RAKTPOV5jZfcDT5LbavtjdPwm8C/iLYMqTajc2LCs4i1RGISTHDGKWm7kshGkRKb+Ojg4yDcd/j/MmH72fVLdSOs5XAv/s7he4+/9x94MA7j4I/H4g1UnVU3AWqbxCSI6NmXFWcBYJzp69e0bD8bim5b7s37+/MgVJYEoJzp3u/suxV5jZ9wHcfVVZq5LIGBuWtRyWSGX09/cDb3abGxLGwMBAmCWJRFZ/fz89R3vgREd8NQEGu3fvrlRZEpBSgvPHx7nuk+UqRKIpnU6DxQF1nEUqpa+vj8bEm0vrNyVy14lI+e3duxcAbz5BxzkGsWmx0ftK9Zpw50Az+wbwX4AzzWzDmJuagdVBFSbRkMlk8HgcS2cUnEUqpK+vj2mJN9/Em2oy9Pb2hliRSHSNdpEnWGMsMy1DW3tb4PVIsIrZcvt24OfA94C/GnN9n7vr8FA5oUwmM9px1qiGSGX09vbSWPPmijZN8Qy9PT0hViQSXe3t7bml6JpOfD9vdvbs2kMmkyEej1ekNim/YkY14kAv8EdA35gTZjYruNIkCtLpNB7L/YDQ+pUilXH0yGGaE29+wtNc6/QcPRJiRSLRtWvXLqzZcmnpRFpyS9LpAMHqVkzHeR1Q+MzPjrnNgTPKWpFESq7jXPPmeREJ3JEjhzk7kaUrk+uNtCSybDpyNOSqRKJp566dZJsnbgx5Sy5KtbW1sXjx4qDLkoBM2HF29+Xufkb+tPyYk0KznFA6nRntOGtUQyR47s7Roz201L4549xS6/QPDJJKpUKsTCR6RkZG6NjfgU8/wYGBBS25L7t27Qq2KAlUMR1nAMzsw+Ndf+wSdSJjpTNp0KiGSMUMDQ0xkkzRUpul8Nlx7jwcPnyY+fPnh1idSLS0t7fj7qOh+IQSuZU1FJyrW9HBGfivY87XAxeTG+P4SFkrkkjJpDOjwVmjGiLB6+rqAmBG3ZsdsJn5893d3QrOImW0c+dOgOI6zkCmOUPr9tYgS5KAFR2c3f3ysZfNbDHwL2WvSCIlnU7jWlVDpGIOHToEwKy6Nz/hmZk/XwjVIlIeO3fuxOI2ujPgRHy6s7d1L8lkktra2mCLk0CUsgHKsfYCv1KuQiSaUuk0xBOAgrNIJXR3dwMwY5zgXAjVIlIeO3bsyI1pFJumpkM2k9UOglWslBnnH/Hm6hox4ELglSCKkujI7RxYC2YKziIVcPDgQeCtHeeWWiceU3AWKbfW7a1kWoofQ/QZuRi1Y8cOzjrrrKDKkgCVMuO8dsz5NPBTd9fOgXJCuVGNGBaLKziLVEBnZyfNtUb9mJ/uMYPZDXDgwIHwChOJmCNHjnD0yFFYUsKDpoHFLdeplqpUSnBeCQy7ewbAzOJm1ujug8GUJlGQTCahPg6xmtx5EQnUgQMHmF3/9g7YnNoUBzo7Q6hIJJoK4bfYAwMBiOXWc96+fXtAVUnQSplxXgU0jLncAPyivOVI1CSTIxCrgVickZGRsMsRibzOjv3MqXv7pzuz67N0dGjHMpFyGQ2/M0p7XHZ6lm2t28pfkFREKcG53t37Cxfy5xvLX5JEhbuTSiZzG6DEahScRQKWzWbp6OhkXuPb10yf35Cl+/ARfR+KlMmOHTuINcagrsQHTofenl4OHz4cSF0SrFKC84CZvbtwwczeAwyVvySJimQymdv0JJbAYzUMDem/i0iQurq6SKZSzG94+6jGvMbcdfv3q+ssUg7bWreVdGBgQeEAQY1rVKdSgvOfAXeb2bNm9hxwJ/DHwZQlUTA4mBt/93gtmViC/v7+CR4hIqdi3759QK67fKzCdXv37q1oTSJRlEql2N2+u7T55oL8aIcOEKxOpWyA8rKZnQecm79qq7ungilLomBgYCB3Jp7A4wn6C5dFJBCFUDx/nFGN0xoVnEXKpb29PbcbbonzzQDUQqwppo5zlZowOJvZR9z9STO78pibzjEz3P3egGqTKtfb2wuA19RBvJaent6QKxKJtvb2dhLx3IGAx2pKONPrTBsviJRBIfQWxi5KlWnJ6ADBKlVMx/nDwJPA5ePc5oCCs4yrp6cHAE/U44kGeo/sC7kikWhrb29nQWOWmI1/+4KGFO1tbRWtSSSKtm/fXtJW28fyGc6ebXsYGRmhrq7UowslTMUE5yP5ryvc/bkgi5FoOXr0KABeU4/X1DMyMqwfEiIB2t22i6WNx99oaGFThpfbd+HumB0nXYvIhFq3t+bmm0s5UmwMn+5kM1na2to499xzJ36ATBrF/JN/Pf/1h0EWItHT3d0NgCca8NrGt1wnIuU1ODhIx4GDnN50/KP8T2/K0Nc/qO9DkVPg7rS2tpKd/vaRqKLlZ6M151x9ignOW8ysFTjPzDaMOW00sw1BFyjV69ChQ1iiHuIJsvngfOjQoZCrEommtvwIxuJpxw/Ohdt27dpViZJEIungwYP09/Wf3IGBBdPAEqbgXIUmHNVw9y+b2WnAY8Bngy9JouLQoUOjnWavbRq9TkTKb+fOncCJg/Oi/G07d+7kve99b0XqEomabdtyB/X5zJM7MBAAy41rvLH1jTJVJZVSzKoaq9z9o2b2mLu3V6IoiYaOzk7SiVxg9trcERSdnZ1hliQSWTt37qQubswdZw3ngpba3MoaWj9W5OS1traCAdNP7XmyM7Js376dbDZLLHaSw9JSccX8Sy0wsw8Al5vZr5nZu8eegi5QqpO7s3/ffryuOXdFPIHVNtDR0RFuYSIRtb21lUXT0sddUaNgSVOS7VoGS+Sktba2Yi1Wwk4YxzEDRoZHtLZ6lSnmn/3/Bf4fYBHwT8fc5sBHyl2UVL/e3l6GhgbJzm0evS5T26ztfkUC4O5s397KxTOOv6JGwdLmDI+1t5NOp6mpOdV3fpGpZ8sbW8hML32r7WMVRj22bt3KkiVLTvn5pDIm7Di7+0p3/yTwA3e/7JiTQrOMq7D1r9e1jF6XrWtmz16t5SxSbgcPHqR/YJAlzRMH5yXNaVLpDO3tmrwTKVVXVxeHuw/DrDI8WQtY3Ni6dWsZnkwqZcLgnN9mG+CRY8c0NKohx1P46CnbMCY417dw6OABkslkWGWJRFJraysAS05wYGDB0vx9Co8RkeIVQu4pHRhYEMtthLJ5y+ZTfy6pmGI+p/sL4A+AfxznNo1qyLj27t0LZm/tONdPz80+79/PsmXLwitOJGK2b9+OUVxwPq0xSyKu9WNFTsbWrVtzBwaeylJ0Y2RnZtm2bRuZTIZ4PF6eJ5VAFbMc3R/kv14WfDkSFXv27MHqmyH25g8Cr58+epuCs0j5tLa2Mr8J6otohcRjsLgpw/bt6jiLlGrTpk3Y9DIcGFgwC5Lbk7S1tXHmmWeW6UklSMUsR3fliW5393vLV45ERVt7O6naFmrb1wCQXHoJ2THBWUTKZ/u2rSxpKn4Easm0NOu2bdPW2yIlyGazbNq8icz8Uz8wsMBn50Y+Nm3apOBcJYpZju7y/On3gRXAV/Kn64HfC640qVbuzt49e/H66cQGuokN5Lf3ranF6hrZvXt3uAWKRMjAwAAdBw6ytLn4N/MlzRn6+gfo6uoKsDKRaNmzZw+DA4PlOTCwoAlidTE2b9acc7UoZlTj6wBm9jjwDnfvyF9eANwYaHVSlbq6uhgZGSZbP53YYPdbbsvUtig4i5RRMTsGHqswC71jxw7mzp0bSF0iUVMIt4UucVkYZGZl2LBxQ/meUwJVylY1iwuhOe8AoIUH5W3GW1GjIFPfwu49WuxdpFwKwbmYAwMLFo8JziJSnA0bNmB1Bs0T37cUPjv3KW1PT095n1gCUUpwXmVmj5nZ18zsa8AjwC+CKUuq2egazvVv34/U66fT23OUgYGBSpclEkk7d+6kocaYXX/8rbaP1ZRwZjW8GbpFZGKvvvYq2VnZ3KoaZeRzch3sjRs3lveJJRBFB2d3/2PgauBd+dO17v4nQRUm1Wvv3r0Qi+O1TW+7LVvf8uZ9ROSUtbW1sbApTanH+J3ekKK9bVcwRYlEzJEjR9i3d99oyC2rWWAxY8MGjWtUg5IWVHH3+4D7xrvNzNa4+yVlqUqqWkdHR24pOnv772Ve1zx6n3PPPbfSpYlETnvbTn61ceIdA4+1sCnDM7t3k81micVK+fBRZOophNpAgnMcfJbz6muvlv+5pezK+dOyvozPJVVs3779pBJv7zZDbtttyAVnETk1fX19HD7Sw+lNpS+PtbApw/BIkoMHDwZQmUi0vPrqq1iNlXdFjTGyc7Js27qNwcHBYF5AyqacwTmAX8OkGnV0dIx2lt+mpg6rqVNwFimDwproCxqLn28uKDxG66qLTGzturVkZ2fLm5rG8HlONpvVuEYV0OdzUlYjIyP09fWOO99c4HVNHDp0qIJViUTT/v37AZjXWHrHeX7+MYXnEJHxHTlyhPa2dnxegP3B2bk55/Xr1wf3GlIW5QzO2n5KRjdUOFFwTtc0clDBWeSUFVawmddQesd5Zp2TiL35HCIyvkKY9bkBBuea3LJ0615ZF9xrSFkUHZzN7E/MbOYJ7vKfylCPVLlCJzl7oo5zbZPmKkXKoKOjg5n1Rl289MfGDOY2uoKzyATWrVuH1RqcKAGVQXZeltZtrfT29gb7QnJKSuk4zwdeNrO7zOwTZm9d/MjdXy9vaVKNDh8+DIAnGo97H69tpLenh0ym9I+XRblZTz4AACAASURBVORNHR0dzKkrfUWNgjl1aQ506ngDkeNxd1548QWyc4Obbx59rfmOu7NunbrOk1kp6zj/D+BsYAXwNaDVzP7OzM4MqDapQkeOHAHAE8dfZMVr6nF3+vr6KlWWSCQd6NzP7PqT/wV0Tn2WAwc6y1iRSLTs3buXQwcP4fMrsP7BTLBa4+WXXw7+teSklfT7k7s70Jk/pcl9cLHSzH4QQG1ShY4cOQJmUFN33Pt4ogF4szstIqXLZrMcOtTFnBJ2DDzW7PosPb39DA0NlbEykeh46aWXACoTnGOQnZvlhRdfIBe3ZDIqZcb5m2a2DvgBsBq4wN2/AbwH+EJA9UmVOXr0KJaoH3fzkwKvyXWje3p6KlWWSOQcPnyYVDpT0lbbxyo8VscciIzvhRdewJoNplXm9fw0p+tQF21tbZV5QSlZKR3nWcCV7v5b7n63u6cA3D0LfCaQ6qTq9Pb2jgbj4/FErhut4Cxy8gphd5aCs0gghoeHeeWVV8icVrnjcXxBrtO8Zs2air2mlKaU4HyGu7ePvcLMbgFw9y1lrUqqVk9PD9l47YnvFM8Fax05LHLyCmF3dt0pBOf8Yw8cOFCWmkSiZP369aRSqdEwWxENYDOM59c8X7nXlJKUEpx/dewFM4uTG9MQGXX0aA/Z+PHnm+HNjrOCs8jJGw3Op9BxnlmfxVDHWWQ8a9asyW2zPaeyr5s5LcPrG1/XAfST1ITB2cy+bWZ9wDvNrDd/6gMOAg8EXqFUlZ7e3hOuqAFArAaL12hUQ+QUdHR0UFdjTEucfDcsEYMZ9UZHh5akExkrm83yzLPPkJ2fhZNYJ/1U+MLc9tsa15icJgzO7v49d28G/o+7t+RPze4+292/XYEapYr09fXCBB1nABL16jiLnIL9+/czvyGDneKerfPqU+zfr01QRMZ64403ONJ9BD89hNUtZkGsIcazzz5b+deWCdVMdAczO8/d3wDuNrN3H3u7u78SSGVSdYaHh0klk/gJlqIr8Hg9R48erUBVItG0b+8e5tWf/OYnBfMaMmzeu7cMFYlEx3PPPQdGZeebCwzSC9K88OILjIyMUFdXRDNKKmbC4Az8BfAHwD+Oc5sDHylrRVK13tw1sGHC+6Zr6jic3yxFREqTTqfp6OzkgoUnP99ccFpjlmc7jjI4OEhj4/F3/BSZKtydp595GuYCExzrHlgNpzsjO0dYu3Ytl156aThFyLgmDM7u/gf5r5cFX45Us0IHuZjg7IkGDh9WcBY5GXv27CGVSrN42qkvk1V4jh07dnDBBRec8vOJVLudO3eyd89esu8+9V9MT9o8sDrjySefVHCeZIoZ1bjyRLe7+73lK0eqWXd3NwBeO3HXyhMNHOlqx92xUx3SFJliWltbAVjafOqjGoXn2L59u4KzCPDkk0/mxjTCmG8uiEFmQYZnn3tW4xqTTDGjGpef4DYHFJwFeHMt2Gxt04T39dpppFMpjhw5wqxZs4IuTSRSWltbScRgQeOpd8Rm1TnTam00jItMZe7OE794AuYBEywQFXgti53htmFefPFFPvzhD4dbjIwqZlTj65UoRKpfZ2cnFq+BCXYOBPC6aaOPUXAWKc3rGzeytDlDTSkr8R+HGSxvTrJxw2un/mQiVW7btm10dnSSvSjEMY2CeWD1xqpVqxScJ5FiRjW+6u63mtmfj3e7u/9T+cuSatTR0YHXNVPM+ljZfHDu6OjgHe94R9CliURGX18fW7Zs4fJlybI95/mzUvy0dQ8HDx5k3rx5ZXtekWrz6KOPYjELd0yjIAaZ03PjGn19fTQ3N4ddkVDczoGFz92bj3M6LjNbbGZPmdlmM9tkZt/MXz/LzJ4ws9b815mn8GeQSWJXWxvpupai7uv108GM9vb2ie8sIqPWr19P1p0LZp36fHNB4bnWrVtXtucUqTapVIrHn3iczMJMaKtpHMuXOelUmqeeeirsUiSvmA1Qrsl//ZvxThM8PA38hbu/A3g/8Edm9g7gr4BV7n42sCp/WapYMplk/759ZBuK/B0oVoPVt9DW1hZoXSJR8+KLL1JfY5w1vXzBefG0DNPrcs8tMlW98MIL9PX24csmQbe5YCbYdONnP/9Z2JVIXtETcmZ2hpk9ZGaHzOygmT1gZmec6DHu3lHYIMXd+4AtwOnAFcBN+bvdBHzu5MqXyWL37t1ks9nigzOQqptO6/btAVYlEi2pVIqnn3qSC2ePlGW+ucAM3jNnhOdXP8fg4GD5nlikivzs5z/DGgzmh13JGAaZJRk2b9rMnj17wq5GKCE4A7cDdwELgIXA3cBPi32wmS0Dfg14EZjv7h35mzo5zn9TM/tDM1trZmsPHTpUQqlSaVu2bAEgO21O0Y/JTpvLvr176evrC6oskUhZs2YNff0DfGjBSNmf+4MLRhgeSfLMM8+U/blFJruuri7WPL+GzJJMacmoAnypQwweeuihsEsRSvvv0ejut7h7On+6lSIXazGzacA9wJ+5e+/Y29zdyS1r9zbufq27X+TuF82dO7eEUqXSNm/ejCXq8SJnnAEy03L/pm+88UZQZYlEyqOPPsqMeji/jPPNBWdPzzC/yXn00Z+X/blFJrtHHnmEbDaLnzGJxjQKGnJbfz/ys0dIJst3ULCcnAmDc/5AvlnAz83sr8xsmZktNbP/Bkw4dGNmCXKh+bYxm6UcMLMF+dsXAAdP/o8gk8HGja+TapxT1IoaBdmm3NH7mzZtCqoskcg4dOgQa9Y8zwfmDRMPoCNmBh+aP8z69a+ye/fu8r+AyCSVyWR44MEHcp99Twu7mvFlz8zS19vHL3/5y7BLmfKK+fG7DlgL/Dbwn4GngKeBbwC/c6IHWm5LuBXAlmOWrXsQuCp//irggZKqlkmlu7ub3bvbybYsKO2BNbV40xzWvfJKMIWJRMjKlSvxbJaPLS7/mEbBZaePkIjBXXfdFdhriEw2L774Il2Husiccepb2AdmHtg04/777w+7kimvmFU1lrv7Gfmvx55OeHAgcCnwn4CPmNmr+dOngL8HPm5mrcDH8pelSr2SD76ZloUlPzbVsoDXX3+d4eHhcpclEhkDAwM8cP/9XDwvybyG4DZmmF7nfHDBCD//+c84cuRIYK8jMpncc889WKPljt6arAwyZ2TYsGEDO3bsCLuaKa2kD/zM7Hwz+20z+93C6UT3d/fn3N3c/Z3ufmH+9DN373b3j7r72e7+MXc/fGp/DAnTunXrsEQd2abZJT8227KQTDrNa69p1zKR43nooYcYHBriU0uD/wXzk0uGSafS3HPPPYG/lkjYdu/ezcsvv0xm+eQ7KPBYvtyxGuPee++d+M4SmFKWo/ufwI/yp8uAHwCfDaguqRLZbJbnVj9Pqvl0sNJ/6mSaF2DxGtasWRNAdSLVb2hoiJ/edivvmJXmjJbgP0pe2JTlorlJVt59N729vRM/QKSK3XvvvVjcJudBgceqhcziDI8+9qi+N0NUStL5IvBRoNPdvw68C5geSFVSNbZs2UJvz1HSM5ec3BPEa0g1L+TZ554jt8CKiIx17733cqSnly+eUbn1la88c4ihoSHuuOOOir2mSKUNDAzwyM8eIXN6psg1wsLnZzupZIqHH3447FKmrFKC85C7Z4G0mbWQWwljcTBlSbV4/vnnwYzMjJP/r5CZuYRDBw+yc+fOMlYmUv36+/u5/bZbedecFOfMqNyBS4unZXn//CQr776Lw4c1SSfR9MgjjzAyPIKfXUVNm+nAPFh5z0rS6fIvSykTKyU4rzWzGcB15FbaeAXQ5+tTmLvz5FNPkWk+DWrqTvp5Ct1qbbwg8lZ33XUXff0DfPGMoYq/9pVnDJFMJrnlllsq/toiQctkMty98m6YA8wKu5rSZM7O0HWoS0vThaSo4JxfVu577n7U3a8GPg5clR/ZkCmqra2NfXv3kp61/NSeKNFItmUBTz39dFnqEomCw4cPc8dPf8rF85Isr8Bs87EWNGX58IIRHrj/fvbv31/x1xcJ0urVqznQeYDM2ZN4CbrjWQDWbFo2MiRFBef87n4/G3O5zd03BFaVVIVChzgzc9kpP1dq5lLa29pob28/5ecSiYKbbrqJZHKE/3Bm5bvNBVeeMYSRYcWKFaHVIBKEO++6E2ua5EvQHY9B5swMmzdv1gZiIShlVOMVM3tvYJVI1Vm16kmyLafhtY2n/FyZfNf6aXWdRdi3bx8PPvgAv7FwhAVNwa3bPJFZ9c5vLRrmF794gu3bt4dWh0g5bdmyhY0bNpI5a/IvQXc8vtyxWnWdw1DKf5n3AWvMbIeZbTCzjWamrvMUtWvXLtrb20jNnGgPnOJ4bRPZ5tP4xapVZXk+kWq2YsUK4jifD2G2+ViXLxumsQauvfaasEsRKYu77roLSxi+vIoOCjxWDWSWZ3j66afp6OgIu5oppZTg/FvAmcBHgMuBz+S/yhT01FNP5VbTmLWsbM+ZmrWc9rY22trayvacItVmx44drFr1C35r8SAz68J/Y29KOJ9eOsgLL7zIxo0bwy5H5JQcOHCAp556KrfhSSLsak6Nn+W4uTYrqrBSgrMf5yRTjLvzxBO/INNcnjGNgsys5WDGKnWdZQq77rrraKgxPrN0JOxSRv3m4hFm1MM1V1+t9dalqt17771kPYufFYH/x42QXZTlwYcepL+/P+xqpoxSgvMjwMP5r6uAncDPgyhKJrdt27axb99e0rPPLOvzem0jmZYFPP7EE3pzlilp06ZNPP/883x6ySBNicnzPVAfhyuWDrJh40ZefvnlsMsROSkDAwPc/8D9ZBdloSnsasrDz3GGh4Z56KGHwi5lyig6OLv7Be7+zvzXs4GL0TrOU9KqVasgFiNdxjGNgvSsM+nYv5+tW7eW/blFJrsVK66npQ5+a8lw2KW8zWWnjzCnAVZcf51+sZWq9NBDDzE0OISfE6H/vzOBeblVQlKpVNjVTAknfTypu79C7oBBmUIymQyPPf4EmZbToab8e5SmZy2DWJwnnnii7M8tMplt3LiRtWvX8eklg9THw67m7WpicMWyAba8sZUXX3wx7HJESpJOp7nzrjthHlW34clEMudkONx9mCeffDLsUqaEooOzmf35mNO3zOx2QKviTzHr16/nyOFuUnPODuYFaupIT1/MY48/oe1EZUr5yU9W0FIHH100eWabj/WhBUnmNsJPVlyvrrNUlVWrVtHd1U3mnCrc8GQip4FNN267/TZ9X1ZAKR3n5jGnOnKzzlcEUZRMXo8++ihWU0cmv012ENJzzqa356hmKWXK2LhxI+vWvcJnJmm3uaAmBp9dOsAbW7fxwgsvhF2OSFGy2Sy33HoLNt3gtLCrCYDlus5tu9r0fVkBpQTnze7+N/nT37r7bWg5uimlv7+fp595huTMZRCrCex1MjMWYYkGHnnkZxPfWSQCbrzhBlrq4COTuNtc8KEFSeY0wI033qDullSF1atXs7t9N5nzMmBhVxMMX+JYk3HTzTfp+zJgpQTnbxd5nUTUE088QXJkhPS884J9oVickTln8dzq5+ju7g72tURCtnnzZl5eu5ZPLZ7c3eaCmhhcvnSALVveYO3atWGXI3JC7s7Nt9yMTTN8UYQDZSzXdd68aTOvvfZa2NVE2oTB2cw+aWY/Ak43sx+OOd0IaAh1inB37n/gAbxpDtmmOYG/XnrueWQzGX72M3WdJdpuuvFGmmvhY1XQbS748MIksxvghht+ou6WTGpr165l6xtbyZxbvdtrF8uXO1af6zpLcIr5b7QfWAsMA+vGnB4kt5ugTAEbNmxg186dJOeeCxb8Z13eMJ1My0Luu+9+HSQokbV582bWvPACn1w8RH1w009ll4jBZ5YM8vrrm9R1lknL3VmxYgXWaPjSKfALXjzXdV63dh0bNmwIu5rImjA4u/tr7n4TcAFwq7vflL/8AFA9LRI5JbfdfjuWaCAd1Goa40iddj5dXYe0xI5E1orrr6e5Fn5z8eRbt3kiv3H6CLMb4Hqt6yyT1EsvvcTmzZtzs81VMAZVDn6mYw3GihUrwi4lskr54OJxoGHM5QbgF+UtRyajXbt28cKaNYzM+xWIV64tlpmxGBpncuttWmJHoue1117j5bVruXzpYFV1mwsSMfjcstys85o12gtLJhd357rrr8vNNi+fQu8fNbmu8/r161m/fn3Y1URSKcG53t1HN0PPn28sf0ky2dx8881YPEFq/jsq+8JmjJx2AW27drF69erKvrZIgNyd6669lhmTfN3miXxoQZJ5jc71111LNpsNuxyRUatXr2bb1m25bnPEZ5uP5Wc61mhce921ajoFoJT/TgNm9u7CBTN7DzBU/pJkMmltbWXVqlWMzH8HJMq/U+BE0rPPgobpXHPNtWQyEVy4Xqak5557jg0bN/K55QPUVfFHyDUx+OIZA2zfsZPHHnss7HJEgNwOt1dfczXWPEVmm48Vh8x5GTa9vonnn38+7Goip5Tg/GfA3Wb2rJk9B9wJ/HEwZclkcc2112KJOlIL3hlOAbEYw6e/h/b2Nh5//PFwahApo1QqxY//v39n4TTnsoXJsMs5Ze+fn+LM6Rmuu/YahobUS5HwPfbYY+xu3036/HQo3WZ71eAocBRiT8dylyvMlzvWYvz46h+r6VRmRf+XcveXgfOAbwD/F/Ar7r4uqMIkfGvXruWlF19k5LR3QU1daHVkZi3Hp83l2uuu0xuzVL3777+fvfv285Wz+olH4CPkmMFXzh6gq/swd955Z9jlyBQ3MjLCdddfB7OA08OpwY4alsqfDhl2NIRdV2KQ/tU0u9t369OgMiv1x/a5wDuAdwNfNrPfLX9JMhkkk0n+8R//CRpaSJ1W4dnmY5kxvOR9dHd1cdNNWp9SqteRI0e44ScrOH9WmnfOjs4yi+fMyHDxvCS33XoLnZ2dYZcjU9g999xDd1c3mXdGd5fAop0OzILrrr+OkZHqPZZisik6OJvZ/wR+lD9dBvwA+GxAdUnI7rjjDvbt28vwkksC3V67WNnm00jNPYc77ryTtra2sMsROSn//u//ztDQIL977kAllkOvqP94ziBk0vzrv/5L2KXIFNXb28vNt9yML3CYG3Y1k4BB5p0Zuru6ueeee8KuJjJK6Th/Efgo0OnuXwfeBUwPpCoJ1b59+7jp5ptJz1qWWxJukkgufi8eq+Ef/uEfdAS/VJ1XXnmFxx9/nM8sGWJhU/T+/86pdz6/fIDVq5/n2WefDbscmYJuvfVWBgcGyV4Qve+vkzYXfEFu2/He3t6wq4mEUoLzkLtngbSZtQAHgcmTqqQsstksf/e975HOGsmll4RdzlslGhhedDEbNmzg/vvvD7sakaIlk0n+6R//gXmNcMXy6tvspFifWDLComlZ/vVf/pnBwcGwy5Ep5MCBA6y8ZyXZpVm19I6RPT/L4MAgt912W9ilREIpwXmtmc0AriO35fYrgFa9j5j77ruPjRs2MLzkfXhtU9jlvE167jlkpi/ixz++mv3794ddjkhRbrvtNnbv2ctV5/RRW8XLz02kJga/d14/Bw918ZOf/CTscmQKueGGG8hkM/j5U3D5uYnMgOzSLHevvJtDhw6FXU3VmzA4m9ml+bP/t7sfdfergY8DV+VHNiQi9u3bx4+vvprM9EUV3Vq7JGaMLP8gyUyW7/3932tkQya9trY2brn5Zj5w2gjvmhOdAwKP55wZGT66aJiVd9/N5s2bwy5HpoDOzk4efexRMssz2pbtOPxXnXQ6zR133BF2KVWvmI7zD/NfR7vL7t7m7huCKUnCkMlk+M53v0sqAyPLP8RkPnLJ66YxvPh9vPbqq6xcuTLsckSOK5vN8oPv/z0N8SxfPWfqLKX4O2cNMaPO+cH3/550Ovq/LEi4br/9dtwdP1fd5uNqguySLPc/cD9Hjx4Nu5qqVkxwTpnZtcAiM/vhsaegC5TK+OlPf8rmTZsYWnoJXjf5RjSOlZ57DpmZS7j6mmu0yoZMWg8++CCvb9rMfzyrn5baqfOm3lgDV53Tz85dbepwSaC6urp4+JGHySxVt3kifp6TSqa4++67wy6lqhUTnD8DPElue+1145ykym3fvp0VK1aQnrWczOwzwy6nOGYML/8gGWr439/5jrpaMukcPnyYa67+Me+YleaDC6p/h8BSvWdeiovmJrnpxhu0trME5t577yWdTuPnTZ1fTE9aC2QXZVl5z0odvHsKJgzO7t7l7ncAn3X3m449VaBGCVAymeQ73/ku2XgdI8sundQjGm+TaGRo2QfY3trKzTffHHY1Im/x4x//mOHhIb4WwTWbi/XVc3NrO//wX/817FIkgrLZLI8+9ih+msO0sKupDn6WMzQ4xPPPPx92KVWrlFU1/szMfnLsKbDKpCJuvPFGdu3aydCyD0KiPuxySpaZtZz07LO4+eabeeONN8IuRwSAV199lccee4xPRXTN5mLNqXc+t3yA51av1hu1lN3GjRvpOtSFL1a3uWhzwBqNJ554IuxKqlYpwflh4JH8aRXQAvQHUZRUxubNm7nttttI5eeFq9XIskvwRAPf+e53ta2ohM7d+bd/+xFzGqK9ZnOxPrlkhIVNzr//6EcaqZKyWrVqFVZj+OkKzkUzyCzK8NJLL2lDlJNUdHB293vGnG4Dfhu4KLjSJEjJZJK/+7vv4bVNJJe8P+xyTk1NHUPLPsSe3bs1siGhW716Ndu2tfL55QPUR3jN5mLVxOA/nDnAnn37WLVqVdjlSES4O08+9SSZ0zJQE3Y11cWXOJlMhtWrV4ddSlUqpeN8rLOBeeUqRCrr1ltvZffudoaXXQo1tWGXc8oyMxaRmnM2t91+O9u3bw+7HJmistksK66/nvmNzgdPC/eAwFu2NtDeF6e9L853107jlq0NodXynrkpljZnueEnK9R1lrLo7u6mt6cX5oZdSRWaAVZj7NixI+xKqlLRwdnM+syst/AVeAj4y+BKk6Ds2rWLW265hfTss8jMiM6u6cml78PjdXzve39PJpMJuxyZgp577jl27NzJ55cPEj+VtkQZtPfFGcrEGMrEeONogva+8NrfMYMrzxhkf0enZiulLPbu3QuAT9OYRskMmAZ79uwJu5KqVMqoRrO7t4z5eo673xNkcVJ+7s4///O/kI0lGFla5SMax6qpZ2jJ+2ht3cbDDz8cdjUyBT388EPMboBL5k+95ecm8u45KRZNy/LwQw+GXYpEQCE40xxuHdUq25SlfXd72GVUpZJ6ImZ2upl9wMw+XDgFVZgE49lnn+XVV9czfPq7q3IVjYlkZp1BtuU0rr3uOvr6+sIuR6aQnp4eXn75ZS6ZNxx6t3kyMoNL5o+w8fVNHDhwIOxypMrt27cvl2C06clJ8WlOZ2cn2ezUXfXnZJUyqvF9YDXwP4D/mj99K6C6JAAjIyP86Ef/Bk2zSM87L+xygmHGyJL309fbx003aZlxqZxnnnmGTCbL+9VtPq7C382TTz4ZciVS7RKJBGQBTWqcHId4PI5N1UXmT0EpfZHPAee6+6fc/fL86bNBFSbl9+ijj3LgQCfDiy4Gi25LLNs0h9Tcs7n3vvvo7u4OuxyZIn75zDMsaHKWNmu+/njmN2Y5Y3qWZ55+OuxSpMrNnDkzd0YrkJ6cYZgxc4aC80koJT3tBBJBFSLBSqfT3Hbb7fi0eWSmnx52OYFLLXwX6XSalStXhl2KTBFtu3ZyZnNyyu4SWKwzm5O0tbXhrlahnDwF51Njw8bsWbPDLqMqlRKcB4FXzewaM/th4RRUYVJezzzzDJ2dHYwsuKC6ttU+SV4/nfTMZdx7733092ufHgnWyMgIB7u6md+oecGJzG/MMjg0RE9PT9ilSBWbPTsf+gbDraNaxYZjCs4nqZTg/CDwHeB5YN2Yk1SBhx56CBpayMxcGnYpFZNa8E6GhgY1TymB6+joAGB+o8Y0JlL4OxpdFUHkJJx11lkkahPYgeg3gspuALzXeec73xl2JVWp6P123F1HWlWpvr4+Xn3tNZLzzo/0bPOxsk1zoL6Z1atX89nPahxfgnP06FEAWhIaP5hI4e+o8HcmcjIaGhp4//vez3PrniN9YTq3NrEUxfbl/rI+9KEPhVxJdSplVY2zzWylmW02s52FU5DFSXm89NJLZDMZMjOXhF1KZZmRmr6EtWvXMTQ0FHY1EmHTp08HoD+td++JDOT/jgp/ZyIn69d//dfxQYfDYVdSXWL7Yiw/YzmLFi0Ku5SqVEr78Qbgx0AauAy4Gbg1iKKkvF555RUsUU922tTbmzQzYzGpVJLNmzeHXYpE2IwZMwDoTU6dT3ROVuHvqPB3JnKyLrnkktySarv1C2vRBoAu+I1f/42wK6lapfyUb3D3VYC5e7u7/y/g08GUJeXU1dVFtm7alBrTKMjW5baV0rJ0EqSWlhbMjJ6k3sAnUvg7UnCWU9Xc3MzHP/5x4rvioA8Vi2JbjHhNnE996lNhl1K1SklSI2YWA1rN7I/N7PPAtIDqkjLq7j5MJh69XQKL4YkGQMFZghWPx1m+bCnbjmrFzolsO1rDvLlzmDZNbx9y6r72ta9hGLZFv7ROqB9ibTE+d8XnmD9/ftjVVK1SgvM3yW1u+afAe4CvAr8bRFFSXr19vVBTG3YZ4YgnwEzbb0vgLv3gh9h6tIb+lN7AjyeZgY2H6/jghz6sjRekLBYuXMinPvkp4m1xLU03AdtsJGoSfOUrXwm7lKpWSnB24BZyy9JdBJwDXBdEUVJe8+fNI5YcCLuMUFhyANz127UE7tJLLyXr8FqXus7Hs+lwgmTGufTSS8MuRSLkqquuImYx7HX9MnZcRyG2O8aVV17JnDlzwq6mqpUSnG8jd4DgF4DP5E+XB1GUlNfSpUuJD/fAFNypy4ZzS14tWTLFVhSRijvvvPOYPWsmqzun6Kc7RVjdWUtjYwMXXnhh2KVIhMyfP58v/c6XiLXH4GDYaQrWWAAAIABJREFU1UxCDvFX4rS0tPDVr3417GqqXinB+ZC7P+juu/IHB7a7e3tglUnZLFu2DE8NY8mpt4NefCA327x06dTZ+EXCEYvFuPILX2RDd4IdPfGwy5l09g/EePFALVdc8TkSCXXlpby+9rWvcdqC06h5pQa0D9Fb2HaDbvjmn35Ty0CWQSnB+X+a2fVm9mUzu7JwCqwyKZsPfOADmBk1h7aFXUpluVPb3cp55/0Ks2bNCrsamQKuvPJKmqc1cf+uhrBLmXTu31VPXV0tX/7yl8MuRSKorq6Ov/xvf4n3ObZZIxujBiG+Kc7FF1/Mxz72sbCriYRSgvPXgQuBT5Ab0bic3LiGTHILFy7kve99L3Vd28CzYZdTMbG+Thg8yuc//7mwS5Epoqmpid/50pdZ35VgZ6+6zgX7B2K8cKCOz1/5BS1DJ4F5z3vewyc+8Qli22KgjSnBIfZKjEQswZ//+Z/rgNwyKSU4v9fdL3L3q9z96/nT7wVWmZTVFVdcgY8MEO/eFXYpFZPofJ3GpiYuu+yysEuRKeQLX/gCM6a3cNPWJrJT77CCt3GHm7c2UV9fz5e+9KWwy5GI+6M/+iNmTJ9Bzcsa2bA2wzqM//yH/5mFCxeGXU5klBKcnzezdwRWiQTqkksu4YwzzqRh70uQSYZdTuDiR/dSc6SdL3/pS9TXT801rCUcTU1N/Ok3/4wdPXEe31MXdjmhe7ajltcP1/CN//JHzJw5M+xyJOKmT5/Ot//q2/jRKT6yMQDx1+Jc+GsX8oUvfCHsaiKllOD8fuBVM9tqZhvMbKOZbQiqMCmvmpoavvWtv8BHBqjduz7scoKVTVO/ew2nn75I85QSio9+9KO8//3v4+6dTRwamno7dhb0jBi3bZ/GBRecz+WXaxEmqYxLLrmET3/608S2xqAr7GpC4BBfG6eupo7//u3/Tiw2dX8GBaGUv81PAGcDv8mb8836SVhFzj//fD7zmc+QOPA6sYHgf5rUtq8hNthNbLCb+s0PU9u+JvDXBEjsexWGevjWt/6C2lotDSaV9/+3d9/hUVZ5G8e/Z2bSOxASagKhC0EggAXEvigIiqCCIiIoNopLtSCo2LHS1t5eRBQbgq4CKghYICBSQq8hEAghfeoz5/0jwXVd0QBJnim/z3VxkTq50czMPec5RSnF2LHjUNZQXs8KzikbWsPb2yJxeS1MmDBRnrxFjbrnnntITEzEttYGHrPT1Cy1U8GR8l00kpOTzY4TcCr9SPb7LehkOzr/dccdd1CrVi0idn0LnuqdsmEpPYYy3CjDjbX4MJbS6j/22lKYQ2jOBnr27EmnTp2q/ecJcTJJSUncfc89bMq3BeWUjRWHQvn5SCi3Dhsu20GKGhcVFcX9991fvsvG5iCaslEC1k1WunTtwpVXXml2moAkQwBBJjY2loenTsXiLCZs78rAOhTFXUbk7u9o2KghY8aMMTuNEPTp04fzzz+P93dGsq84eHbZOFxm4Z3t0Zx9dntZEChM06lTJ/r06VO+y0b1j9uYT4M100p4aDgTxk+QXTSqiRTnINS+fXuGDRuG7dhubEe3mR2namgv4buWY9MeHn3kESIjI81OJARKKSZOnERsXDyzN8fgCoJV/h4vzN4cQ2h4JA8+OBmrNXheMAjfc9ddd1EnsU75lI0Av/+p3eVTNEbeM5K6deuaHSdgSXEOUjfeeCOdO3cmfN8PWEr8/4zSkOxMrIUHuffeMaSlpZkdR4jfxMfH88CDkzlYopi3I/APRvlkdzi7Cy2MnzBRnryF6SIjI5k0cRK6SKOyAngE1g7WjVY6ZXSiV69eZqcJaNVanJVSbyiljiilNv3uY7WUUkuUUjsq/pb9iUxgsVh46KGHSEysQ8TOZeAuMzvSabPm7yE0ZwO9e/emd285k0f4ns6dO9O/f3+WZIfz6zGb2XGqzfYCK5/vi6Bnz55ceOGFZscRAuC3U/Os261Qanaa6qE2Kizawrix42SKRjWr7hHntyjfjeP3JgHLtNbNgWUV7wsTxMXF8fhjjxHidROx81u/PFVQ2QuI2LOClq1aybxm4dNGjBhBakpjXsmKodgVeE9sdg/8a0ssdevWZfTo0WbHEeK/jBgxApvVhvo18O575INln4Xrr7ueBg0amJ0m4FVrcdZarwDy//DhvsDbFW+/Dch5yCZq0aIF48ePw1J0iJDsTLPjnBrDQ+TOb4iOjOCxadNk6znh08LCwpj80BRK3Fbe3Bp4c/Df2xFJnl3x4OSHiIqKMjuOEP8lKSmJGwfdiCXbAkfNTlOFdPlBJ3Hxcdx8881mpwkKZsxxTtJaH6p4+zCQZEIG8Ts9e/akV69ehOZswFqQbXacSgvdtxrsx5ny0EMyl1L4hebNmzP01lv5+UhoQE3Z2F5g5duDYVx/ww2kp6ebHUeIPzVo0CBq16mNdWM1L1h1Q0REBP379yciIgLc1fizcoA8uGPEHbIovoaYujhQa62Bk+6HppS6XSm1Vim19ujRQHqJ6HtGjx5NSmoqEbuXo1y+PwnMmreDkKPbGXzTTXTp0sXsOEJU2vXXX0/DBvV5d3s0Hv+bHfU/vBre2R5Nndq1GDJkiNlxhDip8PBwBt80uHxruurcns4NvXr1YtSoUeUL9aqxOFt3WElKTqJnzz/OihXVxYzinKuUqgdQ8fdJt3TQWr+itc7QWmckJibWWMBgFB4ezrRHH8WmDEL3rvLp/Z2Vq4yIfT/Stm07hg4danYcIU5JaGgoo0aP4VCp4qsAOBhleU4oe4ss3H3PSBnxEj6vZ8+ehEeEl5+uV11CYPHixbz00kssXrwYQqrp5xQAR+HaftfKto81yIzivBA4MSwxBPjMhAziT6SkpDB82DBsx/djPb7X7DgnFbrvB6zKy333TcJmC5zL3SJ4nHPOOZx33rl8sieSUrf/LlZyGfDBrijap7fj4osvNjuOEH8rMjKS3r16Y822gqOafkgI2O12FixYgN1ur7birHYoQsNCZfu5Glbd29HNA34AWiqlspVSw4AngcuUUjuASyveFz5iwIABNE1rRsT+H8HjNDvO/7Ae348tfw+3DBlCo0aNzI4jxGkbOvRWHB5Yfdh/F7WuORJKsQtuGXqrbIEl/Ea/fv3QXo3a48e/sx6wHrDS8x89iYmJMTtNUKnuXTUGaq3raa1DtNYNtdava62Paa0v0Vo311pfqrX+464bwkQ2m41JEyegnaWEHN5sdpz/pjXhB9fSsFFjBg4caHYaIc5IixYtaJbWlOWHws2OctqW54RRPzmJDh06mB1FiEpr2LAhqU1SUUf9uDgfA21oLrjgArOTBB05OVD8j1atWtG16zmEHc0Cr8fsOL+xFh6E0nxuHnwTISHVNWlMiJqhlKL3VX3YW2Rhb5H/zU/MLbOw5biNXlf1wWKRpxLhX9qnt8eSb/mL7Ql8mzqmUErRpk0bs6MEHXm0E3/qhhuuR7vs2PJ2mR3lNyGHN5KQUItLLrnE7ChCVInLLrsMgF/y/O+F4IaKzCf+DUL4k3bt2qHdGgrNTnJ6VJ4iJTWF6Ohos6MEHSnO4k917NiRlJRUQo7tNDsKUL6ThrXwINdcc7WMNouAERMTQ0J8HHkO/3soznNYCA0JISlJtuIX/qdt27YAqHz/nK5hOW6hXdt2ZscISv73aC1qhFKKTp06Yi3L84mt6SyleUB5oRcikCQlJXHMT4tzUt1EWRQo/NJvI7WGuTlOmwdZFGgS/3u0FjWmZcuWaI8b5TD/WpalNA+lFM2aNTM7ihBVKim5Hsdc/ncV5ZjTSt3kZLNjCHFavN6K04f89XWfAu0Dg1rBSIqzOKmmTZsCYLEfNzlJeYa6SclywIIIOKWlpYRa/O8IwTCLpqzU908ZFeLP+H3plOJsGinO4qTsdjsA2mr+PrPaGorDUV271QthDq0127dtJTW6Gs/krSapMR527tqFx+M7O+8IUVkul6v8DX8dcbb87t8gapQUZ3FSBQUFAOgQ8/eZ1SHhFBcV/ufymhAB4MiRIxQVl5Aa63/lMzXWg9vtYd++fWZHEeKUbd5cfk6BjvfPUVsdp9m0eZPZMYKSFGdxUjk5OQDoEPOnR+iQSLxeL0eOHDE7ihBVJjMzE4AmMf63Qim1IvOJf4MQ/iQzMxMVqiDB7CSnx5voZcf2HRQXF5sdJehIcRZ/SmvNl//+Nzo6EUIizI6DEV9+vPZXX31lchIhqobH4+Hdd96mcYyXJrH+V5zrRXppmWAw7725OJ1Os+MIcUp+XvMz3jpev21BOkmjtWbdunVmRwk6fvorI6rb5s2b2bd3L67ElmZHAUCHx2LE1efzRYtkuoYICF999RUHcw5xbdMyLH44z1Ip6N+0jGP5x/nss8/MjiNEpe3atYvcw7nouv45TQOAWqBCFN9//73ZSYKOFGfxp95//32ULQRP7aZmR/mNO7EVR3JzWb58udlRhDgjLpeLt996k6ZxXjrW8b+FgSe0TvBwVi0Pc999h7KyMrPjCFEpb731FipEoRv7cXG2gJFisHTpUrKzs81OE1SkOIv/sWTJElasWIEzOR18YEeNE4yEVHR0HaY/+yzHjh0zO44Qp+2ll17icO4RBjQtxd/PDxmQVkZBYRHPPPOMbI8lfN6OHTtYvnw5RjMDwsxOc2Z0a422aN58802zowQVKc7iv+Tm5jL92efwxiThrt/e7Dj/zWLB3rQHJSVlPPHkk/IkLfzS4sWLWbhwIb1THLSr7X+7afxRsziD/ml2li1bxocffmh2HCH+0uuvv44KVegWAfD8EQ5GWvmo8969e81OEzSkOIvfOBwOpj78MA6XC0fTHqB879dDRyTgaNSZn3/6iblz55odR4hTkpWVxXPPTqdtLQ8D0uxmx6kyfVIdZCS6mD17tixWEj5r7dq1rF69GqO5Ab5zMfWM6JYabDBjxgxZ/1NDfK8ZCVM4HA4mTpzI5s2bsad2R4fHmh3ppDxJbfDUasorr7wiI1zCbxw6dIgHH7ifuBCDu9uWYA2gR1+l4PazSkmONJg65SHZ21n4nIKCAh559BFUbICMNp8QBkZbgzVr1rBgwQKz0wSFAHroFqfL6XRy3/33s379LzibXIDhQwsC/5RSONMuxFMrlRkzZvDxxx+bnUiIv5STk8PIe+6mrCife9OLiAkNoCfuCpE2uLddEdpRxKiR98ilY+EzvF4vjz3+GIVFhXi6esBmdqKqpdM0ur5mzpw5bNu2zew4AU+Kc5ArLi5m0n33kbk2E2fT7ngSm5sdqXIsFpxpF+FJSOGFF17g/ffflznPwicdOHCgvDQXHuO+swtJ8cPDTiqrXpSXBzoWoh2FjBp5D7t27TI7khAsWLCAn378CSPdgHiz01QDBd7OXrxhXqZMnSI73FQzKc5BbO/evdx2++1kZq6rKM0tzI50aixWnM0uxlOrCbNnz+axxx6TgxiET9m/fz+jRt6Do+gY93UoJNUPDzo5VfWjvDzQoRDlLGL0qJHs2LHD7EgiiP3yyy/MmTMHXV+j0wJ4cCUUPF085OTk8PgTj2MYgf9YYxYpzkHq+++/5/YRIzh0NB97qyv9rzSfUFGeXQ078fXXX3PXXXeTm5trdioh2LhxI3ffdSfu0uPc3zGwR5r/6MTIs81TwsiR97BmzRqzI4kgtHv3biZOmog3you3sxf8fOvHv5UI3nQvK5avYObMmXIVtppIcQ4ybrebV199lQceeAC7NZrSNn3xxiabHevMKIW7QQccLS5j5569DBs2nJ9//tnsVCKIff3114wZPYoIo4jJHQtpFB18q92TI7081KmA2tZSxo8fL6cLihqVm5vLP8f+EwcOPN09AbOLxt/RLTTe5l4++ugj3n//fbPjBCQpzkFk165d3Hb7CN59913ciS0oa90LHRZtdqwqYySkUNqmD0UeC+PGjePpp5+WuV6iRmmteeONN5g2bRppMU6mZhRSLyr4SvMJdcI1kzsV0i7BybPPPsvMmTPlErKodsXFxYwbP47jxcfxnO+BSLMT1SzdXuNt6GXOnDksWbLE7DgBR4pzEPB4PLzzzjsMHz6cPQcO4mh+Ga6mF4AlwJYWAzointKz+uKq145FixYx+OabyczMNDuWCAIOh4NHHnmEt956i+71nEzqUEx0iFwqjbTBP9uXcFlDBx988AEPPvgApaWlZscSAaqsrIwJEyewf/9+POd6AnMx4N9RoLtoSITHH3+c1atXm50ooEhxDnA7duzgjjvv5LXXXsMZl0JJ22sxaqWYHat6WWy4G3fF3uYqjhY7uffee3nuuecoLi42O5kIUAcPHuTOO0bwzbJlXJdm5/Y2Zdjk0fU3VgsMaWXn5hZl/LB6NbcPHy7b1YkqZ7fbGT9+PJs3by7fdq6u2YlMZAXjPANvnJcHHnyAH374wexEAUMe2gNUYWEhzz77LMOHD2fH7v04ml2Ms/nFEBJudrQa441JovSsa3Ann8Wnn37KwIGDWLRokZyuJKrU6tWrGT7sVg5n72Xc2cX0aeJA+cAiJLtHERERQf/+/YmIiMDuMT/U5Y2d3NexmMK8g9x+22188803ZkcSAcLhcDBh4gQ2btqIt6sXGpqdyAeEgqe7B29MeXmWtT9VQ4pzgDEMg08//ZQbBg7is4ULcdVtQ0l6f98/1KS6WG24Us7F3vZqCnQ4Tz/9NLePGMGWLVvMTib8nGEYvP7660yaNIlEWymPdi6gfR2P2bF+U+ZR9OrVi1GjRtGrVy/KfKA4A7RO8DCtcwGNIsqYOnUqM2fOxOPxnf9uwv+cOPl2w4YNeLt40Y1kitRvQsFzgQcj2mDSfZNkh5sqEHiTXIPYunXreOmlGezevQtvbD0cbS9DR9YyO5ZP8EbVwd66N9ZjO9mxdy133HEHPXv25LbbbiMxMdHseMLP5OfnM23ao6xdm8kF9Zzc0qqMUKvZqf5bpE2zePFiABYvXkxdm++UiVrhmgc6FvHejgg++OADsrZs5qEpU0lKSjI7mvAzdrud++67j/Xr15eX5sa+83vuMypGnlkBkyZN4oknnqBLly5mp/JbMuIcALKysrj3n/9kzJgx7D6Yi6PZxdhbXSml+Y+UwqjTnJJ21+Kql86/v/6aG24YyKxZsygoKDA7nfATmZmZ3Dr0FjasX8ew1qXc1sb3SjNAhE1jt9tZsGABdrudCB8qzgA2C9zc0s5dbUvYsXULtw69ReZhilNSVlbG+PHjWbd+XXlpTvGt33GfElY+8uyJ8jDpvklyXzsDUpz92N69e3nwwQcZMWIE6zZswtm4K6UnpmX4wiRLX2UNxd24C2XtBlAWl8L8Dz7guuuv580335Tt68RJGYbBG2+8wT//eS/h7uM80rmAixq45K52hs5LdvNo5wJqqWImTpzIrFmzcLvdZscSPq60tJSx48by68ZfpTRXVhh4engwYgzuv/9+Vq5caXYivyTF2Q/l5OTw+OOPM2TIEL5f/SOuBh0pSb8OT712AbnFXHXR4TG40npQ1rYfJRFJvPnmmwy47jrmz5+Pw+EwO57wIXl5edx77xjeeustuiU7eaRzQVAealJd6kV5mZJRyKUNHcyfP5977rmbQ4cOmR1L+Kji4mLu/ee9bN6yGeMcQ6ZnnIqKaRtGvMGDkx/ku+++MzuR35Hi7Ef27t3LtGnTGDRoEF99vQRXUltK0q/D3bAj2ILkWKRqoCMTcDa/FPtZfSlQMcyaNYv+/Qfwf//3f7LfrOCHH35g6C1DyNr4K7e3KWXEWWWE++DUDH8XaoVbWtkZ1a6EvTu3cevQW/j222/NjiV8zInSvG3bNoxzDNk943SEgtHdwJvgZerUqVKeT5EMT/qBHTt28M4777B8xQqUxYarbhvc9dqhQ6PMjhZQvNGJOFpdgaX4MEbOL7zyyiv839y5XDdgAP379yc2NtbsiKIGud1uXn75ZT744AMax3h5oEsx9YP4FMCa0iXJTWpsAbM2RTNlyhQyMzMZOXIkYWFhZkcTJjsxPWPHjh3lh5vUNzuRHwsBo5sBK2Hqw1N5xPIIF1xwgdmp/IKMOPuwTZs2MWHCBIYNG8aKVT/gqpdOSfvrcKWcI6W5GnljknG07In9rL4Uhdbhrbfe4tr+/ZkzZw7Hjh0zO56oAdnZ2dx55x188MEHXNbQwdSMQinNNahuhJfJnYroneJg4cKF3DZ8GHv27DE7ljBRWVkZY8eOZeu2rXjOkdJcJSrKszfey0NTHpI5z5UkxdnHeL1eVq5cyV133c1dd93FT5m/4GrYiZL21+Nu1BlCIsyOGDS80Yk4W1xGWbt+lEY1YN7779N/wACeeuop9u3bZ3Y8UU2WLl3KsFuHcnDvTsaklzCkld0nd80IdDYL3NDczsQOxeQf3s/ttw1n0aJFaC3zWYNNWVkZ48aPY0vWFoyuBjQwO1EACamYthHvZfLkybLbRiUof3kQysjI0GvXrjU7RrVxOp0sWbKE9+bNI/vAAQiPwZl0Fp7ElmANMTveaQnfsghr8eHf3jdiknG06W1iojOjHIWEHNpIaN5OtNfDeeefz6CBA2nXrh1Ktlbwe06nkxkzZrBw4UJaxBvc1baYOuH+8fj4Z6atjWZrwX8eO1rFu3kwo8TERKevwKn41+ZoNuXbuOyyyxg7diyRkZFmxxI1wO12/7blnNHVgEZmJ/p7lu8sqKP/eU7QiRrvhT5+xcoF1u+tWIusPP/c85x99tlmJzKVUipTa53xZ5+TOc4mKy4u5tNPP+WDDxdQWHAcHVUHZ9pFGLWbgJILAr5Eh8fhatINV8NOhORu4YefM1m9ahWtW7dh0KCBdOvWDatVhib9UXZ2NlMemsyOnbvoneKgf5odm9z9fEZ8mGZCh2I+2xPOx0uWsH3bVh55dBpNmjQxO5qoRlprnn76adatW4e3s9cvSrPfqlgwqL5VTLpvEv+a8y9SU1PNTuWT5KnBJDk5Obz44otc068fr776Kvk6CnurKyg7qy9GnTQpzb4sJAJ3xfQZZ8p5ZO3JZvLkyQwcdCMff/wxdrvd7ITiFHz33XcMHzaMnP27Gdu+hBuaS2n2RRYF1zR1MLFjMcdzDzDi9tv497//bXYsUY3eeOMNvvrqK7xnedGp/nv1x2+EgqebB7vXzthxY2VNz0nIiHMN27x5M/Pmvc/3369AK4W7VhruFm3RkbXNjiZOlTUET3IbPEmtsObv5dDhTbzwwgu8+trr9Lvmavr160ft2vL/1Vd5vV5effVV5s6dS1qcl5Fti6kT4eOXUwVta3l4rEv5rhuPP/44WVlZjBw5EptNns4CyaJFi3j77bfxpnrRraU015go8JznIW95HuMnjGfmjJkyLeoP5JGmBhiGwapVq3hv3jy2bN6MCgnDmZyOJ7mN7I4RCJQFo3ZTymo1wVKSi+fQRt59913emzePf1x+Oddff71cUvYxdrudadOm8f3333NRAydDWpbJKLMfSQjT3NehmPk7I/jkk0/IPnCAqQ8/TExMjNnRRBXIyspi+vTpkAS6kwZZQlKzaoGnq4edq3fyzDPPMGXKFLMT+RQpztXI6XTy1VdfMfe99ziUkwPhsThTzvHrBX/iLyiFNyYZZ0wyLkchIYc28eVXX/PFF1/Qtes53HTTjaSnp8tCQpMdOXKESRMnsHv3bm5qUcY/Gjnl2Gw/ZLXAoBZ26kcZvJm5ljvvGMFTTz9Dgway5YI/czqdTHtsGjpclx9wIi9ozVEfvG28LFu2jB49enDhhReanchnSHGuBiUlJRUL/j6k4PhxdHQizmYXY9RKDa65y4aLiIgIevXqxeLFiykxXGYnqjHlCwnPL19IeCSLn9dt4KeffqRNmzbceOONnH/++VgsQfS74CO2bdvGxAnjsRcXMrZ9Me3reMyOJM7QhQ1cJEV6eXHjQW6/bTiPP/Ek7du3NzuWOE2vv/46B/YfwOhugByIayrdSqMOKZ6Z/gzt27cnISHB7Eg+QZ65q1BeXh5z5szh2mv788orr3DMG4m91ZWUtemDUbtpcJVmQHlc9OrVi1GjRtGrVy+UJ3iK829CwnE36PDbQsItuw/wwAMPMPjmm/nyyy9xu91mJwwaO3fu5N4xo7E4Cngoo0BKcwBpneBhakYB0bqE8ePGsWnTJrMjidOwceNG5s+fj7epF5LNTiOwgCfDQ0lJCdOnT5c91CsEV5OrJnl5eTz//PMMuO465r3/PsURydjbXo2j1RV44+oTrNeBtS2UxYsX89JLL7F48WK0LYiHD6w2PMltKE0fgCPtIvYfK+WJJ57guuuv57PPPpMCXc327t3LP+8dQ5i3jAc7FtIoWhYBBprkSC8PdCwk3uZg/LixbNu2zexI4hRorZn+7HSIAp0uBc1nxIFxlsH333/PmjVrzE7jE6Q4n4Hjx48zc+ZMrr/+Bj759DMcCWmUpQ/A2fxivFF1zI5nPmsodrudBQsWlG/RZg3i4nyCsmDUSaPsrKtxtPwHR51Wnn32WQYOGsQXX3yBxyOjoFUtJyeHe8eMRjuKmHR2oeycEcDiwzSTOhQSoe2M/ee9cky3H8nKymLP7j0YLQ2QJUA+RbfQqHDFws8Xmh3FJ0hxPg1FRUW8/PLLXHfd9Xzw4YeUxaVQlt4fV9Pu6PBYs+MJf6AURnwj7K1742j5D3JLDJ588kluGjyYJUuWYBiG2QkDQklJCfeOGY2zpIBJHQqpFyWlOdDVCS8vzxZXCf+8dwz5+flmRxKVsHjxYpRNoRvJaLPPsYDRyGDVylUUFBSYncZ0UpxPgcfjYd68eQwYcB1z575HaXR9ytpdiyuthxRmcXoqCnTZWX1xNL+UnON2Hn30UW4ZOpT169ebnc7vvfrqq+Tm5jI2XaZnBJPkSC8TOxRSWHCcWbNmmR1H/A2Hw8GSpUswGshos6/STTSGYfD111+bHcV0UpwracOGDQwdeitz5sxsf0KJAAAgAElEQVShJKw2Ze2uwdnsYnREvNnRRCBQCqNWKqVtr8HR7CL2HT7G6NGjefTRR2XE7DRt3bqVTz/9hEsbOmgeLyP4waZRtJdeKXaWLFnCunXrzI4j/sKaNWtw2B1yOqAviwMS4JtvvzE7iemkOP+N48eP8/jjjzNy5Ej25R7D0eIyHC3/gY6sZXY0EYiUwqidRmnba3HVP5sly5b9dpS3TN+oPMMwmD79GeJCoX+aHIEerPqmOkiMhOeenS4LcH1YaWlp+RtyQJ1P05GakpISs2OYTorzX1i2bBkDBw3i319/jat+e0rb9sNISDE7lggGVhvuRhmUte1HSUgcL7zwArffPoLs7Gyzk/mFH374ge3bdzCwWSmRslt90Aq1ws0titl/IJtly5aZHUecxG+DAtJIfJq2aBnAQX5N/5TH42HWrFk8/PDDlFiiKWt7De5GneW0P1HjdEQ8jpZX4Gh2ETv37mf4bbfx448/mh3L523evBmrgs51g3DvcPFf2tf2EGFTbNmyxewo4iR+K2PBuXOr/1BIcUaK8/8oKChg7NixzJ8/H3dSG+ytrkRHyGk5wkQnpm+06Usp4UyYOJF33nlHNqP/C9u2baNhtJdQq9lJhNksClJi3GzbmmV2FPF3AqST6XiNDqn4k6jR8YHxWK088soGpDj/l3379jFs+HB+2fArzqYX4Eo9DyzyzCt8gw6Poaz1VXhqNeW1115j8uTJsu/zn9Bas21rFqkxMqdVlGsa42Hnrl1yf/FRGRkZAKj9gVHM9Nka4oF48F7oLX/f37nAkmvhnK7nmJ3EdFKcKxQWFjJ+wgTyCkooa90bT2ILsyMJ8b+sNpxpF+Js3IUVK1bwwgsvyMjzHxiGgcPpJMIq/11EuXCbxuMxcLlk6o4vql+/Pp06dcK61wpyt/VJap9CG5qrrrrK7Cimk+IMuN1uHpw8mdzcI5Q1vxRvdKLZkYQ4OaXw1EvHVa89Cxcu5OOPPzY7kU+x2Wy0atWKHUWyJkGU215go0lqCpGRsm2Dr7rqqqvQpRpyzU4i/ocG614rzVs0p3nz5manMZ0UZ+DFF19kwy+/YG/SHW90XbPjCFEp7kYZeBJSmDFjBmvXrjU7jk9p3/5s9hZZcQTInElx+gwv7CwKJb392WZHEX+hW7duxMbFYt1iBTmryLdkgy7Q9Lmqj9lJfELQF+cDBw6wcOFC3MltMeo0MzuOEJWnFM60HnjDYpg9e47ZaXxKeno6hoZNx2TUOdhtK7Dh8Gjat29vdhTxF0JDQ7l3zL1wDNQvgTHXOSAUgm2tjdatW3PFFVeYncYnBH1xXrRoESiFu1662VGEOHXWUJx127Bz5w62bdtmdhqf0aFDBxo1qM//7YjGLuvBgpbLgLe3R5NYpzbnnCOLmnzdJZdcwvXXX49llwW1V8qz6Vxg+8FGbHQsjz32GKGhoWYn8glBXZzdbjeLv/gST3xjdKjMfRP+yVOnGcpiY/HixWZH8RlhYWFMuv8Bjjlg/s4Is+MIk3yyJ5yDJYoJEycRFRVldhxRCSNGjKBDhw5Y11nhuNlpgpgGy88WVJnisWmPUadOHbMT+YygLs4HDhygqLAAIyHV7CgByRtVG20NQVtDMGKS8UbVNjtSYLKF4Y6tR+a69WYn8Snt2rVjwIDrWJodzub84Ds+MCXGIMLqJcLqpVW8m5SY4JrwvavQyuJ9EVx55ZV07drV7Diikmw2Gw8//DC1a9XGttIG+WYnCkJeUGsU6pBi5D0jSU+XK/K/F9TFOTo6uvwNHVxPKDXFlXIu3sjaeCNr42jTG1fKuWZHCljKaxAbE2N2DJ8zfPhwGjaoz4xNsewqDK492Qe3tJMSY5ASY/BgRgmDW9rNjlRjDpRYeP7XWGrVqsXdd99tdhxxiuLj43nu2eeoHVsb23IbHDI7URBxg3WlFcs+C0OHDqVfv35mJ/I5QV2c4+LiAFBuh8lJhDgzVsNJQkK82TF8Tnh4ONOffY6YWnV5Yn0cW4Jw5DnY7Cq0Mm1dPNaoBJ57/gVi5AWlX0pNTeWVf71C09SmWFdZUbtlznO1s4PtOxuWPAuTJk1i6NChKCX/3f8oqItzWFgYMbGxWIsOghwiIfyUchajHIUkJsr+43+mfv36zJw1m+T6DXnmlxjWH5WdNgLVlnwbT6yPI65WXWbNnkNqaqrZkcQZqFOnDrNmzqJL5y5YMi2oTUoOSKkuhWD7xkaoI5Snn3qaK6+80uxEPiuoizPAbcOHYy06hC1vh9lRhDh1WhO2ZyWhNhsDBw40O43PSkxM5KWZs2jSrDkv/BrNtwdD5bVygFl9OIRnfokhuUEjZs6eQ/369c2OJKpAZGQkTz75JL169cKSZcGyygJOs1MFEA1qr8L2rY348Hhmz5pNly5dzE7l04K+OPfp04f09HTCD/yEcpWZHUeIU2LL24G18CB33DGC5ORks+P4tPj4eF588SXO7tiR17OieOHXKIpcchnS35W6FbM3RTJ7UzQtWrXmpRkzZQeAAGOz2ZgwYQKjRo3CdtSGbYkNDpudKgC4QP2ksKyx0K51O1595VU5GbASgr44WywWJk6ciA1N5NYvsJQeMzuSEH9Pa2y5WYTvW03btu245pprzE7kF6Kionj22ee46667+PV4BPf9FM/6PJn37K8259u47+d4fjwSwbBhw5gxcxbx8TLXPxAppejfvz+vvfoajeo2wvq9FbVBgaztPz15YFtqw3bQxm233caLL75I3bpycnJlBH1xBmjUqBHTpz9DQoSVyC0LsR3eLHOehe/yOAnbuYywvavI6NiBxx6bhsUid+XKslgs3HDDDbzy6mvUrp/Ks7/E8GZWpBzP7UdcBszdHsET62KIqtWAOXPmMGTIEGw2eREU6NLS0njt1dfo27cvlu0WbN/aoMjsVH7EC2qzwvqdlaTYJGbPns3gwYOxWoNr16EzIc+2FTp06MDbb71J165dCNv3A+E7loDstiF8jKX4MFGbPiG08AB33nknzzzzDAkJCWbH8ktpaWm88uprDBw4kG9ywpj0YwI/Hg6R18w+TGvIPBrC/T/H8+X+cK655hpef/NNWrdubXY0UYPCw8MZO3Ysjz/+OFGeKGxLbahtCrxmJ/NxhWD91opli4XLL7ucN994kzZt2pidyu8obdKzhFKqJ/AiYAVe01o/+Vdfn5GRodeuXVvtubTWfPTRR8yaPRuvsuKs0wpPcht0qJw6dTrCtywCwNGmt8lJ/JjWWEpyCTm0Edvx/STXS+bhqVOlLFShDRs28MLzz7Nr926axxvc2LyUZnH+PwQ9bW35XvUPZpSYnOTM7Su2MndHJFvybTRu1JBRo8fIIiZBXl4e05+dzupVq6EWGBkGxJmdCizflY9Lei/0gTbvBZWlsGy1EBsTy71j7uWSSy4xO5VPU0plaq0z/vRzZhRnpZQV2A5cBmQDa4CBWustJ/uemirOJ+zatYu33nqLFStWoJXCXSsNd7126MhaNZYhEEhxPgPaizV/H2GHN6JKjhAVHUO/a65m0KBBcnxwNTAMgy+//JJXX3mZ4wWFnJfs5LpmduqE++8QdCAU5+NOxYJdEazICSMmJppbhw2nT58+Mi1D/EZrzTfffMPzLzxPcXExRisD3Vqbek3dZ4pzPtjW2tCFmssuu4yRI0fKOoBK8MXifC4wVWv9j4r37wPQWj9xsu+p6eJ8Qk5ODh9++CGfL1qEy+nEiGuAq147vLENQDYG/1tSnE+D4cZ2dDthuZvBUURyvfrccP11XHHFFURERJidLuCVlZUxd+5c5r//Ptrr4YpGZfRKcRIV4n8F2p+Ls90DX+0P5/P9kRjawrX9+3PzzTfLgSbipAoKCnjppZdYunQpKk7hyfCASWNdphdnT/lcZssOC7Vr12bC+Amce66c3ltZvlic+wM9tdbDK94fDHTVWt9zsu8xqzifUFRUxMKFC/nwwwUcP54PEXG44lMxajfBG1lbSvRJSHGuJK8Ha8FBbPl7CCncj/a4aHPWWQy84Qa6desmCzdMkJubyyuvvMKSJUuICFFc3rCMno2cxIT6T4H2x+Jc5oGv94fz7+wISlxwwQXdufPOu2jQoIHZ0YSfWL16NU8/8zT5+fl4m3nRbTXU8AUKU4tzLtjW2dAlmj59+nDnnXfKVcpT5LfFWSl1O3A7QOPGjTvt27evxrP+kcvlYtmyZXy9ZAnr163D6/VCeAyuhFSMWk3wRiVKif6d0H0/AOBKkVe6/8NwYy3IrijLB9CGm6joGHpc0J3evXvTtm1bsxMKYOfOnbz99tssX76ccJvi0gZ2rkhxEOcHBfrdbeVXKAa3tJuc5O+VuBVf7Q/jq+wIytxw7rnnMGTILbJ4SZyWkpISXn75ZT777DNUlMLTwQP1au7nq1/Ke4A+uwYfJ5ygNigs+yw0aNiACeMn0KFDh5r7+QHEF4uz30zV+CuFhYWsXLmS75YvZ+2aNRiGgQqPxhWfgqdWE7zRSVKixX8zXFgLDmA7toeQomy04SE2No4ePS7gwgsvpEOHDjJ300ft2bOHd955h2++WUaIBS5p4ODKFAcJYb5foH1ZsUvxxf4wlmZHYvdounfvzs0330zLli3NjiYCwMaNG3nyqSc5sP8A3kbe8iIbbnaqKqZB7VdYf7Wi3IqbbryJwYMHExYWZnYyv+WLxdlG+eLAS4CDlC8OHKS13nyy7/HF4vx7xcXFrF69mu+++46ffv4Zj9uNCovEFdMAI64+3tgG6NBIs2OKmqY1yp6PtfAgtsIcbMWH0F6D+IQELrrwQi688ELatWsnZdmP7Nu3j3fffZelS5dixUuP+g6uaOwkKdIHVs/7kTyH4t/7w/k2JwKXobnwwou4+eabSUtLMzuaCDAul4v33nuPt995G6/Vi9HOQKdqCIRxrVKwZlohF1q3bs3EiRNp2rSp2an8ns8VZwCl1JXAC5RvR/eG1vqxv/p6Xy/Ov1dWVsYPP/zAihUrWLN2LSXFxeWfiKqNO6Y+RlwDjJhksEpZCkTKVYq18CDWwoOEFB9CVxzl3qhRY845pys9evSgbdu2cmiJn8vOzmbu3Ll89e9/4/UaZCS66JXiIC0AtrGrTvuKrXyxL4wfc8PQysIll1zC4MGDSU1NNTuaCHB79+7l6WeeZtPGTZAERicD/HXqrwa1U2HdZCXUFsodI+7g6quvlvUwVcQni/Op8qfi/Hter5cdO3awZs0a1qxZy68bf8XweFAWK57opPISHddAFhj6M8ONtfhweVEuOghlxwGIjY2jc+cMOnfuTEZGhhxnGqDy8vJYsGABn336CaVldloleOiVYqd9bQ8WuUsD5QeXbM63sXh/BBuP2QgPC+OqPn0YMGAAycnJZscTQcTr9bJw4UJmzZ6Fy3BhpBvoJn42+lwC1rVWOApdunRh/PjxJCUlmZ0qoEhx9iEOh4MNGzawZs0afl6zhr179gCgQiNwRydjxNbDiK2HDo+XIu2rvAaWkiNYiw5hKz6EpSQXvF5sISG0T0+nS5cuZGRkkJaWJqPKQaS0tJRFixbxwfz3OZp3jAbRmisbl3F+sgtbkP4aGF746UgIi/dFsq/YQq2EePoPuI6+ffvKtnLCVIcOHeKpp55i3bp1/jP6/LtR5vDQcEaPGs0VV1yBkq5Q5aQ4+7C8vDwyMzNZu3Yta9auJf/YMQBUaOQfinScFGmzeA0sJUexFh/CWpSDreQo2utBKUXTtDS6VIwop6eny2IMgcfj4ZtvvuG9uf/H7j17SQiHno3KuKiBk8ggmZ3lMGD5wTC+PBBJnh0aN2rIDQMHcfnllxMaGmp2PCGA8oNTPv/8c2bMnFE++tzOQDf10dHn340ydz2nK+PHjZermNVIirOf0Fpz8OBBfvnlF9atW0dm5rryPaMBFRb130U6LFaKdHXxGlhKj2ItOlQ+qlx6BG2UF+UmTZvSqWNHOnToQPv27WXUTJyU1po1a9bw3ty5rFu/nogQxSX17fyjceDuxFHkUiw5EMaSg+V7MLdrexY3DBzE+eefL1dfhM86fPgwTz39FJlrM9H1NN7OXvChMRC1V2FdbyUiNILRo0fTs2dPGWWuZlKc/ZTWmuzsbNavX8/69evJXLeOguPl82d/K9JxDTBi66PDok1O68e0F0vpMaxFORUjykfQhhuAJk2a0rFjBzp27Ej79u2JjY01OazwR1lZWcybN48Vy5djUZpuyU6uTHFQPyowduI4Yrfw5b4wlh8q3yHj/PPPY9CgG2nXrp3Z0YSoFK01n3zyCTNmzsAb6sXTxQOJJofygFpXvi9zevt0pjw0hcREs0MFBynOAUJrzYEDB34r0mvXZlJUVFj+yYg43DH1Kop0PbAF2kaVVUhrlKOwvCgXHiSkJBftdgDQOCWVjE7/GVGOj483OawIJNnZ2cyfP58vvliMx+2hU6KLvk0cNIn1z504DpRY+GxPBD8fCcVisXLZ5ZczcOBA2SFD+K3t27fz0JSHyMnJwXuWF93KpKkbBWD7yQbFMGTIEIYMGSI7ZtQgKc4BSmvN7t27K+ZIZ7L+l/U4HeUFUEfVwR1bH29sfdn6jhNbxOVgKcohtDgH7SwFoE5iXbp0zqBTp0507NiR2rVrm5xUBIP8/Hw+/vhjPv5oASWlZXRKdHFtUweNY/yjQOeUWvh4dwQ/5YYSHh5G36uvYcCAATIaJgJCWVkZ06dPZ+nSpeULB7sYNXdoiga1R2H9xUpcbBxTHppCp06dauiHixOkOAcJj8dDVlbWb4sNN2/ejGEYYLFiRNfFiG2AJyEFHREEO3Z4jfI5ygX7CSk+9NsWcdExMWR06kSnij8NGjSQuWLCNCUlJSxYsID3359HWZmdLnVdXNvUToNo35zCcbjMwqe7w1mdG0ZoaBj9BwzghhtukClMIuBorfniiy947vnn8Ng8eLp5IK66fyio9QrLLgsZGRlMnjyZhISEav6h4s9IcQ5SdrudX3/9lczMTH5es5bdu3aWfyI8Fnd8YzwJjfHGJIMKkEU7Hmf5cdbH9xFSdBDtcRESGsrZ7duTkVE+qtysWTNZpCR8TnFxMfPnz+fDD+bjcDg5N9nJNU0c1POROdBH7BY+2xPO94fCCAkJ4Zp+1zJo0CCZyiQC3s6dOxk7biwFJQV4zqvGec8esPxkQeUoBg4cyIgRI+S5ykRSnAUAR48eZfXq1axcuZK1mZnlB7GEhOGObYQnIQUjvgFY/WurKOUoxlqwD9vxfViLD4PWxMXH071bN7p160anTp1kizjhNwoKCnj//ff5+KMFuFwuLqrv5LpmdqJCzHmctnvg490RfH0gHIvNRt++V3PjjTfKlCYRVA4fPszYcWPJPpiN0dlAN6ri+6MTrKusqHzFqFGjuPbaa6v29sUpk+Is/kdZWRk///wzq1atYuWq1ZSWFJefZhhTD0+tVDy1moLNN0u0shdiy9tBaOF+KC3frq9x4xQuuKA7559/Pq1bt5ZX6sKv5efn8+677/LJxx8THaq5sVkp5yW7amyGldaw9mgI7+6IJt8OvXr1YujQobJvrAhaRUVF3HfffWzcuBFvey+6RRV1p1KwrbRhtVuZ8tAUevToUTW3K86IFGfxlzweD5s2bWLVqlWs+P57DuXkoKwhuGo1wZPYCm90ovlzor0erPl7CT26DUvRIZRSpKen061bN84//3waNmxobj4hqsH27duZ/szTbN22nbNqebilZWm1T984arfwzrZI1ueFkNa0CWPHjadt27bV+jOF8AdOp5Np06axfPnyqinPdrB9ayNSRfL0U0/L9o0+RIqzqDStNVu3buXzzz9nyZKlOJ0OiKyFM7ElnjrNwFaz0x5U2XFCjm4l9NgutNtBcnI9+vS5iiuuuEIuF4ugYBgGn3/+OS//aw5Oh52rUuxcleogtIp3pvJ44cv9YXyyJwqLLZRhw4dz7bXXYrMF9448QvyeYRhMmTqFFctXYJxrwOmO2bjBttxGqD2UWTNn0aJFiyrNKc6MFGdxWsrKyli6dCmfLfycHdu3oSw2XLVS8SSdVT4KXV20F1veLkKObsVSnIvVaqVHjx5cddVVdOjQQaZhiKB07NgxZs2axdKlS0mLMxiTXlxlJxAWuxQvbYwm67iN7t27M2rUKJKSkqrktoUINE6nk9GjR5O1LQvPBR6oc4o34C2f02w5YuGpp56ia9eu1ZJTnD4pzuKMbd++nUWLFvHVV19jt5fhqZ2Gq1HnKj+x0FqQTfiBn6DsOA0aNKRv3z707NlTVu8LUeG7777j8cemEaFcjGlXRFrcme39fKDEwnO/xlHgtjF+/AR69uxZRUmFCFwFBQWMuGMEh48dxnORB2Iq+Y0aVKbCssfChAkT6N27d7XmFKdHirOoMmVlZcybN4/33puHx/DiTG6Lu357sIac0e0qewFh+3/CWnCA5Hr1uOfuu+nevbvssSzEn9i5cyf3TZpI/rE8bmtdzHnJ7tO6ncwjIczZEkNUbDyPPf4Ebdq0qeKkQgSu7OxsRtwxghJVgudSD1TiYqjao7CstTB48GBuu+226g8pTosUZ1HlcnNzefnll1m6dCkqLApH/Q54w09vd3hb/h5CjmQRER7B0KG30K9fP0JDfXNHDyF8RUFBAZMffIANv27kmiZ2rk1znNL3f7EvjPd2RNKyRXMef+JJOfVPiNOwcuVK7r//frztKo7n/itOsH1l46wWZzFzxkyZdujDpDiLarNp0yZeemkGW7dmnfZtKKXo06cPw4YNkykZQpwCt9vN9OnT+fLLLxnaqpRLGroq9X2rD4cwe1M0PXr04MEHH5S9zoU4Aw88+AArV63Ec7kH/mL2ovpJEXIwhDfffJPU1NQayydO3V8VZ1kuLc5I27Zt+de/5pCVlYXDcWojXifUq1eP+vXrV3EyIQJfSEgIEyZMID//GO/8/DP1Ir20qeX5y+/ZVWjl1S3RpLdrx0MPPURIyJlNsxIi2I0ZPYY1a9ag12mM7gb82QzDXLDst3DTkJukNPs5GXEWQgg/V1JSwp13jCDvcDYPZxSQHPnnez0fcyimrIknPL4ur7z6mlzhEaKKfPTRR7z44osY3Qyo94dParAtsVEvuh5vv/W2XOHxAzLiLIQQASw6Oponn3qa228bzrjV+qTnFWkNEeFhvPDU01KahahCffr04e133ub4ruN46/3hhWse6ELNzXfeLKU5AEhxFkKIANCgQQNemjGT7777jr+6ktitWzeaNm1ag8mECHwhISH07dOXt99+G0qBqP98Tu1UREVHcfHFF5uWT1QdKc5CCBEg0tLSSEtLMzuGEEHpqquu4t1338Wb5UU3qXjx6gFLjoVe/XsRHh5ubkBRJaQ4CyGEEEKcobp169K9e3eWL18Oe/7zcaUUffv2NS+YqFJSnIUQQgghqsCfnQaYkJBAo0aNTEokqpoUZyGEEEKIKhATE0PXrl3NjiGqkRxbI4QQQgghRCVIcRZCCCGEEKISpDgLIYQQQghRCVKchRBCCCGEqAQpzkIIIYQQQlSCFGchhBBCCCEqQYqzEEIIIYQQlSDFWQghhBBCiEqQ4iyEEEIIIUQlSHEWQgghhBCiEqQ4CyGEEEIIUQlSnIUQQgghhKgEKc5CCCGEEEJUghRnIYQQQgghKkGKsxBCCCGEEJUgxVkIIYQQQohKkOIshBBCCCFEJSittdkZKkUpdRTYZ3YOcVrqAHlmhxAiSMn9TwhzyH3Pf6VorRP/7BN+U5yF/1JKrdVaZ5idQ4hgJPc/Icwh973AJFM1hBBCCCGEqAQpzkIIIYQQQlSCFGdRE14xO4AQQUzuf0KYQ+57AUjmOAshhBBCCFEJMuIshBBCCCFEJUhxFlVKKfWGUuqIUmrT7z42VSl1UCn1S8WfK83MKEQgUko1Ukp9q5TaopTarJQaXfHxARXve5VSssJfiCr2F/e9+b973turlPrF7KzizMlUDVGllFIXACXAO1rrthUfmwqUaK2nm5lNiECmlKoH1NNar1NKxQCZwNWABrzAy8A4rfVaE2MKEXBOdt/TWm/53dc8CxRqrR8xK6eoGjazA4jAorVeoZRKNTuHEMFGa30IOFTxdrFSKgtooLVeAqCUMjOeEAHrZPc9YAuAKr/zXQdcbFpIUWVkqoaoKfcopX6tmMqRYHYYIQJZxYvXDsBP5iYRIric5L7XHcjVWu8wI5OoWlKcRU2YA6QBZ1P+qvxZc+MIEbiUUtHAR8AYrXWR2XmECBZ/cd8bCMwzJ5WoajJVQ1Q7rXXuibeVUq8Ci0yMI0TAUkqFUP7EPVdr/bHZeYQIFie77ymlbEA/oJNZ2UTVkhFnUe0qFk6ccA2w6WRfK4Q4PRXzKF8HsrTWz5mdR4hg8Tf3vUuBrVrr7JpPJqqD7KohqpRSah5wIVAHyAWmVLx/NuWr+/cCIyoWUwghqohSqhvwPbCR8l00AO4HwoAZQCJQAPyitf6HKSGFCEAnu+9prb9QSr0F/Ki1/pdZ+UTVkuIshBBCCCFEJchUDSGEEEIIISpBirMQQgghhBCVIMVZCCGEEEKISpDiLIQQQgghRCVIcRZCCCGEEKISpDgLIYSPUkqVVPPt36KUqv+79/cqpepU588UQgh/JsVZCCGC1y1A/b/7IiGEEOXkyG0hhPAjSqlE4F9A44oPjdFar1JKTa34WNOKv1/QWr9U8T2TgZuAo8ABIJPyw4gygLlKKTtwbsXtjVRKXQWEAAO01ltr4t8lhBD+QEachRDCv7wIPK+17gxcC7z2u8+1Av4BdAGmKKVClFInvq49cAXlZRmt9QJgLXCj1vpsrbW94jbytNYdgTnAuJr4BwkhhL+QEWchhPAvlwJtlFIn3o9VSkVXvL1Ya+0EnEqpI0AScD7wmdbaATiUUp//ze1/XPF3JtCvaqMLIYR/k+IshBD+xQKcU1GEf1NRpLL9bu4AAAD1SURBVJ2/+5DB6T3Gn7iN0/1+IYQIWDJVQwgh/MvXwMgT7yilzv6br18FXKWUCq8Yme79u88VAzFVH1EIIQKTjCYIIYTvilRKZf/u/eeAUcAspdSvlD+GrwDuONkNaK3XKKUWAr8CucBGoLDi028B//rD4kAhhBAnobTWZmcQQghRjZRS0VrrEqVUJOVF+3at9TqzcwkhhL+REWchhAh8ryil2gDhwNtSmoUQ4vTIiLMQQgghhBCVIIsDhRBCCCGEqAQpzkIIIYQQQlSCFGchhBBCCCEqQYqzEEIIIYQQlSDFWQghhBBCiEqQ4iyEEEIIIUQl/D8FDwhH2nFQAAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn \n",
    "import matplotlib \n",
    "from matplotlib import pyplot\n",
    "fig = pyplot.figure(figsize=(12,8))\n",
    "\n",
    "seaborn.violinplot(data=df_concat, x=\"Length\", y=\"manufacturability_difficulty_score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcell_epitopes = pd.read_csv(\"../bcell/accessible-linear-bcell-epitopes-grouped-merged-filtered.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_sources</th>\n",
       "      <th>sources</th>\n",
       "      <th>accessible_subsequence_start</th>\n",
       "      <th>accessible_subsequence_end</th>\n",
       "      <th>accessible_subsequence</th>\n",
       "      <th>accessible_subsequence_length</th>\n",
       "      <th>accessible_subsequence_is_4mer_or_longer</th>\n",
       "      <th>IgG</th>\n",
       "      <th>IgA</th>\n",
       "      <th>IgG_and_IgA</th>\n",
       "      <th>accessible_subsequence_contains_polymorphism</th>\n",
       "      <th>accessible_subsequence_contains_glycosite</th>\n",
       "      <th>near_any_feature</th>\n",
       "      <th>in_any_feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>Wang 2020 S456-460; Dahlke 2020 S449-463; Char...</td>\n",
       "      <td>457</td>\n",
       "      <td>473</td>\n",
       "      <td>RKSNLKPFERDISTEIY</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>Charite 2020 S569-583; Charite 2020 S571-585; ...</td>\n",
       "      <td>580</td>\n",
       "      <td>583</td>\n",
       "      <td>QTLE</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>6</td>\n",
       "      <td>Poh 2020 S809-826; Wang 2020 S806-820; Charite...</td>\n",
       "      <td>809</td>\n",
       "      <td>812</td>\n",
       "      <td>PSKP</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2</td>\n",
       "      <td>Dahlke 2020 S1131-1145; Charite 2020 S1141-1155</td>\n",
       "      <td>1142</td>\n",
       "      <td>1146</td>\n",
       "      <td>QPELD</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    num_sources                                            sources  \\\n",
       "9             3  Wang 2020 S456-460; Dahlke 2020 S449-463; Char...   \n",
       "12            4  Charite 2020 S569-583; Charite 2020 S571-585; ...   \n",
       "17            6  Poh 2020 S809-826; Wang 2020 S806-820; Charite...   \n",
       "21            2    Dahlke 2020 S1131-1145; Charite 2020 S1141-1155   \n",
       "\n",
       "    accessible_subsequence_start  accessible_subsequence_end  \\\n",
       "9                            457                         473   \n",
       "12                           580                         583   \n",
       "17                           809                         812   \n",
       "21                          1142                        1146   \n",
       "\n",
       "   accessible_subsequence  accessible_subsequence_length  \\\n",
       "9       RKSNLKPFERDISTEIY                             17   \n",
       "12                   QTLE                              4   \n",
       "17                   PSKP                              4   \n",
       "21                  QPELD                              5   \n",
       "\n",
       "    accessible_subsequence_is_4mer_or_longer   IgG   IgA  IgG_and_IgA  \\\n",
       "9                                       True  True  True         True   \n",
       "12                                      True  True  True         True   \n",
       "17                                      True  True  True         True   \n",
       "21                                      True  True  True         True   \n",
       "\n",
       "    accessible_subsequence_contains_polymorphism  \\\n",
       "9                                          False   \n",
       "12                                         False   \n",
       "17                                         False   \n",
       "21                                         False   \n",
       "\n",
       "    accessible_subsequence_contains_glycosite  near_any_feature  \\\n",
       "9                                       False              True   \n",
       "12                                      False              True   \n",
       "17                                      False              True   \n",
       "21                                      False              True   \n",
       "\n",
       "    in_any_feature  \n",
       "9             True  \n",
       "12           False  \n",
       "17           False  \n",
       "21           False  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcell_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>HLA-I_peptides</th>\n",
       "      <th>HLA-I_haplotypes</th>\n",
       "      <th>HLA-I_pop_freq</th>\n",
       "      <th>HLA-II_peptides</th>\n",
       "      <th>HLA-II_haplotypes</th>\n",
       "      <th>HLA-II_pop_freq</th>\n",
       "      <th>...</th>\n",
       "      <th>difficult_local_hydrophobicity</th>\n",
       "      <th>too_many_cysteines</th>\n",
       "      <th>manufacturability_difficulty_score</th>\n",
       "      <th>Length</th>\n",
       "      <th>overlapping_bcell_epitope</th>\n",
       "      <th>overlapping_bcell_epitope_start</th>\n",
       "      <th>overlapping_bcell_epitope_end</th>\n",
       "      <th>overlapping_bcell_epitope_length</th>\n",
       "      <th>fraction_of_full_bcell_epitope</th>\n",
       "      <th>full_bcell_epitope</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFVFLVLLPLVSSQC</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FVFLVLLPLVSSQCV</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VFLVLLPLVSSQCVN</td>\n",
       "      <td>S</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>VFLVLLPLVSSQCVN</td>\n",
       "      <td>HLA-DRB1*01:01</td>\n",
       "      <td>0.130464</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLVLLPLVSSQCVNL</td>\n",
       "      <td>S</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>FLVLLPLVSSQCVNL</td>\n",
       "      <td>HLA-DRB1*01:01</td>\n",
       "      <td>0.130464</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LVLLPLVSSQCVNLT</td>\n",
       "      <td>S</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LVLLPLVSSQCVNLT</td>\n",
       "      <td>HLA-DRB1*01:01</td>\n",
       "      <td>0.130464</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1909</th>\n",
       "      <td>DSTQAMGYINVFAFPFTIYSLLLCRMN</td>\n",
       "      <td>N</td>\n",
       "      <td>415</td>\n",
       "      <td>441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>STQAMGYINVFAFPFTIYSLLLCRMNS</td>\n",
       "      <td>N</td>\n",
       "      <td>416</td>\n",
       "      <td>442</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1911</th>\n",
       "      <td>TQAMGYINVFAFPFTIYSLLLCRMNSR</td>\n",
       "      <td>N</td>\n",
       "      <td>417</td>\n",
       "      <td>443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1912</th>\n",
       "      <td>QAMGYINVFAFPFTIYSLLLCRMNSRN</td>\n",
       "      <td>N</td>\n",
       "      <td>418</td>\n",
       "      <td>444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>22.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1913</th>\n",
       "      <td>AMGYINVFAFPFTIYSLLLCRMNSRNY</td>\n",
       "      <td>N</td>\n",
       "      <td>419</td>\n",
       "      <td>445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5742 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Sequence Protein  Start  End HLA-I_peptides  \\\n",
       "0                 MFVFLVLLPLVSSQC       S      1   15            NaN   \n",
       "1                 FVFLVLLPLVSSQCV       S      2   16            NaN   \n",
       "2                 VFLVLLPLVSSQCVN       S      3   17            NaN   \n",
       "3                 FLVLLPLVSSQCVNL       S      4   18            NaN   \n",
       "4                 LVLLPLVSSQCVNLT       S      5   19            NaN   \n",
       "...                           ...     ...    ...  ...            ...   \n",
       "1909  DSTQAMGYINVFAFPFTIYSLLLCRMN       N    415  441            NaN   \n",
       "1910  STQAMGYINVFAFPFTIYSLLLCRMNS       N    416  442            NaN   \n",
       "1911  TQAMGYINVFAFPFTIYSLLLCRMNSR       N    417  443            NaN   \n",
       "1912  QAMGYINVFAFPFTIYSLLLCRMNSRN       N    418  444            NaN   \n",
       "1913  AMGYINVFAFPFTIYSLLLCRMNSRNY       N    419  445            NaN   \n",
       "\n",
       "     HLA-I_haplotypes  HLA-I_pop_freq  HLA-II_peptides HLA-II_haplotypes  \\\n",
       "0                 NaN             0.0              NaN               NaN   \n",
       "1                 NaN             0.0              NaN               NaN   \n",
       "2                 NaN             0.0  VFLVLLPLVSSQCVN    HLA-DRB1*01:01   \n",
       "3                 NaN             0.0  FLVLLPLVSSQCVNL    HLA-DRB1*01:01   \n",
       "4                 NaN             0.0  LVLLPLVSSQCVNLT    HLA-DRB1*01:01   \n",
       "...               ...             ...              ...               ...   \n",
       "1909              NaN             0.0              NaN               NaN   \n",
       "1910              NaN             0.0              NaN               NaN   \n",
       "1911              NaN             0.0              NaN               NaN   \n",
       "1912              NaN             0.0              NaN               NaN   \n",
       "1913              NaN             0.0              NaN               NaN   \n",
       "\n",
       "      HLA-II_pop_freq  ... difficult_local_hydrophobicity too_many_cysteines  \\\n",
       "0            0.000000  ...                           True              False   \n",
       "1            0.000000  ...                           True              False   \n",
       "2            0.130464  ...                           True              False   \n",
       "3            0.130464  ...                           True              False   \n",
       "4            0.130464  ...                           True              False   \n",
       "...               ...  ...                            ...                ...   \n",
       "1909         0.000000  ...                           True              False   \n",
       "1910         0.000000  ...                           True              False   \n",
       "1911         0.000000  ...                           True              False   \n",
       "1912         0.000000  ...                           True              False   \n",
       "1913         0.000000  ...                           True              False   \n",
       "\n",
       "     manufacturability_difficulty_score Length  overlapping_bcell_epitope  \\\n",
       "0                                  19.0     15                       None   \n",
       "1                                  15.0     15                       None   \n",
       "2                                  14.0     15                       None   \n",
       "3                                  14.0     15                       None   \n",
       "4                                  14.0     15                       None   \n",
       "...                                 ...    ...                        ...   \n",
       "1909                               21.0     27                       None   \n",
       "1910                               21.0     27                       None   \n",
       "1911                               21.0     27                       None   \n",
       "1912                               22.0     27                       None   \n",
       "1913                               21.0     27                       None   \n",
       "\n",
       "      overlapping_bcell_epitope_start  overlapping_bcell_epitope_end  \\\n",
       "0                                 NaN                            NaN   \n",
       "1                                 NaN                            NaN   \n",
       "2                                 NaN                            NaN   \n",
       "3                                 NaN                            NaN   \n",
       "4                                 NaN                            NaN   \n",
       "...                               ...                            ...   \n",
       "1909                              NaN                            NaN   \n",
       "1910                              NaN                            NaN   \n",
       "1911                              NaN                            NaN   \n",
       "1912                              NaN                            NaN   \n",
       "1913                              NaN                            NaN   \n",
       "\n",
       "      overlapping_bcell_epitope_length  fraction_of_full_bcell_epitope  \\\n",
       "0                                    0                             0.0   \n",
       "1                                    0                             0.0   \n",
       "2                                    0                             0.0   \n",
       "3                                    0                             0.0   \n",
       "4                                    0                             0.0   \n",
       "...                                ...                             ...   \n",
       "1909                                 0                             0.0   \n",
       "1910                                 0                             0.0   \n",
       "1911                                 0                             0.0   \n",
       "1912                                 0                             0.0   \n",
       "1913                                 0                             0.0   \n",
       "\n",
       "      full_bcell_epitope  \n",
       "0                   None  \n",
       "1                   None  \n",
       "2                   None  \n",
       "3                   None  \n",
       "4                   None  \n",
       "...                  ...  \n",
       "1909                None  \n",
       "1910                None  \n",
       "1911                None  \n",
       "1912                None  \n",
       "1913                None  \n",
       "\n",
       "[5742 rows x 35 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from collections import OrderedDict\n",
    "\n",
    "def annotate_overlapping_bcell_epitopes(df, bcell_epitopes):\n",
    "    cols = OrderedDict([\n",
    "        (\"overlapping_bcell_epitope\", []),\n",
    "        (\"overlapping_bcell_epitope_start\", []),\n",
    "        (\"overlapping_bcell_epitope_end\", []),\n",
    "        (\"overlapping_bcell_epitope_length\", []),\n",
    "        (\"fraction_of_full_bcell_epitope\", []),\n",
    "        (\"full_bcell_epitope\", []),\n",
    "    ])\n",
    "\n",
    "    for (prot, start, end) in zip(df_concat.Protein, df_concat.Start, df_concat.End):\n",
    "        found_bcell_epitope = False\n",
    "        if prot == \"S\":\n",
    "            for (epitope, epitope_start, epitope_end) in zip(\n",
    "                    bcell_epitopes.accessible_subsequence,\n",
    "                    bcell_epitopes.accessible_subsequence_start,\n",
    "                    bcell_epitopes.accessible_subsequence_end):\n",
    "                if end >= epitope_start and start <= epitope_end:\n",
    "                    found_bcell_epitope = True\n",
    "                    overlap_start = max(epitope_start, start)\n",
    "                    overlap_end = min(epitope_end, end)\n",
    "                    overlap_length = overlap_end - overlap_start + 1\n",
    "                    overlapping_bcell_epitope = epitope[\n",
    "                        overlap_start - epitope_start:\n",
    "                        overlap_end - epitope_start + 1]\n",
    "                    overlap_percent_of_full_bcell_epitope = overlap_length / len(epitope)\n",
    "\n",
    "                    cols[\"overlapping_bcell_epitope\"].append(overlapping_bcell_epitope)\n",
    "                    cols[\"overlapping_bcell_epitope_start\"].append(overlap_start)\n",
    "                    cols[\"overlapping_bcell_epitope_end\"].append(overlap_end)\n",
    "                    cols[\"overlapping_bcell_epitope_length\"].append(overlap_length)\n",
    "                    cols[\"fraction_of_full_bcell_epitope\"].append(overlap_percent_of_full_bcell_epitope)\n",
    "                    cols[\"full_bcell_epitope\"].append(epitope)\n",
    "                    break\n",
    "        if not found_bcell_epitope:\n",
    "            cols[\"overlapping_bcell_epitope\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_start\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_end\"].append(None)\n",
    "            cols[\"overlapping_bcell_epitope_length\"].append(0)\n",
    "            cols[\"fraction_of_full_bcell_epitope\"].append(0.0)\n",
    "            cols[\"full_bcell_epitope\"].append(None)\n",
    "        \n",
    "    df_with_bcell_epitopes = df.copy()\n",
    "    for (col_name, col_values) in cols.items():\n",
    "        df_with_bcell_epitopes[col_name] = col_values\n",
    "    return df_with_bcell_epitopes\n",
    "\n",
    "df_with_bcell_epitopes = annotate_overlapping_bcell_epitopes(df_concat, bcell_epitopes);\n",
    "df_with_bcell_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
      "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
      "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
      "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
      "       'num_extremely_unstable_dipeptides',\n",
      "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
      "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
      "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
      "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
      "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df_concat.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Residue Proper Numbering</th>\n",
       "      <th>SiteSpecific Accessiblity</th>\n",
       "      <th>Nude Accessibility</th>\n",
       "      <th>Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>1142</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1116</th>\n",
       "      <td>1143</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>1144</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1118</th>\n",
       "      <td>1145</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119</th>\n",
       "      <td>1146</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1120 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Residue Proper Numbering  SiteSpecific Accessiblity  Nude Accessibility  \\\n",
       "0                           27                       0.94                0.99   \n",
       "1                           28                       0.99                1.00   \n",
       "2                           29                       0.15                0.26   \n",
       "3                           30                       0.36                0.76   \n",
       "4                           31                       0.00                0.00   \n",
       "...                        ...                        ...                 ...   \n",
       "1115                      1142                       0.78                0.89   \n",
       "1116                      1143                       0.66                0.70   \n",
       "1117                      1144                       0.78                0.79   \n",
       "1118                      1145                       0.85                0.86   \n",
       "1119                      1146                       1.00                1.00   \n",
       "\n",
       "      Difference  \n",
       "0           0.05  \n",
       "1           0.01  \n",
       "2           0.11  \n",
       "3           0.40  \n",
       "4           0.00  \n",
       "...          ...  \n",
       "1115        0.11  \n",
       "1116        0.04  \n",
       "1117        0.01  \n",
       "1118        0.00  \n",
       "1119        0.00  \n",
       "\n",
       "[1120 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_accessibility = pd.read_csv(\"../solvent-accessibility/Woods-Glycans-MD-Site-Specific-Accessibility.csv\");\n",
    "df_accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def num_contiguous_accessible_residues(protein, start, end, aa_num_to_accessibility, cutoff=0.2):\n",
    "    if protein != \"S\":\n",
    "        return 0\n",
    "    kmers = []\n",
    "    curr_kmer = 0\n",
    "    for aa_num in range(start, end + 1):\n",
    "        accessibility = aa_num_to_accessibility.get(aa_num, 0)\n",
    "        if accessibility >= cutoff:\n",
    "            curr_kmer += 1\n",
    "        elif curr_kmer > 0:\n",
    "            kmers.append(curr_kmer)\n",
    "            curr_kmer = 0\n",
    "    if curr_kmer > 0:\n",
    "        kmers.append(curr_kmer)\n",
    "    if len(kmers) == 0:\n",
    "        return 0\n",
    "    else:\n",
    "        return max(kmers)\n",
    "\n",
    "def annotate_accesibility(df, cutoff=0.2):\n",
    "    df = df.copy()\n",
    "    df_accessibility = pd.read_csv(\"../solvent-accessibility/Woods-Glycans-MD-Site-Specific-Accessibility.csv\");\n",
    "    aa_num_to_accessibility = dict(\n",
    "        zip(df_accessibility[\"Residue Proper Numbering\"], \n",
    "            df_accessibility[\"SiteSpecific Accessiblity\"]))\n",
    "    \n",
    "    df[\"longest_contiguous_accessible_region_in_spike\"] = [\n",
    "        num_contiguous_accessible_residues(p, s, e, aa_num_to_accessibility, cutoff=cutoff)\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    df[\"max_accessibility\"] = [\n",
    "        max([aa_num_to_accessibility.get(aa_num, 0) for aa_num in range(s, e + 1)])\n",
    "        if p == \"S\"\n",
    "        else 0\n",
    "        for (p, s, e) in zip(df.Protein, df.Start, df.End)\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "def annotate_selection_criteria(df, rounding_decimals=2):\n",
    "    df = df.copy()\n",
    "    df[\"manufacturability_feasibility_score\"] = 2.0 ** (\n",
    "        -df.manufacturability_difficulty_score / 10.0)\n",
    "    df[\"has mouse MHC-I b\"] = ~df[\"Mouse MHC-I b\"].isnull()\n",
    "    df[\"has mouse MHC-II b\"] = ~df[\"Mouse MHC-II b\"].isnull()\n",
    "    df[\"has mouse MHC-I d\"] = ~df[\"Mouse MHC-I d\"].isnull()\n",
    "    df[\"has mouse MHC-II d\"] = ~df[\"Mouse MHC-II d\"].isnull()\n",
    "    df[\"mouse H2-b count\"] = (\n",
    "        df[\"has mouse MHC-I b\"].astype(int) + \n",
    "        df[\"has mouse MHC-II b\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse H2-d count\"] = (\n",
    "        df[\"has mouse MHC-I d\"].astype(int) + \n",
    "        df[\"has mouse MHC-II d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse class I count\"] = (\n",
    "        df[\"has mouse MHC-I b\"].astype(int) + \n",
    "        df[\"has mouse MHC-I d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse class II count\"] = (\n",
    "        df[\"has mouse MHC-II b\"].astype(int) + \n",
    "        df[\"has mouse MHC-II d\"].astype(int) \n",
    "    )\n",
    "    df[\"mouse count\"] = df[\"mouse H2-b count\"] + df[\"mouse H2-d count\"]\n",
    "    df[\"HLA-I_pop_freq_rounded\"] = df[\"HLA-I_pop_freq\"].round(rounding_decimals)\n",
    "    df[\"HLA-II_pop_freq_rounded\"] = df[\"HLA-II_pop_freq\"].round(rounding_decimals)\n",
    "    df[\"Total_frequency_rounded\"] = df[\"Total_frequency\"].round(rounding_decimals)\n",
    "    df[\"manufacturability_feasibility_score_rounded\"] = \\\n",
    "        df[\"manufacturability_feasibility_score\"].round(rounding_decimals)\n",
    "\n",
    "    return annotate_accesibility(df)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>HLA-I_peptides</th>\n",
       "      <th>HLA-I_haplotypes</th>\n",
       "      <th>HLA-I_pop_freq</th>\n",
       "      <th>HLA-II_peptides</th>\n",
       "      <th>HLA-II_haplotypes</th>\n",
       "      <th>HLA-II_pop_freq</th>\n",
       "      <th>...</th>\n",
       "      <th>mouse H2-d count</th>\n",
       "      <th>mouse class I count</th>\n",
       "      <th>mouse class II count</th>\n",
       "      <th>mouse count</th>\n",
       "      <th>HLA-I_pop_freq_rounded</th>\n",
       "      <th>HLA-II_pop_freq_rounded</th>\n",
       "      <th>Total_frequency_rounded</th>\n",
       "      <th>manufacturability_feasibility_score_rounded</th>\n",
       "      <th>longest_contiguous_accessible_region_in_spike</th>\n",
       "      <th>max_accessibility</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>TLLALHRSYLTPGDSSSGWTA</td>\n",
       "      <td>S</td>\n",
       "      <td>240</td>\n",
       "      <td>260</td>\n",
       "      <td>LLALHRSYL</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.171795</td>\n",
       "      <td>TLLALHRSYLTPGDS</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*15:01</td>\n",
       "      <td>0.289063</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.71</td>\n",
       "      <td>16</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>LLALHRSYLTPGDSSSGWTAG</td>\n",
       "      <td>S</td>\n",
       "      <td>241</td>\n",
       "      <td>261</td>\n",
       "      <td>LLALHRSYL</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.171795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>LALHRSYLTPGDSSSGWTAGA</td>\n",
       "      <td>S</td>\n",
       "      <td>242</td>\n",
       "      <td>262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>ALHRSYLTPGDSSSGWTAGAA</td>\n",
       "      <td>S</td>\n",
       "      <td>243</td>\n",
       "      <td>263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LTPGDSSSGWTAGAA</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>LHRSYLTPGDSSSGWTAGAAA</td>\n",
       "      <td>S</td>\n",
       "      <td>244</td>\n",
       "      <td>264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LTPGDSSSGWTAGAA,TPGDSSSGWTAGAAA</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>HRSYLTPGDSSSGWTAGAAAY</td>\n",
       "      <td>S</td>\n",
       "      <td>245</td>\n",
       "      <td>265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>LTPGDSSSGWTAGAA,PGDSSSGWTAGAAAY,TPGDSSSGWTAGAAA</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>RSYLTPGDSSSGWTAGAAAYY</td>\n",
       "      <td>S</td>\n",
       "      <td>246</td>\n",
       "      <td>266</td>\n",
       "      <td>WTAGAAAYY</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>0.235236</td>\n",
       "      <td>GDSSSGWTAGAAAYY,LTPGDSSSGWTAGAA,PGDSSSGWTAGAAA...</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.81</td>\n",
       "      <td>16</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>CTEVPVAIHADQLTPTWRVYS</td>\n",
       "      <td>S</td>\n",
       "      <td>617</td>\n",
       "      <td>637</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.35</td>\n",
       "      <td>16</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>TEVPVAIHADQLTPTWRVYST</td>\n",
       "      <td>S</td>\n",
       "      <td>618</td>\n",
       "      <td>638</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.44</td>\n",
       "      <td>17</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>EVPVAIHADQLTPTWRVYSTG</td>\n",
       "      <td>S</td>\n",
       "      <td>619</td>\n",
       "      <td>639</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.41</td>\n",
       "      <td>18</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>VPVAIHADQLTPTWRVYSTGS</td>\n",
       "      <td>S</td>\n",
       "      <td>620</td>\n",
       "      <td>640</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.44</td>\n",
       "      <td>19</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>PVAIHADQLTPTWRVYSTGSN</td>\n",
       "      <td>S</td>\n",
       "      <td>621</td>\n",
       "      <td>641</td>\n",
       "      <td>HADQLTPTW</td>\n",
       "      <td>HLA-B*35:01</td>\n",
       "      <td>0.111583</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.87</td>\n",
       "      <td>20</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>VAIHADQLTPTWRVYSTGSNV</td>\n",
       "      <td>S</td>\n",
       "      <td>622</td>\n",
       "      <td>642</td>\n",
       "      <td>HADQLTPTW</td>\n",
       "      <td>HLA-B*35:01</td>\n",
       "      <td>0.111583</td>\n",
       "      <td>VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.87</td>\n",
       "      <td>21</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>AIHADQLTPTWRVYSTGSNVF</td>\n",
       "      <td>S</td>\n",
       "      <td>623</td>\n",
       "      <td>643</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF</td>\n",
       "      <td>HLA-DRB1*07:01</td>\n",
       "      <td>0.226271</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1.00</td>\n",
       "      <td>21</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>IHADQLTPTWRVYSTGSNVFQ</td>\n",
       "      <td>S</td>\n",
       "      <td>624</td>\n",
       "      <td>644</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,TPTWRVYSTGSNVFQ</td>\n",
       "      <td>HLA-DRB1*07:01</td>\n",
       "      <td>0.226271</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.06</td>\n",
       "      <td>1.00</td>\n",
       "      <td>21</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>HADQLTPTWRVYSTGSNVFQT</td>\n",
       "      <td>S</td>\n",
       "      <td>625</td>\n",
       "      <td>645</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVFQ</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.318226</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.00</td>\n",
       "      <td>20</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>ADQLTPTWRVYSTGSNVFQTR</td>\n",
       "      <td>S</td>\n",
       "      <td>626</td>\n",
       "      <td>646</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVF...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.318226</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.05</td>\n",
       "      <td>1.00</td>\n",
       "      <td>19</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>DQLTPTWRVYSTGSNVFQTRA</td>\n",
       "      <td>S</td>\n",
       "      <td>627</td>\n",
       "      <td>647</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVF...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.00</td>\n",
       "      <td>18</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>QLTPTWRVYSTGSNVFQTRAG</td>\n",
       "      <td>S</td>\n",
       "      <td>628</td>\n",
       "      <td>648</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.93</td>\n",
       "      <td>17</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>LTPTWRVYSTGSNVFQTRAGC</td>\n",
       "      <td>S</td>\n",
       "      <td>629</td>\n",
       "      <td>649</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.76</td>\n",
       "      <td>16</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NITRFQTLLALHRSYLTPGDSSSGWTA</td>\n",
       "      <td>S</td>\n",
       "      <td>234</td>\n",
       "      <td>260</td>\n",
       "      <td>LLALHRSYL,NITRFQTL,TRFQTLLAL</td>\n",
       "      <td>HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02</td>\n",
       "      <td>0.608186</td>\n",
       "      <td>FQTLLALHRSYLTPG,ITRFQTLLALHRSYL,NITRFQTLLALHRS...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...</td>\n",
       "      <td>0.574336</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.66</td>\n",
       "      <td>16</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>ITRFQTLLALHRSYLTPGDSSSGWTAG</td>\n",
       "      <td>S</td>\n",
       "      <td>235</td>\n",
       "      <td>261</td>\n",
       "      <td>LLALHRSYL,TRFQTLLAL</td>\n",
       "      <td>HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02</td>\n",
       "      <td>0.608186</td>\n",
       "      <td>FQTLLALHRSYLTPG,ITRFQTLLALHRSYL,QTLLALHRSYLTPG...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...</td>\n",
       "      <td>0.574336</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>TRFQTLLALHRSYLTPGDSSSGWTAGA</td>\n",
       "      <td>S</td>\n",
       "      <td>236</td>\n",
       "      <td>262</td>\n",
       "      <td>LLALHRSYL,TRFQTLLAL</td>\n",
       "      <td>HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02</td>\n",
       "      <td>0.608186</td>\n",
       "      <td>FQTLLALHRSYLTPG,QTLLALHRSYLTPGD,RFQTLLALHRSYLT...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...</td>\n",
       "      <td>0.574336</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>RFQTLLALHRSYLTPGDSSSGWTAGAA</td>\n",
       "      <td>S</td>\n",
       "      <td>237</td>\n",
       "      <td>263</td>\n",
       "      <td>LLALHRSYL</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.171795</td>\n",
       "      <td>FQTLLALHRSYLTPG,LTPGDSSSGWTAGAA,QTLLALHRSYLTPG...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...</td>\n",
       "      <td>0.612766</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>FQTLLALHRSYLTPGDSSSGWTAGAAA</td>\n",
       "      <td>S</td>\n",
       "      <td>238</td>\n",
       "      <td>264</td>\n",
       "      <td>LLALHRSYL</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.171795</td>\n",
       "      <td>FQTLLALHRSYLTPG,LTPGDSSSGWTAGAA,QTLLALHRSYLTPG...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...</td>\n",
       "      <td>0.612766</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>QTLLALHRSYLTPGDSSSGWTAGAAAY</td>\n",
       "      <td>S</td>\n",
       "      <td>239</td>\n",
       "      <td>265</td>\n",
       "      <td>LLALHRSYL</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.171795</td>\n",
       "      <td>LTPGDSSSGWTAGAA,PGDSSSGWTAGAAAY,QTLLALHRSYLTPG...</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*07:01,HLA-D...</td>\n",
       "      <td>0.612766</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.66</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>TLLALHRSYLTPGDSSSGWTAGAAAYY</td>\n",
       "      <td>S</td>\n",
       "      <td>240</td>\n",
       "      <td>266</td>\n",
       "      <td>LLALHRSYL,WTAGAAAYY</td>\n",
       "      <td>HLA-B*08:01,HLA-A*01:01</td>\n",
       "      <td>0.366618</td>\n",
       "      <td>GDSSSGWTAGAAAYY,LTPGDSSSGWTAGAA,PGDSSSGWTAGAAA...</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*13:01,HLA-D...</td>\n",
       "      <td>0.437626</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>LLALHRSYLTPGDSSSGWTAGAAAYYV</td>\n",
       "      <td>S</td>\n",
       "      <td>241</td>\n",
       "      <td>267</td>\n",
       "      <td>LLALHRSYL,WTAGAAAYY</td>\n",
       "      <td>HLA-B*08:01,HLA-A*01:01</td>\n",
       "      <td>0.366618</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.491200</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>LALHRSYLTPGDSSSGWTAGAAAYYVG</td>\n",
       "      <td>S</td>\n",
       "      <td>242</td>\n",
       "      <td>268</td>\n",
       "      <td>WTAGAAAYY</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>0.235236</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.557580</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>ALHRSYLTPGDSSSGWTAGAAAYYVGY</td>\n",
       "      <td>S</td>\n",
       "      <td>243</td>\n",
       "      <td>269</td>\n",
       "      <td>WTAGAAAYY</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>0.235236</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.557580</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>LHRSYLTPGDSSSGWTAGAAAYYVGYL</td>\n",
       "      <td>S</td>\n",
       "      <td>244</td>\n",
       "      <td>270</td>\n",
       "      <td>AAAYYVGYL,WTAGAAAYY</td>\n",
       "      <td>HLA-C*03:04,HLA-A*01:01</td>\n",
       "      <td>0.345119</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.557580</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>HRSYLTPGDSSSGWTAGAAAYYVGYLQ</td>\n",
       "      <td>S</td>\n",
       "      <td>245</td>\n",
       "      <td>271</td>\n",
       "      <td>AAAYYVGYL,WTAGAAAYY</td>\n",
       "      <td>HLA-C*03:04,HLA-A*01:01</td>\n",
       "      <td>0.345119</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,GWTAGAAAYYVGYL...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.557580</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.81</td>\n",
       "      <td>17</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>RSYLTPGDSSSGWTAGAAAYYVGYLQP</td>\n",
       "      <td>S</td>\n",
       "      <td>246</td>\n",
       "      <td>272</td>\n",
       "      <td>AAAYYVGYL,WTAGAAAYY</td>\n",
       "      <td>HLA-C*03:04,HLA-A*01:01</td>\n",
       "      <td>0.345119</td>\n",
       "      <td>DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,GWTAGAAAYYVGYL...</td>\n",
       "      <td>HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...</td>\n",
       "      <td>0.557580</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.71</td>\n",
       "      <td>16</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>LYQDVNCTEVPVAIHADQLTPTWRVYS</td>\n",
       "      <td>S</td>\n",
       "      <td>611</td>\n",
       "      <td>637</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.38</td>\n",
       "      <td>16</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>YQDVNCTEVPVAIHADQLTPTWRVYST</td>\n",
       "      <td>S</td>\n",
       "      <td>612</td>\n",
       "      <td>638</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.38</td>\n",
       "      <td>17</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>QDVNCTEVPVAIHADQLTPTWRVYSTG</td>\n",
       "      <td>S</td>\n",
       "      <td>613</td>\n",
       "      <td>639</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.35</td>\n",
       "      <td>18</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>DVNCTEVPVAIHADQLTPTWRVYSTGS</td>\n",
       "      <td>S</td>\n",
       "      <td>614</td>\n",
       "      <td>640</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.38</td>\n",
       "      <td>19</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>VNCTEVPVAIHADQLTPTWRVYSTGSN</td>\n",
       "      <td>S</td>\n",
       "      <td>615</td>\n",
       "      <td>641</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.38</td>\n",
       "      <td>20</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>NCTEVPVAIHADQLTPTWRVYSTGSNV</td>\n",
       "      <td>S</td>\n",
       "      <td>616</td>\n",
       "      <td>642</td>\n",
       "      <td>HADQLTPTW,VPVAIHADQL</td>\n",
       "      <td>HLA-B*35:01,HLA-B*07:02</td>\n",
       "      <td>0.289710</td>\n",
       "      <td>PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>0.197497</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.35</td>\n",
       "      <td>21</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>CTEVPVAIHADQLTPTWRVYSTGSNVF</td>\n",
       "      <td>S</td>\n",
       "      <td>617</td>\n",
       "      <td>643</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02,HLA-B*07:02</td>\n",
       "      <td>0.410640</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PVAIHADQLTPTWRV,VAIHADQLTPTWRVY</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*03:01</td>\n",
       "      <td>0.379080</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.35</td>\n",
       "      <td>22</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>TEVPVAIHADQLTPTWRVYSTGSNVFQ</td>\n",
       "      <td>S</td>\n",
       "      <td>618</td>\n",
       "      <td>644</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02,HLA-B*07:02</td>\n",
       "      <td>0.410640</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PVAIHADQLTPTWRV,TPTWRVYSTGSNVF...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*03:01</td>\n",
       "      <td>0.379080</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>23</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>EVPVAIHADQLTPTWRVYSTGSNVFQT</td>\n",
       "      <td>S</td>\n",
       "      <td>619</td>\n",
       "      <td>645</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02,HLA-B*07:02</td>\n",
       "      <td>0.410640</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01</td>\n",
       "      <td>0.452874</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.41</td>\n",
       "      <td>23</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>VPVAIHADQLTPTWRVYSTGSNVFQTR</td>\n",
       "      <td>S</td>\n",
       "      <td>620</td>\n",
       "      <td>646</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02,HLA-B*07:02</td>\n",
       "      <td>0.410640</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01</td>\n",
       "      <td>0.452874</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.44</td>\n",
       "      <td>23</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>PVAIHADQLTPTWRVYSTGSNVFQTRA</td>\n",
       "      <td>S</td>\n",
       "      <td>621</td>\n",
       "      <td>647</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01,H...</td>\n",
       "      <td>0.567206</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.87</td>\n",
       "      <td>23</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>VAIHADQLTPTWRVYSTGSNVFQTRAG</td>\n",
       "      <td>S</td>\n",
       "      <td>622</td>\n",
       "      <td>648</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01,H...</td>\n",
       "      <td>0.567206</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.87</td>\n",
       "      <td>23</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>AIHADQLTPTWRVYSTGSNVFQTRAGC</td>\n",
       "      <td>S</td>\n",
       "      <td>623</td>\n",
       "      <td>649</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.76</td>\n",
       "      <td>22</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>IHADQLTPTWRVYSTGSNVFQTRAGCL</td>\n",
       "      <td>S</td>\n",
       "      <td>624</td>\n",
       "      <td>650</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.87</td>\n",
       "      <td>21</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>HADQLTPTWRVYSTGSNVFQTRAGCLI</td>\n",
       "      <td>S</td>\n",
       "      <td>625</td>\n",
       "      <td>651</td>\n",
       "      <td>HADQLTPTW,RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-B*35:01,HLA-A*24:02</td>\n",
       "      <td>0.262839</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.460695</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.76</td>\n",
       "      <td>20</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>ADQLTPTWRVYSTGSNVFQTRAGCLIG</td>\n",
       "      <td>S</td>\n",
       "      <td>626</td>\n",
       "      <td>652</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...</td>\n",
       "      <td>HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*01:01,H...</td>\n",
       "      <td>0.531054</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.76</td>\n",
       "      <td>19</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>DQLTPTWRVYSTGSNVFQTRAGCLIGA</td>\n",
       "      <td>S</td>\n",
       "      <td>627</td>\n",
       "      <td>653</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQ...</td>\n",
       "      <td>HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DRB1*04:01,H...</td>\n",
       "      <td>0.531054</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.76</td>\n",
       "      <td>18</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>QLTPTWRVYSTGSNVFQTRAGCLIGAE</td>\n",
       "      <td>S</td>\n",
       "      <td>628</td>\n",
       "      <td>654</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQ...</td>\n",
       "      <td>HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DRB1*04:01,H...</td>\n",
       "      <td>0.531054</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.71</td>\n",
       "      <td>17</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>LTPTWRVYSTGSNVFQTRAGCLIGAEH</td>\n",
       "      <td>S</td>\n",
       "      <td>629</td>\n",
       "      <td>655</td>\n",
       "      <td>RVYSTGSNVF,VYSTGSNVF</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>0.170254</td>\n",
       "      <td>GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,NVFQTRAGCLIGAE...</td>\n",
       "      <td>HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DQA1*05:05/D...</td>\n",
       "      <td>0.531054</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.66</td>\n",
       "      <td>16</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52 rows  45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Sequence Protein  Start  End  \\\n",
       "239        TLLALHRSYLTPGDSSSGWTA       S    240  260   \n",
       "240        LLALHRSYLTPGDSSSGWTAG       S    241  261   \n",
       "241        LALHRSYLTPGDSSSGWTAGA       S    242  262   \n",
       "242        ALHRSYLTPGDSSSGWTAGAA       S    243  263   \n",
       "243        LHRSYLTPGDSSSGWTAGAAA       S    244  264   \n",
       "244        HRSYLTPGDSSSGWTAGAAAY       S    245  265   \n",
       "245        RSYLTPGDSSSGWTAGAAAYY       S    246  266   \n",
       "616        CTEVPVAIHADQLTPTWRVYS       S    617  637   \n",
       "617        TEVPVAIHADQLTPTWRVYST       S    618  638   \n",
       "618        EVPVAIHADQLTPTWRVYSTG       S    619  639   \n",
       "619        VPVAIHADQLTPTWRVYSTGS       S    620  640   \n",
       "620        PVAIHADQLTPTWRVYSTGSN       S    621  641   \n",
       "621        VAIHADQLTPTWRVYSTGSNV       S    622  642   \n",
       "622        AIHADQLTPTWRVYSTGSNVF       S    623  643   \n",
       "623        IHADQLTPTWRVYSTGSNVFQ       S    624  644   \n",
       "624        HADQLTPTWRVYSTGSNVFQT       S    625  645   \n",
       "625        ADQLTPTWRVYSTGSNVFQTR       S    626  646   \n",
       "626        DQLTPTWRVYSTGSNVFQTRA       S    627  647   \n",
       "627        QLTPTWRVYSTGSNVFQTRAG       S    628  648   \n",
       "628        LTPTWRVYSTGSNVFQTRAGC       S    629  649   \n",
       "233  NITRFQTLLALHRSYLTPGDSSSGWTA       S    234  260   \n",
       "234  ITRFQTLLALHRSYLTPGDSSSGWTAG       S    235  261   \n",
       "235  TRFQTLLALHRSYLTPGDSSSGWTAGA       S    236  262   \n",
       "236  RFQTLLALHRSYLTPGDSSSGWTAGAA       S    237  263   \n",
       "237  FQTLLALHRSYLTPGDSSSGWTAGAAA       S    238  264   \n",
       "238  QTLLALHRSYLTPGDSSSGWTAGAAAY       S    239  265   \n",
       "239  TLLALHRSYLTPGDSSSGWTAGAAAYY       S    240  266   \n",
       "240  LLALHRSYLTPGDSSSGWTAGAAAYYV       S    241  267   \n",
       "241  LALHRSYLTPGDSSSGWTAGAAAYYVG       S    242  268   \n",
       "242  ALHRSYLTPGDSSSGWTAGAAAYYVGY       S    243  269   \n",
       "243  LHRSYLTPGDSSSGWTAGAAAYYVGYL       S    244  270   \n",
       "244  HRSYLTPGDSSSGWTAGAAAYYVGYLQ       S    245  271   \n",
       "245  RSYLTPGDSSSGWTAGAAAYYVGYLQP       S    246  272   \n",
       "610  LYQDVNCTEVPVAIHADQLTPTWRVYS       S    611  637   \n",
       "611  YQDVNCTEVPVAIHADQLTPTWRVYST       S    612  638   \n",
       "612  QDVNCTEVPVAIHADQLTPTWRVYSTG       S    613  639   \n",
       "613  DVNCTEVPVAIHADQLTPTWRVYSTGS       S    614  640   \n",
       "614  VNCTEVPVAIHADQLTPTWRVYSTGSN       S    615  641   \n",
       "615  NCTEVPVAIHADQLTPTWRVYSTGSNV       S    616  642   \n",
       "616  CTEVPVAIHADQLTPTWRVYSTGSNVF       S    617  643   \n",
       "617  TEVPVAIHADQLTPTWRVYSTGSNVFQ       S    618  644   \n",
       "618  EVPVAIHADQLTPTWRVYSTGSNVFQT       S    619  645   \n",
       "619  VPVAIHADQLTPTWRVYSTGSNVFQTR       S    620  646   \n",
       "620  PVAIHADQLTPTWRVYSTGSNVFQTRA       S    621  647   \n",
       "621  VAIHADQLTPTWRVYSTGSNVFQTRAG       S    622  648   \n",
       "622  AIHADQLTPTWRVYSTGSNVFQTRAGC       S    623  649   \n",
       "623  IHADQLTPTWRVYSTGSNVFQTRAGCL       S    624  650   \n",
       "624  HADQLTPTWRVYSTGSNVFQTRAGCLI       S    625  651   \n",
       "625  ADQLTPTWRVYSTGSNVFQTRAGCLIG       S    626  652   \n",
       "626  DQLTPTWRVYSTGSNVFQTRAGCLIGA       S    627  653   \n",
       "627  QLTPTWRVYSTGSNVFQTRAGCLIGAE       S    628  654   \n",
       "628  LTPTWRVYSTGSNVFQTRAGCLIGAEH       S    629  655   \n",
       "\n",
       "                                HLA-I_peptides  \\\n",
       "239                                  LLALHRSYL   \n",
       "240                                  LLALHRSYL   \n",
       "241                                        NaN   \n",
       "242                                        NaN   \n",
       "243                                        NaN   \n",
       "244                                        NaN   \n",
       "245                                  WTAGAAAYY   \n",
       "616                       HADQLTPTW,VPVAIHADQL   \n",
       "617                       HADQLTPTW,VPVAIHADQL   \n",
       "618                       HADQLTPTW,VPVAIHADQL   \n",
       "619                       HADQLTPTW,VPVAIHADQL   \n",
       "620                                  HADQLTPTW   \n",
       "621                                  HADQLTPTW   \n",
       "622             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "623             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "624             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "625                       RVYSTGSNVF,VYSTGSNVF   \n",
       "626                       RVYSTGSNVF,VYSTGSNVF   \n",
       "627                       RVYSTGSNVF,VYSTGSNVF   \n",
       "628                       RVYSTGSNVF,VYSTGSNVF   \n",
       "233               LLALHRSYL,NITRFQTL,TRFQTLLAL   \n",
       "234                        LLALHRSYL,TRFQTLLAL   \n",
       "235                        LLALHRSYL,TRFQTLLAL   \n",
       "236                                  LLALHRSYL   \n",
       "237                                  LLALHRSYL   \n",
       "238                                  LLALHRSYL   \n",
       "239                        LLALHRSYL,WTAGAAAYY   \n",
       "240                        LLALHRSYL,WTAGAAAYY   \n",
       "241                                  WTAGAAAYY   \n",
       "242                                  WTAGAAAYY   \n",
       "243                        AAAYYVGYL,WTAGAAAYY   \n",
       "244                        AAAYYVGYL,WTAGAAAYY   \n",
       "245                        AAAYYVGYL,WTAGAAAYY   \n",
       "610                       HADQLTPTW,VPVAIHADQL   \n",
       "611                       HADQLTPTW,VPVAIHADQL   \n",
       "612                       HADQLTPTW,VPVAIHADQL   \n",
       "613                       HADQLTPTW,VPVAIHADQL   \n",
       "614                       HADQLTPTW,VPVAIHADQL   \n",
       "615                       HADQLTPTW,VPVAIHADQL   \n",
       "616  HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF   \n",
       "617  HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF   \n",
       "618  HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF   \n",
       "619  HADQLTPTW,RVYSTGSNVF,VPVAIHADQL,VYSTGSNVF   \n",
       "620             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "621             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "622             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "623             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "624             HADQLTPTW,RVYSTGSNVF,VYSTGSNVF   \n",
       "625                       RVYSTGSNVF,VYSTGSNVF   \n",
       "626                       RVYSTGSNVF,VYSTGSNVF   \n",
       "627                       RVYSTGSNVF,VYSTGSNVF   \n",
       "628                       RVYSTGSNVF,VYSTGSNVF   \n",
       "\n",
       "                                    HLA-I_haplotypes  HLA-I_pop_freq  \\\n",
       "239                                      HLA-B*08:01        0.171795   \n",
       "240                                      HLA-B*08:01        0.171795   \n",
       "241                                              NaN        0.000000   \n",
       "242                                              NaN        0.000000   \n",
       "243                                              NaN        0.000000   \n",
       "244                                              NaN        0.000000   \n",
       "245                                      HLA-A*01:01        0.235236   \n",
       "616                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "617                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "618                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "619                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "620                                      HLA-B*35:01        0.111583   \n",
       "621                                      HLA-B*35:01        0.111583   \n",
       "622                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "623                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "624                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "625                                      HLA-A*24:02        0.170254   \n",
       "626                                      HLA-A*24:02        0.170254   \n",
       "627                                      HLA-A*24:02        0.170254   \n",
       "628                                      HLA-A*24:02        0.170254   \n",
       "233  HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02        0.608186   \n",
       "234  HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02        0.608186   \n",
       "235  HLA-B*08:01,HLA-C*06:02,HLA-C*07:01,HLA-C*07:02        0.608186   \n",
       "236                                      HLA-B*08:01        0.171795   \n",
       "237                                      HLA-B*08:01        0.171795   \n",
       "238                                      HLA-B*08:01        0.171795   \n",
       "239                          HLA-B*08:01,HLA-A*01:01        0.366618   \n",
       "240                          HLA-B*08:01,HLA-A*01:01        0.366618   \n",
       "241                                      HLA-A*01:01        0.235236   \n",
       "242                                      HLA-A*01:01        0.235236   \n",
       "243                          HLA-C*03:04,HLA-A*01:01        0.345119   \n",
       "244                          HLA-C*03:04,HLA-A*01:01        0.345119   \n",
       "245                          HLA-C*03:04,HLA-A*01:01        0.345119   \n",
       "610                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "611                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "612                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "613                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "614                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "615                          HLA-B*35:01,HLA-B*07:02        0.289710   \n",
       "616              HLA-B*35:01,HLA-A*24:02,HLA-B*07:02        0.410640   \n",
       "617              HLA-B*35:01,HLA-A*24:02,HLA-B*07:02        0.410640   \n",
       "618              HLA-B*35:01,HLA-A*24:02,HLA-B*07:02        0.410640   \n",
       "619              HLA-B*35:01,HLA-A*24:02,HLA-B*07:02        0.410640   \n",
       "620                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "621                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "622                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "623                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "624                          HLA-B*35:01,HLA-A*24:02        0.262839   \n",
       "625                                      HLA-A*24:02        0.170254   \n",
       "626                                      HLA-A*24:02        0.170254   \n",
       "627                                      HLA-A*24:02        0.170254   \n",
       "628                                      HLA-A*24:02        0.170254   \n",
       "\n",
       "                                       HLA-II_peptides  \\\n",
       "239                                    TLLALHRSYLTPGDS   \n",
       "240                                                NaN   \n",
       "241                                                NaN   \n",
       "242                                    LTPGDSSSGWTAGAA   \n",
       "243                    LTPGDSSSGWTAGAA,TPGDSSSGWTAGAAA   \n",
       "244    LTPGDSSSGWTAGAA,PGDSSSGWTAGAAAY,TPGDSSSGWTAGAAA   \n",
       "245  GDSSSGWTAGAAAYY,LTPGDSSSGWTAGAA,PGDSSSGWTAGAAA...   \n",
       "616                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "617                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "618                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "619                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "620                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "621                                    VAIHADQLTPTWRVY   \n",
       "622                                    LTPTWRVYSTGSNVF   \n",
       "623                    LTPTWRVYSTGSNVF,TPTWRVYSTGSNVFQ   \n",
       "624    LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVFQ   \n",
       "625  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVF...   \n",
       "626  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,TPTWRVYSTGSNVF...   \n",
       "627  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "628  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "233  FQTLLALHRSYLTPG,ITRFQTLLALHRSYL,NITRFQTLLALHRS...   \n",
       "234  FQTLLALHRSYLTPG,ITRFQTLLALHRSYL,QTLLALHRSYLTPG...   \n",
       "235  FQTLLALHRSYLTPG,QTLLALHRSYLTPGD,RFQTLLALHRSYLT...   \n",
       "236  FQTLLALHRSYLTPG,LTPGDSSSGWTAGAA,QTLLALHRSYLTPG...   \n",
       "237  FQTLLALHRSYLTPG,LTPGDSSSGWTAGAA,QTLLALHRSYLTPG...   \n",
       "238  LTPGDSSSGWTAGAA,PGDSSSGWTAGAAAY,QTLLALHRSYLTPG...   \n",
       "239  GDSSSGWTAGAAAYY,LTPGDSSSGWTAGAA,PGDSSSGWTAGAAA...   \n",
       "240  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...   \n",
       "241  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...   \n",
       "242  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...   \n",
       "243  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,LTPGDSSSGWTAGA...   \n",
       "244  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,GWTAGAAAYYVGYL...   \n",
       "245  DSSSGWTAGAAAYYV,GDSSSGWTAGAAAYY,GWTAGAAAYYVGYL...   \n",
       "610                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "611                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "612                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "613                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "614                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "615                    PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "616    LTPTWRVYSTGSNVF,PVAIHADQLTPTWRV,VAIHADQLTPTWRVY   \n",
       "617  LTPTWRVYSTGSNVF,PVAIHADQLTPTWRV,TPTWRVYSTGSNVF...   \n",
       "618  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...   \n",
       "619  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...   \n",
       "620  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,PVAIHADQLTPTWR...   \n",
       "621  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "622  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "623  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "624  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "625  LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQT,RVYSTGSNVFQTRA...   \n",
       "626  GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQ...   \n",
       "627  GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,PTWRVYSTGSNVFQ...   \n",
       "628  GSNVFQTRAGCLIGA,LTPTWRVYSTGSNVF,NVFQTRAGCLIGAE...   \n",
       "\n",
       "                                     HLA-II_haplotypes  HLA-II_pop_freq  ...  \\\n",
       "239                      HLA-DRB1*13:01,HLA-DRB1*15:01         0.289063  ...   \n",
       "240                                                NaN         0.000000  ...   \n",
       "241                                                NaN         0.000000  ...   \n",
       "242                          HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "243                          HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "244                          HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "245                          HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "616                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "617                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "618                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "619                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "620                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "621                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "622                                     HLA-DRB1*07:01         0.226271  ...   \n",
       "623                                     HLA-DRB1*07:01         0.226271  ...   \n",
       "624                      HLA-DRB1*07:01,HLA-DRB1*04:01         0.318226  ...   \n",
       "625                      HLA-DRB1*07:01,HLA-DRB1*04:01         0.318226  ...   \n",
       "626  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "627  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "628  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "233  HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...         0.574336  ...   \n",
       "234  HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...         0.574336  ...   \n",
       "235  HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...         0.574336  ...   \n",
       "236  HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...         0.612766  ...   \n",
       "237  HLA-DRB1*07:01,HLA-DRB1*11:01,HLA-DRB1*13:01,H...         0.612766  ...   \n",
       "238  HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*07:01,HLA-D...         0.612766  ...   \n",
       "239  HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*13:01,HLA-D...         0.437626  ...   \n",
       "240  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.491200  ...   \n",
       "241  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.557580  ...   \n",
       "242  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.557580  ...   \n",
       "243  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.557580  ...   \n",
       "244  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.557580  ...   \n",
       "245  HLA-DQA1*01:02/DQB1*06:02,HLA-DQA1*03:03/DQB1*...         0.557580  ...   \n",
       "610                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "611                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "612                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "613                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "614                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "615                                     HLA-DRB1*03:01         0.197497  ...   \n",
       "616                      HLA-DRB1*07:01,HLA-DRB1*03:01         0.379080  ...   \n",
       "617                      HLA-DRB1*07:01,HLA-DRB1*03:01         0.379080  ...   \n",
       "618       HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01         0.452874  ...   \n",
       "619       HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01         0.452874  ...   \n",
       "620  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01,H...         0.567206  ...   \n",
       "621  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*03:01,H...         0.567206  ...   \n",
       "622  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "623  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "624  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DQA1*05:05/D...         0.460695  ...   \n",
       "625  HLA-DRB1*07:01,HLA-DRB1*04:01,HLA-DRB1*01:01,H...         0.531054  ...   \n",
       "626  HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DRB1*04:01,H...         0.531054  ...   \n",
       "627  HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DRB1*04:01,H...         0.531054  ...   \n",
       "628  HLA-DRB1*01:01,HLA-DRB1*07:01,HLA-DQA1*05:05/D...         0.531054  ...   \n",
       "\n",
       "    mouse H2-d count mouse class I count mouse class II count mouse count  \\\n",
       "239                0                   0                    0           0   \n",
       "240                0                   0                    0           0   \n",
       "241                0                   0                    0           0   \n",
       "242                0                   0                    0           0   \n",
       "243                0                   0                    1           1   \n",
       "244                0                   0                    1           1   \n",
       "245                0                   0                    1           1   \n",
       "616                0                   0                    0           0   \n",
       "617                0                   0                    0           0   \n",
       "618                0                   0                    0           0   \n",
       "619                0                   0                    0           0   \n",
       "620                0                   0                    0           0   \n",
       "621                0                   0                    0           0   \n",
       "622                1                   1                    1           2   \n",
       "623                1                   1                    1           2   \n",
       "624                1                   1                    1           2   \n",
       "625                1                   1                    1           2   \n",
       "626                1                   1                    1           2   \n",
       "627                1                   1                    1           2   \n",
       "628                1                   1                    1           2   \n",
       "233                1                   1                    1           2   \n",
       "234                1                   0                    1           1   \n",
       "235                1                   0                    1           1   \n",
       "236                1                   0                    1           1   \n",
       "237                1                   0                    2           2   \n",
       "238                0                   0                    1           1   \n",
       "239                0                   0                    1           1   \n",
       "240                0                   0                    1           1   \n",
       "241                0                   0                    1           1   \n",
       "242                0                   0                    1           1   \n",
       "243                1                   2                    1           3   \n",
       "244                1                   2                    1           3   \n",
       "245                1                   2                    1           3   \n",
       "610                0                   0                    0           0   \n",
       "611                0                   0                    0           0   \n",
       "612                0                   0                    0           0   \n",
       "613                0                   0                    0           0   \n",
       "614                0                   0                    0           0   \n",
       "615                0                   0                    0           0   \n",
       "616                1                   1                    1           2   \n",
       "617                1                   1                    1           2   \n",
       "618                1                   1                    1           2   \n",
       "619                1                   1                    1           2   \n",
       "620                1                   1                    1           2   \n",
       "621                1                   1                    1           2   \n",
       "622                1                   1                    1           2   \n",
       "623                1                   1                    1           2   \n",
       "624                1                   1                    1           2   \n",
       "625                1                   1                    1           2   \n",
       "626                1                   1                    1           2   \n",
       "627                1                   1                    1           2   \n",
       "628                1                   1                    1           2   \n",
       "\n",
       "     HLA-I_pop_freq_rounded  HLA-II_pop_freq_rounded  Total_frequency_rounded  \\\n",
       "239                    0.17                     0.29                     0.05   \n",
       "240                    0.17                     0.00                     0.00   \n",
       "241                    0.00                     0.00                     0.00   \n",
       "242                    0.00                     0.21                     0.00   \n",
       "243                    0.00                     0.21                     0.00   \n",
       "244                    0.00                     0.21                     0.00   \n",
       "245                    0.24                     0.21                     0.05   \n",
       "616                    0.29                     0.20                     0.06   \n",
       "617                    0.29                     0.20                     0.06   \n",
       "618                    0.29                     0.20                     0.06   \n",
       "619                    0.29                     0.20                     0.06   \n",
       "620                    0.11                     0.20                     0.02   \n",
       "621                    0.11                     0.20                     0.02   \n",
       "622                    0.26                     0.23                     0.06   \n",
       "623                    0.26                     0.23                     0.06   \n",
       "624                    0.26                     0.32                     0.08   \n",
       "625                    0.17                     0.32                     0.05   \n",
       "626                    0.17                     0.46                     0.08   \n",
       "627                    0.17                     0.46                     0.08   \n",
       "628                    0.17                     0.46                     0.08   \n",
       "233                    0.61                     0.57                     0.35   \n",
       "234                    0.61                     0.57                     0.35   \n",
       "235                    0.61                     0.57                     0.35   \n",
       "236                    0.17                     0.61                     0.11   \n",
       "237                    0.17                     0.61                     0.11   \n",
       "238                    0.17                     0.61                     0.11   \n",
       "239                    0.37                     0.44                     0.16   \n",
       "240                    0.37                     0.49                     0.18   \n",
       "241                    0.24                     0.56                     0.13   \n",
       "242                    0.24                     0.56                     0.13   \n",
       "243                    0.35                     0.56                     0.19   \n",
       "244                    0.35                     0.56                     0.19   \n",
       "245                    0.35                     0.56                     0.19   \n",
       "610                    0.29                     0.20                     0.06   \n",
       "611                    0.29                     0.20                     0.06   \n",
       "612                    0.29                     0.20                     0.06   \n",
       "613                    0.29                     0.20                     0.06   \n",
       "614                    0.29                     0.20                     0.06   \n",
       "615                    0.29                     0.20                     0.06   \n",
       "616                    0.41                     0.38                     0.16   \n",
       "617                    0.41                     0.38                     0.16   \n",
       "618                    0.41                     0.45                     0.19   \n",
       "619                    0.41                     0.45                     0.19   \n",
       "620                    0.26                     0.57                     0.15   \n",
       "621                    0.26                     0.57                     0.15   \n",
       "622                    0.26                     0.46                     0.12   \n",
       "623                    0.26                     0.46                     0.12   \n",
       "624                    0.26                     0.46                     0.12   \n",
       "625                    0.17                     0.53                     0.09   \n",
       "626                    0.17                     0.53                     0.09   \n",
       "627                    0.17                     0.53                     0.09   \n",
       "628                    0.17                     0.53                     0.09   \n",
       "\n",
       "     manufacturability_feasibility_score_rounded  \\\n",
       "239                                         0.71   \n",
       "240                                         0.71   \n",
       "241                                         0.81   \n",
       "242                                         0.81   \n",
       "243                                         0.81   \n",
       "244                                         0.81   \n",
       "245                                         0.81   \n",
       "616                                         0.35   \n",
       "617                                         0.44   \n",
       "618                                         0.41   \n",
       "619                                         0.44   \n",
       "620                                         0.87   \n",
       "621                                         0.87   \n",
       "622                                         1.00   \n",
       "623                                         1.00   \n",
       "624                                         1.00   \n",
       "625                                         1.00   \n",
       "626                                         1.00   \n",
       "627                                         0.93   \n",
       "628                                         0.76   \n",
       "233                                         0.66   \n",
       "234                                         0.71   \n",
       "235                                         0.71   \n",
       "236                                         0.71   \n",
       "237                                         0.71   \n",
       "238                                         0.66   \n",
       "239                                         0.71   \n",
       "240                                         0.71   \n",
       "241                                         0.81   \n",
       "242                                         0.81   \n",
       "243                                         0.81   \n",
       "244                                         0.81   \n",
       "245                                         0.71   \n",
       "610                                         0.38   \n",
       "611                                         0.38   \n",
       "612                                         0.35   \n",
       "613                                         0.38   \n",
       "614                                         0.38   \n",
       "615                                         0.35   \n",
       "616                                         0.35   \n",
       "617                                         0.44   \n",
       "618                                         0.41   \n",
       "619                                         0.44   \n",
       "620                                         0.87   \n",
       "621                                         0.87   \n",
       "622                                         0.76   \n",
       "623                                         0.87   \n",
       "624                                         0.76   \n",
       "625                                         0.76   \n",
       "626                                         0.76   \n",
       "627                                         0.71   \n",
       "628                                         0.66   \n",
       "\n",
       "     longest_contiguous_accessible_region_in_spike  max_accessibility  \n",
       "239                                             16               0.96  \n",
       "240                                             17               0.96  \n",
       "241                                             17               0.96  \n",
       "242                                             17               0.96  \n",
       "243                                             17               0.96  \n",
       "244                                             17               0.96  \n",
       "245                                             16               0.96  \n",
       "616                                             16               0.81  \n",
       "617                                             17               0.81  \n",
       "618                                             18               0.81  \n",
       "619                                             19               0.81  \n",
       "620                                             20               0.81  \n",
       "621                                             21               0.81  \n",
       "622                                             21               0.81  \n",
       "623                                             21               0.81  \n",
       "624                                             20               0.81  \n",
       "625                                             19               0.81  \n",
       "626                                             18               0.81  \n",
       "627                                             17               0.81  \n",
       "628                                             16               0.81  \n",
       "233                                             16               0.96  \n",
       "234                                             17               0.96  \n",
       "235                                             17               0.96  \n",
       "236                                             17               0.96  \n",
       "237                                             17               0.96  \n",
       "238                                             17               0.96  \n",
       "239                                             17               0.96  \n",
       "240                                             17               0.96  \n",
       "241                                             17               0.96  \n",
       "242                                             17               0.96  \n",
       "243                                             17               0.96  \n",
       "244                                             17               0.96  \n",
       "245                                             16               0.96  \n",
       "610                                             16               0.81  \n",
       "611                                             17               0.81  \n",
       "612                                             18               0.81  \n",
       "613                                             19               0.81  \n",
       "614                                             20               0.81  \n",
       "615                                             21               0.81  \n",
       "616                                             22               0.81  \n",
       "617                                             23               0.81  \n",
       "618                                             23               0.81  \n",
       "619                                             23               0.81  \n",
       "620                                             23               0.81  \n",
       "621                                             23               0.81  \n",
       "622                                             22               0.81  \n",
       "623                                             21               0.81  \n",
       "624                                             20               0.81  \n",
       "625                                             19               0.81  \n",
       "626                                             18               0.81  \n",
       "627                                             17               0.84  \n",
       "628                                             16               0.84  \n",
       "\n",
       "[52 rows x 45 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_concat_annot = annotate_selection_criteria(df_concat);\n",
    "df_concat_annot[df_concat_annot.longest_contiguous_accessible_region_in_spike > 15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
       "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
       "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
       "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
       "       'num_extremely_unstable_dipeptides',\n",
       "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
       "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
       "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
       "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
       "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length',\n",
       "       'manufacturability_feasibility_score', 'has mouse MHC-I b',\n",
       "       'has mouse MHC-II b', 'has mouse MHC-I d', 'has mouse MHC-II d',\n",
       "       'mouse H2-b count', 'mouse H2-d count', 'mouse class I count',\n",
       "       'mouse class II count', 'mouse count', 'HLA-I_pop_freq_rounded',\n",
       "       'HLA-II_pop_freq_rounded', 'Total_frequency_rounded',\n",
       "       'manufacturability_feasibility_score_rounded',\n",
       "       'longest_contiguous_accessible_region_in_spike', 'max_accessibility',\n",
       "       'overlapping_bcell_epitope', 'overlapping_bcell_epitope_start',\n",
       "       'overlapping_bcell_epitope_end', 'overlapping_bcell_epitope_length',\n",
       "       'fraction_of_full_bcell_epitope', 'full_bcell_epitope'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell = annotate_overlapping_bcell_epitopes(df_concat_annot, bcell_epitopes);\n",
    "df_concat_annot_bcell.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_alleles(alleles):\n",
    "    if type(alleles) in (list, set, tuple):\n",
    "        alleles = set(alleles)\n",
    "    elif type(alleles) is str:\n",
    "        alleles = set(alleles.split(\",\"))\n",
    "    else:\n",
    "        alleles = set()\n",
    "    return alleles\n",
    "\n",
    "    \n",
    "def population_coverage(alleles, allele_frequencies):\n",
    "    alleles = parse_alleles(alleles)\n",
    "\n",
    "    if len(alleles) > 0:\n",
    "\n",
    "        individual_allele_frequencies = np.array([allele_frequencies[a] for a in alleles])\n",
    "        prob_not_allele_haploid = 1.0 - (individual_allele_frequencies / 100.0)\n",
    "        prob_not_allele_diploid = prob_not_allele_haploid ** 2\n",
    "        prob_missing_all = np.prod(prob_not_allele_diploid)\n",
    "        prob_atleast_one = 1.0 - prob_missing_all\n",
    "\n",
    "        return prob_atleast_one\n",
    "    else:\n",
    "        return 0.0 \n",
    "\n",
    "def update_marginal_coverage(\n",
    "        df, \n",
    "        allele_frequencies, \n",
    "        chosen_class1_alleles=[], \n",
    "        chosen_class2_alleles=[], \n",
    "        decimals=2,\n",
    "        prefix=\"marginal_\"):\n",
    "    \"\"\"\n",
    "    Create or update columns named 'marginal_class1_coverage' & 'marginal_class2_coverage'\n",
    "    by excluding a blacklist of class1 & class2 alleles which have already been \n",
    "    included in a partially constructed list of vaccine peptides.\n",
    "    \n",
    "    Expects allele_frequencies to be a dictionary with values between [0, 100]\n",
    "    and will update DataFrame with values in the same range. \n",
    "    \"\"\"\n",
    "    # based on which class1 and class2 alleles we've already chosen\n",
    "    # compute updated \n",
    "    df = df.copy()\n",
    "    chosen_class1_alleles = set(chosen_class1_alleles)\n",
    "    chosne_class2_alleles = set(chosen_class2_alleles)\n",
    "    class1_coverages = []\n",
    "    class2_coverages = []\n",
    "    for class1_alleles, class2_alleles in zip(df['HLA-I_haplotypes'], df['HLA-II_haplotypes']):\n",
    "        class1_alleles = parse_alleles(class1_alleles)\n",
    "        class2_alleles = parse_alleles(class2_alleles)\n",
    "        extra_class1_alleles = class1_alleles.difference(chosen_class1_alleles)\n",
    "        extra_class2_alleles = class2_alleles.difference(chosen_class2_alleles)\n",
    "        class1_coverages.append(population_coverage(extra_class1_alleles, allele_frequencies))\n",
    "        class2_coverages.append(population_coverage(extra_class2_alleles, allele_frequencies))\n",
    "    \n",
    "    class1_key = \"%sclass1_coverage\" % prefix\n",
    "    df[class1_key] = class1_coverages\n",
    "    df[class1_key + \"_rounded\"] = df[class1_key].round(decimals)\n",
    "    class2_key = \"%sclass2_coverage\" % prefix\n",
    "    df[class2_key] = class2_coverages\n",
    "    df[class2_key + \"_rounded\"] = df[class2_key].round(decimals)\n",
    "    combined_key = \"%scombined_coverage\" % prefix\n",
    "    df[combined_key] = np.array(class1_coverages) * np.array(class2_coverages)\n",
    "    df[combined_key + \"_rounded\"] = df[combined_key].round(decimals)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class\n",
      "I     164.655275\n",
      "II    139.423328\n",
      "Name: Frequency, dtype: float64\n",
      "Class\n",
      "I     17\n",
      "II    15\n",
      "Name: Frequency, dtype: int64\n",
      "Alleles in frequency table, not in vaccine peptides:\n",
      "==> HLA-DQA1*03:01/DQB1*03:02 (9.62)\n",
      "==> HLA-DQA1*01:01/DQB1*05:01 (10.85)\n",
      "==> HLA-DQA1*02:01/DQB1*02:02 (11.08)\n",
      "==> HLA-DQA1*01:03/DQB1*06:03 (5.66)\n",
      "==> HLA-B*15:01 (4.02)\n",
      "Alleles in vaccine peptides, not in frequency table:\n"
     ]
    }
   ],
   "source": [
    "df_freq = pd.read_csv(\"HLA_freq_formated.txt\");\n",
    "df_freq[\"Class\"] = df_freq.Haplotype.map(lambda x: \"II\" if \"D\" in x else \"I\");\n",
    "print(df_freq.groupby(\"Class\")[\"Frequency\"].sum())\n",
    "print(df_freq.groupby(\"Class\")[\"Frequency\"].count())\n",
    "\n",
    "allele_to_frequency = {a: f for (a, f) in zip(df_freq[\"Haplotype\"], df_freq[\"Frequency\"])}\n",
    "unique_alleles_in_vaccine_peptides = set([])\n",
    "for class1, class2 in zip(df_concat[\"HLA-I_haplotypes\"], df_concat[\"HLA-II_haplotypes\"]):\n",
    "    if type(class1) is str:\n",
    "        unique_alleles_in_vaccine_peptides.update(class1.split(\",\"))\n",
    "    if type(class2) is str:\n",
    "        unique_alleles_in_vaccine_peptides.update(class2.split(\",\"))\n",
    "len(unique_alleles_in_vaccine_peptides), len(allele_to_frequency)\n",
    "print(\"Alleles in frequency table, not in vaccine peptides:\")\n",
    "for a, f in allele_to_frequency.items():\n",
    "    if a not in unique_alleles_in_vaccine_peptides:\n",
    "        print(\"==> %s (%0.2f)\" % (a, f))\n",
    "print(\"Alleles in vaccine peptides, not in frequency table:\")\n",
    "for a in unique_alleles_in_vaccine_peptides:\n",
    "    if a not in allele_to_frequency:\n",
    "        print(\"==> %s\" % a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Should be near zero 7.771561172376096e-16 5.551115123125783e-16\n",
      "Should be >0.4 0.41749339987056 5.551115123125783e-16\n"
     ]
    }
   ],
   "source": [
    "\n",
    "        \n",
    "df_concat_with_marginal_freqs = update_marginal_coverage(\n",
    "    df=df_concat,\n",
    "    allele_frequencies=allele_to_frequency)\n",
    "# sanity check that the population frequency calculation matches the one \n",
    "# that Christof annotated vaccine peptides with in the absence of any \n",
    "# \"chosen\" / excluded alleles.\n",
    "print(\n",
    "    \"Should be near zero\",\n",
    "    (df_concat_with_marginal_freqs[\"HLA-I_pop_freq\"] - df_concat_with_marginal_freqs[\"marginal_class1_coverage\"]).abs().max(),\n",
    "    (df_concat_with_marginal_freqs[\"HLA-II_pop_freq\"] - df_concat_with_marginal_freqs[\"marginal_class2_coverage\"]).abs().max()\n",
    ")\n",
    "df_concat_with_marginal_freqs_a0201 = update_marginal_coverage(\n",
    "    df=df_concat,\n",
    "    chosen_class1_alleles={\"HLA-A*02:01\"},\n",
    "    allele_frequencies=allele_to_frequency)\n",
    "# sanity check that the population frequency calculation matches the one \n",
    "# that Christof annotated vaccine peptides with in the absence of any \n",
    "# \"chosen\" / excluded alleles.\n",
    "print(\n",
    "    \"Should be >0.4\",\n",
    "    (df_concat_with_marginal_freqs[\"HLA-I_pop_freq\"] - df_concat_with_marginal_freqs_a0201[\"marginal_class1_coverage\"]).abs().max(),\n",
    "    (df_concat_with_marginal_freqs[\"HLA-II_pop_freq\"] - df_concat_with_marginal_freqs_a0201[\"marginal_class2_coverage\"]).abs().max()\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def select_bcell_vaccine_peptides(df):\n",
    "    df = annotate_selection_criteria(df)\n",
    "    \n",
    "    sort_columns = [\n",
    "        \"fraction_of_full_bcell_epitope\",\n",
    "        \"HLA-II_pop_freq_rounded\",\n",
    "        \"mouse class II count\",\n",
    "        \"manufacturability_feasibility_score_rounded\",\n",
    "        \"HLA-I_pop_freq_rounded\",\n",
    "        \"mouse class I count\",\n",
    "    ]\n",
    "\n",
    "    selected_peptides = set()\n",
    "    dataframes = []\n",
    "    for length, df_length in df.groupby(\"Length\"):\n",
    "        for epitope, df_epitope in df_length.groupby(\"full_bcell_epitope\"):\n",
    "            df_sorted = df_epitope.sort_values(by=sort_columns, ascending=False)\n",
    "            \n",
    "            selected_peptides.add(df_sorted.Sequence.iloc[0])\n",
    "            dataframes.append(df_sorted.iloc[:1])\n",
    "            \n",
    "            has_class2_h2b = ~df_sorted['Mouse MHC-II b'].isnull()\n",
    "            has_class2_h2d = ~df_sorted['Mouse MHC-II d'].isnull()\n",
    "            has_both = has_class2_h2b & has_class2_h2d\n",
    "            \n",
    "            both = df_sorted[has_both]\n",
    "            if len(both) > 0 and both.Sequence.iloc[0] not in selected_peptides:\n",
    "                selected_peptides.add(both.Sequence.iloc[0])\n",
    "                dataframes.append(both.iloc[:1])\n",
    "            \n",
    "            h2b = df_sorted[has_class2_h2b]\n",
    "            if len(h2b) > 0 and h2b.Sequence.iloc[0] not in selected_peptides:\n",
    "                selected_peptides.add(h2b.Sequence.iloc[0])\n",
    "                dataframes.append(h2b.iloc[:1])\n",
    "            \n",
    "            h2d = df_sorted[has_class2_h2d]\n",
    "            if len(h2d) > 0 and h2d.Sequence.iloc[0] not in selected_peptides:\n",
    "                selected_peptides.add(h2d.Sequence.iloc[0])\n",
    "                dataframes.append(h2b.iloc[:1])\n",
    "    return pd.concat(dataframes).sort_values([\"Length\", \"Protein\", \"Start\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>HLA-I_peptides</th>\n",
       "      <th>HLA-I_haplotypes</th>\n",
       "      <th>HLA-I_pop_freq</th>\n",
       "      <th>HLA-II_peptides</th>\n",
       "      <th>HLA-II_haplotypes</th>\n",
       "      <th>HLA-II_pop_freq</th>\n",
       "      <th>...</th>\n",
       "      <th>mouse H2-d count</th>\n",
       "      <th>mouse class I count</th>\n",
       "      <th>mouse class II count</th>\n",
       "      <th>mouse count</th>\n",
       "      <th>HLA-I_pop_freq_rounded</th>\n",
       "      <th>HLA-II_pop_freq_rounded</th>\n",
       "      <th>Total_frequency_rounded</th>\n",
       "      <th>manufacturability_feasibility_score_rounded</th>\n",
       "      <th>longest_contiguous_accessible_region_in_spike</th>\n",
       "      <th>max_accessibility</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>KSNLKPFERDISTEI</td>\n",
       "      <td>S</td>\n",
       "      <td>458</td>\n",
       "      <td>472</td>\n",
       "      <td>KPFERDISTEI,KSNLKPFER</td>\n",
       "      <td>HLA-B*07:02,HLA-A*11:01</td>\n",
       "      <td>0.286880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>7</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>ISTEIYQAGSTPCNG</td>\n",
       "      <td>S</td>\n",
       "      <td>468</td>\n",
       "      <td>482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ISTEIYQAGSTPCNG</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.44</td>\n",
       "      <td>6</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>IADTTDAVRDPQTLE</td>\n",
       "      <td>S</td>\n",
       "      <td>569</td>\n",
       "      <td>583</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>GFNFSQILPDPSKPS</td>\n",
       "      <td>S</td>\n",
       "      <td>799</td>\n",
       "      <td>813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GFNFSQILPDPSKPS</td>\n",
       "      <td>HLA-DRB1*01:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.233805</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>IVNNTVYDPLQPELD</td>\n",
       "      <td>S</td>\n",
       "      <td>1132</td>\n",
       "      <td>1146</td>\n",
       "      <td>VYDPLQPEL</td>\n",
       "      <td>HLA-C*04:01</td>\n",
       "      <td>0.231730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>6</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>YRLFRKSNLKPFERDISTEIY</td>\n",
       "      <td>S</td>\n",
       "      <td>453</td>\n",
       "      <td>473</td>\n",
       "      <td>FRKSNLKPF,KPFERDISTEI,KSNLKPFER,RLFRKSNL,RLFRK...</td>\n",
       "      <td>HLA-C*07:02,HLA-B*07:02,HLA-A*11:01,HLA-B*08:0...</td>\n",
       "      <td>0.780469</td>\n",
       "      <td>YRLFRKSNLKPFERD</td>\n",
       "      <td>HLA-DRB1*07:01</td>\n",
       "      <td>0.226271</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.18</td>\n",
       "      <td>1.00</td>\n",
       "      <td>9</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>KPFERDISTEIYQAGSTPCNG</td>\n",
       "      <td>S</td>\n",
       "      <td>462</td>\n",
       "      <td>482</td>\n",
       "      <td>KPFERDISTEI</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>0.200499</td>\n",
       "      <td>ISTEIYQAGSTPCNG</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.44</td>\n",
       "      <td>6</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>QTLEILDITPCSFGGVSVITP</td>\n",
       "      <td>S</td>\n",
       "      <td>580</td>\n",
       "      <td>600</td>\n",
       "      <td>ILDITPCSF</td>\n",
       "      <td>HLA-C*05:01</td>\n",
       "      <td>0.133120</td>\n",
       "      <td>DITPCSFGGVSVITP</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.62</td>\n",
       "      <td>4</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>IKDFGGFNFSQILPDPSKPSK</td>\n",
       "      <td>S</td>\n",
       "      <td>794</td>\n",
       "      <td>814</td>\n",
       "      <td>ILPDPSKPSK</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>0.214307</td>\n",
       "      <td>FGGFNFSQILPDPSK,FNFSQILPDPSKPSK,GFNFSQILPDPSKP...</td>\n",
       "      <td>HLA-DRB1*04:01,HLA-DRB1*01:01</td>\n",
       "      <td>0.233805</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.05</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1125</th>\n",
       "      <td>CDVVIGIVNNTVYDPLQPELD</td>\n",
       "      <td>S</td>\n",
       "      <td>1126</td>\n",
       "      <td>1146</td>\n",
       "      <td>VYDPLQPEL</td>\n",
       "      <td>HLA-C*04:01</td>\n",
       "      <td>0.231730</td>\n",
       "      <td>CDVVIGIVNNTVYDP,DVVIGIVNNTVYDPL</td>\n",
       "      <td>HLA-DRB1*01:01</td>\n",
       "      <td>0.130464</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.29</td>\n",
       "      <td>6</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>YNYLYRLFRKSNLKPFERDISTEIYQA</td>\n",
       "      <td>S</td>\n",
       "      <td>449</td>\n",
       "      <td>475</td>\n",
       "      <td>FRKSNLKPF,KPFERDISTEI,KSNLKPFER,RLFRKSNL,RLFRK...</td>\n",
       "      <td>HLA-C*07:02,HLA-B*07:02,HLA-A*11:01,HLA-B*08:0...</td>\n",
       "      <td>0.780469</td>\n",
       "      <td>LKPFERDISTEIYQA,LYRLFRKSNLKPFER,NYLYRLFRKSNLKP...</td>\n",
       "      <td>HLA-DRB1*04:01,HLA-DRB1*13:01,HLA-DRB1*11:01,H...</td>\n",
       "      <td>0.457626</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.00</td>\n",
       "      <td>9</td>\n",
       "      <td>0.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>FRKSNLKPFERDISTEIYQAGSTPCNG</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>482</td>\n",
       "      <td>FRKSNLKPF,KPFERDISTEI,KSNLKPFER</td>\n",
       "      <td>HLA-C*07:02,HLA-B*07:02,HLA-A*11:01</td>\n",
       "      <td>0.459618</td>\n",
       "      <td>ISTEIYQAGSTPCNG,LKPFERDISTEIYQA</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*04:01</td>\n",
       "      <td>0.302979</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.44</td>\n",
       "      <td>9</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>AVRDPQTLEILDITPCSFGGVSVITPG</td>\n",
       "      <td>S</td>\n",
       "      <td>575</td>\n",
       "      <td>601</td>\n",
       "      <td>ILDITPCSF</td>\n",
       "      <td>HLA-C*05:01</td>\n",
       "      <td>0.133120</td>\n",
       "      <td>DITPCSFGGVSVITP,ITPCSFGGVSVITPG</td>\n",
       "      <td>HLA-DQA1*05:05/DQB1*03:01</td>\n",
       "      <td>0.208968</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.76</td>\n",
       "      <td>4</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>IYKTPPIKDFGGFNFSQILPDPSKPSK</td>\n",
       "      <td>S</td>\n",
       "      <td>788</td>\n",
       "      <td>814</td>\n",
       "      <td>ILPDPSKPSK,IYKTPPIKDF</td>\n",
       "      <td>HLA-A*03:01,HLA-A*24:02</td>\n",
       "      <td>0.348075</td>\n",
       "      <td>FGGFNFSQILPDPSK,FNFSQILPDPSKPSK,GFNFSQILPDPSKP...</td>\n",
       "      <td>HLA-DRB1*04:01,HLA-DRB1*01:01</td>\n",
       "      <td>0.233805</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>808</th>\n",
       "      <td>PSKPSKRSFIEDLLFNKVTLADAGFIK</td>\n",
       "      <td>S</td>\n",
       "      <td>809</td>\n",
       "      <td>835</td>\n",
       "      <td>LLFNKVTLA,SFIEDLLFNK,FIEDLLFNK,FIEDLLFNKV,KVTL...</td>\n",
       "      <td>HLA-A*02:01,HLA-A*11:01,HLA-B*08:01,HLA-A*03:01</td>\n",
       "      <td>0.661908</td>\n",
       "      <td>EDLLFNKVTLADAGF,IEDLLFNKVTLADAG,FIEDLLFNKVTLAD...</td>\n",
       "      <td>HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*07:01</td>\n",
       "      <td>0.398583</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.87</td>\n",
       "      <td>4</td>\n",
       "      <td>0.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1140</th>\n",
       "      <td>LQPELDSFKEELDKYFKNHTSPDVDLG</td>\n",
       "      <td>S</td>\n",
       "      <td>1141</td>\n",
       "      <td>1167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>DKYFKNHTSPDVDLG,LDKYFKNHTSPDVDL</td>\n",
       "      <td>HLA-DRB1*01:01,HLA-DRB1*04:01,HLA-DRB1*07:01</td>\n",
       "      <td>0.407173</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.81</td>\n",
       "      <td>6</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows  51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Sequence Protein  Start   End  \\\n",
       "457               KSNLKPFERDISTEI       S    458   472   \n",
       "467               ISTEIYQAGSTPCNG       S    468   482   \n",
       "568               IADTTDAVRDPQTLE       S    569   583   \n",
       "798               GFNFSQILPDPSKPS       S    799   813   \n",
       "1131              IVNNTVYDPLQPELD       S   1132  1146   \n",
       "452         YRLFRKSNLKPFERDISTEIY       S    453   473   \n",
       "461         KPFERDISTEIYQAGSTPCNG       S    462   482   \n",
       "579         QTLEILDITPCSFGGVSVITP       S    580   600   \n",
       "793         IKDFGGFNFSQILPDPSKPSK       S    794   814   \n",
       "1125        CDVVIGIVNNTVYDPLQPELD       S   1126  1146   \n",
       "448   YNYLYRLFRKSNLKPFERDISTEIYQA       S    449   475   \n",
       "455   FRKSNLKPFERDISTEIYQAGSTPCNG       S    456   482   \n",
       "574   AVRDPQTLEILDITPCSFGGVSVITPG       S    575   601   \n",
       "787   IYKTPPIKDFGGFNFSQILPDPSKPSK       S    788   814   \n",
       "808   PSKPSKRSFIEDLLFNKVTLADAGFIK       S    809   835   \n",
       "1140  LQPELDSFKEELDKYFKNHTSPDVDLG       S   1141  1167   \n",
       "\n",
       "                                         HLA-I_peptides  \\\n",
       "457                               KPFERDISTEI,KSNLKPFER   \n",
       "467                                                 NaN   \n",
       "568                                                 NaN   \n",
       "798                                                 NaN   \n",
       "1131                                          VYDPLQPEL   \n",
       "452   FRKSNLKPF,KPFERDISTEI,KSNLKPFER,RLFRKSNL,RLFRK...   \n",
       "461                                         KPFERDISTEI   \n",
       "579                                           ILDITPCSF   \n",
       "793                                          ILPDPSKPSK   \n",
       "1125                                          VYDPLQPEL   \n",
       "448   FRKSNLKPF,KPFERDISTEI,KSNLKPFER,RLFRKSNL,RLFRK...   \n",
       "455                     FRKSNLKPF,KPFERDISTEI,KSNLKPFER   \n",
       "574                                           ILDITPCSF   \n",
       "787                               ILPDPSKPSK,IYKTPPIKDF   \n",
       "808   LLFNKVTLA,SFIEDLLFNK,FIEDLLFNK,FIEDLLFNKV,KVTL...   \n",
       "1140                                                NaN   \n",
       "\n",
       "                                       HLA-I_haplotypes  HLA-I_pop_freq  \\\n",
       "457                             HLA-B*07:02,HLA-A*11:01        0.286880   \n",
       "467                                                 NaN        0.000000   \n",
       "568                                                 NaN        0.000000   \n",
       "798                                                 NaN        0.000000   \n",
       "1131                                        HLA-C*04:01        0.231730   \n",
       "452   HLA-C*07:02,HLA-B*07:02,HLA-A*11:01,HLA-B*08:0...        0.780469   \n",
       "461                                         HLA-B*07:02        0.200499   \n",
       "579                                         HLA-C*05:01        0.133120   \n",
       "793                                         HLA-A*03:01        0.214307   \n",
       "1125                                        HLA-C*04:01        0.231730   \n",
       "448   HLA-C*07:02,HLA-B*07:02,HLA-A*11:01,HLA-B*08:0...        0.780469   \n",
       "455                 HLA-C*07:02,HLA-B*07:02,HLA-A*11:01        0.459618   \n",
       "574                                         HLA-C*05:01        0.133120   \n",
       "787                             HLA-A*03:01,HLA-A*24:02        0.348075   \n",
       "808     HLA-A*02:01,HLA-A*11:01,HLA-B*08:01,HLA-A*03:01        0.661908   \n",
       "1140                                                NaN        0.000000   \n",
       "\n",
       "                                        HLA-II_peptides  \\\n",
       "457                                                 NaN   \n",
       "467                                     ISTEIYQAGSTPCNG   \n",
       "568                                                 NaN   \n",
       "798                                     GFNFSQILPDPSKPS   \n",
       "1131                                                NaN   \n",
       "452                                     YRLFRKSNLKPFERD   \n",
       "461                                     ISTEIYQAGSTPCNG   \n",
       "579                                     DITPCSFGGVSVITP   \n",
       "793   FGGFNFSQILPDPSK,FNFSQILPDPSKPSK,GFNFSQILPDPSKP...   \n",
       "1125                    CDVVIGIVNNTVYDP,DVVIGIVNNTVYDPL   \n",
       "448   LKPFERDISTEIYQA,LYRLFRKSNLKPFER,NYLYRLFRKSNLKP...   \n",
       "455                     ISTEIYQAGSTPCNG,LKPFERDISTEIYQA   \n",
       "574                     DITPCSFGGVSVITP,ITPCSFGGVSVITPG   \n",
       "787   FGGFNFSQILPDPSK,FNFSQILPDPSKPSK,GFNFSQILPDPSKP...   \n",
       "808   EDLLFNKVTLADAGF,IEDLLFNKVTLADAG,FIEDLLFNKVTLAD...   \n",
       "1140                    DKYFKNHTSPDVDLG,LDKYFKNHTSPDVDL   \n",
       "\n",
       "                                      HLA-II_haplotypes  HLA-II_pop_freq  ...  \\\n",
       "457                                                 NaN         0.000000  ...   \n",
       "467                           HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "568                                                 NaN         0.000000  ...   \n",
       "798                       HLA-DRB1*01:01,HLA-DRB1*04:01         0.233805  ...   \n",
       "1131                                                NaN         0.000000  ...   \n",
       "452                                      HLA-DRB1*07:01         0.226271  ...   \n",
       "461                           HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "579                           HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "793                       HLA-DRB1*04:01,HLA-DRB1*01:01         0.233805  ...   \n",
       "1125                                     HLA-DRB1*01:01         0.130464  ...   \n",
       "448   HLA-DRB1*04:01,HLA-DRB1*13:01,HLA-DRB1*11:01,H...         0.457626  ...   \n",
       "455            HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*04:01         0.302979  ...   \n",
       "574                           HLA-DQA1*05:05/DQB1*03:01         0.208968  ...   \n",
       "787                       HLA-DRB1*04:01,HLA-DRB1*01:01         0.233805  ...   \n",
       "808        HLA-DRB1*13:01,HLA-DRB1*01:01,HLA-DRB1*07:01         0.398583  ...   \n",
       "1140       HLA-DRB1*01:01,HLA-DRB1*04:01,HLA-DRB1*07:01         0.407173  ...   \n",
       "\n",
       "     mouse H2-d count mouse class I count mouse class II count mouse count  \\\n",
       "457                 0                   0                    0           0   \n",
       "467                 0                   0                    1           1   \n",
       "568                 0                   0                    0           0   \n",
       "798                 0                   0                    1           1   \n",
       "1131                1                   1                    0           1   \n",
       "452                 0                   1                    0           1   \n",
       "461                 0                   0                    1           1   \n",
       "579                 0                   0                    0           0   \n",
       "793                 0                   0                    1           1   \n",
       "1125                1                   1                    0           1   \n",
       "448                 0                   1                    0           1   \n",
       "455                 0                   0                    1           1   \n",
       "574                 0                   0                    0           0   \n",
       "787                 0                   0                    1           1   \n",
       "808                 1                   1                    1           2   \n",
       "1140                0                   0                    0           0   \n",
       "\n",
       "      HLA-I_pop_freq_rounded  HLA-II_pop_freq_rounded  \\\n",
       "457                     0.29                     0.00   \n",
       "467                     0.00                     0.21   \n",
       "568                     0.00                     0.00   \n",
       "798                     0.00                     0.23   \n",
       "1131                    0.23                     0.00   \n",
       "452                     0.78                     0.23   \n",
       "461                     0.20                     0.21   \n",
       "579                     0.13                     0.21   \n",
       "793                     0.21                     0.23   \n",
       "1125                    0.23                     0.13   \n",
       "448                     0.78                     0.46   \n",
       "455                     0.46                     0.30   \n",
       "574                     0.13                     0.21   \n",
       "787                     0.35                     0.23   \n",
       "808                     0.66                     0.40   \n",
       "1140                    0.00                     0.41   \n",
       "\n",
       "      Total_frequency_rounded  manufacturability_feasibility_score_rounded  \\\n",
       "457                      0.00                                         1.00   \n",
       "467                      0.00                                         0.44   \n",
       "568                      0.00                                         1.00   \n",
       "798                      0.00                                         1.00   \n",
       "1131                     0.00                                         0.81   \n",
       "452                      0.18                                         1.00   \n",
       "461                      0.04                                         0.44   \n",
       "579                      0.03                                         0.62   \n",
       "793                      0.05                                         1.00   \n",
       "1125                     0.03                                         0.29   \n",
       "448                      0.36                                         1.00   \n",
       "455                      0.14                                         0.44   \n",
       "574                      0.03                                         0.76   \n",
       "787                      0.08                                         1.00   \n",
       "808                      0.26                                         0.87   \n",
       "1140                     0.00                                         0.81   \n",
       "\n",
       "      longest_contiguous_accessible_region_in_spike  max_accessibility  \n",
       "457                                               7               0.71  \n",
       "467                                               6               0.84  \n",
       "568                                               4               0.92  \n",
       "798                                               4               0.82  \n",
       "1131                                              6               1.00  \n",
       "452                                               9               0.71  \n",
       "461                                               6               0.84  \n",
       "579                                               4               0.92  \n",
       "793                                               4               0.82  \n",
       "1125                                              6               1.00  \n",
       "448                                               9               0.93  \n",
       "455                                               9               0.84  \n",
       "574                                               4               0.92  \n",
       "787                                               4               0.82  \n",
       "808                                               4               0.82  \n",
       "1140                                              6               1.00  \n",
       "\n",
       "[16 rows x 51 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_selected_bcell_vaccine_peptides = select_bcell_vaccine_peptides(df_with_bcell_epitopes)\n",
    "for length, df_length in df_selected_bcell_vaccine_peptides.groupby(\"Length\"):\n",
    "    df_length.to_csv(\"selected-bcell-vaccine-%dmer.csv\" % length, index=False)\n",
    "df_selected_bcell_vaccine_peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence', 'Protein', 'Start', 'End', 'HLA-I_peptides',\n",
       "       'HLA-I_haplotypes', 'HLA-I_pop_freq', 'HLA-II_peptides',\n",
       "       'HLA-II_haplotypes', 'HLA-II_pop_freq', 'Mouse MHC-I b',\n",
       "       'Mouse MHC-I d', 'Mouse MHC-II b', 'Mouse MHC-II d', 'Total_frequency',\n",
       "       'num_extremely_unstable_dipeptides',\n",
       "       'num_moderately_unstable_dipeptides', 'num_cysteines',\n",
       "       'num_methionines', 'num_thiols', 'difficult_c_terminal',\n",
       "       'difficult_n_terminal', 'gravy', 'max_kmer_gravy',\n",
       "       'extreme_local_hydrophobicity', 'difficult_local_hydrophobicity',\n",
       "       'too_many_cysteines', 'manufacturability_difficulty_score', 'Length',\n",
       "       'overlapping_bcell_epitope', 'overlapping_bcell_epitope_start',\n",
       "       'overlapping_bcell_epitope_end', 'overlapping_bcell_epitope_length',\n",
       "       'fraction_of_full_bcell_epitope', 'full_bcell_epitope'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_with_bcell_epitopes.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5742"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "def choose_sort_columns(class1, class2, bcell): \n",
    "    class1_sort_columns = [\n",
    "        \"marginal_class1_coverage_rounded\", \n",
    "        \"mouse class I count\",\n",
    "        \"HLA-I_pop_freq_rounded\",\n",
    "    ]\n",
    "    class2_sort_columns = [\n",
    "        \"marginal_class2_coverage_rounded\", \n",
    "        \"mouse class II count\",\n",
    "        \"HLA-II_pop_freq_rounded\"\n",
    "    ]\n",
    "    \n",
    "    if class1 and class2:\n",
    "        primary_sort_columns = [\n",
    "            \"marginal_combined_coverage_rounded\", \n",
    "            \"mouse count\",\n",
    "            \"Total_frequency_rounded\",\n",
    "        ]\n",
    "        secondary_sort_columns = class1_sort_columns + class2_sort_columns\n",
    "    elif class1:\n",
    "        primary_sort_columns = class1_sort_columns\n",
    "        secondary_sort_columns = class2_sort_columns\n",
    "    elif class2:\n",
    "        primary_sort_columns = class2_sort_columns\n",
    "        secondary_sort_columns = class1_sort_columns\n",
    "    \n",
    "    if bcell:\n",
    "        primary_sort_columns = ['fraction_of_full_bcell_epitope'] + primary_sort_columns\n",
    "    \n",
    "    tie_breakers = [\n",
    "        \"Total_frequency\",\n",
    "        \"HLA-I_pop_freq\",\n",
    "        \"HLA-II_pop_freq\",\n",
    "        \"manufacturability_feasibility_score\",\n",
    "        \"Start\",\n",
    "    ]\n",
    "    return (\n",
    "        primary_sort_columns + \n",
    "        [\"manufacturability_feasibility_score_rounded\"] + \n",
    "        secondary_sort_columns + \n",
    "        tie_breakers\n",
    "    )\n",
    "    \n",
    "    \n",
    "def apply_filters(df, class1, class2, bcell, H2b, H2d, length, exclude_accessible_spike_regions, verbose):\n",
    "    masks = []\n",
    "    \n",
    "    if H2b:\n",
    "        if class1:\n",
    "            H2b_class1 = df[\"has mouse MHC-I b\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class I H2b\" % (H2b_class1.sum(), len(df)))\n",
    "            masks.append(H2b_class1)\n",
    "        if class2:\n",
    "            H2b_class2 = df[\"has mouse MHC-II b\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class II H2b\" % (H2b_class2.sum(), len(df)))\n",
    "            masks.append(H2b_class2)\n",
    "    if H2d:\n",
    "        if class1:\n",
    "            H2d_class1 = df[\"has mouse MHC-I d\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class I H2d\" % (H2d_class1.sum(), len(df)))\n",
    "            masks.append(H2d_class1)\n",
    "        if class2:\n",
    "            H2d_class2 = df[\"has mouse MHC-II d\"]\n",
    "            if verbose:\n",
    "                print(\"==> %d/%d peptides have class II H2d\" % (H2d_class2.sum(), len(df)))\n",
    "            masks.append(H2d_class2)\n",
    "            \n",
    "    if length:\n",
    "        length_mask = df.Sequence.str.len() == length\n",
    "        if verbose:\n",
    "            print(\"==> %d/%d peptides are %dmer\" % (length_mask.sum(), len(df), length))\n",
    "        masks.append(length_mask)\n",
    "    \n",
    "    if bcell:\n",
    "        bcell_mask = df.overlapping_bcell_epitope_length >= 4\n",
    "        if verbose:\n",
    "            print(\"==> %d/%d peptides overlap B-cell epitopes\" % (bcell_mask.sum(), len(bcell_mask)))\n",
    "        masks.append(bcell_mask)\n",
    "        \n",
    "    if exclude_accessible_spike_regions:\n",
    "        accessible_spike_mask = (\n",
    "            (df.longest_contiguous_accessible_region_in_spike > 3) |\n",
    "            (df.max_accessibility > 0.5)\n",
    "        )\n",
    "        not_accessible_mask = ~accessible_spike_mask\n",
    "        if verbose:\n",
    "            print(\"%d/%d peptides do not contain accessible regions of the spike protein\" % (\n",
    "                not_accessible_mask.sum(),\n",
    "                len(not_accessible_mask)))\n",
    "        masks.append(not_accessible_mask)\n",
    "        \n",
    "    combined_mask = np.ones(len(df), dtype=bool)\n",
    "    for m in masks:\n",
    "        combined_mask = combined_mask & m.values\n",
    "    df_subset = df[combined_mask]\n",
    "    if verbose:\n",
    "        print(\"==> Keeping %d/%d peptides\" % (len(df_subset), len(df)))\n",
    "    return df_subset\n",
    "\n",
    "def select_vaccine_peptides(\n",
    "        df, \n",
    "        length=None,\n",
    "        class1=True,\n",
    "        class2=True,\n",
    "        H2b=True,\n",
    "        H2d=True,\n",
    "        bcell=False,\n",
    "        bcell_epitopes=None,\n",
    "        min_population_coverage=0.9,\n",
    "        min_per_protein={\"S\": 1, \"M\": 1, \"N\": 1},\n",
    "        max_kmer_overlap=9,\n",
    "        verbose=False):\n",
    "    \n",
    "    df = annotate_selection_criteria(df)\n",
    "    if bcell:\n",
    "        assert bcell_epitopes is not None\n",
    "        df = annotate_overlapping_bcell_epitopes(df, bcell_epitopes=bcell_epitopes)\n",
    "        exclude_accessible_spike_regions = False\n",
    "    else:\n",
    "        exclude_accessible_spike_regions = True\n",
    "        \n",
    "    sort_columns = choose_sort_columns(class1, class2, bcell)\n",
    "    df = apply_filters(\n",
    "        df=df,\n",
    "        class1=class1, \n",
    "        class2=class2, \n",
    "        bcell=bcell, \n",
    "        H2b=H2b, \n",
    "        H2d=H2d, \n",
    "        length=length,\n",
    "        exclude_accessible_spike_regions=exclude_accessible_spike_regions,\n",
    "        verbose=verbose)\n",
    "    \n",
    "    k = max_kmer_overlap + 1\n",
    "\n",
    "    \n",
    "    class1_population_coverage = 0.0\n",
    "    class2_population_coverage = 0.0\n",
    "    \n",
    "    selected_rows = []\n",
    "    seen_kmers = set()\n",
    "    protein_counts = Counter()\n",
    " \n",
    "    chosen_class1_alleles = set()\n",
    "    chosen_class2_alleles = set()\n",
    "    need_more_class1 = class1\n",
    "    need_more_class2 = class2\n",
    "    missing_proteins = set(min_per_protein.keys())\n",
    "    \n",
    "    if bcell:\n",
    "        missing_bcell_epitopes = set(df.full_bcell_epitope.unique())\n",
    "    else:\n",
    "        missing_bcell_epitopes = set()\n",
    "    \n",
    "    while need_more_class1 or need_more_class2 or len(missing_proteins) > 0 or len(missing_bcell_epitopes) > 0:\n",
    "        df_with_marginal_coverage = update_marginal_coverage(\n",
    "            df=df,\n",
    "            allele_frequencies=allele_to_frequency, \n",
    "            chosen_class1_alleles=chosen_class1_alleles, \n",
    "            chosen_class2_alleles=chosen_class2_alleles)\n",
    "        \n",
    "        \n",
    "        df_sorted = df_with_marginal_coverage.sort_values(sort_columns, ascending=False)\n",
    "        \n",
    "        missing_proteins = {\n",
    "            p \n",
    "            for p, min_count in min_per_protein.items() \n",
    "            if protein_counts.get(p, 0) < min_count \n",
    "        }\n",
    "        \n",
    "        \n",
    "        seen_bcell_epitopes = {row.full_bcell_epitope for row in selected_rows}\n",
    "        missing_bcell_epitopes = missing_bcell_epitopes.difference(seen_bcell_epitopes)\n",
    "                \n",
    "        # if we don't care about B-cell epitopes then get rid of \n",
    "        # any peptides which don't contribute some amount of marginal\n",
    "        # coverage for either Class I or Class II \n",
    "        if need_more_class1 and need_more_class2:\n",
    "            marginal_coverage_mask = (\n",
    "                (df_sorted.marginal_class1_coverage > 0).values &\n",
    "                (df_sorted.marginal_class2_coverage > 0).values\n",
    "            )\n",
    "        elif need_more_class1:\n",
    "            marginal_coverage_mask = (df_sorted.marginal_class1_coverage > 0).values\n",
    "        elif need_more_class2:\n",
    "            marginal_coverage_mask = (df_sorted.marginal_class2_coverage > 0).values\n",
    "        else:\n",
    "            marginal_coverage_mask = np.zeros(len(df_sorted), dtype=bool)\n",
    "        \n",
    "        \n",
    "        if bcell:\n",
    "            missing_bcell_epitopes_mask = df_sorted.full_bcell_epitope.isin(missing_bcell_epitopes).values\n",
    "            assert len(missing_bcell_epitopes_mask) == len(marginal_coverage_mask)\n",
    "            # all combinations of masks\n",
    "            candidate_subset_order = [    \n",
    "                df_sorted[missing_bcell_epitopes_mask & marginal_coverage_mask],\n",
    "                df_sorted[missing_bcell_epitopes_mask & ~marginal_coverage_mask],\n",
    "                df_sorted[~missing_bcell_epitopes_mask & marginal_coverage_mask],\n",
    "            ]\n",
    "        else:\n",
    "            df_sorted = df_sorted[marginal_coverage_mask]\n",
    "            missing_proteins_mask = df_sorted.Protein.isin(missing_proteins)     \n",
    "            candidate_subset_order = [\n",
    "                df_sorted[missing_proteins_mask],\n",
    "                df_sorted[~missing_proteins_mask],\n",
    "            ]\n",
    "            \n",
    "        added_peptide = False\n",
    "        # first try adding peptides from proteins we don't have enough of before \n",
    "        # looking at peptides from other proteins\n",
    "        \n",
    "        for df_candidates in candidate_subset_order:\n",
    "            if added_peptide:\n",
    "                    break\n",
    "            for i, row in df_candidates.iterrows():\n",
    "                if added_peptide:\n",
    "                    break\n",
    "            \n",
    "                peptide = row.Sequence\n",
    "                    \n",
    "                peptide_kmers = {\n",
    "                    peptide[peptide_start:peptide_start + k]\n",
    "                    for peptide_start in range(len(peptide) - k + 1)\n",
    "                }\n",
    "                already_seen_sequence = any([kmer in seen_kmers for kmer in peptide_kmers])\n",
    "                \n",
    "                if already_seen_sequence:\n",
    "                    continue\n",
    "    \n",
    "                # accept peptide \n",
    "                seen_kmers.update(peptide_kmers)\n",
    "                protein_counts[row.Protein] += 1\n",
    "                selected_rows.append(row)\n",
    "                \n",
    "                class1_alleles = parse_alleles(row[\"HLA-I_haplotypes\"])\n",
    "                chosen_class1_alleles.update(class1_alleles)\n",
    "\n",
    "                class2_alleles = parse_alleles(row[\"HLA-II_haplotypes\"])\n",
    "                chosen_class2_alleles.update(class2_alleles)\n",
    "                added_peptide = True\n",
    "                \n",
    "\n",
    "        class1_population_coverage = population_coverage(\n",
    "            chosen_class1_alleles,\n",
    "            allele_frequencies=allele_to_frequency)\n",
    "        class2_population_coverage = population_coverage(\n",
    "            chosen_class2_alleles,\n",
    "            allele_frequencies=allele_to_frequency)\n",
    "                \n",
    "        if added_peptide:\n",
    "            if verbose:\n",
    "                print(\"Selected %d peptides, covering %0.2f%% of Class I (%d alleles) and %0.2f%% of Class II (%d alleles)\" % (\n",
    "                    len(selected_rows),\n",
    "                    class1_population_coverage * 100,\n",
    "                    len(chosen_class1_alleles),\n",
    "                    class2_population_coverage * 100,\n",
    "                    len(chosen_class2_alleles),\n",
    "                ))\n",
    "            need_more_class1 = class1 and (class1_population_coverage < min_population_coverage)\n",
    "            need_more_class2 = class2 and (class2_population_coverage < min_population_coverage)\n",
    "        else:\n",
    "            if verbose:\n",
    "                print(\n",
    "                    \"Unable to achieve target population coverage of %0.2f, current class1=%0.2f, class2=%0.2f\" % (\n",
    "                        min_population_coverage,\n",
    "                        class1_population_coverage,\n",
    "                        class2_population_coverage))\n",
    "            need_more_class1 = need_more_class2 = False\n",
    "            # give up on filling out protein list\n",
    "            missing_proteins = set()\n",
    "    \n",
    "    if len(selected_rows) == 0:\n",
    "        print(\"Failed to select any peptides\")\n",
    "        return None \n",
    "\n",
    "    df_result = pd.DataFrame.from_records(selected_rows).sort_values([\"Length\", \"Protein\", \"Start\"])\n",
    "    print(\"\\t%d peptides\\n\\tClass I coverage = %0.2f%% (%d alleles)\\n\\tClass II coverage %0.2f%% (%d alleles)\\n\\tProtein counts: %s\" % (\n",
    "        len(df_result),\n",
    "        class1_population_coverage * 100,\n",
    "        len(chosen_class1_alleles),\n",
    "        class2_population_coverage * 100,\n",
    "        len(chosen_class2_alleles),\n",
    "        dict(protein_counts.items())\n",
    "    ))\n",
    "\n",
    "    return df_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1) Class1=False, Class2=True, H2b=False, H2d=False, Length=15\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-15mer.csv\n",
      "\n",
      "2) Class1=False, Class2=True, H2b=False, H2d=False, Length=21\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-21mer.csv\n",
      "\n",
      "3) Class1=False, Class2=True, H2b=False, H2d=False, Length=27\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.30% (11 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-27mer.csv\n",
      "\n",
      "4) Class1=False, Class2=True, H2b=False, H2d=True, Length=15\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-15mer.csv\n",
      "\n",
      "5) Class1=False, Class2=True, H2b=False, H2d=True, Length=21\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-21mer.csv\n",
      "\n",
      "6) Class1=False, Class2=True, H2b=False, H2d=True, Length=27\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.30% (11 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'N': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2d-27mer.csv\n",
      "\n",
      "7) Class1=False, Class2=True, H2b=True, H2d=False, Length=15\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-15mer.csv\n",
      "\n",
      "8) Class1=False, Class2=True, H2b=True, H2d=False, Length=21\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-21mer.csv\n",
      "\n",
      "9) Class1=False, Class2=True, H2b=True, H2d=False, Length=27\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 91.30% (11 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-27mer.csv\n",
      "\n",
      "10) Class1=False, Class2=True, H2b=True, H2d=True, Length=15\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 83.95% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-15mer.csv\n",
      "\n",
      "11) Class1=False, Class2=True, H2b=True, H2d=True, Length=21\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-21mer.csv\n",
      "\n",
      "12) Class1=False, Class2=True, H2b=True, H2d=True, Length=27\n",
      "False True\n",
      "False True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 91.30% (11 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-h2b-h2d-27mer.csv\n",
      "\n",
      "13) Class1=True, Class2=False, H2b=False, H2d=False, Length=15\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.72% (13 alleles)\n",
      "\tClass II coverage 22.63% (1 alleles)\n",
      "\tProtein counts: {'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-15mer.csv\n",
      "\n",
      "14) Class1=True, Class2=False, H2b=False, H2d=False, Length=21\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.90% (13 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-21mer.csv\n",
      "\n",
      "15) Class1=True, Class2=False, H2b=False, H2d=False, Length=27\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-27mer.csv\n",
      "\n",
      "16) Class1=True, Class2=False, H2b=False, H2d=True, Length=15\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.31% (12 alleles)\n",
      "\tClass II coverage 22.63% (1 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-15mer.csv\n",
      "\n",
      "17) Class1=True, Class2=False, H2b=False, H2d=True, Length=21\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.90% (13 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-21mer.csv\n",
      "\n",
      "18) Class1=True, Class2=False, H2b=False, H2d=True, Length=27\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.90% (13 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-h2d-27mer.csv\n",
      "\n",
      "19) Class1=True, Class2=False, H2b=True, H2d=False, Length=15\n",
      "True False\n",
      "True False\n",
      "False False\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 46.49% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-15mer.csv\n",
      "\n",
      "20) Class1=True, Class2=False, H2b=True, H2d=False, Length=21\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 93.31% (12 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-21mer.csv\n",
      "\n",
      "21) Class1=True, Class2=False, H2b=True, H2d=False, Length=27\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 94.72% (14 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-27mer.csv\n",
      "\n",
      "22) Class1=True, Class2=False, H2b=True, H2d=True, Length=15\n",
      "True False\n",
      "True False\n",
      "False False\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 46.49% (3 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-15mer.csv\n",
      "\n",
      "23) Class1=True, Class2=False, H2b=True, H2d=True, Length=21\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 49.50% (4 alleles)\n",
      "\tProtein counts: {'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-21mer.csv\n",
      "\n",
      "24) Class1=True, Class2=False, H2b=True, H2d=True, Length=27\n",
      "True False\n",
      "False False\n",
      "\t2 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 55.50% (5 alleles)\n",
      "\tProtein counts: {'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd8-h2b-h2d-27mer.csv\n",
      "\n",
      "25) Class1=True, Class2=True, H2b=False, H2d=False, Length=15\n",
      "True True\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t4 peptides\n",
      "\tClass I coverage = 90.22% (11 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1, 'N': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-15mer.csv\n",
      "\n",
      "26) Class1=True, Class2=True, H2b=False, H2d=False, Length=21\n",
      "True True\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "False True\n",
      "\t5 peptides\n",
      "\tClass I coverage = 92.07% (12 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 2, 'N': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-21mer.csv\n",
      "\n",
      "27) Class1=True, Class2=True, H2b=False, H2d=False, Length=27\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "False True\n",
      "\t4 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 88.45% (11 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-27mer.csv\n",
      "\n",
      "28) Class1=True, Class2=True, H2b=False, H2d=True, Length=15\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "False True\n",
      "\t4 peptides\n",
      "\tClass I coverage = 90.17% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 3}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-15mer.csv\n",
      "\n",
      "29) Class1=True, Class2=True, H2b=False, H2d=True, Length=21\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'M': 2, 'S': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-21mer.csv\n",
      "\n",
      "30) Class1=True, Class2=True, H2b=False, H2d=True, Length=27\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.30% (11 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 2, 'M': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2d-27mer.csv\n",
      "\n",
      "31) Class1=True, Class2=True, H2b=True, H2d=False, Length=15\n",
      "True True\n",
      "True True\n",
      "\t2 peptides\n",
      "\tClass I coverage = 81.77% (8 alleles)\n",
      "\tClass II coverage 62.49% (6 alleles)\n",
      "\tProtein counts: {'M': 1, 'S': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-15mer.csv\n",
      "\n",
      "32) Class1=True, Class2=True, H2b=True, H2d=False, Length=21\n",
      "True True\n",
      "False True\n",
      "False True\n",
      "False True\n",
      "\t4 peptides\n",
      "\tClass I coverage = 93.90% (13 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 2, 'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-21mer.csv\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "33) Class1=True, Class2=True, H2b=True, H2d=False, Length=27\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 92.24% (12 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 1, 'N': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-27mer.csv\n",
      "\n",
      "34) Class1=True, Class2=True, H2b=True, H2d=True, Length=15\n",
      "True True\n",
      "\t1 peptides\n",
      "\tClass I coverage = 77.20% (7 alleles)\n",
      "\tClass II coverage 36.16% (3 alleles)\n",
      "\tProtein counts: {'M': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-15mer.csv\n",
      "\n",
      "35) Class1=True, Class2=True, H2b=True, H2d=True, Length=21\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'M': 2, 'S': 1}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-21mer.csv\n",
      "\n",
      "36) Class1=True, Class2=True, H2b=True, H2d=True, Length=27\n",
      "True True\n",
      "True True\n",
      "False True\n",
      "\t3 peptides\n",
      "\tClass I coverage = 91.19% (10 alleles)\n",
      "\tClass II coverage 84.69% (10 alleles)\n",
      "\tProtein counts: {'S': 1, 'M': 2}\n",
      "\tFilename: selected-tcell-cd4-cd8-h2b-h2d-27mer.csv\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for class1 in [False, True]:\n",
    "    for class2 in [False, True]:\n",
    "        if not (class1 or class2):\n",
    "            continue\n",
    "        for h2b in [False, True]:\n",
    "            for h2d in [False, True]:\n",
    "                for length in df_concat.Length.unique():\n",
    "                    print(\"\\n%d) Class1=%s, Class2=%s, H2b=%s, H2d=%s, Length=%d\" % (\n",
    "                        i + 1,\n",
    "                        class1, \n",
    "                        class2, \n",
    "                        h2b, \n",
    "                        h2d, \n",
    "                        length))\n",
    "                    i += 1\n",
    "                    df_tcell_vaccine_peptides = \\\n",
    "                        select_vaccine_peptides(\n",
    "                            df_with_bcell_epitopes,\n",
    "                            class1=class1,\n",
    "                            class2=class2,\n",
    "                            H2b=h2b,\n",
    "                            H2d=h2d,\n",
    "                            length=length)\n",
    "                    filename = \"selected-tcell\"\n",
    "                    if class2:\n",
    "                        filename += \"-cd4\"\n",
    "                    if class1:\n",
    "                        filename += \"-cd8\"\n",
    "                    if h2b:\n",
    "                        filename += \"-h2b\"\n",
    "                    if h2d: \n",
    "                        filename += \"-h2d\"\n",
    "                    filename += \"-%dmer.csv\" % length\n",
    "                    print(\"\\tFilename: %s\" % filename)\n",
    "                    df_tcell_vaccine_peptides.to_csv(filename, index=False)\n",
    "                    \n",
    "                    \"\"\"\n",
    "                    print(df_tcell_vaccine_peptides[[\n",
    "                        \"Protein\", \"Start\", \"End\", \"Sequence\", \n",
    "                        \"HLA-I_pop_freq_rounded\", \n",
    "                        \"HLA-II_pop_freq_rounded\",\n",
    "                        \"longest_contiguous_accessible_region_in_spike\",\n",
    "                        \"max_accessibility\"]])\n",
    "                    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1) B-Cell Class1=False, Class2=True, H2b=False, H2d=False, Length=15\n",
      "==> 1914/5742 peptides are 15mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 62/5742 peptides\n",
      "Selected 1 peptides, covering 0.00% of Class I (0 alleles) and 23.38% of Class II (2 alleles)\n",
      "Selected 2 peptides, covering 72.54% of Class I (6 alleles) and 40.72% of Class II (3 alleles)\n",
      "Selected 3 peptides, covering 72.54% of Class I (6 alleles) and 40.72% of Class II (3 alleles)\n",
      "Selected 4 peptides, covering 78.90% of Class I (7 alleles) and 40.72% of Class II (3 alleles)\n",
      "Selected 5 peptides, covering 78.90% of Class I (7 alleles) and 52.84% of Class II (5 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.79, class2=0.53\n",
      "\t5 peptides\n",
      "\tClass I coverage = 78.90% (7 alleles)\n",
      "\tClass II coverage 52.84% (5 alleles)\n",
      "\tProtein counts: {'S': 5}\n",
      "\tFilename: selected-bcell-cd4-15mer.csv\n",
      "\n",
      "2) B-Cell Class1=False, Class2=True, H2b=False, H2d=False, Length=21\n",
      "==> 1914/5742 peptides are 21mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 86/5742 peptides\n",
      "Selected 1 peptides, covering 21.43% of Class I (1 alleles) and 23.38% of Class II (2 alleles)\n",
      "Selected 2 peptides, covering 78.05% of Class I (7 alleles) and 40.72% of Class II (3 alleles)\n",
      "Selected 3 peptides, covering 80.97% of Class I (8 alleles) and 53.11% of Class II (4 alleles)\n",
      "Selected 4 peptides, covering 85.38% of Class I (9 alleles) and 53.11% of Class II (4 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.85, class2=0.53\n",
      "\t4 peptides\n",
      "\tClass I coverage = 85.38% (9 alleles)\n",
      "\tClass II coverage 53.11% (4 alleles)\n",
      "\tProtein counts: {'S': 4}\n",
      "\tFilename: selected-bcell-cd4-21mer.csv\n",
      "\n",
      "3) B-Cell Class1=False, Class2=True, H2b=False, H2d=False, Length=27\n",
      "==> 1914/5742 peptides are 27mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 110/5742 peptides\n",
      "Selected 1 peptides, covering 78.05% of Class I (7 alleles) and 45.76% of Class II (4 alleles)\n",
      "Selected 2 peptides, covering 80.97% of Class I (8 alleles) and 57.10% of Class II (5 alleles)\n",
      "Selected 3 peptides, covering 88.91% of Class I (9 alleles) and 62.69% of Class II (6 alleles)\n",
      "Selected 4 peptides, covering 88.91% of Class I (9 alleles) and 62.69% of Class II (6 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.89, class2=0.63\n",
      "\t4 peptides\n",
      "\tClass I coverage = 88.91% (9 alleles)\n",
      "\tClass II coverage 62.69% (6 alleles)\n",
      "\tProtein counts: {'S': 4}\n",
      "\tFilename: selected-bcell-cd4-27mer.csv\n",
      "\n",
      "4) B-Cell Class1=False, Class2=True, H2b=False, H2d=True, Length=15\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 15mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 0/5742 peptides\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.00\n",
      "Failed to select any peptides\n",
      "\n",
      "5) B-Cell Class1=False, Class2=True, H2b=False, H2d=True, Length=21\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 21mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 0/5742 peptides\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.00\n",
      "Failed to select any peptides\n",
      "\n",
      "6) B-Cell Class1=False, Class2=True, H2b=False, H2d=True, Length=27\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 27mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 3/5742 peptides\n",
      "Selected 1 peptides, covering 66.19% of Class I (4 alleles) and 39.86% of Class II (3 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.66, class2=0.40\n",
      "\t1 peptides\n",
      "\tClass I coverage = 66.19% (4 alleles)\n",
      "\tClass II coverage 39.86% (3 alleles)\n",
      "\tProtein counts: {'S': 1}\n",
      "\tFilename: selected-bcell-cd4-h2d-27mer.csv\n",
      "\n",
      "7) B-Cell Class1=False, Class2=True, H2b=True, H2d=False, Length=15\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 1914/5742 peptides are 15mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 5/5742 peptides\n",
      "Selected 1 peptides, covering 0.00% of Class I (0 alleles) and 23.38% of Class II (2 alleles)\n",
      "Selected 2 peptides, covering 0.00% of Class I (0 alleles) and 39.39% of Class II (3 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.39\n",
      "\t2 peptides\n",
      "\tClass I coverage = 0.00% (0 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-15mer.csv\n",
      "\n",
      "8) B-Cell Class1=False, Class2=True, H2b=True, H2d=False, Length=21\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 1914/5742 peptides are 21mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 17/5742 peptides\n",
      "Selected 1 peptides, covering 21.43% of Class I (1 alleles) and 23.38% of Class II (2 alleles)\n",
      "Selected 2 peptides, covering 37.18% of Class I (2 alleles) and 39.39% of Class II (3 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.37, class2=0.39\n",
      "\t2 peptides\n",
      "\tClass I coverage = 37.18% (2 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-21mer.csv\n",
      "\n",
      "9) B-Cell Class1=False, Class2=True, H2b=True, H2d=False, Length=27\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 1914/5742 peptides are 27mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 29/5742 peptides\n",
      "Selected 1 peptides, covering 45.96% of Class I (3 alleles) and 30.30% of Class II (2 alleles)\n",
      "Selected 2 peptides, covering 64.77% of Class I (5 alleles) and 39.39% of Class II (3 alleles)\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.65, class2=0.39\n",
      "\t2 peptides\n",
      "\tClass I coverage = 64.77% (5 alleles)\n",
      "\tClass II coverage 39.39% (3 alleles)\n",
      "\tProtein counts: {'S': 2}\n",
      "\tFilename: selected-bcell-cd4-h2b-27mer.csv\n",
      "\n",
      "10) B-Cell Class1=False, Class2=True, H2b=True, H2d=True, Length=15\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 15mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 0/5742 peptides\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.00\n",
      "Failed to select any peptides\n",
      "\n",
      "11) B-Cell Class1=False, Class2=True, H2b=True, H2d=True, Length=21\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 21mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 0/5742 peptides\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.00\n",
      "Failed to select any peptides\n",
      "\n",
      "12) B-Cell Class1=False, Class2=True, H2b=True, H2d=True, Length=27\n",
      "==> 1472/5742 peptides have class II H2b\n",
      "==> 889/5742 peptides have class II H2d\n",
      "==> 1914/5742 peptides are 27mer\n",
      "==> 258/5742 peptides overlap B-cell epitopes\n",
      "==> Keeping 0/5742 peptides\n",
      "Unable to achieve target population coverage of 0.50, current class1=0.00, class2=0.00\n",
      "Failed to select any peptides\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for class1 in [False]:\n",
    "    for class2 in [False, True]:\n",
    "        if not (class1 or class2):\n",
    "            continue\n",
    "        for h2b in [False, True]:\n",
    "            for h2d in [False, True]:\n",
    "                for length in df_concat.Length.unique():\n",
    "                    print(\"\\n%d) B-Cell Class1=%s, Class2=%s, H2b=%s, H2d=%s, Length=%d\" % (\n",
    "                        i + 1,\n",
    "                        class1, \n",
    "                        class2, \n",
    "                        h2b, \n",
    "                        h2d, \n",
    "                        length))\n",
    "                    i += 1\n",
    "                    df_bcell_vaccine_peptides = \\\n",
    "                        select_vaccine_peptides(\n",
    "                            df_with_bcell_epitopes,\n",
    "                            class1=class1,\n",
    "                            class2=class2,\n",
    "                            bcell=True,\n",
    "                            bcell_epitopes=bcell_epitopes,\n",
    "                            H2b=h2b,\n",
    "                            H2d=h2d,\n",
    "                            length=length,\n",
    "                            min_population_coverage=0.5,\n",
    "                            max_kmer_overlap=length - 3,\n",
    "                            verbose=True)\n",
    "                    if df_bcell_vaccine_peptides is None:\n",
    "                        continue\n",
    "                    filename = \"selected-bcell\"\n",
    "                    if class2:\n",
    "                        filename += \"-cd4\"\n",
    "                    if class1:\n",
    "                        filename += \"-cd8\"\n",
    "                    if h2b:\n",
    "                        filename += \"-h2b\"\n",
    "                    if h2d: \n",
    "                        filename += \"-h2d\"\n",
    "                    filename += \"-%dmer.csv\" % length\n",
    "                    print(\"\\tFilename: %s\" % filename)\n",
    "                    df_bcell_vaccine_peptides.to_csv(filename, index=False)\n",
    "\n",
    "                    \"\"\"\n",
    "                    print(df_tcell_vaccine_peptides[[\n",
    "                        \"Protein\", \"Start\", \"End\", \"Sequence\", \n",
    "                        \"HLA-I_pop_freq_rounded\", \n",
    "                        \"HLA-II_pop_freq_rounded\",\n",
    "                        \"longest_contiguous_accessible_region_in_spike\",\n",
    "                        \"max_accessibility\"]])\n",
    "                    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_concat_annot_S = df_concat_annot[df_concat_annot.Protein == \"S\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.36213668499607227"
      ]
     },
     "execution_count": 363,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df_concat_annot_S.longest_contiguous_accessible_region_in_spike > 3).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    0\n",
       "2    0\n",
       "0    1\n",
       "Name: longest_contiguous_accessible_region_in_spike, dtype: int64"
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcell_vaccine_peptides.longest_contiguous_accessible_region_in_spike"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Residue Proper Numbering</th>\n",
       "      <th>SiteSpecific Accessiblity</th>\n",
       "      <th>Nude Accessibility</th>\n",
       "      <th>Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>245</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>246</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>247</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>248</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>249</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>250</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>251</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>252</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>253</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>254</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>255</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>256</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>257</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>258</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>259</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>260</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>261</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>262</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>263</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>264</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>265</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>266</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Residue Proper Numbering  SiteSpecific Accessiblity  Nude Accessibility  \\\n",
       "218                       245                       0.34                0.35   \n",
       "219                       246                       0.88                0.93   \n",
       "220                       247                       0.66                0.67   \n",
       "221                       248                       0.59                0.60   \n",
       "222                       249                       0.53                0.55   \n",
       "223                       250                       0.83                0.87   \n",
       "224                       251                       0.94                0.98   \n",
       "225                       252                       0.76                0.80   \n",
       "226                       253                       0.96                0.97   \n",
       "227                       254                       0.79                0.80   \n",
       "228                       255                       0.90                0.91   \n",
       "229                       256                       0.91                0.91   \n",
       "230                       257                       0.48                0.49   \n",
       "231                       258                       0.78                0.80   \n",
       "232                       259                       0.64                0.64   \n",
       "233                       260                       0.60                0.61   \n",
       "234                       261                       0.32                0.32   \n",
       "235                       262                       0.11                0.11   \n",
       "236                       263                       0.00                0.00   \n",
       "237                       264                       0.00                0.00   \n",
       "238                       265                       0.00                0.00   \n",
       "239                       266                       0.00                0.00   \n",
       "\n",
       "     Difference  \n",
       "218        0.01  \n",
       "219        0.05  \n",
       "220        0.01  \n",
       "221        0.01  \n",
       "222        0.02  \n",
       "223        0.04  \n",
       "224        0.04  \n",
       "225        0.04  \n",
       "226        0.01  \n",
       "227        0.01  \n",
       "228        0.00  \n",
       "229        0.00  \n",
       "230        0.01  \n",
       "231        0.02  \n",
       "232        0.00  \n",
       "233        0.00  \n",
       "234        0.00  \n",
       "235        0.00  \n",
       "236        0.00  \n",
       "237        0.00  \n",
       "238        0.00  \n",
       "239        0.00  "
      ]
     },
     "execution_count": 408,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_accessibility.iloc[218:240]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mouse class II count</th>\n",
       "      <th>has mouse MHC-II b</th>\n",
       "      <th>has mouse MHC-II d</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>333</td>\n",
       "      <td>141.0</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>283</td>\n",
       "      <td>202.0</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1745</td>\n",
       "      <td>1129.0</td>\n",
       "      <td>616.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         mouse class II count  has mouse MHC-II b  has mouse MHC-II d\n",
       "Protein                                                              \n",
       "M                         333               141.0               192.0\n",
       "N                         283               202.0                81.0\n",
       "S                        1745              1129.0               616.0"
      ]
     },
     "execution_count": 439,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell.groupby(\"Protein\")[[\"mouse class II count\", \"has mouse MHC-II b\", \"has mouse MHC-II d\"]].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "S    444\n",
       "M    125\n",
       "N     44\n",
       "Name: Protein, dtype: int64"
      ]
     },
     "execution_count": 440,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell[df_concat_annot_bcell[\"mouse class II count\"] > 1].Protein.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    31.000000\n",
       "mean      0.235473\n",
       "std       0.040004\n",
       "min       0.118847\n",
       "25%       0.208968\n",
       "50%       0.233805\n",
       "75%       0.233805\n",
       "max       0.302979\n",
       "Name: HLA-II_pop_freq, dtype: float64"
      ]
     },
     "execution_count": 474,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell[\n",
    "    (df_concat_annot_bcell.Protein == \"S\") & \n",
    "    (df_concat_annot_bcell.Length == 27) & \n",
    "    df_concat_annot_bcell[\"has mouse MHC-II b\"] & \n",
    "    ~df_concat_annot_bcell.full_bcell_epitope.isnull()][\"HLA-II_pop_freq\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sequence                                                      FRKSNLKPFERDISTEIYQAGSTPCNG\n",
       "Protein                                                                                 S\n",
       "Start                                                                                 456\n",
       "End                                                                                   482\n",
       "HLA-I_peptides                                            FRKSNLKPF,KPFERDISTEI,KSNLKPFER\n",
       "HLA-I_haplotypes                                      HLA-C*07:02,HLA-B*07:02,HLA-A*11:01\n",
       "HLA-I_pop_freq                                                                   0.459618\n",
       "HLA-II_peptides                                           ISTEIYQAGSTPCNG,LKPFERDISTEIYQA\n",
       "HLA-II_haplotypes                                HLA-DQA1*05:05/DQB1*03:01,HLA-DRB1*04:01\n",
       "HLA-II_pop_freq                                                                  0.302979\n",
       "Mouse MHC-I b                                                                         NaN\n",
       "Mouse MHC-I d                                                                         NaN\n",
       "Mouse MHC-II b                                                            ISTEIYQAGSTPCNG\n",
       "Mouse MHC-II d                                                                        NaN\n",
       "Total_frequency                                                                  0.139255\n",
       "num_extremely_unstable_dipeptides                                                       1\n",
       "num_moderately_unstable_dipeptides                                                      0\n",
       "num_cysteines                                                                           1\n",
       "num_methionines                                                                         0\n",
       "num_thiols                                                                              1\n",
       "difficult_c_terminal                                                                False\n",
       "difficult_n_terminal                                                                False\n",
       "gravy                                                                           -0.896296\n",
       "max_kmer_gravy                                                                        0.8\n",
       "extreme_local_hydrophobicity                                                        False\n",
       "difficult_local_hydrophobicity                                                      False\n",
       "too_many_cysteines                                                                  False\n",
       "manufacturability_difficulty_score                                                     12\n",
       "Length                                                                                 27\n",
       "manufacturability_feasibility_score                                              0.435275\n",
       "has mouse MHC-I b                                                                   False\n",
       "has mouse MHC-II b                                                                   True\n",
       "has mouse MHC-I d                                                                   False\n",
       "has mouse MHC-II d                                                                  False\n",
       "mouse H2-b count                                                                        1\n",
       "mouse H2-d count                                                                        0\n",
       "mouse class I count                                                                     0\n",
       "mouse class II count                                                                    1\n",
       "mouse count                                                                             1\n",
       "HLA-I_pop_freq_rounded                                                               0.46\n",
       "HLA-II_pop_freq_rounded                                                               0.3\n",
       "Total_frequency_rounded                                                              0.14\n",
       "manufacturability_feasibility_score_rounded                                          0.44\n",
       "longest_contiguous_accessible_region_in_spike                                           9\n",
       "max_accessibility                                                                    0.84\n",
       "overlapping_bcell_epitope                                               RKSNLKPFERDISTEIY\n",
       "overlapping_bcell_epitope_start                                                       457\n",
       "overlapping_bcell_epitope_end                                                         473\n",
       "overlapping_bcell_epitope_length                                                       17\n",
       "fraction_of_full_bcell_epitope                                                          1\n",
       "full_bcell_epitope                                                      RKSNLKPFERDISTEIY\n",
       "Name: 455, dtype: object"
      ]
     },
     "execution_count": 476,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_concat_annot_bcell[\n",
    "    (df_concat_annot_bcell.Protein == \"S\") & \n",
    "    (df_concat_annot_bcell.Length == 27) & \n",
    "    df_concat_annot_bcell[\"has mouse MHC-II b\"] & \n",
    "    ~df_concat_annot_bcell.full_bcell_epitope.isnull()].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
